---
title: "Quality Control Reads"
author: "Sergio E. Palma-Vera"
output: github_document
---
# Intro
This analysis processes FastQC results on reads before and after quality-trimming and adapter-removal with fastp.

The aim of this analysis is to have an overview of the read quality and coverage (before alignment).

# Load packages
```{r, message=F}
library(dplyr)
library(ggplot2)
library(fastqcr)
library(here)
library(reshape2)
library(stringr)
library(pheatmap)
```

# Load sample information
```{r}
sample_info <- read.csv(here("sample_info","sample_info_batch3","sample_info.csv"), header = T, stringsAsFactors = F) %>% mutate(name_short = str_remove(name, "-S1")) 

sample_info$Linie[sample_info$Linie == "HLB"] <- "DUHLB"

line_lev <- c("DUK","DUC","DU6","DU6P","DUHLB","FZTDU")

sample_info %>% head()
sample_info %>% dim()
```



# Genome mmu length
```{r}
genome_length <- 	2730871774 # golden path
```
http://www.ensembl.org/Mus_musculus/Info/Annotation

# Import fastqc data
Raw reads
```{r}
qc_raw_aggr <- qc_aggregate(here("batch3","01_quality_control","output"), progressbar = F)
qc_raw_stats <- qc_stats(qc_raw_aggr)
```

Post-fastp reads
```{r}
qc_fastp_aggr <- qc_aggregate(here("batch3","02_quality_trimming_adapter_removal","output_fastqc"), progressbar = F)
qc_fastp_stats <- qc_stats(qc_fastp_aggr)
```

# Count number of files that passed and failed at each module
```{r}
summary(qc_raw_aggr)
```

```{r}
summary(qc_fastp_aggr)
```


# Visualize files that passed and failed at each module
## Recode status (pass, warn, fail) into integers (2,1,0)
Raw reads
```{r}
qc_raw_aggr$status_recoded[qc_raw_aggr$status == "PASS"] <- 2L
qc_raw_aggr$status_recoded[qc_raw_aggr$status == "WARN"] <- 1L
qc_raw_aggr$status_recoded[qc_raw_aggr$status == "FAIL"] <- 0L

qc_raw_aggr_wide <- qc_raw_aggr %>% 
  dplyr::select(module,sample,status_recoded) %>% 
  dcast(module ~ sample, value.var = "status_recoded")

qc_raw_aggr_wide %>% dim()

qc_raw_aggr_wide[1:5,1:5]
```

Post fastp reads
```{r}
qc_fastp_aggr$status_recoded[qc_fastp_aggr$status == "PASS"] <- 2L
qc_fastp_aggr$status_recoded[qc_fastp_aggr$status == "WARN"] <- 1L
qc_fastp_aggr$status_recoded[qc_fastp_aggr$status == "FAIL"] <- 0L

qc_fastp_aggr_wide <- qc_fastp_aggr %>% 
  dplyr::select(module,sample,status_recoded) %>% 
  dcast(module ~ sample, value.var = "status_recoded")

qc_fastp_aggr_wide %>% dim()

qc_fastp_aggr_wide[1:5,1:5]
```

## Combine raw and post-fastp reads
```{r}
qc_module_status_ma <- left_join(qc_raw_aggr_wide, qc_fastp_aggr_wide, by = "module")

qc_module_status_ma %>% dim()
```


## Add rownames
```{r}
rownames(qc_module_status_ma) <- qc_module_status_ma$module %>% str_replace_all(pattern = " ","_")

qc_module_status_ma$module <- NULL

qc_module_status_ma %>% dim()
```

## Prepare column meta information
Get sample name
```{r}
samp_nm <- names(qc_module_status_ma) %>% str_split("-") %>% sapply(function(x) x[1])
```

Get read info
```{r}
read_info <- names(qc_module_status_ma) %>% str_split("_") %>% sapply(function(x) x[4])
```

Get "stage" info (pre/post fastp)
```{r}
qc_stage <- names(qc_module_status_ma) %>% str_split("[.]") %>% sapply(function(x) x[2])

qc_stage[is.na(qc_stage)] <- "raw"

qc_stage[qc_stage == "corrected"] <- "after_fastp"
```

```{r}
col_data <- data.frame(read_info = read_info, qc_stage = qc_stage)

rownames(col_data) <- names(qc_module_status_ma)
```

## Visualize in a heatmap all 360 files (90 samples * 2 stages * 2 reads = 360)
```{r eval = F}
png(here("batch3/03_quality_control_analysis/figures/modules_360_files.png"), 
    width = 2000, height = 1500, units = "px", res = 300)
pheatmap(qc_module_status_ma,
         main = "Module status all fastq files (n=360)\n(PASS=2 / WARN=1 / FAIL=0)",
         annotation_col = col_data, 
         cluster_rows = F, 
         cluster_cols = F, 
         show_colnames = F)
dev.off()

```

## Visualize in a heatmap raw files (90 samples * 1 stages * 2 reads = 180)
```{r, eval=F}
rownm_tmp <- rownames(col_data)[col_data$qc_stage == "raw"] # need to adjust coldata
tmp <- filter(col_data, qc_stage == "raw") 
rownames(tmp) <- rownm_tmp

png(here("batch3/03_quality_control_analysis/figures/modules_180_raw_files.png"), 
    width = 2000, height = 1500, units = "px", res = 300)
pheatmap(qc_module_status_ma[,-grep("corrected",names(qc_module_status_ma))],
         main = "Module status raw fastq files (n=180)\n(PASS=2 / WARN=1 / FAIL=0)",
         annotation_col = tmp, 
         cluster_rows = F, 
         cluster_cols = F, 
         show_colnames = F)
dev.off()

```


## Visualize in a heatmap corrected files (90 samples * 1 stages * 2 reads = 180)
```{r, eval=F}
rownm_tmp <- rownames(col_data)[col_data$qc_stage == "after_fastp"] # need to adjust coldata
tmp <- filter(col_data, qc_stage == "after_fastp") 
rownames(tmp) <- rownm_tmp

png(here("batch3/03_quality_control_analysis/figures/modules_180_corrected_files.png"), 
    width = 2000, height = 1500, units = "px", res = 300)
pheatmap(qc_module_status_ma[,grep("corrected",names(qc_module_status_ma))],
         main = "Module status corrected fastq files (n=180)\n(PASS=2 / WARN=1 / FAIL=0)",
         annotation_col = tmp, 
         cluster_rows = F, 
         cluster_cols = F, 
         show_colnames = F)
dev.off()

```

## Visualize all R1 files
```{r, eval=F}
rownm_tmp <- rownames(col_data)[col_data$read_info == "R1"] # need to adjust coldata
tmp <- filter(col_data, read_info == "R1") 
rownames(tmp) <- rownm_tmp

png(here("batch3/03_quality_control_analysis/figures/modules_180_R1_files.png"), 
    width = 2000, height = 1500, units = "px", res = 300)
pheatmap(qc_module_status_ma[,grep("R1",names(qc_module_status_ma))],
         main = "Module status R1 fastq files (n=180)\n(PASS=2 / WARN=1 / FAIL=0)",
         annotation_col = tmp, 
         cluster_rows = F, 
         cluster_cols = F, 
         show_colnames = F)
dev.off()

```


## Visualize all R2 files
```{r, eval=F}
rownm_tmp <- rownames(col_data)[col_data$read_info == "R2"] # need to adjust coldata
tmp <- filter(col_data, read_info == "R2") 
rownames(tmp) <- rownm_tmp

png(here("batch3/03_quality_control_analysis/figures/modules_180_R2_files.png"), 
    width = 2000, height = 1500, units = "px", res = 300)
pheatmap(qc_module_status_ma[,grep("R2",names(qc_module_status_ma))],
         main = "Module status R2 fastq files (n=180)\n(PASS=2 / WARN=1 / FAIL=0)",
         annotation_col = tmp, 
         cluster_rows = F, 
         cluster_cols = F, 
         show_colnames = F)
dev.off()

```

# Prepare qc_stats
## Prepare raw data
```{r}
qc_raw_stats <- qc_raw_stats %>% 
  mutate(tot.seq = as.numeric(tot.seq),
         seq.length = as.numeric(seq.length),
         tot_bp = tot.seq * seq.length,
         read_type = str_split(sample,"_") %>% sapply(function(x) x[4]),
         sample_name = str_split(sample,"_") %>% sapply(function(x) x[1]),
         sample_name = str_remove(sample_name, "-L1")) %>% 
  left_join(dplyr::select(sample_info, name_short, Linie), 
            by = c("sample_name"="name_short")) %>% 
  mutate(Linie = factor(Linie, levels = line_lev))
```

## Prepare corrected data
Fix columns
```{r}
qc_fastp_stats <- qc_fastp_stats %>% 
  mutate(tot.seq = as.numeric(tot.seq),
         seq.length = as.numeric(seq.length),
         tot_bp = tot.seq * seq.length,
         read_type = str_split(sample,"_") %>% sapply(function(x) x[4]),
         sample_name = str_split(sample,"_") %>% sapply(function(x) x[1]),
         sample_name = str_remove(sample_name, "-L1")) %>% 
  left_join(dplyr::select(sample_info, name_short, Linie), 
            by = c("sample_name"="name_short")) %>% 
  mutate(Linie = factor(Linie, levels = line_lev))

```

The columns `seq.length` and `tot_bp` are NAs. The former due to the fact that after fastp reads are of variable sizes - nothing to do about that. However, for the later, I can add the number of bp in file, that have been pre-computed.

Load total base pairs per fastq:
```{r}
nbp <- list.files(here("batch3/02_quality_trimming_adapter_removal/output_n_bp"), pattern = ".NBP") %>% 
  lapply(function(x) read.table(file.path(here("batch3/02_quality_trimming_adapter_removal/output_n_bp"),x))) %>% 
  bind_rows()

nbp <- nbp %>% 
  mutate(file = list.files(here("batch3/02_quality_trimming_adapter_removal/output_n_bp"), pattern = ".NBP")) %>% 
  rename(tot_bp = V1) %>% 
  dplyr::select(file, tot_bp) %>% 
  mutate(file = str_remove(file, ".NBP"))
```

Add new total bp info to `qc_fastp_stats`
```{r}
qc_fastp_stats <- left_join(qc_fastp_stats, nbp, by = c("sample" = "file"))
qc_fastp_stats <- qc_fastp_stats %>% 
  dplyr::select(-tot_bp.x) %>% 
  rename(tot_bp = tot_bp.y)
head(qc_fastp_stats)
dim(qc_fastp_stats)
```



# Evaluate number of pairs, number of reads and cvg per sample from raw fastq files
Same number of reads in R1 and R2 files?
```{r} 
qc_raw_stats %>% 
  dcast(sample_name+Linie ~ read_type, value.var = "tot.seq") %>% 
  apply(1, function(x) x[3]==x[4]) %>% all()
```
Yes

Visualize number of read pairs per sample
```{r, eval = F} 
png("./batch3/03_quality_control_analysis/figures/n_read_pairs_raw_fastq.png", 
    height = 2500, width = 4000, units = "px", res = 300)
qc_raw_stats %>% 
  filter(read_type == "R1") %>% 
  ggplot(aes(x = sample_name, y = tot.seq/1e6, fill = Linie)) + 
    geom_bar(stat = "identity") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("N Read Pairs (in mill)") +
    ggtitle("Number of read pairs in million in raw fastq files")
dev.off()
```

Visualize number of reads per sample
```{r, eval = F}
png("./batch3/03_quality_control_analysis/figures/n_reads_raw_fastq.png", 
    height = 2500, width = 4000, units = "px", res = 300)
qc_raw_stats %>% 
  dcast(sample_name+Linie ~ read_type, value.var = "tot.seq") %>%
  mutate(tot_reads_samp = R1 + R2) %>% 
  ggplot(aes(x = sample_name, y = tot_reads_samp, fill = Linie)) +
    geom_bar(stat = "identity") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("total read per sample (in mill)") +
    ggtitle("Number of reads in million in raw fastq files")
dev.off()
```

One of the samples (I34772, DUK) has much less number of pair reads than the rest of the samples. 
```{r, eval = F}
png("./batch3/03_quality_control_analysis/figures/avg_cvg_raw_fastq.png", 
    height = 2500, width = 4000, units = "px", res = 300)
qc_raw_stats %>% 
  dcast(sample_name+Linie ~ read_type, value.var = "tot_bp") %>% 
  mutate(tot_bp = R1 + R2, avg_cvg = tot_bp/genome_length) %>% 
  ggplot(aes(x = sample_name, y = avg_cvg, fill = Linie)) +
    geom_bar(stat = "identity") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("Average Genome Coverage per sample") +
    ggtitle("Average Genome Coverage per sample from raw fastq files") +
    geom_hline(yintercept = 5, color = "red", linetype = "dashed")
dev.off()    
```

Number of samples below avg cvg 5x
```{r}
qc_raw_stats %>% 
  dcast(sample_name+Linie ~ read_type, value.var = "tot_bp") %>% 
  mutate(tot_bp = R1 + R2, avg_cvg = tot_bp/genome_length) %>% 
  group_by(Linie) %>% 
  summarise(n_below_5x = sum(avg_cvg < 5))
```
There are 7 samples below the targeted avg-cvg

Check coverage of those samples
```{r}
qc_raw_stats %>% 
  dcast(sample_name+Linie ~ read_type, value.var = "tot_bp") %>% 
  mutate(tot_bp = R1 + R2, avg_cvg = tot_bp/genome_length) %>% 
  filter(avg_cvg < 5)
```
Out of the 7 samples, 6 samples are above 4x, while one is only 0.24x






# Summarize GC content

# Summarize pct duplication


# Conclusions

* Raw fastq files and corrected fastq files each corresponds to 180 files. All modules were visualized on a heatmap colored according to the module-status (PASS, WARN, FAIL). 

* Overall these modules had the PASS status before and after correction with fasp: "adapter content", "basic statisitcs", "per base n content", "per base sequence content", "per base sequence quality", "per sequence quality scores", "sequence duplication levels" 

* The module "sequence length distribution" changed from PASS to WARN in all samples after fastp-correction. But this is nothing to worry about; it is a consequence of the quality trimming and adapter removal, because reads become shorter.

* The module "kmer content" was partially fixed after correction.

* The "overrepresented sequences" WARN/FAIL was partially solved by fastp (only a few WARN remeained).

* "Per sequence GC content" fails before and after correction, because there is a peak and a "shoulder". This happens in most of the samples. The peak aligns with the expected distribution's peak. I had a similar problem before, and it is not something to worry about (https://www.biostars.org/p/341611/). We also discussed this in one of the status meetings and we agreed that it required no further attention. The worst case scenario is that it is due to contamination, but those sequences would no align to the reference. Reads have a mean GC content (as seen in the fastp reports) close to the expected GC content for mouse (42%, https://bionumbers.hms.harvard.edu/bionumber.aspx?&id=102409&ver=9). Since this situation is very similar to what was observed before, it requires no further attention. 

* Other than the GC-distribution issue, which requires no further measures, the data seems good enough to proceed.

* In terms of raw avg-cvg, there were 7 samples below 5x, out of which 6 samples were above 4x and one sample (DUK) only had 0.24x







