---
title: "Analysis of WGS DU-mice data"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---

```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(here)
library(plotly)
library(DT)
#library(venn)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
library(knitr)
```

```{r sample_info, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

```

```{r metrics_cohort_final_detail, include=F}

final_vcf_metrics_detail <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)

```

```{r metrics_cohort_raw, include=F}
raw_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "raw")

```

```{r metrics_cohort_after_vqsr95, include=F}
mid_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "site_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```

```{r metrics_cohort_final, include=F}
final_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(data_set = "gt_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```

```{r pca_data, include=F}

pc_pct <- read.csv(here("batches123_06_genetic_structure/data/pc_pct.csv")) %>% dplyr::select(-X)

pc_eigenscores <- read.csv(here("batches123_06_genetic_structure/data/pc_eigenscores.csv")) %>% 
  dplyr::select(-X) %>% 
  mutate(sample.id = str_remove(sample.id, "-L1")) %>% 
  left_join(sample_info, by = c("sample.id"="sample_id")) %>% 
  dplyr::select(Linie, sample.id, target_cvg, everything())
  

```

```{r win_fst_data_sel_vs_ctrl, include=F}

win_fst_sel_vs_ctrl <- lapply(
  
  list.files(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),pattern = "*.fst"), 
  
  function(fl){
    
    read.table(
      here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl),
      header = T,
      stringsAsFactors = F
      )
    
    }
  
  )

names(win_fst_sel_vs_ctrl) <- list.files(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),pattern = "*.fst"
  ) %>% str_remove(".windowed.weir.fst")

```



Read Alignment
===============


Variant Calling 
================

Column {data-width=300}
------------------------

### **Overview** 
* GATK4 best practices
* Site level filtering: 
  * bi-allelic SNPs
  * 95% truth sensitivity (VQSR)
* Genotype level filtering: 
  * DP: 4 - 77 (mean DP + 4*SD)
  * GQ >= 20
  * Else missing (./.)


### **Number of Variants**
```{r}
vcfs_main_summary <- bind_rows(
  raw_vcf_metrics_summary,
  mid_vcf_metrics_summary,
  final_vcf_metrics_summary
) %>% 
  dplyr::select(data_set, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV, TOTAL_INDELS, TOTAL_MULTIALLELIC_SNPS) %>% 
  melt(id.vars = "data_set") %>% 
  dcast(variable ~ data_set, fill = "value") %>%
  mutate(raw = as.numeric(raw), 
         site_filtered = as.numeric(site_filtered), 
         gt_filtered = as.numeric(gt_filtered)) %>% 
  dplyr::select(variable, raw, site_filtered, gt_filtered) %>% 
  knitr::kable(format.args = list(big.mark = ","), digits = 2)

vcfs_main_summary
```


Column {data-width=700, .tabset}
------------------------

### **SNPs per sample**
```{r}
p1 <- final_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_SNPS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

#p1

ggplotly(p1) 
```

### **Sample metrics**
```{r}
final_vcf_metrics_detail %>% 
  dplyr::select(Linie, SAMPLE_ALIAS, target_cvg, everything(),
                -PCT_GQ0_VARIANTS, -TOTAL_GQ0_VARIANTS) %>% 
  datatable(rownames = F, options = list(pageLength = 25))
  #kable()
```

### **Population metrics**
```{r}
here("batches123_04_FinalVCF/metrics") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  lapply(function(x){
    d <- read.table(here("batches123_04_FinalVCF/metrics",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() %>% 
  datatable(rownames = F)
```

> Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT=="RR") and all-missing sites (GT=="./.").


### **Intersections (upsetr)**
```{r prepare data for upsetr, eval = F}
fls <- here("batches123_04_FinalVCF/output/") %>% list.files(pattern = "*.SNPids") %>% 
  .[-grep("cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.SNPids",.)]

snp_ids <- lapply(fls, function(fl){
  ln <- str_remove(fl, "cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
    str_remove(".SNPids")
  d <- vroom(here("batches123_04_FinalVCF/output",fl), col_names = F) %>% mutate(line = ln, snp_id = paste0(X1,"_",X3))
  return(d)
})

names(snp_ids) <- str_remove(fls, "cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
    str_remove(".SNPids")


all_ids <- snp_ids %>% bind_rows() %>% select(snp_id) %>% unique() %>% unlist()

sets <- data.frame(ids = all_ids,
		   duk = (all_ids %chin% snp_ids$DUK$snp_id),
		   duc = (all_ids %chin% snp_ids$DUC$snp_id),
		   du6 = (all_ids %chin% snp_ids$DU6$snp_id),
		   du6p = (all_ids %chin% snp_ids$DU6P$snp_id),
		   duhlb = (all_ids %chin% snp_ids$DUhLB$snp_id),
		   fztdu = (all_ids %chin% snp_ids$FZTDU$snp_id))

write.csv(sets, here("00_dashboard/data","snp_id_sets.csv"))
```

```{r make upsetr plot, fig.width=10}
set_dat <- vroom(here("00_dashboard/data","snp_id_sets.csv")) %>% as.data.frame()

set_dat_bool <- data.frame(
  duk = as.integer(set_dat$duk),
	duc = as.integer(set_dat$duc),
	du6 = as.integer(set_dat$du6),
	du6p = as.integer(set_dat$du6p),
	duhlb = as.integer(set_dat$duhlb),
	fztdu = as.integer(set_dat$fztdu)
	)

rownames(set_dat_bool) <- set_dat$ids

upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  intersections =  list(
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    list("fztdu","duhlb","du6p","du6","duc"),
    list("fztdu","duhlb","du6p","du6","duk"),
    list("fztdu","duhlb","du6p","duc","duk"),
    list("fztdu","duhlb","du6","duc","duk"),
    list("fztdu","du6p","du6","duc","duk"),
    list("duhlb","du6p","du6","duc","duk"),
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    )
)

```

### **Missingness**

```{r}
include_graphics(here("batches123_04_FinalVCF/figures_tables","miss_cts.png"))

include_graphics(here("batches123_04_FinalVCF/figures_tables","miss_cts_facet.png"))
```

```{r}

n_snps <- mid_vcf_metrics_summary$TOTAL_SNPS

tmp <- read.csv(here("batches123_04_FinalVCF/figures_tables","n_snps_by_min_n_called_per_line.csv")) 

fraction_snps_by_min_n_called_per_line <- lapply(tmp[3:ncol(tmp)], function(x) round(x/n_snps, 2)) %>%
  bind_cols() %>%
  mutate(min_called_samples = tmp$min_called_samples) %>%
  dplyr::select(min_called_samples, everything())

fraction_snps_by_min_n_called_per_line[2:26,] %>% kable(align = "l")

rm(tmp)

```

> Figures produced externally from site-filtered biSNPs with script "./batches123_04_FinalVCF/scripts/15_visualize_missingness.R".

### **More about GT-filtering**

```{r}
data.frame(
  min_N_non_miss_per_group = 10:25,
  N_SNPs = c(8875347, 7384239, 5794507, 4415506, 3279141, 2364751, 1645183,
             1091424, 683508, 396131, 206405, 93424, 34033, 9207, 1277, 32)
) %>% kable(row.names = F, align = "l")
```

* If 1/6 groups with min 20/25 non-miss samples: **7064770 SNPs**

* If 6/6 groups with min 20/25 non-miss samples: **34033 SNPs**




SNP Annotations
===============
```{r}

```


LDD & SFS 
========================

Column
-------------------

### Allele frequency state per population
```{r}

```

### LD Decay

Column
-------------------

### Alternative Allele Frequency Distribution
```{r}

```

### Minor Allele Frequency Distribution
```{r}

```

Diversity
===========================

Genetic Structure
==============================

Column {.tabset}
----------------

### PC1 - PC2
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```


### PC2 - PC3
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC3 - PC4
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```


### PC4 - PC5
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC5 - PC6
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### Scree Plot
```{r, fig.width = 7, fig.height=7}
pc_pct %>% 
  ggplot(aes(x=PC,y=varprop)) +
    geom_bar(stat = "identity") +
    xlab("Principal Component") +
    ylab("% Variance") +
    scale_x_continuous(breaks = 1:7) +
    theme_bw()
```


Column
---------------

### Hierarchical Clustering
```{r, fig.width = 10, fig.height=5}

dend <- readRDS(here("batches123_06_genetic_structure/data/dendrogram.rds"))

dend_line <- readRDS(here("batches123_06_genetic_structure/data/Linie.rds"))

plot(dend, leaflab="none")
legend("bottom", legend = levels(dend_line), col = 1:nlevels(dend_line), pch=19,ncol=6, xpd=TRUE, inset=c(0, -.15), cex=.8)

```


### Admixture 
```{r import_fam_info}
# About sample ordering: https://www.biostars.org/p/221817/

fam_fl <- read.table(here("batches123_06_genetic_structure/data/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.fam"))
```

```{r make_ancestry_barplot_function}
make_ancestry_barplot <- function(k,q_fl){
  
  
#k=5 
#q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.5.Q")
  

    
  k_dat <- read.table(q_fl) %>% 
    bind_cols(
      dplyr::select(fam_fl, V2) %>% dplyr::rename(sample_id = V2)
      ) %>% 
    mutate(sample_id = str_remove(sample_id, "-L1")) %>% 
    left_join(sample_info, by = "sample_id") %>% 
    reshape2::melt(id.vars = c("sample_id","Linie","target_cvg")) %>% 
    mutate(Linie = factor(Linie, levels = c("FZTDU","DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
    arrange(Linie, sample_id)
  
  k_dat <- k_dat %>% 
    mutate(sample_id = factor(sample_id, levels = unique(k_dat$sample_id)))

  p <- k_dat %>% 
        ggplot(aes(x = sample_id, y = value,fill = variable)) +
          geom_bar(stat = "identity", position = "stack") +
          theme_bw(base_size = 15) +
          theme(
            axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            legend.position = "none",
            plot.title = element_text(hjust = 0.5)
            ) +
          xlab(NULL) +
          ylab("Ancestry Fraction") +
          annotate("text",x = seq(12.5,150,25), y = 1.02, label = levels(k_dat$Linie)) +
          ggtitle(paste0("K=",k)) +
          ggsave("tst.png", height = 7, width = 15)
  return(p)
}


```

```{r barplot_k5, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.5.Q")
) 
```


Admixture detailed
==============================

Column
--------

### K2
```{r barplot_k2, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=2, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.2.Q")
) 
```


### K3
```{r barplot_k3, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=3, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.3.Q")
) 
```

Column
--------

### K4
```{r barplot_k4, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=4, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.4.Q")
) 
```


### K5
```{r barplot_k5_again, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.5.Q")
) 
```


Genetic Differentiation
==========================


```{r prepare_zscores, inlcude = F}

win_fst_sel_vs_ctrl_prep <- lapply(win_fst_sel_vs_ctrl, function(d){
  d %>% 
    # filter by min number of SNPs in window (min 10 SNPs)
    filter(N_VARIANTS >= 10) %>% 
    # separate x chromosome
    mutate(is_x = CHROM == "X") %>% 
    # group by autosomes/chrX
    group_by(is_x) %>% 
    # compute zscores
    mutate(z_win_fst_score = scale(MEAN_FST))   
}) %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(contrast, levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU")))
```


Column {.tabset}
-----------------------

### Summary
```{r summary_win_fst}
summary_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast, is_x) %>% 
  summarise(
    n_win = n(),
    min = min(MEAN_FST),
    median = median(MEAN_FST),
    mean = mean(MEAN_FST),
    q95 = quantile(MEAN_FST, 0.95),
    q99 = quantile(MEAN_FST, 0.99),
    max = max(MEAN_FST)
    ) %>% 
  arrange(is_x)

summary_win_fst %>%
  mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window Fst")
```

```{r summary_win_zfst}
summary_z_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(z_win_fst_score),
    median = median(z_win_fst_score),
    mean = mean(z_win_fst_score),
    q95 = quantile(z_win_fst_score, 0.95),
    q99 = quantile(z_win_fst_score, 0.99),
    max = max(z_win_fst_score)
    ) 

summary_z_win_fst %>% kable(digits = 2, caption = "z(Mean Window Fst)")
```

### Windowed Fst Distribution
```{r, fig.width=10,fig.height=7}

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = MEAN_FST)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("Mean Window Fst")

p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = MEAN_FST)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("Mean Window Fst")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed z-Fst Distribution
```{r, fig.width=10,fig.height=7}
q <- summary_z_win_fst %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q95","q99"), variable.name = "quantile", value.name = "zFst")


p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = z_win_fst_score)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("z(Mean Window Fst)") +
    geom_vline(q, mapping = aes(xintercept = zFst, color = quantile))



p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = z_win_fst_score)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("z(Mean Window Fst)") +
    geom_hline(q, mapping = aes(yintercept = zFst, color = quantile))

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Genomewide Manhattan
```{r prepare_data_mahattans, include=F}
manhattan_dat <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  group_by(contrast, CHROM) %>% 
  arrange(contrast, CHROM, BIN_START) %>% 
  mutate(
    genomic_index = NA, 
    chr_index = 1:n(), 
    center = BIN_START + (BIN_END - BIN_START)/2,
    center_mb = center/1e6,
    chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    )

manhattan_dat$genomic_index <- 1:nrow(manhattan_dat)

ticks <- manhattan_dat %>% 
  group_by(contrast, CHROM) %>% 
  summarise(chr_median = median(genomic_index)) 

```

```{r create_export_genomewide_manhattan, eval = F}
# Create genomewide manhattan and export
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = MEAN_FST, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank()
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("Mean Window Fst")
    ggsave(
      filename = "manhattan_genomewide_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
```

```{r import_genomewide_manhattan}
# import genomewide manhattan already exported
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst.png"))

```

```{r create_chr_manhattans, eval = F}

for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst")
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
    }
```

### Chromosomes Manhattan
```{r, fig.width=12}
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr1.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr2.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr3.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr4.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr5.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr6.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr7.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr8.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr9.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr10.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr11.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr12.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr13.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr14.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr15.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr16.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr17.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr18.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr19.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chrX.png"))
```

