
Read Alignment
===============

### Mean CVG summary table
```{r}
cvg_dat %>% 
  group_by(target_cvg, data_set) %>% 
  summarise(min = min(CVG), mean = mean(CVG), median = median(CVG), max = max(CVG)) %>% 
  kable(digits = 2)
```


### Percentage of Genome covered summary 
```{r}
genome_cvg_pct_dat %>% 
  group_by(target_cvg, data_set, cvg_class) %>% 
  summarise(min = min(pct_cvg), mean = mean(pct_cvg), median = median(pct_cvg), max = max(pct_cvg)) %>% 
  # prepare datatable output
  datatable(
    rownames = F, 
    options = list(
      columnDefs = list(list(className = 'dt-center', targets = "_all")),
      pageLength = 100
      )
    ) %>% 
  formatRound(columns = c("min","mean","median","max"), digits=2)

```



SNP Annotations
===============

### summary table 
```{r}
snpeff_data$summary_table %>% kable()
```

### Change rate by chr
```{r}
snpeff_data$change_rate_by_chr %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### vars_by_type
```{r}
snpeff_data$vars_by_type %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_func_class (supp. table 11)
```{r}
snpeff_data$eff_by_func_class %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_impact (supp. table 11)
```{r}
snpeff_data$eff_by_impact %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```



Genetic Differentiation
========================

### Chr-Manhattan Windowed Fst
```{r, eval = F}
# run once to export and then import with next chunk
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst")
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}
```

```{r, fig.width=12}
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr1.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr2.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr3.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr4.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr5.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr6.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr7.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr8.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr9.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr10.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr11.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr12.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr13.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr14.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr15.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr16.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr17.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr18.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr19.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chrX.png"))
```


### Chr-Manhattan z(Windowed Fst)
```{r, eval = F}
# run once to export and then import with next chunk
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_z_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = z_win_fst_score)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst") +
      geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
      geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}
```


```{r, fig.width=12}
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr1.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr2.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr3.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr4.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr5.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr6.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr7.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr8.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr9.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr10.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr11.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr12.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr13.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr14.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr15.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr16.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr17.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr18.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr19.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chrX.png"))
```



Diversity ratio
==================
```{r pi_ratio_tables,include=F}
fls <- list.files(here("batches123_07_selection_analysis/02_pi_ratio/output"), pattern = ".csv$", full.names = T)

pi_ratio_dat <- lapply(fls, function(fl){
  read.csv(fl) %>% dplyr::select(CHROM, BIN_START, BIN_END, pi_ratio)
})

names(pi_ratio_dat) <- basename(fls) %>% str_remove("pi_ratio_") %>% str_remove(".csv")

pi_ratio_dat <- pi_ratio_dat %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(pi_ratio_log2 = log2(pi_ratio))
```



### Summary
```{r summary_win_pi}
summary_win_pi_ratio <- pi_ratio_dat %>% 
  mutate(is_x = (CHROM == "X")) %>% 
  group_by(contrast, is_x) %>% 
  summarise(
    n_win = n(),
    min = min(pi_ratio),
    q1 = quantile(pi_ratio, 0.01),
    q5 = quantile(pi_ratio, 0.05),
    median = median(pi_ratio),
    mean = mean(pi_ratio),
    max = max(pi_ratio)
    ) %>% 
  arrange(is_x)

summary_win_pi_ratio %>%
  mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window pi_ratio")
```

### Distribution
```{r, fig.width=10,fig.height=7}

p1 <- pi_ratio_dat %>% 
  ggplot(aes(x = pi_ratio)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("pi-ratio")

p2 <- pi_ratio_dat %>% 
  ggplot(aes(x = NA, y = pi_ratio)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("pi-ratio")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```


###########################
###########################
###########################


Genetic Structure
==============================

Column {.tabset}
----------------

### PC1 - PC2
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```


### PC2 - PC3
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC3 - PC4
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```


### PC4 - PC5
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC5 - PC6
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### Scree Plot
```{r, fig.width = 7, fig.height=7}
pc_pct %>% 
  ggplot(aes(x=PC,y=varprop)) +
    geom_bar(stat = "identity") +
    xlab("Principal Component") +
    ylab("% Variance") +
    scale_x_continuous(breaks = 1:7) +
    theme_bw()
```

Column
---------------

### Hierarchical Clustering
```{r, fig.width = 10, fig.height=5}

dend <- readRDS(here("batches123_06_genetic_structure/data/dendrogram.rds"))

dend_line <- readRDS(here("batches123_06_genetic_structure/data/Linie.rds"))

plot(dend, leaflab="none")
legend("bottom", legend = levels(dend_line), col = 1:nlevels(dend_line), pch=19,ncol=6, xpd=TRUE, inset=c(0, -.15), cex=.8)

```


### Admixture 
```{r import_fam_info}
# About sample ordering: https://www.biostars.org/p/221817/

fam_fl <- read.table(here("batches123_06_genetic_structure/data/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.fam"))
```

```{r barplot_k5, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) 
```

Admixture detailed
==============================

Column
--------

### K2
```{r barplot_k2, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=2, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.2.Q")
) 
```


### K3
```{r barplot_k3, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=3, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.3.Q")
) 
```

Column
--------

### K4
```{r barplot_k4, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=4, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.4.Q")
) 
```


### K5
```{r barplot_k5_again, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) 
```







SFS - LDD
========================

Column  {.tabset}
-------------------

### Alternative Allele Frequency 
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_ALT_frq.png"))
```

### Minor Allele Frequency (excl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_excl_maf0.png"))
```

### Minor Allele Frequency (incl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_incl_maf0.png"))
```

### Linkage Disequilibrium Decay
```{r}
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max5Mb.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max250Kb.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)
```

### Allele Frequency Composition
```{r}
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state_pct.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)

```

Genetic Differentiation
==========================

Column {.tabset}
-----------------------

### Summary
```{r summary_win_fst}
summary_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast, is_x) %>% 
  summarise(
    n_win = n(),
    min = min(MEAN_FST),
    median = median(MEAN_FST),
    mean = mean(MEAN_FST),
    q95 = quantile(MEAN_FST, 0.95),
    q99 = quantile(MEAN_FST, 0.99),
    max = max(MEAN_FST)
    ) %>% 
  arrange(is_x)

summary_win_fst %>%
  mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window Fst")
```

```{r summary_win_zfst}
summary_z_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(z_win_fst_score),
    median = median(z_win_fst_score),
    mean = mean(z_win_fst_score),
    q95 = quantile(z_win_fst_score, 0.95),
    q99 = quantile(z_win_fst_score, 0.99),
    max = max(z_win_fst_score)
    ) 

summary_z_win_fst %>% kable(digits = 2, caption = "z(Mean Window Fst)")
```

### Windowed Fst Distribution
```{r, fig.width=10,fig.height=7}

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = MEAN_FST)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("Mean Window Fst")

p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = MEAN_FST)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("Mean Window Fst")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed z-Fst Distribution
```{r, fig.width=10,fig.height=7}
q <- summary_z_win_fst %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q95","q99"), variable.name = "quantile", value.name = "zFst")

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = z_win_fst_score)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("z(Mean Window Fst)") +
    geom_vline(q, mapping = aes(xintercept = zFst, color = quantile), linetype = "dashed")


p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = z_win_fst_score)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("z(Mean Window Fst)") +
    geom_hline(q, mapping = aes(yintercept = zFst, color = quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Genomewide Manhattan
```{r prepare_data_mahattans, include=F}
manhattan_dat <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  group_by(contrast, CHROM) %>% 
  arrange(contrast, CHROM, BIN_START) %>% 
  mutate(
    genomic_index = NA, 
    chr_index = 1:n(), 
    center = BIN_START + (BIN_END - BIN_START)/2,
    center_mb = center/1e6,
    chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    )

manhattan_dat$genomic_index <- 1:nrow(manhattan_dat)

ticks <- manhattan_dat %>% 
  group_by(contrast, CHROM) %>% 
  summarise(chr_median = median(genomic_index)) 

```

```{r create_export_genomewide_manhattan, eval = F}
# run this once to create genomewide manhattan and export

# Manhattan of win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = MEAN_FST, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("Mean Window Fst") +
    ggtitle("Windowed Fst - Selected vs Control") +
    ggsave(
      filename = "manhattan_genomewide_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
    
# Manhattan of z-win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = z_win_fst_score, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("z(Mean Window Fst)") +
    geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
    geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
    ggtitle("z(Windowed Fst) - Selected vs Control") +
  ggsave(
      filename = "manhattan_genomewide_z_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
```

```{r import_genomewide_manhattan, fig.width=12}
# import genomewide manhattan created above
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst.png"))

include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst.png"))

```

### Genomewide mean per SNP Fst
```{r visualize_pairwise_global_fst}
#library(pheatmap)

vis_pairwise_fst <- function(contrast_data){

#contrast_data = global_fst

  pops <- contrast_data$pop1 %>% as.character() %>% c(contrast_data$pop2) %>% unique()

  ma <- matrix(rep(NA, length(pops)*length(pops)), nrow = 6, ncol = 6)
  
  rownames(ma) <- pops
  colnames(ma) <- pops

  for(i in 1:nrow(contrast_data)){
    
    col_i <- contrast_data$pop1[i]
    row_i <- contrast_data$pop2[i]
    
    ma[row_i,col_i] <- contrast_data[i,3] # 3rd col should contain values
    ma[col_i,row_i] <- contrast_data[i,3]
    }

  min_fst <- min(ma, na.rm = T)-0.01
  max_fst <- max(ma, na.rm = T)+0.01
  mid_fst <- min_fst + (max_fst - min_fst)/2
  
  p <- ggcorrplot(ma, lab=T, type = "lower") +
        scale_fill_gradient2(
          limit = c(min_fst,max_fst), low = "blue", high = "red", mid = "white", midpoint = mid_fst
          )
  return(p)
  }

vis_pairwise_fst(global_fst)

```

### Regions of Extreme Differentiation
```{r}
#win_fst_sel_vs_ctrl_prep %>% head()

win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  mutate(
    is_99th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.99) ),
    is_95th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.95) ),
      )

#win_fst_sel_vs_ctrl_prep %>% summarise(sum(is_99th_quantile)/n()) # it checks out :)


get_REDs <- function(contr){
  #contr="DU6_vs_FZTDU"
  
  #-----------------------
  # prepare q99 intervals
  #-----------------------
  # subset data q99
  ss1 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr1 <- GRanges(seqnames = ss1$CHROM, IRanges(start = ss1$BIN_START, end = ss1$BIN_END))
  
  # merge adjacent overlapping ranges
  gr_red1 <- GenomicRanges::reduce(gr1, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  #-----------------------------------------
  # prepare <99th - 95th quantile intervals
  #-----------------------------------------
  # subset data <99th - 95th
  ss2 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_95th_quantile & !is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr2 <- GRanges(seqnames = ss2$CHROM, IRanges(start = ss2$BIN_START, end = ss2$BIN_END))

  # merge adjacent overlapping ranges
  gr_red2 <- GenomicRanges::reduce(gr2, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  # Only keep <99th - 95th regions adjacent-overlapping q99 regions
  gr_red2_ov <- subsetByOverlaps(gr_red2, gr_red1, maxgap = 1)
  
  #-----------------------------------------------------------
  # combine both quantile sets and merge
  #-----------------------------------------------------------
  # Combine q99 and overlapping/adjacent <99th - 95th regions
  res <- c(gr_red1, gr_red2_ov) %>% sort()
    
  # merge adjacent/overlapping regions
  res <- reduce(res, min.gapwidth=1L)
  return(res)
  }

# apply function for each contrast
REDs <- lapply(unique(win_fst_sel_vs_ctrl_prep$contrast), function(contr) get_REDs(contr = contr))
names(REDs) <- unique(win_fst_sel_vs_ctrl_prep$contrast) # add contrast name


# Summarize Mb per contrast and chromosome
lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% 
    as.data.frame() %>% 
    group_by(seqnames) %>% 
    summarise(Mb = sum(width_Mb)) %>% 
    mutate(seqnames = factor(seqnames, levels = c(1:19,"X"))) %>% 
    arrange(seqnames)
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(seqnames ~ contrast, value.var = "Mb") %>% 
  kable(caption = "Mb Regions of Extreme Differentiation per chromosome")

# Summarize Mb per contrast
lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% as.data.frame() %>% summarise(Mb = sum(width_Mb))
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(. ~ contrast, value.var = "Mb") %>% 
  .[,-1] %>% 
  kable(caption = "Mb Regions of Extreme Differentiation per contrast")
  
```


### Genes 
```{r}
# Get genes
RED_genes <- lapply(REDs, function(x) subsetByOverlaps(mmu_gene_set_gr, x, maxgap = 1))


# Number of features per contrast
RED_genes %>% lapply(length) %>% bind_rows() %>% kable(caption = "All features")

# Number of protein_coding per contrast
lapply(RED_genes, function(x) x[x$mcols.gene_biotype == "protein_coding"]) %>% 
  lapply(length) %>% bind_rows() %>% kable(caption = "Protein coding Genes")

# break down by biotype
lapply(RED_genes, function(x){
  as.data.frame(x) %>% 
    group_by(mcols.gene_biotype) %>% 
    summarise(n = n())
}) %>% 
  bind_rows(.id = "contrast") %>% 
  reshape2::dcast(mcols.gene_biotype ~ contrast, value.var = "n") %>% 
  dplyr::rename(gene_biotype = mcols.gene_biotype) %>% 
  kable(caption = "Features by biotype")
```


