---
title: "Analysis of WGS DU-mice data"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---

Intro
==========

```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(tidyr)
library(here)
library(plotly)
library(DT)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
library(knitr)
library(ggcorrplot)
library(tibble)
library(png)
library(grid)
library(gridExtra)
library(fastqcr)
library(GenomicRanges)
library(kableExtra)
library(ape)
library(WebGestaltR)
library(pheatmap)
library(limma)
```

```{r sample_info_and_genome_length, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

#http://www.ensembl.org/Mus_musculus/Info/Annotation 
genome_length <- 2730871774 # genome length (Golden Path Length	) 
```

```{r function_import_fastqc_reports, include=F}
get_fastqc_data <- function(qc.dir){
  qc <- qc_aggregate(qc.dir)
  qc.stats <- qc_stats(qc)
  
  fields <- str_split(qc.stats$sample, "_")
  samp <- sapply(fields, function(x) x[1])
  R_pair <- sapply(fields, function(x) x[4])
  
  qc.stats$sample_name <- samp
  qc.stats$R_pair <- R_pair
  
  qc.stats$pct.dup <- as.numeric(qc.stats$pct.dup)
  qc.stats$pct.gc <- as.numeric(qc.stats$pct.gc)
  qc.stats$tot.seq <- as.numeric(qc.stats$tot.seq)
  qc.stats$seq.length <- as.numeric(qc.stats$seq.length)

  res <- list(qc=qc, qc.stats=qc.stats)
  return(res)
}

```

```{r batch1_prealignment_raw, include=F}
# import
fastqc_raw_60samp_9Lanes <- get_fastqc_data(file.path(here(),"batch1/01_quality_control/res_concat"))

# Add a column with only sample name to qc.stats
fastqc_raw_60samp_9Lanes$qc.stats <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
  mutate(sample_name = str_sub(sample, 1, 6))

# Add column for read pair
fastqc_raw_60samp_9Lanes$qc.stats <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
  mutate(R_pair = str_sub(sample, 8, 9))

# Calculate coverage and GB for each sample
NR_RL_GB_CVG_60samp_9Lanes_raw <- lapply(unique(fastqc_raw_60samp_9Lanes$qc.stats$sample_name),
#x="H07738"
                               function(x){
                                 # subset for sample
                                 dat <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
                                   filter(sample_name == x)
                                 # prepare results read number and read length
                                 res <- data.frame(sample = x,
                                                   NR_R1 = dat$tot.seq[dat$R_pair == "R1"],
                                                   NR_R2 = dat$tot.seq[dat$R_pair == "R2"],
                                                   RL_R1 = dat$seq.length[dat$R_pair == "R1"],
                                                   RL_R2 = dat$seq.length[dat$R_pair == "R2"])
                                 # add GB, CVG and genome length
                                 res <- res %>% 
                                   mutate(GB = ((NR_R1*RL_R1) + (NR_R2*RL_R2))*10^-9,
                                          CVG=GB/( genome_length*10^-9 ),
                                          genome_length = genome_length)
                                 return(res)
                               }) %>% do.call(bind_rows,.)

# add sample info
NR_RL_GB_CVG_60samp_9Lanes_raw <- inner_join(NR_RL_GB_CVG_60samp_9Lanes_raw,
                                  sample_info,
                                  by = c("sample"="sample_id"))

```

```{r batch1_prealignment_clean, include=F}
setwd(here("batch1/03_quality_control_analysis/output_n_bp"))

NR_GB_CVG_60samp_9Lanes_trimm <- lapply(NR_RL_GB_CVG_60samp_9Lanes_raw$sample, 
#sample="H07738"       
       function(sample){
        ## Define files
        fls <- list.files()[grep(sample, list.files())]
        fls_NR <- fls[grep("NR", fls)]
        fls_GB <- fls[grep("GB", fls)]
        
        ## Load and aggregate files
        NR_R1 <- lapply(fls_NR[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        NR_R2 <- lapply(fls_NR[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        GB_R1 <- lapply(fls_GB[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)
        GB_R2 <- lapply(fls_GB[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)

        # summarize results for sample
        res <- data.frame(sample = sample, 
                          NR_R1=NR_R1$total_reads, 
                          NR_R2=NR_R2$total_reads,
                          GB=GB_R1+GB_R2)
        res <- res %>% mutate(CVG=GB/(genome_length*10^-9))

      }
      ) %>% do.call(bind_rows, .)

NR_GB_CVG_60samp_9Lanes_trimm <- inner_join(NR_GB_CVG_60samp_9Lanes_trimm,
                                  sample_info,
                                  by = c("sample"="sample_id"))
```

```{r batch1_CollectWgsMetrics, include=F}
path <- file.path(here(),"batch1/06_quality_control_alignments/output")

fls <- list.files(path)

dat_list <- lapply(fls, function(x){ 
  read.delim(file.path(path, x), stringsAsFactors = FALSE, dec = ",", skip = 6, nrow = 1)
  })

names(dat_list) <- substr(x = fls, start = 1, stop = 9)

CVG_post_align_60samp_9Lanes <- dat_list[grep("NoFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))


CVG_post_align_qual_60samp_9Lanes <- dat_list[grep("WithFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))
```

```{r batch2_prealignment_raw, include=F}
fastqc_raw_10samp_10Lanes <- get_fastqc_data(file.path(here(),"batch2/01_quality_control/output_concat"))

fastqc_raw_10samp_10Lanes$qc.stats <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
  mutate(sample_name = str_sub(sample, 1, 6))

fastqc_raw_10samp_10Lanes$qc.stats <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
  mutate(R_pair = str_sub(sample, 19, 20))

NR_RL_GB_CVG_10samp_10Lanes_raw <- lapply(unique(fastqc_raw_10samp_10Lanes$qc.stats$sample_name),
#x="H07738"
                               function(x){ 
                                 # subset for sample
                                 dat <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
                                   filter(sample_name == x)
                                 # prepare results read number and read length
                                 res <- data.frame(sample = x,
                                                   NR_R1 = dat$tot.seq[dat$R_pair == "R1"],
                                                   NR_R2 = dat$tot.seq[dat$R_pair == "R2"],
                                                   RL_R1 = dat$seq.length[dat$R_pair == "R1"],
                                                   RL_R2 = dat$seq.length[dat$R_pair == "R2"])
                                 # add GB, CVG and genome length
                                 res <- res %>% 
                                   mutate(GB = ((NR_R1*RL_R1) + (NR_R2*RL_R2))*10^-9,
                                          CVG=GB/( genome_length*10^-9 ),
                                          genome_length = genome_length)
                                 return(res)
                               }) %>% do.call(bind_rows,.)

NR_RL_GB_CVG_10samp_10Lanes_raw <- inner_join(NR_RL_GB_CVG_10samp_10Lanes_raw,
                                  sample_info,
                                  by = c("sample"="sample_id"))
```

```{r batch2_prealignment_clean, include=F}
setwd(file.path(here(),"batch2/03_quality_control_analysis/GB_NR"))

NR_GB_CVG_10samp_1lane_trimm <- lapply(NR_RL_GB_CVG_10samp_10Lanes_raw$sample, 
#sample="H07739"       
       function(sample){  
        ## Define files reseq
        fls <- list.files()[grep(sample, list.files())]
        fls_NR <- fls[grep("NR", fls)]
        fls_GB <- fls[grep("GB", fls)]
        
        ## Load and aggregate files
        NR_R1 <- lapply(fls_NR[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        NR_R2 <- lapply(fls_NR[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        GB_R1 <- lapply(fls_GB[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)
        GB_R2 <- lapply(fls_GB[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)

        # summarize results for sample
        res <- data.frame(sample = sample, 
                          NR_R1=NR_R1$total_reads, 
                          NR_R2=NR_R2$total_reads,
                          GB=GB_R1+GB_R2)
        res <- res %>% mutate(CVG=GB/(genome_length*10^-9))

      }
      ) %>% do.call(bind_rows, .)

# 1+9=10 lanes 
NR_GB_CVG_10samp_10Lanes_trimm <- inner_join(NR_GB_CVG_10samp_1lane_trimm,
                                             NR_GB_CVG_60samp_9Lanes_trimm,
                                             by = "sample",
                                             suffix = c("_1lane", "_9Lanes")) %>% 
                                  select(sample, 
                                         NR_R1_9Lanes, NR_R2_9Lanes, NR_R1_1lane, NR_R2_1lane,
                                         GB_9Lanes, GB_1lane,
                                         CVG_9Lanes, CVG_1lane) %>% 
                                  mutate(CVG_10Lanes = CVG_9Lanes + CVG_1lane) %>% 
                                  inner_join(sample_info,by = c("sample"="sample_id"))

```

```{r batch2_CollectWgsMetrics, include=F}
path <- file.path(here(),"batch2/06_quality_control_alignments/output")

fls <- list.files(path)

dat_list <- lapply(fls, function(x){ 
  read.delim(file.path(path, x), stringsAsFactors = FALSE, dec = ",", skip = 6, nrow = 1)
  })

names(dat_list) <- substr(x = fls, start = 1, stop = 9)

CVG_post_align_10samp_10Lanes <- dat_list[grep("NoFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))


CVG_post_align_qual_10samp_10Lanes <- dat_list[grep("WithFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))  
  
```

```{r batch3_prealignment_data, include=F}
# objects previously created, just importing
# follow link to see how objects were created

qc_raw_stats_avg_cvg <- readRDS(here("batch3/03_quality_control_analysis/r_objects/qc_raw_stats_avg_cvg.rds"))
qc_fastp_stats_avg_cvg <- readRDS(here("batch3/03_quality_control_analysis/r_objects/qc_fastp_stats_avg_cvg.rds"))
```

```{r batch3_CollectWgsMetrics, include=F}
fls <- list.files(here("batch3/06_quality_control_alignments/output"), pattern = "/*_pt1_points.txt")

cvg_summary <- lapply(fls, function(fl){
  
#fl=fls[1]
  # Prepare sample name (taking into account droput-reseqd sample)
  cond <- fl == "I34772-L1_S19_L004.sorted.RG.dedup.bqsr.CollectWgsMetrics.default_pt1_points.txt" | fl == "I34772-L1_S19_L004.sorted.RG.dedup.bqsr.CollectWgsMetrics.Q0.M0_pt1_points.txt"
  if(cond){
    s <- "I34772_dropout_reseqd"
  }else{
    s <- str_split(fl,"-") %>% sapply(function(x) x[1])
  }

  # Load sample histogram information
  d <- read.table(file.path(here("batch3/06_quality_control_alignments/output"),fl), header = T, stringsAsFactors = F)
  
  # Extract mode of metrics (with or without bp and mapping quality filters)
  m <- str_split(fl,"_") %>% sapply(function(x) x[3]) %>% 
    str_remove("sorted.RG.dedup.bqsr.CollectWgsMetrics.") %>% 
    str_remove("L00[1-9].")
  
  # Add sample and mode information (default = mapping and bp quality of min 20)
  d <- mutate(d, sample = s, mode = m, mode = str_replace(mode, "default","Q20.M20"))
  
  # add mouse line information
  d <- sample_info %>% 
    right_join(d, by = c("sample_id"="sample")) %>% 
    dplyr::rename(sample_name = sample_id)
  
}) %>% 
  # bind rows
  bind_rows()

# Add mouse line group to drop out sample and prepare columns for integration
cvg_summary <- cvg_summary %>% 
  mutate(Linie = ifelse(sample_name == "I34772_dropout_reseqd", "DUK", Linie))  %>% 
  # fix droput/resequenced sample
  mutate(target_cvg = ifelse(is.na(target_cvg), "5x", target_cvg))
  
```

```{r batch1_batch2_flagstat, include=F}
# batch1

## get files to parse
fls <- list.files(
  file.path(here(),"batch1/06_quality_control_alignments/output"),pattern = "flagstat", full.names = T
  )

## Get nr of properly mapped reads (line nr 5 in samtools flagstat output)
pct_properly_paired_reads_60samp_9Lanes <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

pct_properly_paired_reads_60samp_9Lanes <- inner_join(
  pct_properly_paired_reads_60samp_9Lanes,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

rm(fls)

# batch2 (boosting of low covered sampels in batch1)

## get files to parse
fls <- list.files(
  file.path(here(),"batch2/06_quality_control_alignments/output"),pattern = "flagstat", full.names = T
  )

## Get nr of properly mapped reads (line nr 5 in samtools flagstat output)
pct_properly_paired_reads_10samp_10Lanes <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

pct_properly_paired_reads_10samp_10Lanes <- inner_join(
  pct_properly_paired_reads_10samp_10Lanes,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

rm(fls)

# combine data
NR_alignment_dat_target_cvg_30x <- pct_properly_paired_reads_60samp_9Lanes %>% 
  # remove low covered samples from batch1
  filter(
    !(sample_name %in% pct_properly_paired_reads_10samp_10Lanes$sample_name)
    ) %>% 
  # replace same samples boosted (aka batch2)
  bind_rows(
    pct_properly_paired_reads_10samp_10Lanes
  )

```

```{r batch3_flagstat, include=F}

fls <- list.files(
  file.path(here(),"batch3/06_quality_control_alignments/output"),pattern = "flagstat.tab$", full.names = T
  )


NR_alignment_dat_target_cvg_5x <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

NR_alignment_dat_target_cvg_5x <- inner_join(
  NR_alignment_dat_target_cvg_5x,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

#remove droput, keep only resequenced sample (row 64)
NR_alignment_dat_target_cvg_5x <- NR_alignment_dat_target_cvg_5x[-64,]
```

```{r collect_cvg_data, include=F}

boosted_samples <- NR_RL_GB_CVG_10samp_10Lanes_raw$sample

cvg_dat_prealignment <- bind_rows(
  
  # raw
  bind_rows(
    NR_RL_GB_CVG_60samp_9Lanes_raw %>% 
      filter(!(sample %in% boosted_samples)) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    NR_RL_GB_CVG_10samp_10Lanes_raw %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    qc_raw_stats_avg_cvg %>% 
      mutate(target_cvg = "5x") %>% 
      dplyr::rename(sample = sample_name, CVG = avg_cvg) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG)
  ) %>% 
    mutate(data_set = "cvg_raw"),
  
  #clean
  bind_rows(
    NR_GB_CVG_60samp_9Lanes_trimm %>% 
      filter(!(sample %in% boosted_samples)) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    NR_GB_CVG_10samp_10Lanes_trimm %>% 
      dplyr::select(target_cvg, Linie, sample, CVG_10Lanes) %>% 
      dplyr::rename(CVG = CVG_10Lanes),
    qc_fastp_stats_avg_cvg %>% 
      mutate(target_cvg = "5x") %>% 
      dplyr::rename(sample = sample_name, CVG = avg_cvg) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG)
  ) %>% 
    mutate(data_set = "cvg_clean")
)


cvg_dat_postalignment <- bind_rows(
  
  # raw
  bind_rows(
    CVG_post_align_60samp_9Lanes %>% filter(!(sample_name %in% boosted_samples)),
    CVG_post_align_10samp_10Lanes,
    cvg_summary %>% 
      filter(mode == "Q0.M0") %>% 
      dplyr::select(names(CVG_post_align_10samp_10Lanes)) 
  ) %>% mutate(data_set = "m0q0"),
  
  # qual
  bind_rows(
    CVG_post_align_qual_60samp_9Lanes %>% filter(!(sample_name %in% boosted_samples)),
    CVG_post_align_qual_10samp_10Lanes,
    cvg_summary %>% 
      filter(mode == "Q20.M20") %>% 
      dplyr::select(names(CVG_post_align_qual_10samp_10Lanes))
  ) %>% mutate(data_set = "m20q20")
)

# combine cvg data before and after alignment
cvg_dat <- bind_rows(
  
  cvg_dat_prealignment %>% 
    dplyr::select(target_cvg, data_set, Linie, sample, CVG) %>% 
    # remove extra sample in 5x data set (the one discarded, aka dropout)
    filter(!(sample == "I34772_dropout")) %>% 
    # re-label sample that sample
    mutate(sample = ifelse(sample == "I34772_dropout_reseq", "I34772", sample)) %>% 
    # rename sample col
    dplyr::rename(sample_name = sample),
  
  cvg_dat_postalignment %>% 
    dplyr::select(target_cvg, data_set, Linie, sample_name, MEAN_COVERAGE) %>% 
    dplyr::rename(CVG = MEAN_COVERAGE) %>% 
    # remove extra sample in 5x data set (the one discarded, aka dropout)
    filter(!(sample_name == "I34772")) 
  ) %>% 
  # rename data set information
  mutate(
    data_set = ifelse(data_set == "cvg_raw", "raw_reads",data_set),
    data_set = ifelse(data_set == "cvg_clean", "clean_reads",data_set),
    data_set = ifelse(data_set == "m0q0", "aligned_m0q0",data_set),
    data_set = ifelse(data_set == "m20q20", "aligned_m20q20",data_set)
  ) %>% 
  # order levels for visualization
  mutate(
    Linie = ifelse(Linie == "DUHLB", "DUhLB", Linie),
    Linie = factor(Linie, levels = c("FZTDU","DUK","DUC","DU6","DU6P","DUhLB")),
    data_set = factor(data_set, levels = c("raw_reads","clean_reads","aligned_m0q0","aligned_m20q20"))
  )
```

```{r CollectInsertSizeMetrics}
# batch1
fls <- list.files(
  file.path(here(),"batch1/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.tab",
  full.names = T
  ) 

insert_size_batch1 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_remove("-L1.merged.sorted.dedup.CollectInsertSizeMetrics.tab")
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())


# batch2
fls <- list.files(
  file.path(here(),"batch2/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.tab",
  full.names = T
  ) 

insert_size_batch2 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_remove(".merged.sorted.dedup.CollectInsertSizeMetrics.tab")
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())

# batch1-2 (replace boosted samples)
insert_size_batch1_batch2 <- insert_size_batch1 %>% 
  filter(!(sample_name %in% insert_size_batch2$sample_name)) %>% 
  bind_rows(insert_size_batch2)

# batch3
fls <- list.files(
  file.path(here(),"batch3/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.txt",
  full.names = T
  ) %>% 
  # remove dropout file
  .[-grep("I34772-L1_S63_L003",.)]

insert_size_batch3 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_sub(1,6)
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())

# collect data
insert_size_dat <- bind_rows(insert_size_batch1_batch2, insert_size_batch3)
```

```{r vcf_metrics_cohort_raw, include=F}
raw_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "raw")
```

```{r vcf_metrics_cohort_after_vqsr95, include=F}
mid_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "site_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```

```{r vcf_metrics_cohort_final, include=F}

final_vcf_metrics_detail <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)

final_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(data_set = "gt_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)

```

```{r prepare_data_for_upsetr_snp_sets, eval = F, include=F}
fls <- here("batches123_04_final_VCF/output/") %>% 
  list.files(pattern = "*.SNPids") %>% 
  .[-grep("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids",.)]

snp_ids <- lapply(fls, function(fl){
  ln <- str_remove(fl, "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
    str_remove(".SNPids")
  d <- vroom(here("batches123_04_final_VCF/output",fl), col_names = F) %>% mutate(line = ln, snp_id = paste0(X1,"_",X3))
  return(d)
})

names(snp_ids) <- str_remove(fls, "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
    str_remove(".SNPids")


all_ids <- snp_ids %>% bind_rows() %>% select(snp_id) %>% unique() %>% unlist()

sets <- data.frame(ids = all_ids,
		   duk = (all_ids %chin% snp_ids$DUK$snp_id),
		   duc = (all_ids %chin% snp_ids$DUC$snp_id),
		   du6 = (all_ids %chin% snp_ids$DU6$snp_id),
		   du6p = (all_ids %chin% snp_ids$DU6P$snp_id),
		   duhlb = (all_ids %chin% snp_ids$DUhLB$snp_id),
		   fztdu = (all_ids %chin% snp_ids$FZTDU$snp_id))

write.csv(sets, here("00_dashboard/data","snp_id_sets.csv"))
rm(sets)
```

```{r snpeff_tables, include=F}
fls <- list.files(
  here("batches123_05_annotation/output"),
  pattern = ".csv"
) 

snpeff_data <- lapply(
  fls, function(fl) read.csv(here("batches123_05_annotation/output",fl))
  )

names(snpeff_data) <- fls %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.summary_") %>% 
  str_remove(".csv")

rm(fls)
```

```{r pca_data, include=F}

pc_pct <- read.csv(here("batches123_06_genetic_structure/data/pc_pct.csv")) %>% dplyr::select(-X)

pc_eigenscores <- read.csv(here("batches123_06_genetic_structure/data/pc_eigenscores.csv")) %>% 
  dplyr::select(-X) %>% 
  mutate(sample.id = str_remove(sample.id, "-L1")) %>% 
  left_join(sample_info, by = c("sample.id"="sample_id")) %>% 
  dplyr::select(Linie, sample.id, target_cvg, everything())
  

```

```{r make_ancestry_barplot_function}
make_ancestry_barplot <- function(k,q_fl){

#k=5 
#q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.5.Q")

  k_dat <- read.table(q_fl) %>% 
    bind_cols(
      dplyr::select(fam_fl, V2) %>% dplyr::rename(sample_id = V2)
      ) %>% 
    mutate(sample_id = str_remove(sample_id, "-L1")) %>% 
    left_join(sample_info, by = "sample_id") %>% 
    reshape2::melt(id.vars = c("sample_id","Linie","target_cvg")) %>% 
    mutate(Linie = factor(Linie, levels = c("FZTDU","DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
    arrange(Linie, sample_id)
  
  k_dat <- k_dat %>% 
    mutate(sample_id = factor(sample_id, levels = unique(k_dat$sample_id)))

  p <- k_dat %>% 
        ggplot(aes(x = sample_id, y = value,fill = variable)) +
          geom_bar(stat = "identity", position = "stack") +
          theme_bw(base_size = 15) +
          theme(
            axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            legend.position = "none",
            plot.title = element_text(hjust = 0.5)
            ) +
          xlab(NULL) +
          ylab("Ancestry Fraction") +
          annotate("text",x = seq(12.5,150,25), y = 1.02, label = levels(k_dat$Linie)) +
          ggtitle(paste0("K=",k)) +
          ggsave("tst.png", height = 7, width = 15)
  return(p)
}
```

```{r win_fst_data_sel_vs_ctrl_zscore, include=F}

win_fst_sel_vs_ctrl <- lapply(
  
  list.files(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
             pattern = "windowed.weir.fst$"), 
  
  function(fl){
    
    read.table(
      here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl),
      header = T,
      stringsAsFactors = F
      )
    }
  )

names(win_fst_sel_vs_ctrl) <- list.files(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
  pattern = "windowed.weir.fst$"
  ) %>% str_remove(".windowed.weir.fst")

# prepare data (i.e. zscore transformation) for visualizations
win_fst_sel_vs_ctrl_prep <- lapply(win_fst_sel_vs_ctrl, function(d){
  d %>% 
    # filter by min number of SNPs in window (min 10 SNPs)
    dplyr::filter(N_VARIANTS >= 10) %>% 
    # separate x chromosome
    mutate(is_x = CHROM == "X") %>% 
    # group by autosomes/chrX
    group_by(is_x) %>% 
    # compute zscores
    mutate(z_win_fst_score = scale(MEAN_FST))   
}) %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    )

# label extreme scores
win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  mutate(
    is_99th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.99) ),
    is_95th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.95) ),
      )

#win_fst_sel_vs_ctrl_prep %>% summarise(sum(is_99th_quantile)/n()) # it checks out :)



```

```{r mean_genomewide_fst_pairwise, include=F}

global_fst <- read.table(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/global_fst.tab")
  ) %>% 
  dplyr::rename(global_fst = V2) %>% 
  separate(col = V1, sep = "_vs_", into = c("pop1","pop2")) %>% 
  mutate(pop1 = factor(pop1, levels = c("DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
  arrange(pop1)
```

```{r ensembl_gene_set, include=F}
# Load gene set
mmu_gene_set <- read.table(here("reference_genome_ensembl/Mus_musculus.GRCm38.93_gene.gtf"), 
                           header = F, stringsAsFactors = F) 
  
names(mmu_gene_set) <- c("seqname","start","end","gene_id","gene_id2","gene_biotype") # as in readme

mmu_gene_set <-mmu_gene_set %>%   filter(seqname %in% c(1:19,"X")) # subset for autosomes and X

# Turn gene set into a genomics range object
mmu_gene_set_gr <- GRanges(seqnames = mmu_gene_set$seqname,
                           ranges = IRanges(start = mmu_gene_set$start,
                                            end = mmu_gene_set$end),
                           mcols = dplyr::select(mmu_gene_set, gene_id, gene_biotype))

```

```{r function_run_WebGestaltR, eval = F}
#---------------------------
# function_run_WebGestaltR
#---------------------------

run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}

```

```{r import_process_WebGestaltResults, include=F}
#-----------
# GOBP
#-----------

# DUC had no significant results

gobp_DUK_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DUK_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DU6_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DU6_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DU6P_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DU6P_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DUhLB_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DUhLB_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 
gobp_FERT_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_FERT_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

#-----------
# KEGG
#-----------

# DUK,DUC and DU6 had no significant results

kegg_DU6P_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_DU6P_vs_FZTDU/pathway_KEGG_sig_results.csv")
  ) 

kegg_DUhLB_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_DUhLB_vs_FZTDU/pathway_KEGG_sig_results.csv")
  )

kegg_FERT_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_FERT_vs_FZTDU/pathway_KEGG_sig_results.csv")
  ) 
  

#-----------------
# Combine results
#-----------------

sig_gobp_kegg <- list(
  gobp_DUK_vs_FZTDU %>% mutate(contrast = "DUK_vs_FZTDU", db = "gobp"), 
  gobp_DU6_vs_FZTDU %>% mutate(contrast = "DU6_vs_FZTDU", db = "gobp"), 
  gobp_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "gobp"), 
  gobp_DUhLB_vs_FZTDU %>% mutate(contrast = "DUhLB_vs_FZTDU", db = "gobp"), 
  gobp_FERT_vs_FZTDU %>% mutate(contrast = "FERT_vs_FZTDU", db = "gobp"), 
  kegg_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "kegg"), 
  kegg_DUhLB_vs_FZTDU %>% mutate(contrast = "DUhLB_vs_FZTDU", db = "kegg"), 
  kegg_FERT_vs_FZTDU %>% mutate(contrast = "FERT_vs_FZTDU", db = "kegg"), 
  kegg_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "kegg")
  ) %>% 
  bind_rows() 

#------------------
# one gene per row
#------------------
sig_gobp_kegg_one_gene_per_row <- lapply(1:nrow(sig_gobp_kegg), function(i){
  
  gene_ids <- sig_gobp_kegg$overlapId[i] %>% str_split("[;]") %>% unlist()
  
  gene_ensembl <- sig_gobp_kegg$userId[i] %>% str_split("[;]") %>% unlist()
  
  data.frame(
    gene_ids = gene_ids,
    gene_ensembl = gene_ensembl
  ) %>% 
    mutate(
      contrast = sig_gobp_kegg$contrast[i],
      db = sig_gobp_kegg$db[i],
      description = sig_gobp_kegg$description[i]
      ) %>% 
    dplyr::select(contrast, db, description, gene_ensembl)
}) %>% 
  bind_rows()

```

```{r snpeff_HIGH_snps_genes_Fst_prep, eval = F}
# prepare data once, export table and load subset in next chunk

snpeff_snps_HIGH <- read.table(
  here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.HIGH.tab"),
  header = F
  ) %>% 
  dplyr::rename(CHROM = V1, POS = V2, effect = V3, impact = V4, gene = V5, gene_ensembl = V6) %>% 
  # add Fst data per SNP
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUK_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUK_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUC_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUC_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DU6_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>%
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6P_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DU6P_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUhLB_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUhLB_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/FERT_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(FERT_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST)

#write.csv(snpeff_snps_HIGH, here("00_dashboard/for_manuscript/supplementary_table_7.csv"))

```

```{r snpeff_HIGH_snps_genes_Fst_load_data, include=F}

# load table prepared in previous chunk
snpeff_snps_HIGH <- read.csv(here("00_dashboard/for_manuscript/supplementary_table_7.csv"))


```



Read Alignment
===============

Column {.tabset}
-----------------

### Mean Genomewide Coverage
```{r visualize_cvg_data, fig.width=15, fig.height=10}

cvg_dat_means <- cvg_dat %>% 
  group_by(target_cvg, data_set) %>% summarise(mean = mean(CVG)) 

cvg_dat %>% 
  ggplot(aes(x = Linie, y = CVG, fill = target_cvg)) +
    geom_boxplot() +
    geom_hline(data = cvg_dat_means, aes(yintercept = mean, color = target_cvg), linetype = "dashed") +
    facet_wrap(~data_set, nrow = 1) +
    scale_y_continuous(breaks = seq(0,80,5)) +
    theme_bw() +
    xlab(NULL) +
    ylab("Mean Genome Coverage") +
    labs(
      title = "Mean Genomewide Coverage",
      subtitle = "By Mouse line and data set",
      caption = "Dashed lines = mean data set"
      ) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5)
    ) 

```

### Percentage of Genome Territory Covered
```{r pct_genome_cvg_visualization, fig.width=15, fig.height=10}
genome_cvg_pct_dat <- 
  # prepare data
  cvg_dat_postalignment %>% 
    # select columns
    dplyr::select(target_cvg, data_set, Linie, sample_name, ends_with("X")) %>% 
    # convert to long table
    reshape2::melt(
      id.vars = c("target_cvg","data_set", "Linie", "sample_name"),
      value.name = "pct_cvg",
      variable.name = "cvg_class"
    ) %>% 
    # fix x-axis and <-axis
    mutate(
      cvg_class = str_remove(cvg_class, "PCT_"),
      cvg_class = factor(cvg_class, paste0(c(1, seq(5,30,5), seq(40,100,10) ), "X") ),
      pct_cvg = pct_cvg*100
      ) %>% 
    #remove dropout sample (keep only re-sequenced one)
    filter(!(sample_name == "I34772"))

genome_cvg_pct_dat %>% 
  ggplot(aes(x = cvg_class, y = pct_cvg)) +
    geom_boxplot() +
    geom_point(alpha = 0.3) +
    facet_grid(target_cvg ~ data_set) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5)
    ) +
    ylab("Percentage Genome") +
    xlab(NULL) +
    scale_y_continuous(breaks = seq(0,100,10)) +
    labs(
      title = "Percentage Genome Covered X times",
      subtitle = "By target cvg and data set"
      ) 
```


### Mean CVG summary table
```{r}
cvg_dat %>% 
  group_by(target_cvg, data_set) %>% 
  summarise(min = min(CVG), mean = mean(CVG), median = median(CVG), max = max(CVG)) %>% 
  kable(digits = 2) %>% kable_styling(full_width = F)
```

### Percentage of Genome covered summary 
```{r}
genome_cvg_pct_dat %>% 
  group_by(target_cvg, data_set, cvg_class) %>% 
  summarise(min = min(pct_cvg), mean = mean(pct_cvg), median = median(pct_cvg), max = max(pct_cvg)) %>% 
  # prepare datatable output
  datatable(
    rownames = F, 
    options = list(
      columnDefs = list(list(className = 'dt-center', targets = "_all")),
      pageLength = 100
      )
    ) %>% 
  formatRound(columns = c("min","mean","median","max"), digits=2)

```

### Insert size
```{r, fig.width=15, fig.height=10}
insert_size_dat %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB", "FZTDU"))) %>% 
  ggplot(aes(x = sample_name, y = MEAN_INSERT_SIZE)) +
    geom_bar(stat = "identity") +
    geom_errorbar(
      aes(x = sample_name, 
          ymax = MEAN_INSERT_SIZE + STANDARD_DEVIATION, 
          ymin = MEAN_INSERT_SIZE - STANDARD_DEVIATION),
      width=0.2
      ) +
    facet_wrap(target_cvg ~ Linie, scales = "free", nrow = 2) +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5)
    ) +
    ylab("Mean Insert Size (+/- SD)") +
    xlab(NULL) +
    ggtitle("Per Sample Mean Insert Size")
```

### Per-sample metrics
```{r}
# raw nr of reads
d1 <- 
  # raw nr ofreads
  bind_rows(
    # batch1 and 2
    NR_RL_GB_CVG_60samp_9Lanes_raw %>% 
      # replace boosted samples
      filter( !(sample %in% NR_RL_GB_CVG_10samp_10Lanes_raw$sample) ),
      NR_RL_GB_CVG_10samp_10Lanes_raw
    ) %>% 
      # prepare col names
      dplyr::select(target_cvg, Linie, sample, NR_R1, NR_R2) %>%
      dplyr::rename(sample_name = sample, R1 = NR_R1, R2 = NR_R2) %>% 
      bind_rows(
        #batch3
        qc_raw_stats_avg_cvg %>% 
          # remove dorpout sample
          filter( !(sample_name %in% "I34772_dropout") ) %>% 
          mutate(
            # rename resequenced sample and add target cvg column
            sample_name = ifelse(sample_name == "I34772_dropout_reseq", "I34772",sample_name),
            target_cvg = "5x"
            ) %>% 
          # arrange columns for binding
          dplyr::select(target_cvg, Linie, sample_name, R1_NR, R2_NR) %>% 
          dplyr::rename(R1 = R1_NR, R2 = R2_NR)
      ) %>% 
  # fix duhlb name
  mutate(Linie = ifelse(Linie == "DUHLB", "DUhLB", Linie))

# clean nr of reads and pairs aligned
d2 <- bind_rows(
        NR_alignment_dat_target_cvg_30x, 
        NR_alignment_dat_target_cvg_5x
      )

# mean coverage
d3 <- cvg_dat %>% 
  filter(data_set == "aligned_m0q0") %>% dplyr::select(target_cvg, Linie, sample_name, CVG) %>% 
  mutate(sample_name = ifelse(sample_name == "I34772_dropout_reseqd", "I34772",sample_name))

# coverage pct at 5x
d4 <- genome_cvg_pct_dat %>% filter(data_set == "m0q0" & cvg_class == "5X") %>% 
  dplyr::select(target_cvg, Linie, sample_name, pct_cvg) %>% 
  dplyr::rename(pct_cvg_5x = pct_cvg) %>% 
  mutate(sample_name = ifelse(sample_name == "I34772_dropout_reseqd", "I34772",sample_name))

per_sample_metrics <- left_join(
  d1,d2,by=c("target_cvg", "Linie", "sample_name")
) %>% 
  left_join(d3, by=c("target_cvg", "Linie", "sample_name")) %>% 
  left_join(d4, by=c("target_cvg", "Linie", "sample_name")) %>% 
  mutate(NR_raw = R1 + R2) %>% 
  dplyr::rename(
    NR_clean=paired_in_seq,
    mean_CVG = CVG
    ) %>% 
  dplyr::select(-pct_properly_paired, -R1, -R2) %>% 
  dplyr::select(target_cvg, Linie, sample_name, NR_raw, everything()) %>% 
  mutate(
    NR_clean_pct = (NR_clean/NR_raw)*100,
    properly_paired_pct = (properly_paired/NR_raw)*100
  ) %>% 
  #add insert size information
  inner_join(
    dplyr::select(insert_size_dat, target_cvg, sample_name, sample_name, MEAN_INSERT_SIZE),
    by = c("target_cvg","sample_name","sample_name")
    ) 

per_sample_metrics %>% 
    datatable(
      rownames = F, 
      options = list(
        columnDefs = list(list(className = 'dt-center', targets = "_all")),
        pageLength = 150
        )
      ) %>% 
    formatRound(columns = c("mean_CVG","pct_cvg_5x","NR_clean_pct","properly_paired_pct"), digits=2)

#write.csv(per_sample_metrics, here("00_dashboard/for_manuscript/supplementary_table_1.csv"))
```

### Summary metrics
```{r}
per_sample_metrics %>% 
  group_by(target_cvg, Linie) %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by target-cvg and by line")  

tab <- per_sample_metrics %>% 
  group_by(target_cvg) %>% 
  summarise(
    sample_size = n(),
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) 
tab %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by target-cvg")  
#write.csv(tab, here("00_dashboard/for_manuscript/table_1.csv")) 

per_sample_metrics %>% 
  group_by(Linie) %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by line")

per_sample_metrics %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "full data set")

```

Variant Calling 
================

Column {data-width=300}
------------------------

### Overview
* GATK4 best practices
* Site level filtering: 
  * bi-allelic SNPs
  * 95% truth sensitivity (VQSR)
* Genotype level filtering: 
  * DP: 4 - max-DP (max-DP = mean_DP_library + 3*SD)
  * GQ >= 20
  * Else missing (./.)
  * Filtering by min called samples per population: 15 (DUK,DUC,DU6P,DUhLB,FZTDU) or 12 (DU6)
  
### Number of Variants
```{r}
vcfs_main_summary <- bind_rows(
  raw_vcf_metrics_summary,
  mid_vcf_metrics_summary,
  final_vcf_metrics_summary
) %>% 
  dplyr::select(data_set, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV, TOTAL_INDELS, TOTAL_MULTIALLELIC_SNPS) %>% 
  reshape2::melt(id.vars = "data_set") %>% 
  reshape2::dcast(variable ~ data_set, fill = "value") %>%
  mutate(raw = as.numeric(raw), 
         site_filtered = as.numeric(site_filtered), 
         gt_filtered = as.numeric(gt_filtered)
         ) %>% 
  dplyr::select(variable, 
                raw, 
                site_filtered, 
                gt_filtered
                )

vcfs_main_summary %>% 
  knitr::kable(format.args = list(big.mark = ","), digits = 2) %>% 
  kable_styling(full_width = F)

#write.csv(vcfs_main_summary, here("00_dashboard/for_manuscript/supplementary_table_2.csv"))
```


Column  {data-width=700, .tabset}
-----------------------------------

### Missingness
```{r}
include_graphics(here("batches123_04_final_VCF/figures_tables","miss_cts.png"))

include_graphics(here("batches123_04_final_VCF/figures_tables","miss_cts_facet.png"))

#> Figures produced externally from site-filtered biSNPs with script #"./batches123_04_final_VCF/scripts/15_visualize_missingness.R".
```

### Missingness
```{r}
read.csv(
  here("batches123_04_final_VCF/figures_tables","fraction_snps_by_min_n_called_per_line.csv")
  ) %>% 
  dplyr::select(-X) %>% 
  filter(!(min_called_samples == 0)) %>% 
  kable(align = "l", digits = 2, 
        caption = "Fraction of mouse line with at least N calls") %>% 
  kable_styling(full_width = F)
```

### SNPs per sample
```{r}
p1 <- final_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_SNPS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

#p1

ggplotly(p1) 
```

### Sample metrics
```{r}
final_vcf_metrics_detail %>% 
  dplyr::select(Linie, SAMPLE_ALIAS, target_cvg, everything(),
                -PCT_GQ0_VARIANTS, -TOTAL_GQ0_VARIANTS) %>% 
  datatable(rownames = F, options = list(pageLength = 150))
  #kable()
```

### Population metrics
```{r}
tab <- here("batches123_04_final_VCF/output") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  lapply(function(x){
    d <- read.table(here("batches123_04_final_VCF/output",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() 

tab %>% 
  kable(caption = "Per-Line SNP summary", digits = 2) %>% 
  kable_styling(full_width = F) %>% 
  footnote(general = "Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT==RR) and all-missing sites (GT==./.).")

#write.csv(tab, here("00_dashboard/for_manuscript/table_2.csv"))
```

### Intersections minimal
```{r make_upsetr_plot_minimal, fig.width=10}

set_dat <- vroom(here("00_dashboard/data","snp_id_sets.csv")) %>% as.data.frame()

set_dat_bool <- data.frame(
  duk = as.integer(set_dat$duk),
	duc = as.integer(set_dat$duc),
	du6 = as.integer(set_dat$du6),
	du6p = as.integer(set_dat$du6p),
	duhlb = as.integer(set_dat$duhlb),
	fztdu = as.integer(set_dat$fztdu)
	)

rownames(set_dat_bool) <- set_dat$ids


upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  intersections =  list(
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    list("fztdu","duhlb","du6p","du6","duc"),
    list("fztdu","duhlb","du6p","du6","duk"),
    list("fztdu","duhlb","du6p","duc","duk"),
    list("fztdu","duhlb","du6","duc","duk"),
    list("fztdu","du6p","du6","duc","duk"),
    list("duhlb","du6p","du6","duc","duk"),
    list("duk","duc"),
    list("du6","du6p"),
    list("duhlb","fztdu"),
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    )
)

```

### Intersections expanded
```{r make_upsetr_plot_expanded, fig.width=10}
#png(here("00_dashboard/for_manuscript/figure_1.png"), res = 300, height = 3000, width = 4000, units = "px")

upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  intersections =  list(
    # full intersection
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    # 5 line intersctions
    list("fztdu","duhlb","du6p","du6","duc"),
    list("fztdu","duhlb","du6p","du6","duk"),
    list("fztdu","duhlb","du6p","duc","duk"),
    list("fztdu","duhlb","du6","duc","duk"),
    list("fztdu","du6p","du6","duc","duk"),
    list("duhlb","du6p","du6","duc","duk"),
    # fertility lines vs fztdu
    list("duk","duc","fztdu"),
    # body size lines vs fztdu
    list("du6","du6p","fztdu"),
    # lines vs fztdu
    list("duk","fztdu"),
    list("duc","fztdu"),
    list("du6","fztdu"),
    list("du6p","fztdu"),
    list("duhlb","fztdu"),
    # phenotype related intersection
    list("duk","duc"),
    list("du6","du6p"),
    #list("duhlb","fztdu"), #done already
    # private SNPs
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    )
)

grid.text("Intersections SNP sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))

#dev.off()
```

### Intersection Sel vs FZTDU
```{r export_venn_sel_vs_fztdu, eval=F}

png(here("00_dashboard","for_manuscript","supplementary_figure_6.png"), res = 300, height = 2000, width = 2000, units = "px")


par(mfrow = c(3, 2))

for(sel in c("duk","duc","du6","du6p","duhlb")){
  
  #sel="duk"
  
  d <- data.frame(
    set_dat$ids %in% set_dat$ids[ set_dat[,sel] ],
    set_dat$ids %in% set_dat$ids[ set_dat$fztdu ]
  )
  
  names(d) <- c(sel,"fztdu")
  
  a <- vennCounts(d)
  
  vennDiagram(a, circle.col = c("red","blue"), mar = rep(0.1,4), cex = 0.7)
  }

dev.off()

```



```{r import_venn_sel_vs_fztdu, fig.width=10, fig.height=10}
include_graphics(here("00_dashboard","for_manuscript","supplementary_figure_6.png"))
```

### Intx sel-fztdu
```{r for_manuscript_pct_intersection_sel_vs_fztdu}

intersections <- sapply(c("duk","duc","du6","du6p","duhlb","fztdu"), function(sel){
  
  x <- set_dat$ids[ set_dat[,sel] ]
  y <- set_dat$ids[ set_dat$fztdu ]
  
  xy <- intersect(x,y) %>% length()

  })

set_sizes <- sapply(c("duk","duc","du6","du6p","duhlb","fztdu"), function(sel){
  
  x <- set_dat$ids[ set_dat[,sel] ]

  length(x)

  })

tab <- data.frame(
  line = c("duk","duc","du6","du6p","duhlb","fztdu"),
  set_size = set_sizes,
  intersection_sel_x_fztdu = intersections,
  pct_sel_in_fztdu = (intersections/set_sizes)*100,
  pct_fztdu_in_sel = (intersections/ length(set_dat$ids[ set_dat$fztdu ]) )*100
)

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_9.csv"))

tab %>% 
  kable(digits = 2, row.names = FALSE) %>% kable_styling(full_width = F)

```


SNP Annotations
===============

Column {.tabset}
----------------

### summary table 
```{r}
snpeff_data$summary_table %>% kable() %>% kable_styling(full_width = F)
```

### Counts by effect
```{r}
impact_snpeff <- read.csv2(here("00_dashboard/data/impact_snpeff.csv")) %>% 
  dplyr::select(Effect, Impact)

tab <- snpeff_data$cts_by_eff %>% 
  mutate(Type = str_trim(Type)) %>% 
  left_join(impact_snpeff, by = c("Type" = "Effect")) %>% 
  unique() %>% 
  mutate(
    Impact = ifelse(Type == "5_prime_UTR_premature_start_codon_gain_variant",
                  "LOW",
                  Impact)
  ) 

tab %>% 
  datatable(
    rownames = F, 
    options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all")))
    )


#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_4.csv"))
```

### Change rate by chr
```{r}
snpeff_data$change_rate_by_chr %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### counts by genotype region
```{r}
tab <- snpeff_data$cts_by_geno_region 

tab %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_3.csv")) 
```

### vars_by_type
```{r}
snpeff_data$vars_by_type %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_func_class
```{r}
snpeff_data$eff_by_func_class %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_impact
```{r}
snpeff_data$eff_by_impact %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

Genetic Structure
==============================

Column {.tabset}
----------------

### PC1 - PC2
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC2 - PC3
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

#png(here("00_dashboard/for_manuscript/figure_2A.png"), res = 300, height = 3000, width = 4000, units = "px")
#p1
#dev.off()

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC3 - PC4
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC4 - PC5
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC5 - PC6
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

#png(here("00_dashboard/for_manuscript/figure_2B.png"), res = 300, height = 3000, width = 4000, units = "px")
#p1
#dev.off()

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### Scree Plot
```{r, fig.width = 7, fig.height=7}
pc_pct %>% 
  ggplot(aes(x=PC,y=varprop)) +
    geom_bar(stat = "identity") +
    xlab("Principal Component") +
    ylab("% Variance") +
    scale_x_continuous(breaks = 1:7) +
    theme_bw()
```

Column
---------------

### Hierarchical Clustering
```{r, fig.width = 10, fig.height=5}
dend <- readRDS(here("batches123_06_genetic_structure/data/dendrogram.rds")) %>% as.hclust()

# change labels to group to avoid legend
#identical(sample_info[,"sample_id"], str_remove(dend[["labels"]], "-L1")) #check order of labels is the same! yes!
dend[["labels"]] <- sample_info$Linie


#png(here("00_dashboard/for_manuscript/figure_2C.png"), res = 300, height = 3500, width = 4000, units = "px")

colors <- c("red", "blue", "green", "black", "purple","orange")
clus6 <- cutree(dend, 6)
plot(as.phylo(dend), 
     type = "fan", cex = 0.6,
     no.margin = TRUE,
     tip.color = colors[clus6])

#dev.off()

# good reference: http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning
```

### Admixture 
```{r import_fam_info}
# About sample ordering: https://www.biostars.org/p/221817/

fam_fl <- read.table(here("batches123_06_genetic_structure/data/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.fam"))
```

```{r barplot_k5, fig.width=20, fig.height=7}
#png(here("00_dashboard/for_manuscript/figure_2D.png"), res = 300, height = 3000, width = 6000, units = "px")

make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) +
  theme(axis.text.x = element_blank()) +
  scale_y_continuous(expand = c(0,0.02))

#dev.off()
```

Admixture detailed
==============================

Column
--------

### K2
```{r barplot_k2, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=2, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.2.Q")
) 
```


### K3
```{r barplot_k3, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=3, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.3.Q")
) 
```

Column
--------

### K4
```{r barplot_k4, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=4, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.4.Q")
) 
```


### K5
```{r barplot_k5_again, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) 
```







SFS - LDD
========================

Column  {.tabset}
-------------------

### Linkage Disequilibrium Decay
```{r}
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max5Mb.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max250Kb.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)

#png(here("00_dashboard/for_manuscript/figure_3A.png"), res = 300, height = 3000, width = 3000, units = "px")
#grid.arrange(img1, ncol = 1)
#dev.off()
```

### Allele Frequency Composition
```{r}
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state_pct.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)

#png(here("00_dashboard/for_manuscript/figure_3B.png"), res = 300, height = 3000, width = 3000, units = "px")
#grid.arrange(img2, ncol = 1)
#dev.off()
```

### Alternative Allele Frequency 
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_ALT_frq.png"))

#png(here("00_dashboard/for_manuscript/figure_3C.png"), res = 300, height = 3000, width = 3000, units = "px")
#img1 <- rasterGrob(as.raster(readPNG(here("batches123_08_LDD_SFS/figures","hist_ALT_frq.png")))) 
#grid.arrange(img1,ncol = 1)
#dev.off()
```

### Minor Allele Frequency (excl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_excl_maf0.png"))
```

### Minor Allele Frequency (incl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_incl_maf0.png"))
```

Genetic Differentiation
==========================

Column {.tabset}
-----------------------

### Summary
```{r summary_win_fst}
summary_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast, is_x) %>% 
  summarise(
    n_win = n(),
    min = min(MEAN_FST),
    median = median(MEAN_FST),
    mean = mean(MEAN_FST),
    q95 = quantile(MEAN_FST, 0.95),
    q99 = quantile(MEAN_FST, 0.99),
    max = max(MEAN_FST)
    ) %>% 
  arrange(is_x)

summary_win_fst %>%
  mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window Fst") %>% 
  kable_styling(full_width = F) %>% kable_styling(full_width = F)
```

```{r summary_win_zfst}
summary_z_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(z_win_fst_score),
    median = median(z_win_fst_score),
    mean = mean(z_win_fst_score),
    q95 = quantile(z_win_fst_score, 0.95),
    q99 = quantile(z_win_fst_score, 0.99),
    max = max(z_win_fst_score)
    ) 

summary_z_win_fst %>% kable(digits = 2, caption = "z(Mean Window Fst)") %>% kable_styling(full_width = F)
```

### Windowed Fst Distribution
```{r, fig.width=10,fig.height=7}

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = MEAN_FST)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("Mean Window Fst")

p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = MEAN_FST)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("Mean Window Fst")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed Fst Distribution (violin-boxplot)
```{r}
ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = MEAN_FST, y = contrast), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = MEAN_FST, y = contrast), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw() +
  ylab(NULL) +
  xlab("Mean Window Fst") +
  labs(
    title = "Windowed Fst Distribution"
  ) +
  theme(
    plot.title = element_text(hjust = 0.5)
  ) +
  scale_y_discrete(limits = rev(levels(win_fst_sel_vs_ctrl_prep$contrast)))
```

### Windowed z-Fst Distribution
```{r, fig.width=10,fig.height=7}
q <- summary_z_win_fst %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q95","q99"), variable.name = "quantile", value.name = "zFst")

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = z_win_fst_score)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("z(Mean Window Fst)") +
    geom_vline(q, mapping = aes(xintercept = zFst, color = quantile), linetype = "dashed")


p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = z_win_fst_score)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("z(Mean Window Fst)") +
    geom_hline(q, mapping = aes(yintercept = zFst, color = quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed z-Fst Distribution (violin-boxplot)
```{r, fig.width=7,fig.height=7}
#png(here("00_dashboard/for_manuscript/figure_4B.png"), res = 300, height = 2000, width = 2000, units = "px")

ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = z_win_fst_score, y = NA), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = z_win_fst_score, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw(
    base_size = 15
  ) +
  ylab(NULL) +
  xlab("Mean Window Fst") +
  labs(
    title = "Windowed z-Fst Distribution"
  ) +
  theme(
    plot.title = element_text(hjust = 0.5),
    strip.text = element_text(size = 7),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  geom_vline(q, mapping = aes(xintercept = zFst, color = quantile), linetype = "dashed") +
  facet_wrap(~contrast, ncol = 1, strip.position = "right") 

#dev.off()
```

### Genomewide Manhattan
```{r prepare_data_mahattans, include=F}
manhattan_dat <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  group_by(contrast, CHROM) %>% 
  arrange(contrast, CHROM, BIN_START) %>% 
  mutate(
    genomic_index = NA, 
    chr_index = 1:n(), 
    center = BIN_START + (BIN_END - BIN_START)/2,
    center_mb = center/1e6,
    chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    )

manhattan_dat$genomic_index <- 1:nrow(manhattan_dat)

ticks <- manhattan_dat %>% 
  group_by(contrast, CHROM) %>% 
  summarise(chr_median = median(genomic_index)) 

```

```{r create_export_genomewide_manhattan, eval = F}
# run this once to create genomewide manhattan and export

# Manhattan of win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = MEAN_FST, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("Mean Window Fst") +
    ggtitle("Windowed Fst - Selected vs Control") +
    ggsave(
      filename = "manhattan_genomewide_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
    
# Manhattan of z-win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = z_win_fst_score, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("z(Mean Window Fst)") +
    geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
    geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
    ggtitle("z(Windowed Fst) - Selected vs Control") +
  ggsave(
      filename = "manhattan_genomewide_z_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
```

```{r import_genomewide_manhattan, fig.width=12}
# import genomewide manhattan created above
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst.png"))

include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst.png"))

```

### Chr-Manhattan Windowed Fst
```{r, eval = F}
# run once to export and then import with next chunk
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst")
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}
```

```{r, fig.width=12, eval = F}
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr1.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr2.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr3.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr4.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr5.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr6.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr7.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr8.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr9.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr10.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr11.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr12.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr13.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr14.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr15.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr16.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr17.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr18.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chr19.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst_chrX.png"))
```

### Chr-Manhattan z(Windowed Fst)
```{r, eval = F}
# run once to export and then import with next chunk
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_z_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = z_win_fst_score)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst") +
      geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
      geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}
```

```{r, fig.width=12, eval = F}
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr1.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr2.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr3.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr4.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr5.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr6.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr7.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr8.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr9.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr10.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr11.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr12.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr13.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr14.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr15.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr16.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr17.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr18.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chr19.png"))
include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst_chrX.png"))
```

### Genomewide mean per SNP Fst
```{r visualize_pairwise_global_fst}
#library(pheatmap)

vis_pairwise_fst <- function(contrast_data){

#contrast_data = global_fst

  pops <- contrast_data$pop1 %>% as.character() %>% c(contrast_data$pop2) %>% unique()

  ma <- matrix(rep(NA, length(pops)*length(pops)), nrow = 6, ncol = 6)
  
  rownames(ma) <- pops
  colnames(ma) <- pops

  for(i in 1:nrow(contrast_data)){
    
    col_i <- contrast_data$pop1[i]
    row_i <- contrast_data$pop2[i]
    
    ma[row_i,col_i] <- contrast_data[i,3] # 3rd col should contain values
    ma[col_i,row_i] <- contrast_data[i,3]
    }

  min_fst <- min(ma, na.rm = T)-0.01
  max_fst <- max(ma, na.rm = T)+0.01
  mid_fst <- min_fst + (max_fst - min_fst)/2
  
  p <- ggcorrplot(ma, lab=T, type = "lower") +
        scale_fill_gradient2(
          limit = c(min_fst,max_fst), low = "blue", high = "red", mid = "white", midpoint = mid_fst
          )
  return(p)
  }

#png(here("00_dashboard/for_manuscript/figure_4A.png"), res = 300, height = 2000, width = 2000, units = "px")
vis_pairwise_fst(global_fst)
#dev.off()

```

### Intersection 99th quantile windows (excl FERT)
```{r instersection_q99th_windows_excl_FERT, fig.width=10}
win_q99 <- win_fst_sel_vs_ctrl_prep %>% 
  filter(is_99th_quantile, contrast != "FERT_vs_FZTDU") %>% 
  ungroup() %>% 
  mutate(win_id = paste(CHROM, BIN_START, BIN_END, sep = "_"))
  
win_q99_ids <- win_q99 %>% 
  dplyr::select(win_id) %>% 
  unique() %>% 
  unlist() %>% 
  as.character()

win_q99_set <- data.frame(
  DU6_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DU6_vs_FZTDU"],
  DU6P_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DU6P_vs_FZTDU"],
  DUK_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUK_vs_FZTDU"],
  DUC_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUC_vs_FZTDU"],
  DUhLB_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUhLB_vs_FZTDU"]
)

win_q99_set <- lapply(win_q99_set, as.integer) %>% bind_cols() %>% as.data.frame()

rownames(win_q99_set) <- win_q99_ids


#png(here("00_dashboard/for_manuscript/supplementary_figure_1.png"), res = 300, height = 3000, width = 4000, units = "px")
upset(win_q99_set, nintersects = NA, empty.intersections = T)
grid.text("Intersections 99th quantile windows",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()
```

### Intersection 99th quantile windows (incl FERT)
```{r instersection_q99th_windows_incl_FERT, fig.width=10}
win_q99 <- win_fst_sel_vs_ctrl_prep %>% 
  filter(is_99th_quantile) %>% 
  ungroup() %>% 
  mutate(win_id = paste(CHROM, BIN_START, BIN_END, sep = "_"))
  
win_q99_ids <- win_q99 %>% 
  dplyr::select(win_id) %>% 
  unique() %>% 
  unlist() %>% 
  as.character()

win_q99_set <- data.frame(
  DU6_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DU6_vs_FZTDU"],
  DU6P_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DU6P_vs_FZTDU"],
  DUK_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUK_vs_FZTDU"],
  DUC_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUC_vs_FZTDU"],
  DUhLB_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "DUhLB_vs_FZTDU"],
  FERT_vs_FZTDU = win_q99_ids %in% win_q99$win_id[win_q99$contrast == "FERT_vs_FZTDU"]
)

win_q99_set <- lapply(win_q99_set, as.integer) %>% bind_cols() %>% as.data.frame()

rownames(win_q99_set) <- win_q99_ids


#png(here("00_dashboard/for_manuscript/supplementary_figure_1.png"), res = 300, height = 2000, width = 4000, units = "px")
upset(win_q99_set,nsets = 6, nintersects = NA, empty.intersections = T)
grid.text("Intersections 99th quantile windows",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()

```

### REDs & RED-genes
```{r get_REDs_and_RED_genes}
get_REDs <- function(contr){
  #contr="DU6_vs_FZTDU"
  
  #-----------------------
  # prepare q99 intervals
  #-----------------------
  # subset data q99
  ss1 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr1 <- GRanges(seqnames = ss1$CHROM, IRanges(start = ss1$BIN_START, end = ss1$BIN_END))
  
  # merge adjacent overlapping ranges
  gr_red1 <- GenomicRanges::reduce(gr1, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  #-----------------------------------------
  # prepare <99th - 95th quantile intervals
  #-----------------------------------------
  # subset data <99th - 95th
  ss2 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_95th_quantile & !is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr2 <- GRanges(seqnames = ss2$CHROM, IRanges(start = ss2$BIN_START, end = ss2$BIN_END))

  # merge adjacent overlapping ranges
  gr_red2 <- GenomicRanges::reduce(gr2, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  # Only keep <99th - 95th regions adjacent-overlapping q99 regions
  gr_red2_ov <- subsetByOverlaps(gr_red2, gr_red1, maxgap = 1)
  
  #-----------------------------------------------------------
  # combine both quantile sets and merge
  #-----------------------------------------------------------
  # Combine q99 and overlapping/adjacent <99th - 95th regions
  res <- c(gr_red1, gr_red2_ov) %>% sort()
    
  # merge adjacent/overlapping regions
  res <- reduce(res, min.gapwidth=1L)
  return(res)
  }

# apply function for each contrast
REDs <- lapply(unique(win_fst_sel_vs_ctrl_prep$contrast), function(contr) get_REDs(contr = contr))
names(REDs) <- unique(win_fst_sel_vs_ctrl_prep$contrast) # add contrast name

# get_genes_in_REDs
RED_genes <- lapply(REDs, function(x) subsetByOverlaps(mmu_gene_set_gr, x, maxgap = 1))

```

```{r REDs_Mb_tables}
# Summarize Mb per contrast and chromosome
tab <- lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% 
    as.data.frame() %>% 
    group_by(seqnames) %>% 
    summarise(Mb = sum(width_Mb)) %>% 
    mutate(seqnames = factor(seqnames, levels = c(1:19,"X"))) %>% 
    arrange(seqnames)
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(seqnames ~ contrast, value.var = "Mb") %>% 
  janitor::adorn_totals("row") %>% 
  dplyr::rename(CHROM = seqnames) %>% 
  dplyr::select(CHROM, DUK_vs_FZTDU, DUC_vs_FZTDU, DU6_vs_FZTDU, DU6P_vs_FZTDU, DUhLB_vs_FZTDU, FERT_vs_FZTDU)

#write.csv(tab, here("00_dashboard/for_manuscript/table_3.csv"))

tab %>% 
  #kable(caption = "Mb Regions of Extreme Differentiation per chromosome") %>% 
  #kable_styling(full_width = F)
  datatable(caption = "Mb Regions of Extreme Differentiation per chromosome", 
            options = list(pageLength = 21))

# Summarize Mb per contrast
lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% as.data.frame() %>% summarise(Mb = sum(width_Mb))
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(. ~ contrast, value.var = "Mb") %>% 
  .[,-1] %>% 
  kable(caption = "Mb Regions of Extreme Differentiation per contrast") %>% 
  kable_styling(full_width = F)
```


Gene Annotation
================

Column {.tabset}
----------------

### Gene tables
```{r RED_genes_tables}
# Number of features per contrast
RED_genes %>% lapply(length) %>% 
  bind_rows() %>% 
  kable(caption = "All features") %>% 
  kable_styling(full_width = F)

# Number of protein_coding per contrast
lapply(RED_genes, function(x) x[x$mcols.gene_biotype == "protein_coding"]) %>% 
  lapply(length) %>% bind_rows() %>% 
  kable(caption = "Protein coding Genes") %>% 
  kable_styling(full_width = F)

# break down by biotype
tab <- lapply(RED_genes, function(x){
  as.data.frame(x) %>% 
  group_by(mcols.gene_biotype) %>% 
  summarise(n = n())
  }) %>% 
  bind_rows(.id = "contrast") %>% 
  reshape2::dcast(mcols.gene_biotype ~ contrast, value.var = "n") %>% 
  dplyr::rename(gene_biotype = mcols.gene_biotype) 

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_5.csv"))
tab %>% 
  kable(caption = "Features by biotype") %>% 
  kable_styling(full_width = F)
```

### Intersection Gene Sets (excl. FERT)
```{r upsetr_RED_genes_exclFERT, fig.width=10}
RED_genes_ids <- RED_genes[1:5] %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_genes_sets <- lapply(RED_genes, function(x){
  (RED_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_genes_sets) <- RED_genes_ids$mcols.gene_id

names(RED_genes_sets) <- names(RED_genes)

upset(RED_genes_sets, nintersects = NA, empty.intersections = T)
grid.text("Intersections RED-gene sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
```

### Intersection Gene Sets (incl FERT)
```{r upsetr_RED_genes_inclFERT, fig.width=10}
RED_genes_ids <- RED_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_genes_sets <- lapply(RED_genes, function(x){
  (RED_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_genes_sets) <- RED_genes_ids$mcols.gene_id

names(RED_genes_sets) <- names(RED_genes)

#png(here("00_dashboard/for_manuscript/supplementary_figure_2.png"), res = 300, height = 2000, width = 4000, units = "px")
upset(RED_genes_sets, nintersects = NA, nsets = 6, empty.intersections = T)
grid.text("Intersections RED-gene sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()

```

### Intersection Gene Sets (protein coding, excl FERT)
```{r upsetr_RED_genes_protein_coding_exclFERT, fig.width=10}
RED_pc_genes_ids <- RED_genes[1:5] %>% 
  lapply(function(x) as.data.frame(x) %>% filter(mcols.gene_biotype == "protein_coding")  ) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_pc_genes_sets <- lapply(RED_genes, function(x){
  (RED_pc_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_pc_genes_sets) <- RED_pc_genes_ids$mcols.gene_id

names(RED_pc_genes_sets) <- names(RED_genes)

upset(RED_pc_genes_sets, nintersects = NA, empty.intersections = T)
grid.text("Intersections RED-gene (protein coding) sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))

```

### Intersection Gene Sets (protein coding, incl FERT)
```{r upsetr_RED_genes_protein_coding_inclFERT, fig.width=10}
RED_pc_genes_ids <- RED_genes %>% 
  lapply(function(x) as.data.frame(x) %>% filter(mcols.gene_biotype == "protein_coding")  ) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_pc_genes_sets <- lapply(RED_genes, function(x){
  (RED_pc_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_pc_genes_sets) <- RED_pc_genes_ids$mcols.gene_id

names(RED_pc_genes_sets) <- names(RED_genes)

#png(here("00_dashboard/for_manuscript/supplementary_figure_3.png"), res = 300, height = 2000, width = 4000, units = "px")
upset(RED_pc_genes_sets, nintersects = NA, nsets = 6, empty.intersections = T)
grid.text("Intersections RED-gene (protein coding) sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()

```

ORA
========================

```{r run_WebGestaltR_gobp_once_and_export, eval = F}
#--------
# GOBP
#--------

run_WebGestaltR(
  interestGene=RED_genes$DUK_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DUC_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
) # no sig!

run_WebGestaltR(
  interestGene=RED_genes$DU6_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DU6P_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DUhLB_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$FERT_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

#-------
# KEGG
#-------
run_WebGestaltR(
  interestGene=RED_genes$DUK_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig!

run_WebGestaltR(
  interestGene=RED_genes$DUC_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig

run_WebGestaltR(
  interestGene=RED_genes$DU6_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig

run_WebGestaltR(
  interestGene=RED_genes$DU6P_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

run_WebGestaltR(
  interestGene=RED_genes$DUhLB_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

run_WebGestaltR(
  interestGene=RED_genes$FERT_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

```


Column {.tabset}
-----------------

### gobp_DUK_vs_FZTDU
```{r}
gobp_DUK_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)

```

### gobp_DU6_vs_FZTDU
```{r}
gobp_DU6_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_DU6P_vs_FZTDU
```{r}
gobp_DU6P_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_DUhLB_vs_FZTDU
```{r}
gobp_DUhLB_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_FERT_vs_FZTDU
```{r}
gobp_FERT_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_DU6P_vs_FZTDU
```{r}
kegg_DU6P_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_DUhLB_vs_FZTDU
```{r}
kegg_DUhLB_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_FERT_vs_FZTDU
```{r}
kegg_FERT_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```


```{r supp_table_6, eval=F}

write.csv(gobp_DUK_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_1.csv"))
write.csv(gobp_DU6_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_2.csv"))
write.csv(gobp_DU6P_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_3.csv"))
write.csv(gobp_DUhLB_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_4.csv"))
write.csv(kegg_DU6P_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_5.csv"))
write.csv(kegg_DUhLB_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_6.csv"))

write.csv(gobp_FERT_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_7.csv"))
write.csv(kegg_FERT_vs_FZTDU, here("00_dashboard/for_manuscript/supplementary_table_6_8.csv"))

```

### Top Results
```{r, fig.width=15, fig.height=10}
gobp_sig_top <- list(
  DUK_vs_FZTDU = gobp_DUK_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DU6_vs_FZTDU = gobp_DU6_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DU6P_vs_FZTDU = gobp_DU6P_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DUhLB_vs_FZTDU = gobp_DUhLB_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  FERT_vs_FZTDU = gobp_FERT_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio))
  ) %>% 
    bind_rows(.id = "contrast") %>% 
    #dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio) %>% 
    #remove NAs introduced when there are no 5 sig entries
    filter(!is.na(description)) %>% 
    mutate(db = "GOBP")

kegg_sig <- list(
  DU6P_vs_FZTDU = kegg_DU6P_vs_FZTDU,
  DUhLB_vs_FZTDU = kegg_DUhLB_vs_FZTDU,
  FERT_vs_FZTDU = kegg_FERT_vs_FZTDU
  ) %>% 
    bind_rows(.id = "contrast") %>% 
    #dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio)  %>% 
    mutate(db = "KEGG")

#png(here("00_dashboard/for_manuscript/figure_5.png"), res = 300, height = 3000, width = 4000, units = "px")

d <- bind_rows(gobp_sig_top, kegg_sig) %>% 
  mutate(
    contrast = factor(contrast, levels = c("DUK_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
  ) %>% 
  arrange(contrast, enrichmentRatio) %>% 
  mutate(
    description = factor(description, levels = .$description)#,
    #contrast = factor(contrast, levels = c("DUK_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    ) %>% 
  dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio)

ggplot(data = d, aes(x = enrichmentRatio, y = description, fill = contrast)) +
  geom_bar(stat = "identity") +
  geom_text(aes(x = 1.5, y = description, label = geneSet)) +
  ylab(NULL) +
  theme_bw(base_size = 15) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) +
  labs(
    title = "Top Enriched Terms and Pathways"
  ) +
  scale_y_discrete(limits = rev(levels(d$description)))

#dev.off()

```




Candidate Genes
================

```{r get_RED_genes_with_HIGH_impact_snps}

RED_genes_HIGH <- lapply(RED_genes, function(contr){
  contr %>% 
    as.data.frame() %>% 
    # intersect with high impact SNPs
    inner_join(
      snpeff_snps_HIGH,
      by = c("mcols.gene_id"="gene_ensembl")
    ) %>% 
    # prepare snp and gene loci columns
    mutate(
      gene_locus = paste0(seqnames,":",start,"-",end),
      snp_locus = paste0(CHROM,":",POS)
      ) %>% 
    # keep more informative columns
    dplyr::select(snp_locus, impact, effect, gene, mcols.gene_id, mcols.gene_biotype, gene_locus, ends_with("_Fst")) %>% 
    # rename columns for clarity
    dplyr::rename(
      snp_impact = impact, 
      snp_effect = effect, 
      gene_name = gene, 
      gene_ensembl = mcols.gene_id,
      gene_biotype = mcols.gene_biotype
      ) 
  })  %>% 
  bind_rows(.id = "contrast") %>% 
  # set levels contrasts
  mutate(
    contrast = factor(
      contrast, 
      levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU")
      )
    )
``` 

```{r matrices_heatmaps}
#---------------------------------------------------------
# matrix for heatmap all RED-genes with a high-impact SNP
#---------------------------------------------------------

# remove duplicate entries due to multiple snp-gene_isoform hits
ss <- RED_genes_HIGH %>% dplyr::select(gene_name, snp_locus, ends_with("_Fst")) %>% unique()

ma <- ss %>% dplyr::select(ends_with("_Fst")) %>% as.matrix()
rownames(ma) <- paste0(ss$gene_name, " | snp_locus:",ss$snp_locus)

# set NAs to smallest observed Fst (NA = all samples with same genotype for SNP, thus Fst is lowest)
# needed for clustering
ma[is.na(ma)] <- min(ma, na.rm=T)

ma <- ma[,c("DUK_vs_FZTDU_Fst","DUC_vs_FZTDU_Fst","FERT_vs_FZTDU_Fst","DU6_vs_FZTDU_Fst", "DU6P_vs_FZTDU_Fst","DUhLB_vs_FZTDU_Fst")]

colnames(ma) <- str_remove(colnames(ma), "_Fst")

#------------------------------------------------------------------------------------
# matrix excluding high-impact SNPs (in RED-genes) with all contrast with Fst < 0.3
#------------------------------------------------------------------------------------

ma2 <- ma[apply(ma,1,function(x) any(x > 0.3)),]


```


Column {.tabset}
-----------------

### Tables
```{r tabulate_HIGHsnps_and_RED_high_genes}

# tabulate high-snp-genes by effect type
snpeff_snps_HIGH %>% 
  group_by(effect) %>% 
  summarise(n_effects=n()) %>% 
  mutate(pct_effect = n_effects/sum(n_effects) * 100) %>% 
  arrange(desc(n_effects)) %>% 
  janitor::adorn_totals() %>% 
  kable(caption = "Number of HIGH-impact effects by effect-class", digits = 3) %>% 
  kable_styling(full_width = F)

# Total number of genes affected by HIGH impact SNPs
snpeff_snps_HIGH %>% 
  dplyr::select(gene_ensembl) %>% 
  unique() %>% 
  summarise(n_genes = n()) %>% 
  kable(caption = "Number of genes containing HIGH-impact SNPs") %>% 
  kable_styling(full_width = F)

# tabulate high-snp-genes intersecting with RED-genes
RED_genes_HIGH %>% 
  dplyr::select(contrast, gene_ensembl) %>% 
  unique() %>% 
  group_by(contrast) %>% 
  summarise(gene_number = n()) %>% 
  reshape2::dcast(.~contrast, value.var = "gene_number") %>% 
  .[,-1] %>% 
  kable(caption = "Number of RED-genes containing HIGH-impact SNPs by contrast") %>% 
  kable_styling(full_width = F)


```

### Gene intersection
```{r, fig.width=10}
RED_genes_HIGH_ids <- RED_genes_HIGH %>% 
  dplyr::select(contrast, gene_ensembl) %>% 
  dplyr::select(gene_ensembl) %>% 
  unique() %>% 
  unlist() %>% 
  as.character()

RED_genes_HIGH_sets <- lapply(
  unique(RED_genes_HIGH$contrast), function(contr){
    
    gns <- RED_genes_HIGH %>% 
      filter(contrast == contr) %>% 
      dplyr::select(gene_ensembl) %>% 
      unlist() %>% 
      as.character()

    (RED_genes_HIGH_ids %in% gns) %>% as.integer()
    
  }
) %>% 
  bind_cols() %>% 
  as.data.frame()

names(RED_genes_HIGH_sets) <- unique(RED_genes_HIGH$contrast)
rownames(RED_genes_HIGH_sets) <- RED_genes_HIGH_ids

#png(here("00_dashboard/for_manuscript/supplementary_figure_4.png"), res = 300, height = 2000, width = 3000, units = "px")
upset(RED_genes_HIGH_sets, nsets = 6)
grid.text("Intersections RED-genes with HIGH-impact-SNPs",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()
```

### Fst Heatmap
```{r fst_heatmap, fig.width=10, fig.height=20}

#pheatmap(ma,na_col = "darkgrey",cluster_cols = F, cluster_rows = F, cellwidth = 20, cellheight = 10, main = "Fst scores per (HIGH-impact)SNP in RED-genes")

pheatmap(ma,na_col = "darkgrey",cluster_cols = F, cluster_rows = T, main = "Fst scores per (HIGH-impact)SNP in RED-genes")

```

### Fst Heatmap (any Fst > 0.3)
```{r fst_heatmap_low_scores_removed, fig.width=10, fig.height=13}

#png(here("00_dashboard/for_manuscript/supplementary_figure_5.png"), res = 300, height = 3500, width = 2500, units = "px")

pheatmap(ma2,na_col = "darkgrey",cluster_cols = T, cluster_rows = T, display_numbers = T, angle_col = 45,main = "Fst scores (>0.3 any) per (HIGH-impact)SNP in RED-genes")

#dev.off()


```

```{r for_manuscript_nr_snps_genes_after_filtering_by_fst, eval = F}

# Get nr of high-impact SNPs after removing low Fst contrasts (all < 0.3)
rownames(ma2) %>% str_split("[|]") %>% sapply(function(x) x[2]) %>% str_trim() %>% unique() %>% length()

# Get nr of RED genes after removing low Fst contrasts (all < 0.3)
rownames(ma2) %>% str_split("[|]") %>% sapply(function(x) x[1]) %>% str_trim() %>% unique() %>% length()

# Get number of SNPs with min 0.7 in at least one contrast
rownames(ma2)[apply(ma2,1, function(x) any(x >= 0.7))] %>% str_split("[|]") %>% sapply(function(x) x[2]) %>% str_trim() %>% unique() %>% length()

rownames(ma2)[apply(ma2,1, function(x) any(x >= 0.7))] %>% str_split("[|]") %>% sapply(function(x) x[1]) %>% str_trim() %>% unique() %>% length()
```


### Genes in sig. GOBP-KEGG
```{r genes_in_sig_GOBP_KEGG}
names(RED_genes_HIGH) <- str_remove(names(RED_genes_HIGH), "_Fst")

tab <- RED_genes_HIGH %>% 
  # filter RED-genes with high-impact SNP
  filter(
    # filter by loci: high-impact SNPs "any contrast Fst > 0.2"
    snp_locus %in% (rownames(ma2) %>% lapply(function(x) str_split(x,"snp_locus:") %>% unlist() %>% .[2]))
  ) %>% 
  # select relevant columns
  dplyr::select(contrast, snp_locus, gene_ensembl, gene_name, snp_effect, snp_impact, contains("_vs_")) %>% 
  # intersect with genes in significant pathway terms
  inner_join(
    sig_gobp_kegg_one_gene_per_row,
    by = c("contrast","gene_ensembl")
  ) %>% 
  dplyr::select(contrast, snp_locus, gene_ensembl, gene_name, snp_effect, snp_impact, db, description, contains("_vs_"))

write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_8.csv"))

tab %>% 
  kable(caption = "Number of RED-genes containing HIGH-impact SNPs in Significant GOBP or KEGG") %>% 
  kable_styling(full_width = F) 

```



Diversity ratio
==================

Column {.tabset}
----------------

```{r pi_ratio_tables,include=F}
fls <- list.files(here("batches123_07_selection_analysis/02_pi_ratio/output"), pattern = ".csv$", full.names = T)

pi_ratio_dat <- lapply(fls, function(fl){
  read.csv(fl) %>% dplyr::select(CHROM, BIN_START, BIN_END, pi_ratio)
})

names(pi_ratio_dat) <- basename(fls) %>% str_remove("pi_ratio_") %>% str_remove(".csv")

pi_ratio_dat <- pi_ratio_dat %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(pi_ratio_log2 = log2(pi_ratio))
```

### Summary
```{r summary_win_pi}
summary_win_pi_ratio <- pi_ratio_dat %>% 
  mutate(is_x = (CHROM == "X")) %>% 
  group_by(contrast, is_x) %>% 
  summarise(
    n_win = n(),
    min = min(pi_ratio),
    q1 = quantile(pi_ratio, 0.01),
    q5 = quantile(pi_ratio, 0.05),
    median = median(pi_ratio),
    mean = mean(pi_ratio),
    max = max(pi_ratio)
    ) %>% 
  arrange(is_x)

summary_win_pi_ratio %>%
  mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window pi_ratio") %>% 
  kable_styling(full_width = F)
```

### Distribution
```{r, fig.width=10,fig.height=7}

p1 <- pi_ratio_dat %>% 
  ggplot(aes(x = pi_ratio)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("pi-ratio")

p2 <- pi_ratio_dat %>% 
  ggplot(aes(x = NA, y = pi_ratio)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("pi-ratio")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```



