---
title: "WGS DU-mice"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---


```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(tidyr)
library(here)
library(plotly)
library(DT)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
library(knitr)
library(ggcorrplot)
library(tibble)
library(png)
library(grid)
library(gridExtra)
library(fastqcr)
library(GenomicRanges)
library(kableExtra)
library(ape)
library(WebGestaltR)
library(pheatmap)
library(limma)
library(ggrepel)
library(bumphunter) #BiocManager::install("bumphunter")
library("TxDb.Mmusculus.UCSC.mm10.ensGene") #BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(biomaRt)
library(Gviz)
```

```{r sample_info_and_genome_length, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

#http://www.ensembl.org/Mus_musculus/Info/Annotation 
genome_length <- 2730871774 # genome length (Golden Path Length	) 
```

```{r function_import_fastqc_reports, include=F}
get_fastqc_data <- function(qc.dir){
  qc <- qc_aggregate(qc.dir)
  qc.stats <- qc_stats(qc)
  
  fields <- str_split(qc.stats$sample, "_")
  samp <- sapply(fields, function(x) x[1])
  R_pair <- sapply(fields, function(x) x[4])
  
  qc.stats$sample_name <- samp
  qc.stats$R_pair <- R_pair
  
  qc.stats$pct.dup <- as.numeric(qc.stats$pct.dup)
  qc.stats$pct.gc <- as.numeric(qc.stats$pct.gc)
  qc.stats$tot.seq <- as.numeric(qc.stats$tot.seq)
  qc.stats$seq.length <- as.numeric(qc.stats$seq.length)

  res <- list(qc=qc, qc.stats=qc.stats)
  return(res)
}

```

```{r batch1_prealignment_raw, include=F}
# import
fastqc_raw_60samp_9Lanes <- get_fastqc_data(file.path(here(),"batch1/01_quality_control/res_concat"))

# Add a column with only sample name to qc.stats
fastqc_raw_60samp_9Lanes$qc.stats <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
  mutate(sample_name = str_sub(sample, 1, 6))

# Add column for read pair
fastqc_raw_60samp_9Lanes$qc.stats <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
  mutate(R_pair = str_sub(sample, 8, 9))

# Calculate coverage and GB for each sample
NR_RL_GB_CVG_60samp_9Lanes_raw <- lapply(unique(fastqc_raw_60samp_9Lanes$qc.stats$sample_name),
#x="H07738"
                               function(x){
                                 # subset for sample
                                 dat <- fastqc_raw_60samp_9Lanes$qc.stats %>% 
                                   filter(sample_name == x)
                                 # prepare results read number and read length
                                 res <- data.frame(sample = x,
                                                   NR_R1 = dat$tot.seq[dat$R_pair == "R1"],
                                                   NR_R2 = dat$tot.seq[dat$R_pair == "R2"],
                                                   RL_R1 = dat$seq.length[dat$R_pair == "R1"],
                                                   RL_R2 = dat$seq.length[dat$R_pair == "R2"])
                                 # add GB, CVG and genome length
                                 res <- res %>% 
                                   mutate(GB = ((NR_R1*RL_R1) + (NR_R2*RL_R2))*10^-9,
                                          CVG=GB/( genome_length*10^-9 ),
                                          genome_length = genome_length)
                                 return(res)
                               }) %>% do.call(bind_rows,.)

# add sample info
NR_RL_GB_CVG_60samp_9Lanes_raw <- inner_join(NR_RL_GB_CVG_60samp_9Lanes_raw,
                                  sample_info,
                                  by = c("sample"="sample_id"))

```

```{r batch1_prealignment_clean, include=F}
setwd(here("batch1/03_quality_control_analysis/output_n_bp"))

NR_GB_CVG_60samp_9Lanes_trimm <- lapply(NR_RL_GB_CVG_60samp_9Lanes_raw$sample, 
#sample="H07738"       
       function(sample){
        ## Define files
        fls <- list.files()[grep(sample, list.files())]
        fls_NR <- fls[grep("NR", fls)]
        fls_GB <- fls[grep("GB", fls)]
        
        ## Load and aggregate files
        NR_R1 <- lapply(fls_NR[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        NR_R2 <- lapply(fls_NR[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        GB_R1 <- lapply(fls_GB[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)
        GB_R2 <- lapply(fls_GB[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)

        # summarize results for sample
        res <- data.frame(sample = sample, 
                          NR_R1=NR_R1$total_reads, 
                          NR_R2=NR_R2$total_reads,
                          GB=GB_R1+GB_R2)
        res <- res %>% mutate(CVG=GB/(genome_length*10^-9))

      }
      ) %>% do.call(bind_rows, .)

NR_GB_CVG_60samp_9Lanes_trimm <- inner_join(NR_GB_CVG_60samp_9Lanes_trimm,
                                  sample_info,
                                  by = c("sample"="sample_id"))
```

```{r batch1_CollectWgsMetrics, include=F}
path <- file.path(here(),"batch1/06_quality_control_alignments/output")

fls <- list.files(path)

dat_list <- lapply(fls, function(x){ 
  read.delim(file.path(path, x), stringsAsFactors = FALSE, dec = ",", skip = 6, nrow = 1)
  })

names(dat_list) <- substr(x = fls, start = 1, stop = 9)

CVG_post_align_60samp_9Lanes <- dat_list[grep("NoFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))


CVG_post_align_qual_60samp_9Lanes <- dat_list[grep("WithFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))
```

```{r batch2_prealignment_raw, include=F}
fastqc_raw_10samp_10Lanes <- get_fastqc_data(file.path(here(),"batch2/01_quality_control/output_concat"))

fastqc_raw_10samp_10Lanes$qc.stats <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
  mutate(sample_name = str_sub(sample, 1, 6))

fastqc_raw_10samp_10Lanes$qc.stats <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
  mutate(R_pair = str_sub(sample, 19, 20))

NR_RL_GB_CVG_10samp_10Lanes_raw <- lapply(unique(fastqc_raw_10samp_10Lanes$qc.stats$sample_name),
#x="H07738"
                               function(x){ 
                                 # subset for sample
                                 dat <- fastqc_raw_10samp_10Lanes$qc.stats %>% 
                                   filter(sample_name == x)
                                 # prepare results read number and read length
                                 res <- data.frame(sample = x,
                                                   NR_R1 = dat$tot.seq[dat$R_pair == "R1"],
                                                   NR_R2 = dat$tot.seq[dat$R_pair == "R2"],
                                                   RL_R1 = dat$seq.length[dat$R_pair == "R1"],
                                                   RL_R2 = dat$seq.length[dat$R_pair == "R2"])
                                 # add GB, CVG and genome length
                                 res <- res %>% 
                                   mutate(GB = ((NR_R1*RL_R1) + (NR_R2*RL_R2))*10^-9,
                                          CVG=GB/( genome_length*10^-9 ),
                                          genome_length = genome_length)
                                 return(res)
                               }) %>% do.call(bind_rows,.)

NR_RL_GB_CVG_10samp_10Lanes_raw <- inner_join(NR_RL_GB_CVG_10samp_10Lanes_raw,
                                  sample_info,
                                  by = c("sample"="sample_id"))
```

```{r batch2_prealignment_clean, include=F}
setwd(file.path(here(),"batch2/03_quality_control_analysis/GB_NR"))

NR_GB_CVG_10samp_1lane_trimm <- lapply(NR_RL_GB_CVG_10samp_10Lanes_raw$sample, 
#sample="H07739"       
       function(sample){  
        ## Define files reseq
        fls <- list.files()[grep(sample, list.files())]
        fls_NR <- fls[grep("NR", fls)]
        fls_GB <- fls[grep("GB", fls)]
        
        ## Load and aggregate files
        NR_R1 <- lapply(fls_NR[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        NR_R2 <- lapply(fls_NR[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(total_reads = sum(V1))
        GB_R1 <- lapply(fls_GB[grep("R1",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)
        GB_R2 <- lapply(fls_GB[grep("R2",fls_NR)], function(x) read.delim(x, header = FALSE)) %>% 
          do.call(bind_rows,.) %>% 
          summarise(GB = sum(V1)*10^-9)

        # summarize results for sample
        res <- data.frame(sample = sample, 
                          NR_R1=NR_R1$total_reads, 
                          NR_R2=NR_R2$total_reads,
                          GB=GB_R1+GB_R2)
        res <- res %>% mutate(CVG=GB/(genome_length*10^-9))

      }
      ) %>% do.call(bind_rows, .)

# 1+9=10 lanes 
NR_GB_CVG_10samp_10Lanes_trimm <- inner_join(NR_GB_CVG_10samp_1lane_trimm,
                                             NR_GB_CVG_60samp_9Lanes_trimm,
                                             by = "sample",
                                             suffix = c("_1lane", "_9Lanes")) %>% 
                                  dplyr::select(sample, 
                                         NR_R1_9Lanes, NR_R2_9Lanes, NR_R1_1lane, NR_R2_1lane,
                                         GB_9Lanes, GB_1lane,
                                         CVG_9Lanes, CVG_1lane) %>% 
                                  mutate(CVG_10Lanes = CVG_9Lanes + CVG_1lane) %>% 
                                  inner_join(sample_info,by = c("sample"="sample_id"))

```

```{r batch2_CollectWgsMetrics, include=F}
path <- file.path(here(),"batch2/06_quality_control_alignments/output")

fls <- list.files(path)

dat_list <- lapply(fls, function(x){ 
  read.delim(file.path(path, x), stringsAsFactors = FALSE, dec = ",", skip = 6, nrow = 1)
  })

names(dat_list) <- substr(x = fls, start = 1, stop = 9)

CVG_post_align_10samp_10Lanes <- dat_list[grep("NoFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))


CVG_post_align_qual_10samp_10Lanes <- dat_list[grep("WithFilters.txt",fls)] %>% 
  bind_rows(.id = "sample_name") %>% 
  mutate(sample_name = substr(sample_name, 1, 6)) %>% 
  inner_join(sample_info, by = c("sample_name"="sample_id"))  
  
```

```{r batch3_prealignment_data, include=F}
# objects previously created, just importing
# follow link to see how objects were created

qc_raw_stats_avg_cvg <- readRDS(here("batch3/03_quality_control_analysis/r_objects/qc_raw_stats_avg_cvg.rds"))
qc_fastp_stats_avg_cvg <- readRDS(here("batch3/03_quality_control_analysis/r_objects/qc_fastp_stats_avg_cvg.rds"))
```

```{r batch3_CollectWgsMetrics, include=F}
fls <- list.files(here("batch3/06_quality_control_alignments/output"), pattern = "/*_pt1_points.txt")

cvg_summary <- lapply(fls, function(fl){
  
#fl=fls[1]
  # Prepare sample name (taking into account droput-reseqd sample)
  cond <- fl == "I34772-L1_S19_L004.sorted.RG.dedup.bqsr.CollectWgsMetrics.default_pt1_points.txt" | fl == "I34772-L1_S19_L004.sorted.RG.dedup.bqsr.CollectWgsMetrics.Q0.M0_pt1_points.txt"
  if(cond){
    s <- "I34772_dropout_reseqd"
  }else{
    s <- str_split(fl,"-") %>% sapply(function(x) x[1])
  }

  # Load sample histogram information
  d <- read.table(file.path(here("batch3/06_quality_control_alignments/output"),fl), header = T, stringsAsFactors = F)
  
  # Extract mode of metrics (with or without bp and mapping quality filters)
  m <- str_split(fl,"_") %>% sapply(function(x) x[3]) %>% 
    str_remove("sorted.RG.dedup.bqsr.CollectWgsMetrics.") %>% 
    str_remove("L00[1-9].")
  
  # Add sample and mode information (default = mapping and bp quality of min 20)
  d <- mutate(d, sample = s, mode = m, mode = str_replace(mode, "default","Q20.M20"))
  
  # add mouse line information
  d <- sample_info %>% 
    right_join(d, by = c("sample_id"="sample")) %>% 
    dplyr::rename(sample_name = sample_id)
  
}) %>% 
  # bind rows
  bind_rows()

# Add mouse line group to drop out sample and prepare columns for integration
cvg_summary <- cvg_summary %>% 
  mutate(Linie = ifelse(sample_name == "I34772_dropout_reseqd", "DUK", Linie))  %>% 
  # fix droput/resequenced sample
  mutate(target_cvg = ifelse(is.na(target_cvg), "5x", target_cvg))
  
```

```{r batch1_batch2_flagstat, include=F}
# batch1

## get files to parse
fls <- list.files(
  file.path(here(),"batch1/06_quality_control_alignments/output"),pattern = "flagstat", full.names = T
  )

## Get nr of properly mapped reads (line nr 5 in samtools flagstat output)
pct_properly_paired_reads_60samp_9Lanes <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

pct_properly_paired_reads_60samp_9Lanes <- inner_join(
  pct_properly_paired_reads_60samp_9Lanes,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

rm(fls)

# batch2 (boosting of low covered sampels in batch1)

## get files to parse
fls <- list.files(
  file.path(here(),"batch2/06_quality_control_alignments/output"),pattern = "flagstat", full.names = T
  )

## Get nr of properly mapped reads (line nr 5 in samtools flagstat output)
pct_properly_paired_reads_10samp_10Lanes <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

pct_properly_paired_reads_10samp_10Lanes <- inner_join(
  pct_properly_paired_reads_10samp_10Lanes,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

rm(fls)

# combine data
NR_alignment_dat_target_cvg_30x <- pct_properly_paired_reads_60samp_9Lanes %>% 
  # remove low covered samples from batch1
  filter(
    !(sample_name %in% pct_properly_paired_reads_10samp_10Lanes$sample_name)
    ) %>% 
  # replace same samples boosted (aka batch2)
  bind_rows(
    pct_properly_paired_reads_10samp_10Lanes
  )

```

```{r batch3_flagstat, include=F}

fls <- list.files(
  file.path(here(),"batch3/06_quality_control_alignments/output"),pattern = "flagstat.tab$", full.names = T
  )


NR_alignment_dat_target_cvg_5x <- lapply(fls, function(fl){
                  # load flagstat output
                  dat <- read.delim(fl, header = FALSE)
                  
                  # paired in sequencing
                  paired_in_seq <- dat[6,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # properly paired
                  properly_paired <- dat[9,] %>% str_split("[+]") %>% unlist() %>% .[1] %>% as.numeric()
                  
                  # pct properly paired
                  pct_properly_paired <- round((properly_paired/paired_in_seq)*100, 2)
                  
                  # prepare sample name
                  samp <- fl %>% basename() %>% substr(., 1, 6)
                  
                  # return data
                  data.frame(paired_in_seq = paired_in_seq, 
                             properly_paired = properly_paired,
                             pct_properly_paired = pct_properly_paired,
                             sample_name = samp)
                  }) %>% do.call(bind_rows, .)

NR_alignment_dat_target_cvg_5x <- inner_join(
  NR_alignment_dat_target_cvg_5x,
  sample_info, 
  by = c("sample_name"="sample_id")
  )

#remove droput, keep only resequenced sample (row 64)
NR_alignment_dat_target_cvg_5x <- NR_alignment_dat_target_cvg_5x[-64,]
```

```{r collect_cvg_data, include=F}

boosted_samples <- NR_RL_GB_CVG_10samp_10Lanes_raw$sample

cvg_dat_prealignment <- bind_rows(
  
  # raw
  bind_rows(
    NR_RL_GB_CVG_60samp_9Lanes_raw %>% 
      filter(!(sample %in% boosted_samples)) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    NR_RL_GB_CVG_10samp_10Lanes_raw %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    qc_raw_stats_avg_cvg %>% 
      mutate(target_cvg = "5x") %>% 
      dplyr::rename(sample = sample_name, CVG = avg_cvg) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG)
  ) %>% 
    mutate(data_set = "cvg_raw"),
  
  #clean
  bind_rows(
    NR_GB_CVG_60samp_9Lanes_trimm %>% 
      filter(!(sample %in% boosted_samples)) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG),
    NR_GB_CVG_10samp_10Lanes_trimm %>% 
      dplyr::select(target_cvg, Linie, sample, CVG_10Lanes) %>% 
      dplyr::rename(CVG = CVG_10Lanes),
    qc_fastp_stats_avg_cvg %>% 
      mutate(target_cvg = "5x") %>% 
      dplyr::rename(sample = sample_name, CVG = avg_cvg) %>% 
      dplyr::select(target_cvg, Linie, sample, CVG)
  ) %>% 
    mutate(data_set = "cvg_clean")
)


cvg_dat_postalignment <- bind_rows(
  
  # raw
  bind_rows(
    CVG_post_align_60samp_9Lanes %>% filter(!(sample_name %in% boosted_samples)),
    CVG_post_align_10samp_10Lanes,
    cvg_summary %>% 
      filter(mode == "Q0.M0") %>% 
      dplyr::select(names(CVG_post_align_10samp_10Lanes)) 
  ) %>% mutate(data_set = "m0q0"),
  
  # qual
  bind_rows(
    CVG_post_align_qual_60samp_9Lanes %>% filter(!(sample_name %in% boosted_samples)),
    CVG_post_align_qual_10samp_10Lanes,
    cvg_summary %>% 
      filter(mode == "Q20.M20") %>% 
      dplyr::select(names(CVG_post_align_qual_10samp_10Lanes))
  ) %>% mutate(data_set = "m20q20")
)

# combine cvg data before and after alignment
cvg_dat <- bind_rows(
  
  cvg_dat_prealignment %>% 
    dplyr::select(target_cvg, data_set, Linie, sample, CVG) %>% 
    # remove extra sample in 5x data set (the one discarded, aka dropout)
    filter(!(sample == "I34772_dropout")) %>% 
    # re-label sample that sample
    mutate(sample = ifelse(sample == "I34772_dropout_reseq", "I34772", sample)) %>% 
    # rename sample col
    dplyr::rename(sample_name = sample),
  
  cvg_dat_postalignment %>% 
    dplyr::select(target_cvg, data_set, Linie, sample_name, MEAN_COVERAGE) %>% 
    dplyr::rename(CVG = MEAN_COVERAGE) %>% 
    # remove extra sample in 5x data set (the one discarded, aka dropout)
    filter(!(sample_name == "I34772")) 
  ) %>% 
  # rename data set information
  mutate(
    data_set = ifelse(data_set == "cvg_raw", "raw_reads",data_set),
    data_set = ifelse(data_set == "cvg_clean", "clean_reads",data_set),
    data_set = ifelse(data_set == "m0q0", "aligned_m0q0",data_set),
    data_set = ifelse(data_set == "m20q20", "aligned_m20q20",data_set)
  ) %>% 
  # order levels for visualization
  mutate(
    Linie = ifelse(Linie == "DUHLB", "DUhLB", Linie),
    Linie = factor(Linie, levels = c("FZTDU","DUK","DUC","DU6","DU6P","DUhLB")),
    data_set = factor(data_set, levels = c("raw_reads","clean_reads","aligned_m0q0","aligned_m20q20"))
  )
```

```{r CollectInsertSizeMetrics}
# batch1
fls <- list.files(
  file.path(here(),"batch1/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.tab",
  full.names = T
  ) 

insert_size_batch1 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_remove("-L1.merged.sorted.dedup.CollectInsertSizeMetrics.tab")
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())


# batch2
fls <- list.files(
  file.path(here(),"batch2/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.tab",
  full.names = T
  ) 

insert_size_batch2 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_remove(".merged.sorted.dedup.CollectInsertSizeMetrics.tab")
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())

# batch1-2 (replace boosted samples)
insert_size_batch1_batch2 <- insert_size_batch1 %>% 
  filter(!(sample_name %in% insert_size_batch2$sample_name)) %>% 
  bind_rows(insert_size_batch2)

# batch3
fls <- list.files(
  file.path(here(),"batch3/06_quality_control_alignments/output"),
  pattern = ".CollectInsertSizeMetrics.txt",
  full.names = T
  ) %>% 
  # remove dropout file
  .[-grep("I34772-L1_S63_L003",.)]

insert_size_batch3 <- lapply(fls, function(fl){
  d <- read.delim(fl, skip = 6, nrows = 1, header = T, dec = ",")
  s <- fl %>% basename() %>% str_sub(1,6)
  d %>% 
    mutate(sample_name = s) %>% 
    dplyr::select(sample_name, everything())
}) %>% 
  bind_rows() %>% 
  inner_join(
    sample_info,
    by = c("sample_name"="sample_id")
  ) %>% 
  dplyr::select(target_cvg, Linie, sample_name, everything())

# collect data
insert_size_dat <- bind_rows(insert_size_batch1_batch2, insert_size_batch3)
```

```{r vcf_metrics_cohort_raw, include=F}
raw_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "raw")
```

```{r vcf_metrics_cohort_after_vqsr95, include=F}
mid_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "site_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```

```{r vcf_metrics_cohort_final, include=F} 

final_vcf_metrics_detail <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)

final_vcf_metrics_summary <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(data_set = "gt_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)

```

```{r vcf_final_SNP_ids}

x <- vroom(
  here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids"), 
  col_names = FALSE, 
  delim = " ",
  col_types = c(X1 = "c")
  ) %>% 
  dplyr::select(-X2) %>% 
  dplyr::rename(seqname = X1, start = X3) %>% 
  mutate(end = start) 

# turn it into granges
vcf_final_snp_ids_gr <- GRanges(seqnames = x$seqname, IRanges(start = x$start, end = x$end))

rm(x)
```

```{r export_data_for_upsetr_snp_sets, eval = F} 
fls <- here("batches123_04_final_VCF/output/") %>% 
  list.files(pattern = "*.SNPids") %>% 
  .[-grep("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids",.)]

snp_ids <- lapply(fls, function(fl){ 
  ln <- str_remove(fl, "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
    str_remove(".SNPids")
  d <- vroom(here("batches123_04_final_VCF/output",fl), col_names = F) %>% mutate(line = ln, snp_id = paste0(X1,"_",X3))
  return(d)
})

names(snp_ids) <- str_remove(fls, "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
    str_remove(".SNPids")

#------------------------------------------------------------------------------------------------------
# export final list of private SNPs for fztdu, duc and duk so we can proceed with the IVF genotypings:
#------------------------------------------------------------------------------------------------------ 

# get private SNPs
priv_duk <- setdiff(snp_ids$DUK$snp_id,
                    unique(c(snp_ids$DUC$snp_id,snp_ids$DU6$snp_id,snp_ids$DU6P$snp_id,snp_ids$DUhLB$snp_id,snp_ids$FZTDU$snp_id)))

priv_duc <- setdiff(snp_ids$DUC$snp_id,
                    unique(c(snp_ids$DUK$snp_id,snp_ids$DU6$snp_id,snp_ids$DU6P$snp_id,snp_ids$DUhLB$snp_id,snp_ids$FZTDU$snp_id)))

priv_du6 <- setdiff(snp_ids$DU6$snp_id,
                    unique(c(snp_ids$DUK$snp_id,snp_ids$DUC$snp_id,snp_ids$DU6P$snp_id,snp_ids$DUhLB$snp_id,snp_ids$FZTDU$snp_id)))

priv_du6p <- setdiff(snp_ids$DU6P$snp_id,
                    unique(c(snp_ids$DUK$snp_id,snp_ids$DUC$snp_id,snp_ids$DU6$snp_id,snp_ids$DUhLB$snp_id,snp_ids$FZTDU$snp_id)))

priv_duhlb <- setdiff(snp_ids$DUhLB$snp_id,
                    unique(c(snp_ids$DUK$snp_id,snp_ids$DUC$snp_id,snp_ids$DU6$snp_id,snp_ids$DU6P$snp_id,snp_ids$FZTDU$snp_id)))

priv_fztdu <- setdiff(snp_ids$FZTDU$snp_id,
                    unique(c(snp_ids$DUK$snp_id,snp_ids$DUC$snp_id,snp_ids$DU6$snp_id,snp_ids$DU6P$snp_id,snp_ids$DUhLB$snp_id)))

priv_duk %>% length() #52442 ... corroborate with upset plot later ...
priv_duc %>% length() #60817
priv_du6 %>% length() #66231
priv_du6p %>% length() #70969
priv_duhlb %>% length() #116610
priv_fztdu %>% length() #437432


# add allele frequencies and export

dat_alt_frq <- lapply(list.files(here("batches123_08_LDD_SFS/output"), pattern = "*.ALTfrq2"), function(fl){
  
  pop <- fl %>% 
    str_remove(
      "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords."
    ) %>% 
    str_remove(".ALTfrq2")
  
  vroom(here("batches123_08_LDD_SFS/output",fl)) %>% 
    mutate(Linie = pop) %>% 
    separate(col = POS,into = c("tmp","alt_frq"),sep = " ") %>% 
    dplyr::rename(POS = tmp)
})  %>% 
  bind_rows()

dat_alt_frq %>% 
  filter(Linie == "DUK") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_duk) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/duk_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

dat_alt_frq %>% 
  filter(Linie == "DUC") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_duc) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/duc_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

dat_alt_frq %>% 
  filter(Linie == "DU6") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_du6) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/du6_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

dat_alt_frq %>% 
  filter(Linie == "DU6P") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_du6p) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/du6p_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

dat_alt_frq %>% 
  filter(Linie == "DUhLB") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_duhlb) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/duhlb_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

dat_alt_frq %>% 
  filter(Linie == "FZTDU") %>% 
  mutate(snp_id = paste0(CHROM,"_",POS)) %>% 
  filter(snp_id %in% priv_fztdu) %>% 
  dplyr::select(CHROM,POS,alt_frq) %>% 
  write.table(here("00_dashboard/data/fztdu_private.tab"),quote = FALSE,row.names = FALSE, col.names = TRUE)

## ... number of records per file was corroborated to number of private snps

#------------------------------------------------------------------------------------------------------ 
# Prepare set data for upset and export
#------------------------------------------------------------------------------------------------------
all_ids <- snp_ids %>% bind_rows() %>% select(snp_id) %>% unique() %>% unlist()

sets <- data.frame(ids = all_ids,
		   duk = (all_ids %chin% snp_ids$DUK$snp_id),
		   duc = (all_ids %chin% snp_ids$DUC$snp_id),
		   du6 = (all_ids %chin% snp_ids$DU6$snp_id),
		   du6p = (all_ids %chin% snp_ids$DU6P$snp_id),
		   duhlb = (all_ids %chin% snp_ids$DUhLB$snp_id),
		   fztdu = (all_ids %chin% snp_ids$FZTDU$snp_id))

write.csv(sets, here("00_dashboard/data","snp_id_sets.csv"))
rm(sets)
```

```{r import_data_for_upsetr_snp_sets, include=F}
set_dat <- vroom(here("00_dashboard/data","snp_id_sets.csv")) %>% as.data.frame()

set_dat_bool <- data.frame(
  duk = as.integer(set_dat$duk),
	duc = as.integer(set_dat$duc),
	du6 = as.integer(set_dat$du6),
	du6p = as.integer(set_dat$du6p),
	duhlb = as.integer(set_dat$duhlb),
	fztdu = as.integer(set_dat$fztdu)
	)

rownames(set_dat_bool) <- set_dat$ids
```

```{r snpeff_tables, include=F} 
fls <- list.files(
  here("batches123_05_annotation/output"),
  pattern = ".csv"
) %>% 
  .[grep("biallelicSNPs_VQSR95",.)]

snpeff_data <- lapply(
  fls, function(fl) read.csv(here("batches123_05_annotation/output",fl))
  )

names(snpeff_data) <- fls %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.summary_") %>% 
  str_remove(".csv")

rm(fls)
```

```{r impact_snpeff}
impact_snpeff <- read.csv2(here("00_dashboard/data/impact_snpeff.csv")) %>% 
  dplyr::select(Effect, Impact)
```

```{r pca_data, include=F}

pc_pct <- read.csv(here("batches123_06_genetic_structure/data/pc_pct.csv")) %>% dplyr::select(-X)

pc_eigenscores <- read.csv(here("batches123_06_genetic_structure/data/pc_eigenscores.csv")) %>% 
  dplyr::select(-X) %>% 
  mutate(sample.id = str_remove(sample.id, "-L1")) %>% 
  left_join(sample_info, by = c("sample.id"="sample_id")) %>% 
  dplyr::select(Linie, sample.id, target_cvg, everything())
  

```

```{r make_ancestry_barplot_function}
make_ancestry_barplot <- function(k,q_fl){

#k=5 
#q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.ldpruned.5.Q")

  k_dat <- read.table(q_fl) %>% 
    bind_cols(
      dplyr::select(fam_fl, V2) %>% dplyr::rename(sample_id = V2)
      ) %>% 
    mutate(sample_id = str_remove(sample_id, "-L1")) %>% 
    left_join(sample_info, by = "sample_id") %>% 
    reshape2::melt(id.vars = c("sample_id","Linie","target_cvg")) %>% 
    mutate(Linie = factor(Linie, levels = c("FZTDU","DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
    arrange(Linie, sample_id)
  
  k_dat <- k_dat %>% 
    mutate(sample_id = factor(sample_id, levels = unique(k_dat$sample_id)))

  p <- k_dat %>% 
        ggplot(aes(x = sample_id, y = value,fill = variable)) +
          geom_bar(stat = "identity", position = "stack") +
          theme_bw(base_size = 20) +
          theme(
            axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            legend.position = "none",
            plot.title = element_text(hjust = 0.5)
            ) +
          xlab(NULL) +
          ylab("Ancestry Fraction") +
          annotate("text",x = seq(12.5,150,25), y = 1.02, label = levels(k_dat$Linie), size = 8) +
          ggtitle(paste0("K=",k)) #+
          #ggsave("tst.png", height = 7, width = 15)
  return(p)
}
```

```{r win_fst_data_sel_vs_ctrl_zscore, include=F} 

win_fst_sel_vs_ctrl <- lapply(
  
  list.files(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
             pattern = "windowed.weir.fst$"), 
  
  function(fl){
    
    read.table(
      here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl),
      header = T,
      stringsAsFactors = F
      )
    }
  )

names(win_fst_sel_vs_ctrl) <- list.files(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
  pattern = "windowed.weir.fst$"
  ) %>% str_remove(".windowed.weir.fst")

# prepare data (i.e. zscore transformation) for visualizations
win_fst_sel_vs_ctrl_prep <- lapply(win_fst_sel_vs_ctrl, function(d){
  d %>% 
    # filter by min number of SNPs in window (min 10 SNPs)
    dplyr::filter(N_VARIANTS >= 10) %>% 
    # separate x chromosome
    mutate(is_x = CHROM == "X") %>% 
    # group by autosomes/chrX
    group_by(is_x) %>% 
    # compute zscores
    mutate(z_win_fst_score = scale(MEAN_FST))   
}) %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    )

# label extreme scores
win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  mutate(
    is_1th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.01) ),
    is_5th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.05) ),
    is_10th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.10) ),
    is_25th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.25) ),
    is_90th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.90) ),
    is_95th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.95) ),
    is_99th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.99) )
      )

#win_fst_sel_vs_ctrl_prep %>% summarise(sum(is_99th_quantile)/n()) # it checks out :)



```

```{r mean_genomewide_fst_pairwise, include=F}

global_fst <- read.table(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/global_fst.tab")
  ) %>% 
  dplyr::rename(global_fst = V2) %>% 
  separate(col = V1, sep = "_vs_", into = c("pop1","pop2")) %>% 
  mutate(pop1 = factor(pop1, levels = c("DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
  arrange(pop1)
```

```{r function_get_mean_fst_gene_ranges, include=F} 
get_mean_fst_gene_ranges <- function(query_genes, contrast){
  
    #query_genes=ss_genes_gr
    #contrast="DUK_vs_FZTDU"
    
    # create name of file containing per-snp fst scores
    fl_nm <- paste0(contrast,".weir.fst")
    # load per snp fst data
    subject_snp_fst <- vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl_nm)) %>% 
      # NAs correspond to X chromosome, fix this
      mutate(CHROM = ifelse(is.na(CHROM), "X", CHROM))

    # tuen per SNP info into granges
    subject_snp_fst_gr <- GRanges(
      seqnames = subject_snp_fst$CHROM, 
      IRanges(
        start = subject_snp_fst$POS, 
        end = subject_snp_fst$POS
        )
      )
    subject_snp_fst_gr$WEIR_AND_COCKERHAM_FST <- subject_snp_fst$WEIR_AND_COCKERHAM_FST # add snp fst
    
    # find overlaps between genes and snp-fst scores
    ov <- findOverlaps(query = query_genes, subject = subject_snp_fst_gr, maxgap = 0)
    
    # extract gene information from overalps
    query_genes_fst <- query_genes[queryHits(ov)]
    
    # add snp-fst to gene overlaps
    query_genes_fst$WEIR_AND_COCKERHAM_FST <- subject_snp_fst_gr[subjectHits(ov)]$WEIR_AND_COCKERHAM_FST
    
    # turn granges into a data frame
    query_genes_fst <- query_genes_fst %>% as.data.frame()
    
    # group by gene and get the gene-mean fst
    mean_gene_fst <- query_genes_fst %>% 
      dplyr::select(seqnames, start, end, gene_name, WEIR_AND_COCKERHAM_FST) %>% 
      group_by(seqnames, start, end, gene_name) %>% 
        summarise(n = n(), mean_gene_fst = mean(WEIR_AND_COCKERHAM_FST, na.rm=TRUE)) %>% 
      as.data.frame() %>% 
      mutate(contrast = contrast) %>% 
      dplyr::select(contrast, everything())
    
    return(mean_gene_fst)
  }
```

```{r ensembl_gene_set, include=F}
# Load gene set
mmu_gene_set <- read.table(here("reference_genome_ensembl/Mus_musculus.GRCm38.93_gene.gtf"), 
                           header = F, stringsAsFactors = F) 
  
names(mmu_gene_set) <- c("seqname","start","end","strand","gene_id","gene_id2","gene_biotype") # as in readme

mmu_gene_set <-mmu_gene_set %>% 
  filter(seqname %in% c(1:19,"X")) %>% # subset for autosomes and X
  dplyr::rename(gene_symbol = gene_id2)

# Turn gene set into a genomics range object
mmu_gene_set_gr <- GRanges(seqnames = mmu_gene_set$seqname,
                           ranges = IRanges(start = mmu_gene_set$start,
                                            end = mmu_gene_set$end),
                           strand = mmu_gene_set$strand,
                           mcols = dplyr::select(mmu_gene_set, gene_id, gene_symbol, gene_biotype))

```

```{r function_run_WebGestaltR, include = F}
run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}


run_WebGestaltR2 <- function( # allow to define fdr threshold
  interestGene, # vector of genes as input
  fdrThr, #fdr between >0 and 1
  topThr = 100, # how many results to write in results table
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = fdrThr,
    topThr = topThr,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}
```

```{r import_process_WebGestaltResults, include=F}
#-----------
# GOBP
#-----------

# DUC had no significant results

gobp_DUK_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DUK_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DU6_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DU6_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DU6P_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DU6P_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

gobp_DUhLB_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_DUhLB_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 
gobp_FERT_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/GOBP/Project_WebGestaltR_FERT_vs_FZTDU/geneontology_Biological_Process_sig_results.csv")
  ) 

#-----------
# KEGG
#-----------

# DUK,DUC and DU6 had no significant results

kegg_DU6P_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_DU6P_vs_FZTDU/pathway_KEGG_sig_results.csv")
  ) 

kegg_DUhLB_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_DUhLB_vs_FZTDU/pathway_KEGG_sig_results.csv")
  )

kegg_FERT_vs_FZTDU <- read.csv(
  here("00_dashboard/data/WebGestaltR/KEGG/Project_WebGestaltR_FERT_vs_FZTDU/pathway_KEGG_sig_results.csv")
  ) 
  

#-----------------
# Combine results
#-----------------

sig_gobp_kegg <- list(
  gobp_DUK_vs_FZTDU %>% mutate(contrast = "DUK_vs_FZTDU", db = "gobp"), 
  gobp_DU6_vs_FZTDU %>% mutate(contrast = "DU6_vs_FZTDU", db = "gobp"), 
  gobp_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "gobp"), 
  gobp_DUhLB_vs_FZTDU %>% mutate(contrast = "DUhLB_vs_FZTDU", db = "gobp"), 
  gobp_FERT_vs_FZTDU %>% mutate(contrast = "FERT_vs_FZTDU", db = "gobp"), 
  kegg_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "kegg"), 
  kegg_DUhLB_vs_FZTDU %>% mutate(contrast = "DUhLB_vs_FZTDU", db = "kegg"), 
  kegg_FERT_vs_FZTDU %>% mutate(contrast = "FERT_vs_FZTDU", db = "kegg"), 
  kegg_DU6P_vs_FZTDU %>% mutate(contrast = "DU6P_vs_FZTDU", db = "kegg")
  ) %>% 
  bind_rows() 

#------------------
# one gene per row
#------------------
sig_gobp_kegg_one_gene_per_row <- lapply(1:nrow(sig_gobp_kegg), function(i){
  
  gene_ids <- sig_gobp_kegg$overlapId[i] %>% str_split("[;]") %>% unlist()
  
  gene_ensembl <- sig_gobp_kegg$userId[i] %>% str_split("[;]") %>% unlist()
  
  data.frame(
    gene_ids = gene_ids,
    gene_ensembl = gene_ensembl
  ) %>% 
    mutate(
      contrast = sig_gobp_kegg$contrast[i],
      db = sig_gobp_kegg$db[i],
      description = sig_gobp_kegg$description[i]
      ) %>% 
    dplyr::select(contrast, db, description, gene_ensembl)
}) %>% 
  bind_rows()

```

```{r snpeff_HIGH_snps_genes_Fst_prep, eval = F}
# prepare data once, export table and load subset in next chunk

snpeff_snps_HIGH <- read.table(
  here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.HIGH.tab"),
  header = F
  ) %>% 
  dplyr::rename(CHROM = V1, POS = V2, effect = V3, impact = V4, gene = V5, gene_ensembl = V6) %>% 
  # add Fst data per SNP
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUK_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUK_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUC_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUC_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DU6_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>%
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6P_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DU6P_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUhLB_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(DUhLB_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST) %>% 
  inner_join(
    vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/FERT_vs_FZTDU.weir.fst")),
    by = c("CHROM","POS")
    ) %>% 
  dplyr::rename(FERT_vs_FZTDU_Fst = WEIR_AND_COCKERHAM_FST)

#write.csv(snpeff_snps_HIGH, here("00_dashboard/for_manuscript/supplementary_table_14.csv"))

```

```{r snpeff_HIGH_snps_genes_Fst_load_data, include=F}

# load table prepared in previous chunk
snpeff_snps_HIGH <- read.csv(here("00_dashboard/for_manuscript/supplementary_table_14.csv"))

```

```{r snpeff_HIGH_snps_GTcounts_function, include=FALSE}

# import Genotypes
HIGHimpact_GT <- vroom(here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.HIGHimpact.table")) #%>%  head(5)

# clean-up columns
names(HIGHimpact_GT) <- str_remove(names(HIGHimpact_GT),".GT") %>% str_remove("-L1")

tabulate_genotypes_per_contrast <- function(pop){
  
  #pop <- "DUK"

  # Get sample ids for each line
  samps_pop <- sample_info %>% filter(Linie %in% pop) %>% .$sample_id
  
  # tabulate genotypes of selected line
  sel_GT_cts <- HIGHimpact_GT %>% 
    dplyr::select(all_of(samps_pop)) %>% 
    apply(1, function(x){
      tb <- table(x)
      tb_nms <- names(tb)
      paste0(tb, "(", tb_nms, ")") %>% paste(collapse = ";")
    }) 

  # add locus information
  ss <- HIGHimpact_GT %>% dplyr::select(CHROM, POS)
  
  # add column with GT counts
  new_col <- paste0(paste(pop, collapse = "_"),"_GT_counts")
  ss$tmp <- sel_GT_cts
  names(ss)[names(ss) == "tmp"] <- new_col
  
  # merge GT counts with snp annotations ans fst scores
  read.table(
    here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.HIGH.tab"),
    header = F
  ) %>% 
    inner_join(
      ss, by = c("V1"="CHROM","V2"="POS")
    ) %>% 
    dplyr::rename(
      CHROM=V1, SNP_POS=V2, REF=V3, ALT=V4, snp_effect=V5, snp_impact=V6, gene_name=V7, gene_ensembl=V8
    ) %>% 
    unique()
}
```

```{r find_distinct_genes, include=FALSE}
get_distinct_genes <- function(ma, cont1, rest_cont, upper_bound, lower_bound){

  # get entries of matrix with high fst for target contrast
  cond1 <- ma[,cont1] >= upper_bound
  
  # get entries of matrix in which remaining contrast are low 
  cond2 <-  ma[,rest_cont] %>% apply(1, function(x) all(x < lower_bound) )
  
  # turn NAs as FALSE
  cond1[is.na(cond1)] <- FALSE
  cond2[is.na(cond2)] <- FALSE
  
  # combine both conditions into one indexing vector
  idx <- cbind(cond1,cond2) %>% apply(1, all)
  
  # subset matrix
  ma[idx,] %>% as.data.frame()

}
```

```{r load_regulatory_regions_ensembl, include = FALSE} 

# https://www.ensembl.org/info/website/upload/gff.html --> gff format explained

## Regulatory_Build
x <- vroom(
  here("reference_genome_ensembl/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff"), col_names = FALSE
)

regulatory_features <- GRanges(
  seqnames = x$X1, 
  IRanges(x$X4, x$X5),
  mcols = dplyr::select(x,X2, X3, X9) %>% dplyr::rename(source = X2, feature = X3, attribute = X9)
  )


## motiffeatures
y <- vroom(
  here("reference_genome_ensembl/mus_musculus.GRCm38.motiffeatures.20180516.gff"), col_names = FALSE
)


motif_features <- GRanges(
  seqnames = y$X1, 
  IRanges(y$X4, y$X5),
  mcols = dplyr::select(y,X2, X3, X6, X7, X9) %>% dplyr::rename(source = X2, feature = X3, score = X6, strand = X7, attribute = X9)
  )

rm(x,y)
```


Intro
====================================

* This is the analysis of the WGS data of DU mouse lines

* For every DU-line (DUK, DUC, DU6, DU6P, DUhLB, FZTDU) genomes of 25 animals were sequenced. Total of 150 genomes.

* Out the 25 genomes in one line, 10 were sequenced at ~25x and 15 at ~5x.

* This document begins with the evaluation the alignment metrics and concludes with the identification of genomic regions of distinct genetic differentiation for each selected line.

Bash_scripts
====================================

Column {.tabset}
-----------------

### Bash scripts

**Intro**

Text Text Text TextText TextText TextText TextText TextText TextText TextText TextText TextText Text


### Quality control 1

````
insert script

````


### Trimming

````
insert script

````


### cont....

Alignments
====================================

Column {.tabset}
-----------------

### Coverage-1
```{r visualize_cvg_data_1, fig.width=15, fig.height=10}

cvg_dat_means <- cvg_dat %>% 
  group_by(target_cvg, data_set) %>% summarise(mean = mean(CVG)) 

cvg_dat %>% 
  ggplot(aes(x = Linie, y = CVG, fill = target_cvg)) +
    geom_boxplot(outlier.size = 2, size = 1) +
    geom_hline(data = cvg_dat_means, aes(yintercept = mean, color = target_cvg), 
               linetype = "dashed", size = 1, show.legend = FALSE) +
    facet_wrap(~data_set, nrow = 1) +
    scale_y_continuous(breaks = seq(0,80,5)) +
    theme_bw(base_size = 20) +
    xlab(NULL) +
    ylab("Mean Genome Coverage") +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5)
    )
```

### Coverage-2
```{r visualize_cvg_data_2, fig.width=15, fig.height=10}
cvg_dat %>% 
  ggplot(aes(x = Linie, y = CVG)) +
    geom_boxplot(outlier.size = 2, size = 1) +
    geom_hline(data = cvg_dat_means, aes(yintercept = mean), 
               linetype = "dashed", color = "red", size = 1, show.legend = FALSE) +
    facet_grid(target_cvg~data_set, scales = "free") +
    scale_y_continuous(breaks = seq(0,80,5)) +
    theme_bw(base_size = 20) +
    xlab(NULL) +
    ylab("Mean Genome Coverage") +
    labs(
      caption = "Dashed lines = mean data set"
      ) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5)
    )
```


### Percentage of Genome Territory Covered
```{r pct_genome_cvg_visualization, fig.width=15, fig.height=10}
genome_cvg_pct_dat <- 
  # prepare data
  cvg_dat_postalignment %>% 
    # select columns
    dplyr::select(target_cvg, data_set, Linie, sample_name, ends_with("X")) %>% 
    # convert to long table
    reshape2::melt(
      id.vars = c("target_cvg","data_set", "Linie", "sample_name"),
      value.name = "pct_cvg",
      variable.name = "cvg_class"
    ) %>% 
    # fix x-axis and <-axis
    mutate(
      cvg_class = str_remove(cvg_class, "PCT_"),
      cvg_class = factor(cvg_class, paste0(c(1, seq(5,30,5), seq(40,100,10) ), "X") ),
      pct_cvg = pct_cvg*100
      ) %>% 
    #remove dropout sample (keep only re-sequenced one)
    filter(!(sample_name == "I34772"))

genome_cvg_pct_dat %>% 
  ggplot(aes(x = cvg_class, y = pct_cvg)) +
    geom_boxplot() +
    geom_point(alpha = 0.3) +
    facet_grid(target_cvg ~ data_set) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5)
    ) +
    ylab("Percentage Genome") +
    xlab(NULL) +
    scale_y_continuous(breaks = seq(0,100,10)) +
    labs(
      title = "Percentage Genome Covered X times",
      subtitle = "By target cvg and data set"
      ) 
```


### Mean CVG summary table
```{r}
cvg_dat %>% 
  group_by(target_cvg, data_set) %>% 
  summarise(min = min(CVG), mean = mean(CVG), median = median(CVG), max = max(CVG)) %>% 
  kable(digits = 2) %>% kable_styling(full_width = F)
```

### Percentage of Genome covered summary 
```{r}
genome_cvg_pct_dat %>% 
  group_by(target_cvg, data_set, cvg_class) %>% 
  summarise(min = min(pct_cvg), mean = mean(pct_cvg), median = median(pct_cvg), max = max(pct_cvg)) %>% 
  # prepare datatable output
  datatable(
    rownames = F, 
    options = list(
      columnDefs = list(list(className = 'dt-center', targets = "_all")),
      pageLength = 100
      )
    ) %>% 
  formatRound(columns = c("min","mean","median","max"), digits=2)

```

### Insert size
```{r, fig.width=15, fig.height=10}
insert_size_dat %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB", "FZTDU"))) %>% 
  ggplot(aes(x = sample_name, y = MEAN_INSERT_SIZE)) +
    geom_bar(stat = "identity") +
    geom_errorbar(
      aes(x = sample_name, 
          ymax = MEAN_INSERT_SIZE + STANDARD_DEVIATION, 
          ymin = MEAN_INSERT_SIZE - STANDARD_DEVIATION),
      width=0.2
      ) +
    facet_wrap(target_cvg ~ Linie, scales = "free", nrow = 2) +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.title = element_text(hjust = 0.5)
    ) +
    ylab("Mean Insert Size (+/- SD)") +
    xlab(NULL) +
    ggtitle("Per Sample Mean Insert Size")
```

### Per-sample metrics
```{r}
# raw nr of reads
d1 <- 
  # raw nr ofreads
  bind_rows(
    # batch1 and 2
    NR_RL_GB_CVG_60samp_9Lanes_raw %>% 
      # replace boosted samples
      filter( !(sample %in% NR_RL_GB_CVG_10samp_10Lanes_raw$sample) ),
      NR_RL_GB_CVG_10samp_10Lanes_raw
    ) %>% 
      # prepare col names
      dplyr::select(target_cvg, Linie, sample, NR_R1, NR_R2) %>%
      dplyr::rename(sample_name = sample, R1 = NR_R1, R2 = NR_R2) %>% 
      bind_rows(
        #batch3
        qc_raw_stats_avg_cvg %>% 
          # remove dorpout sample
          filter( !(sample_name %in% "I34772_dropout") ) %>% 
          mutate(
            # rename resequenced sample and add target cvg column
            sample_name = ifelse(sample_name == "I34772_dropout_reseq", "I34772",sample_name),
            target_cvg = "5x"
            ) %>% 
          # arrange columns for binding
          dplyr::select(target_cvg, Linie, sample_name, R1_NR, R2_NR) %>% 
          dplyr::rename(R1 = R1_NR, R2 = R2_NR)
      ) %>% 
  # fix duhlb name
  mutate(Linie = ifelse(Linie == "DUHLB", "DUhLB", Linie))

# clean nr of reads and pairs aligned
d2 <- bind_rows(
        NR_alignment_dat_target_cvg_30x, 
        NR_alignment_dat_target_cvg_5x
      )

# mean coverage
d3 <- cvg_dat %>% 
  filter(data_set == "aligned_m0q0") %>% dplyr::select(target_cvg, Linie, sample_name, CVG) %>% 
  mutate(sample_name = ifelse(sample_name == "I34772_dropout_reseqd", "I34772",sample_name))

# coverage pct at 5x
d4 <- genome_cvg_pct_dat %>% filter(data_set == "m0q0" & cvg_class == "5X") %>% 
  dplyr::select(target_cvg, Linie, sample_name, pct_cvg) %>% 
  dplyr::rename(pct_cvg_5x = pct_cvg) %>% 
  mutate(sample_name = ifelse(sample_name == "I34772_dropout_reseqd", "I34772",sample_name))

per_sample_metrics <- left_join(
  d1,d2,by=c("target_cvg", "Linie", "sample_name")
) %>% 
  left_join(d3, by=c("target_cvg", "Linie", "sample_name")) %>% 
  left_join(d4, by=c("target_cvg", "Linie", "sample_name")) %>% 
  mutate(NR_raw = R1 + R2) %>% 
  dplyr::rename(
    NR_clean=paired_in_seq,
    mean_CVG = CVG
    ) %>% 
  dplyr::select(-pct_properly_paired, -R1, -R2) %>% 
  dplyr::select(target_cvg, Linie, sample_name, NR_raw, everything()) %>% 
  mutate(
    NR_clean_pct = (NR_clean/NR_raw)*100,
    properly_paired_pct = (properly_paired/NR_raw)*100
  ) %>% 
  #add insert size information
  inner_join(
    dplyr::select(insert_size_dat, target_cvg, sample_name, sample_name, MEAN_INSERT_SIZE),
    by = c("target_cvg","sample_name")
    ) 

per_sample_metrics %>% 
    datatable(
      rownames = F, 
      options = list(
        columnDefs = list(list(className = 'dt-center', targets = "_all")),
        pageLength = 150
        )
      ) %>% 
    formatRound(columns = c("mean_CVG","pct_cvg_5x","NR_clean_pct","properly_paired_pct"), digits=2)

#write.csv(per_sample_metrics, here("00_dashboard/for_manuscript/supplementary_table_1.csv"))
```

### Summary metrics
```{r}
per_sample_metrics %>% 
  group_by(target_cvg, Linie) %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by target-cvg and by line")  

tab <- per_sample_metrics %>% 
  group_by(target_cvg) %>% 
  summarise(
    sample_size = n(),
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) 
tab %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by target-cvg")  
#write.csv(tab, here("00_dashboard/for_manuscript/table_1.csv")) 

per_sample_metrics %>% 
  group_by(Linie) %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "data set split by line")

per_sample_metrics %>% 
  summarise(
    mean_properly_paired_pct = mean(properly_paired_pct),
    mean_insert_size = mean(MEAN_INSERT_SIZE),
    mean_cvg = mean(mean_CVG),
    mean_pct_cvg_5x = mean(pct_cvg_5x)
    ) %>% 
  kable(digits = 2) %>% 
  kable_styling(full_width = F) %>%
  footnote(general = "full data set")

```

SNP_Calling
====================================

Column {data-width=300}
------------------------

### Overview
* GATK4 best practices
* Site level filtering: 
  * bi-allelic SNPs
  * 95% truth sensitivity (VQSR)
* Genotype level filtering: 
  * DP: 4 - max-DP (max-DP = mean_DP_library + 3*SD)
  * GQ >= 20
  * Else missing (./.)
  * Filtering by min called samples per population: 15 (DUK,DUC,DU6P,DUhLB,FZTDU) or 12 (DU6)
  
### Number of Variants
```{r} 
vcfs_main_summary <- bind_rows(
  raw_vcf_metrics_summary,
  mid_vcf_metrics_summary,
  final_vcf_metrics_summary
) %>% 
  dplyr::select(data_set, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV, TOTAL_INDELS, TOTAL_MULTIALLELIC_SNPS) %>% 
  reshape2::melt(id.vars = "data_set") %>% 
  reshape2::dcast(variable ~ data_set, fill = "value") %>%
  mutate(raw = as.numeric(raw), 
         site_filtered = as.numeric(site_filtered), 
         gt_filtered = as.numeric(gt_filtered)
         ) %>% 
  dplyr::select(variable, 
                raw, 
                site_filtered, 
                gt_filtered
                )

vcfs_main_summary %>% 
  knitr::kable(format.args = list(big.mark = ","), digits = 2) %>% 
  kable_styling(full_width = F)

#write.csv(vcfs_main_summary, here("00_dashboard/for_manuscript/supplementary_table_2.csv"))
```


Column  {data-width=700, .tabset}
-----------------------------------

### Missingness-1
```{r visualize_missingness, fig.width=15, fig.height=10}
dat_miss <- read.csv(
  here("batches123_04_final_VCF/figures_tables","fraction_snps_by_min_n_called_per_line.csv")
  ) %>% 
  dplyr::select(-X)


dat_miss %>% 
  reshape2::melt(
    id.vars = "min_called_samples",
    variable.name = "line",
    value.name = "fraction_snps"
  ) %>% 
  ggplot(aes(x = min_called_samples, y = fraction_snps, color = line)) +
    geom_line(size = 1.5) +
    scale_x_continuous(breaks = seq(0,25,1)) +
    theme_bw(base_size = 12) +
    labs(title = "Fraction of SNPs with at least N calls") +
    theme(plot.title = element_text(hjust = 0.5))
```

### Missingness-2
```{r}
dat_miss %>% 
  filter(!(min_called_samples == 0)) %>% 
  kable(align = "l", digits = 2, 
        caption = "Fraction of SNPs with at least N calls") %>% 
  kable_styling(full_width = F)
```

### SNPs per sample
```{r} 
p1 <- final_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_SNPS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

#p1

ggplotly(p1) 
```

### Sample metrics
```{r}
final_vcf_metrics_detail %>% 
  dplyr::select(Linie, SAMPLE_ALIAS, target_cvg, everything(),
                -PCT_GQ0_VARIANTS, -TOTAL_GQ0_VARIANTS) %>% 
  datatable(rownames = F, options = list(pageLength = 150))
  #kable()
```

### Population metrics-1
```{r  snps_per_pop, fig.width=15, fig.height=10} 
tab <- here("batches123_04_final_VCF/output") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  .[grep("biallelicSNPs_VQSR95",.)] %>% 
  lapply(function(x){ 
    d <- read.table(here("batches123_04_final_VCF/output",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() %>% 
  mutate(line = factor(line, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(line)

tab %>% 
  kable(caption = "Per-Line SNP summary", digits = 2) %>% 
  kable_styling(full_width = F) %>% 
  footnote(general = "Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT==RR) and all-missing sites (GT==./.).")

#tab %>% dplyr::select(line, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV) %>% write.csv(here("00_dashboard/for_manuscript/table_2.csv"))
```

### Population metrics-2 
```{r} 
tab %>% 
  ggplot(aes(x = line, y = TOTAL_SNPS)) +
    geom_bar(stat = "identity") +
      theme_bw(base_size = 12) +
      xlab(NULL)
```


### Intersections minimal 
```{r make_upsetr_plot_minimal_run_once_and_export, eval = FALSE} 
#png(here("00_dashboard/for_manuscript/figure_1.png"), res = 300, height = 3000, width = 3800, units = "px")
upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  text.scale = c(
    2, #intersection size title
    1.5, #intersection size tick labels
    1.5, #set size title
    1, #set size tick labels
    2, #set names
    1.5 #numbers above bars
    ),
  number.angles = 45,
  intersections =  list(
    # full intersection
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    # selected line intersctions
    list("duhlb","du6p","du6","duc","duk"),
    # fertility lines vs fztdu
    list("duk","duc","fztdu"),
    # body size lines vs fztdu
    list("du6","du6p","fztdu"),
    # lines vs fztdu
    list("duk","fztdu"),
    list("duc","fztdu"),
    list("du6","fztdu"),
    list("du6p","fztdu"),
    list("duhlb","fztdu"),
    # phenotype related intersection
    list("duk","duc"),
    list("du6","du6p"),
    #list("duhlb","fztdu"), #done already
    # private SNPs
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    ),
  queries = list(
    # Private SNPs per line (red)
    list(query = intersects, params = list("duk"),color = "red", active = T),
    list(query = intersects, params = list("duc"),color = "red", active = T),
    list(query = intersects, params = list("du6"),color = "red", active = T),
    list(query = intersects, params = list("du6p"),color = "red", active = T),
    list(query = intersects, params = list("duhlb"),color = "red", active = T),
    list(query = intersects, params = list("fztdu"),color = "red", active = T),
    
    # private SNPs of selected lines in FZTDU (blue)
    list(query = intersects, params = list("duk","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duc","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6p","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duhlb","fztdu"),color = "blue", active = T),
    
    # SNPs found only in related lines (yellow)
    list(query = intersects, params = list("duk","duc"),color = "orange", active = T),
    list(query = intersects, params = list("du6","du6p"),color = "orange", active = T),
    
    # SNPs found only in related lines and FZTDU (green)
    list(query = intersects, params = list("duk","duc","fztdu"),color = "green", active = T),
    list(query = intersects, params = list("du6","du6p","fztdu"),color = "green", active = T),
    
    # SNPs found between 5 lines (purple) 
     list(query = intersects, params = list("duhlb","du6p","du6","duc","duk"),color = "purple", active = T),
    
    # common SNPs ()
    list(query = intersects, params = list("fztdu","duhlb","du6p","du6","duc","duk"),color = "black", active = T)
    )

)

grid.text("Intersections SNP sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()
```

```{r upsetr_plot_minimal_import, fig.width=10} 

include_graphics(here("00_dashboard/for_manuscript/figure_1.png"))

```


### Intersections expanded
```{r make_upsetr_plot_expanded_export, eval = FALSE}

png(here("00_dashboard/figures/snp_intersections_expanded.png"), res = 300, height = 3000, width = 3800, units = "px")

upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  text.scale = c(
    2, #intersection size title
    1.5, #intersection size tick labels
    1.5, #set size title
    1, #set size tick labels
    2, #set names
    1.5 #numbers above bars
    ),
  number.angles = 45,
  intersections =  list(
    # full intersection
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    # 5 line intersctions
    list("fztdu","duhlb","du6p","du6","duc"),
    list("fztdu","duhlb","du6p","du6","duk"),
    list("fztdu","duhlb","du6p","duc","duk"),
    list("fztdu","duhlb","du6","duc","duk"),
    list("fztdu","du6p","du6","duc","duk"),
    list("duhlb","du6p","du6","duc","duk"),
    # fertility lines vs fztdu
    list("duk","duc","fztdu"),
    # body size lines vs fztdu
    list("du6","du6p","fztdu"),
    # lines vs fztdu
    list("duk","fztdu"),
    list("duc","fztdu"),
    list("du6","fztdu"),
    list("du6p","fztdu"),
    list("duhlb","fztdu"),
    # phenotype related intersection
    list("duk","duc"),
    list("du6","du6p"),
    #list("duhlb","fztdu"), #done already
    # private SNPs
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    ),
  queries = list(
    # Private SNPs per line (red)
    list(query = intersects, params = list("duk"),color = "red", active = T),
    list(query = intersects, params = list("duc"),color = "red", active = T),
    list(query = intersects, params = list("du6"),color = "red", active = T),
    list(query = intersects, params = list("du6p"),color = "red", active = T),
    list(query = intersects, params = list("duhlb"),color = "red", active = T),
    list(query = intersects, params = list("fztdu"),color = "red", active = T),
    
    # private SNPs of selected lines in FZTDU (blue)
    list(query = intersects, params = list("duk","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duc","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6p","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duhlb","fztdu"),color = "blue", active = T),
    
    # SNPs found only in related lines (yellow)
    list(query = intersects, params = list("duk","duc"),color = "orange", active = T),
    list(query = intersects, params = list("du6","du6p"),color = "orange", active = T),
    
    # SNPs found only in related lines and FZTDU (green)
    list(query = intersects, params = list("duk","duc","fztdu"),color = "green", active = T),
    list(query = intersects, params = list("du6","du6p","fztdu"),color = "green", active = T),
    
    # SNPs found between 5 lines (purple) 
    list(query = intersects, params = list("fztdu","duhlb","du6p","du6","duc"),color = "purple", active = T),
    list(query = intersects, params = list("fztdu","duhlb","du6p","du6","duk"),color = "purple", active = T),
    list(query = intersects, params = list("fztdu","duhlb","du6p","duc","duk"),color = "purple", active = T),
    list(query = intersects, params = list("fztdu","duhlb","du6","duc","duk"),color = "purple", active = T),
    list(query = intersects, params = list("fztdu","du6p","du6","duc","duk"),color = "purple", active = T),
    list(query = intersects, params = list("duhlb","du6p","du6","duc","duk"),color = "purple", active = T),
    
    # common SNPs ()
    list(query = intersects, params = list("fztdu","duhlb","du6p","du6","duc","duk"),color = "black", active = T)
    )

)

grid.text("Intersections SNP sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))

dev.off()

```


```{r upsetr_plot_expanded_import, fig.width=10}

include_graphics(here("00_dashboard/figures/snp_intersections_expanded.png"))

```



### Intersection Sel vs FZTDU
```{r export_venn_sel_vs_fztdu, eval=F}

#png(here("00_dashboard","for_manuscript","supplementary_figure_6.png"), res = 300, height = 2000, width = 2000, units = "px")


par(mfrow = c(3, 2))

for(sel in c("duk","duc","du6","du6p","duhlb")){
  
  #sel="duk"
  
  d <- data.frame(
    set_dat$ids %in% set_dat$ids[ set_dat[,sel] ],
    set_dat$ids %in% set_dat$ids[ set_dat$fztdu ]
  )
  
  names(d) <- c(sel,"fztdu")
  
  a <- vennCounts(d)
  
  vennDiagram(a, circle.col = c("red","blue"), mar = rep(0.1,4), cex = 0.7)
  }

#dev.off()

```

```{r import_venn_sel_vs_fztdu, fig.width=10, fig.height=10}
include_graphics(here("00_dashboard","for_manuscript","supplementary_figure_6.png"))
```

### Intx sel-fztdu-1
```{r for_manuscript_pct_intersection_sel_vs_fztdu}

intersections <- sapply(c("duk","duc","du6","du6p","duhlb","fztdu"), function(sel){
  
  x <- set_dat$ids[ set_dat[,sel] ]
  y <- set_dat$ids[ set_dat$fztdu ]
  
  xy <- intersect(x,y) %>% length()

  })

set_sizes <- sapply(c("duk","duc","du6","du6p","duhlb","fztdu"), function(sel){
  
  x <- set_dat$ids[ set_dat[,sel] ]

  length(x)

  })

tab <- data.frame(
  line = c("duk","duc","du6","du6p","duhlb","fztdu"),
  set_size = set_sizes,
  intersection_sel_x_fztdu = intersections,
  pct_sel_in_fztdu = (intersections/set_sizes)*100,
  pct_fztdu_in_sel = (intersections/ length(set_dat$ids[ set_dat$fztdu ]) )*100
)

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_16.csv"))

tab %>% 
  kable(digits = 2, row.names = FALSE) %>% kable_styling(full_width = F)

```

### Intx sel-fztdu-2 
```{r, fig.width=10, fig.height=10}
#https://www.r-bloggers.com/how-to-make-a-pie-chart-in-r/

tab %>% 
  dplyr::rename(in_fztdu = pct_sel_in_fztdu) %>% 
  mutate(not_in_fztdu = 100 - in_fztdu) %>% 
  dplyr::select(line, in_fztdu, not_in_fztdu) %>% 
  filter(!(line == "fztdu")) %>% 
  reshape2::melt(id.vars = "line", variable.name = "class", value.name = "pct") %>% 
  mutate(
    pct = round(pct,0),
    line = factor(line, levels = c("duk","duc","du6","du6p","duhlb"))
    ) %>% 
  ggplot(aes(x="", y=pct, fill=class)) +
    geom_bar(width = 1, stat = "identity", color = "white") +
    coord_polar("y", start = 0) +
    geom_text(
      aes(label = paste0(pct, "%")), 
      position = position_stack(vjust = 0.5)
      ) +
    scale_fill_manual(values = c("#0073C2FF", "#EFC000FF")) +
    labs(
      title = "Percentage of SNPs in FZTDU"
    ) +
    theme_void() +
    theme(
      strip.text = element_text(size = 15),
      legend.title = element_blank(),
      legend.text = element_text(size = 12),
      plot.title = element_text(hjust = 0.5)
    ) +
    facet_wrap(~line, ncol = 3)
```


SNPeff
====================================

Column {.tabset}
----------------

### summary table 
```{r} 
snpeff_data$summary_table %>% kable() %>% kable_styling(full_width = F)
```

### Counts by effect
```{r} 
tab <- snpeff_data$cts_by_eff %>% 
  mutate(Type = str_trim(Type)) %>% 
  left_join(impact_snpeff, by = c("Type" = "Effect")) %>% 
  unique() %>% 
  mutate(
    Impact = ifelse(Type == "5_prime_UTR_premature_start_codon_gain_variant",
                  "LOW",
                  Impact)
  ) 

tab %>% 
  datatable(
    rownames = F, 
    options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all")))
    )


#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_4.csv"))
```

### Change rate by chr
```{r} 
snpeff_data$change_rate_by_chr %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### counts by genotype region 
```{r}
tab <- snpeff_data$cts_by_geno_region 

tab %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_3.csv")) 
```

### vars_by_type
```{r}
snpeff_data$vars_by_type %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_func_class
```{r}
snpeff_data$eff_by_func_class %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

### eff_by_impact
```{r}
snpeff_data$eff_by_impact %>% 
  datatable(rownames = F, options = list(pageLength = 25, columnDefs = list(list(className = 'dt-center', targets = "_all"))))
```

Genetic_Str
====================================

Column {.tabset}
----------------

### PC1 - PC2
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

#png(here("00_dashboard/for_manuscript/figure_2A.png"), res = 300, height = 3000, width = 4000, units = "px")
#pc_eigenscores %>% 
#  ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) + 
#    geom_point(size = 4, alpha = 0.7) + 
#    theme_bw(base_size = 20) + 
#    theme(legend.title = element_blank())
#dev.off()


p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC2 - PC3
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC2,y=PC3,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC3 - PC4
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC3,y=PC4,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC4 - PC5
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC4,y=PC5,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### PC5 - PC6
```{r, fig.width = 5, fig.height=7}
p1 <- pc_eigenscores %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())

#png(here("00_dashboard/for_manuscript/figure_2B.png"), res = 300, height = 3000, width = 4000, units = "px")
#pc_eigenscores %>% 
#  ggplot(data = ., aes(x=PC5,y=PC6,color=Linie)) + 
#    geom_point(size = 4, alpha = 0.7) + 
#    theme_bw(base_size = 20) + 
#    theme(legend.title = element_blank())
#dev.off()

p2 <- pc_eigenscores %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC5,y=PC6,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### Scree Plot
```{r, fig.width = 7, fig.height=7}
pc_pct %>% 
  ggplot(aes(x=PC,y=varprop)) +
    geom_bar(stat = "identity") +
    xlab("Principal Component") +
    ylab("% Variance") +
    scale_x_continuous(breaks = 1:7) +
    theme_bw()
```

Column
---------------

### Hierarchical Clustering
```{r, fig.width = 10, fig.height=5}
dend <- readRDS(here("batches123_06_genetic_structure/data/dendrogram.rds")) %>% as.hclust()

# change labels to group to avoid legend
#identical(sample_info[,"sample_id"], str_remove(dend[["labels"]], "-L1")) #check order of labels is the same! yes!
dend[["labels"]] <- sample_info$Linie


colors <- c("red", "blue", "green", "black", "purple","orange")
clus6 <- cutree(dend, 6)
plot(as.phylo(dend), 
     type = "fan", 
     cex = 0.6,
     no.margin = TRUE,
     tip.color = colors[clus6])

# good reference: http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning
```


```{r for_paper_figure2C, evel = F}

#png(here("00_dashboard/for_manuscript/figure_2C.png"), res = 300, height = 3500, width = 4000, units = "px")

colors <- c("red", "blue", "green", "black", "purple","orange")
clus6 <- cutree(dend, 6)
plot(as.phylo(dend), 
     type = "fan", 
     cex = 1.5,
     no.margin = TRUE,
     tip.color = colors[clus6])

#dev.off()

```


### Admixture 
```{r import_fam_info}
# About sample ordering: https://www.biostars.org/p/221817/

fam_fl <- read.table(here("batches123_06_genetic_structure/data/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.fam"))
```

```{r barplot_k5, fig.width=20, fig.height=7}
make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) 
```


```{r for_paper_fig2D, eval = F}
#png(here("00_dashboard/for_manuscript/figure_2D.png"), res = 300, height = 3000, width = 6000, units = "px")

make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ldpruned.5.Q")
) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  scale_y_continuous(expand = c(0,0.02))

#dev.off()
```



Misc
====================================

Column {.tabset}
-------------------

### Linkage Disequilibrium Decay
```{r}
# the code to make this figures is not in this document ... where is it??
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max5Mb.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","trends_max250Kb.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)


```

```{r for_manuscript_alt_allele_frq_distr, eval = F}
# same as above, just tweaked for manuscript

fls <- list.files(here("batches123_08_LDD_SFS/output"),pattern = ".ld")

dat <- lapply(fls, function(fl){
  d <- read.table(here("batches123_08_LDD_SFS/output",fl),header = TRUE, stringsAsFactors = FALSE) 
	d <- mutate(d, dist = BP_B - BP_A)
	return(d)
	})

names(dat) <- str_remove(fls, "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered_thinned.recode.") %>% str_remove(".plink.ld")
names(dat)

dat2 <- dat %>% bind_rows(.id = "line")
head(dat2)

dat2$line <- factor(dat2$line, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))

# Full data set (max separation any 2 SNPs = 5Mb)

png(here("00_dashboard/for_manuscript/figure_3A.png"), res = 300, height = 3000, width = 3000, units = "px")
ggplot(data = dat2, aes(x=dist/1000, y=R2, color = line)) +
  geom_smooth(method = "auto") +
    xlab("Pairwise distance (Kb)") +
    ylab(expression(LD~(r^{2}))) +
    theme_bw(base_size = 20)+
    theme(legend.title=element_blank(), plot.title = element_text(hjust = 0.5)) +
    ggtitle("Linkage Disequilibrium Decay") 
dev.off()
```

### Allele Frequency Composition
```{r}
# code for this figures not here, where is it?
img1 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state.png"))
    )
  )
img2 <-  rasterGrob(
  as.raster(
    readPNG(here("batches123_08_LDD_SFS/figures","allele_frq_state_pct.png"))
    )
  )
grid.arrange(img1, img2, ncol = 2)
```


```{r classify_alleles_by_AF, eval = F}
dat <- lapply(list.files(here("batches123_08_LDD_SFS/output"), pattern = "*.ALTfrq$"), function(fl){
  
  pop <- fl %>% 
    str_remove(
      "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords."
    ) %>% 
    str_remove(".ALTfrq")
  
  read.table(here("batches123_08_LDD_SFS/output",fl)) %>% 
    mutate(Linie = pop)
}) %>% 
  bind_rows() 

dat <- dat %>% 
  mutate(state = NA) %>%
  mutate(state = ifelse(V1 > 0 & V1 < 1, "polymorphic",state)) %>% 
  mutate(state = ifelse(V1 == 0, "fixed_REF",state)) %>% 
  mutate(state = ifelse(V1 == 1, "fixed_ALT",state))

dat[sample(1:nrow(dat), 10),] %>% arrange(state) # corroborate

dat %>% group_by(Linie, state) %>% summarise(n())

```

```{r for_manuscript_fig3B, eval=F}
# same as above but improved

n_snps_final_vcf <- 5099945 # the number of SNPs in the final VCF 

dat_summ <- dat %>% 
  group_by(Linie, state) %>% 
  summarise(n = n(), pct = (n/n_snps_final_vcf)*100) 

dat_summ %>% ungroup() %>% group_by(Linie) %>% summarise(sum(n), sum(pct)) # corroborate all records appear (5099945)

dat_summ

# visualize

dat_summ <- dat_summ %>% 
  ungroup() %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU")))

#png(here("00_dashboard/for_manuscript/figure_3B.png"), res = 300, height = 3000, width = 3000, units = "px")
dat_summ %>% 
  ggplot(aes(x = Linie, y = pct, fill = state)) +
    geom_bar(stat = "identity", position = "stack") +
    theme_bw(base_size = 20) +
    ylab("% SNP counts") +
    xlab(NULL) +
    theme(
      plot.title = element_text(hjust = 0.5),
      legend.title = element_blank()
    ) +
    ggtitle("Allele Frequency Classification") 
#dev.off()

```

### Polymorphic AF Fractions

```{r plot_and_export_polymorphic_AF_distribution, eval = F}
p <- dat %>% 
  filter(state == "polymorphic") %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  ggplot(aes(x = V1)) +
    geom_histogram(bins = 30) +
    facet_wrap(~Linie) +
    theme_bw(base_size = 15) +
    xlab(NULL) +
    scale_x_continuous(breaks = seq(0,1,0.1))

png(here("00_dashboard/figures/histogram_polymorphic_AF_distribution.png"), res = 300, width = 3500, height = 2000, units = "px")
p
dev.off()
```


```{r plot_and_export_bined_AF_barplot, eval = F}

# create bins to classify AFs
bins <- data.frame(start_bin = seq(from = 1, to = 0.1, by = -0.1), 
                   end_bin = seq(from = 0.9, to = 0.0, by = -0.1)) %>% 
  mutate(frq_lab = paste0("(< ", start_bin, ") & (>= ", end_bin,")"))


# keep only polym alleles
dat_polymorphic <- dat %>% filter(state == "polymorphic")

# classify AF per pop
frq_class_data <- lapply(1:nrow(bins), function(i){
  
  d <- dat_polymorphic %>% 
    filter(V1 < bins[i,"start_bin"] & V1 >=  bins[i, "end_bin"] ) %>% 
    mutate(frq_lab = bins[i,"frq_lab"])

  return(d)
  }) %>% 
  bind_rows()

# get number of polym alleles per line
n_polym_snps <- frq_class_data %>% 
  group_by(Linie) %>% 
  summarise(n=n()) 


# Count SNPs and calculate percentages for each AF bin
frq_class_data_summ <- lapply(1:nrow(n_polym_snps), function(i){
  
  pop <- n_polym_snps$Linie[i]
  
  n_snps <- n_polym_snps$n[i]
  
  frq_class_data %>% 
    filter(Linie == pop) %>% 
    group_by(Linie, frq_lab) %>% 
    summarise(n = n(), pct =( n/n_snps)*100)
  
  }) %>% 
  bind_rows() %>% 
  mutate(
  Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU")),
  frq_lab = factor(frq_lab, levels = bins$frq_lab)
  ) 

# Set color palette
clrs <- brewer.pal(n = nrow(bins), name = 'Paired') 

# make bar plots
p <- frq_class_data_summ %>% 
  dplyr::rename(n_polymorphic_snps = n, pct_polymorphic_snps = pct) %>% 
  reshape2::melt(id.vars = c("Linie","frq_lab")) %>%
  ggplot(aes(x = Linie, y = value, fill = frq_lab)) +
    geom_bar(stat = "identity", position = "stack") +
    theme_bw(base_size = 20) +
    ylab(NULL) +
    xlab(NULL) +
    theme(
      plot.title = element_text(hjust = 0.5),
      legend.title = element_blank(),
      axis.text.x = element_text(angle = 90)
    ) +
    scale_fill_manual(values = clrs) +
    #ggtitle("Polymorphic Allele Frequency Classification") +
    facet_wrap(~variable, nrow = 1, scales = "free")


# export 
png(here("00_dashboard/figures/barplot_polymorphic_AF_bins.png"), res = 300, width = 3000, height = 2000, units = "px")
p
dev.off()
```


```{r display_polymorphic_AF_plots, out.width="70%", fig.align="center"}

include_graphics(c(here("00_dashboard/figures/histogram_polymorphic_AF_distribution.png"),
                   here("00_dashboard/figures/barplot_polymorphic_AF_bins.png")))

```


### Alternative Allele Frequency 
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_ALT_frq.png"))
```


```{r for_manuscript_fig3C, eval = F}

dat <- lapply(list.files(here("batches123_08_LDD_SFS/output"), pattern = "*.ALTfrq"), function(fl){
  
  pop <- fl %>% 
    str_remove(
      "cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords."
      ) %>% 
    str_remove(".ALTfrq")
  
  read.table(here("batches123_08_LDD_SFS/output", fl)) %>% 
    mutate(Linie = pop)
}) %>% 
  bind_rows() 

dat <- dat %>% mutate(Linie = factor(Linie, c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU")))


png(here("00_dashboard/for_manuscript/figure_3C.png"), res = 300, height = 3000, width = 2200, units = "px")

dat %>% 
  ggplot(aes(x = V1)) +
    geom_histogram(bins = 50) +
    facet_wrap(~Linie, ncol = 2) +
    theme_bw(base_size = 20) +
    ggtitle("Alternative Allele Frequency Distribution") +
    theme(
      plot.title = element_text(hjust = 0.5),
      axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5)
    ) +
    xlab(NULL) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE))

dev.off()

```


### Minor Allele Frequency (excl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_excl_maf0.png"))
```

### Minor Allele Frequency (incl MAF=0)
```{r, fig.width = 10}
include_graphics(here("batches123_08_LDD_SFS/figures","hist_MAF_incl_maf0.png"))
```



Fst
====================================

Column {.tabset}
-----------------------

### Summary
```{r summary_win_fst}
summary_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  #group_by(contrast, is_x) %>%
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(MEAN_FST),
    median = median(MEAN_FST),
    mean = mean(MEAN_FST),
    q1 = quantile(MEAN_FST, 0.01),
    q5 = quantile(MEAN_FST, 0.05),
    q10 = quantile(MEAN_FST, 0.10),
    q25 = quantile(MEAN_FST, 0.25),
    q90 = quantile(MEAN_FST, 0.90),
    q95 = quantile(MEAN_FST, 0.95),
    q99 = quantile(MEAN_FST, 0.99),
    max = max(MEAN_FST)
    ) #%>% 
  #arrange(is_x)

summary_win_fst %>%
  #mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  #dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window Fst") %>% 
  kable_styling(full_width = F) %>% kable_styling(full_width = F)

q_fst <- summary_win_fst %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q1","q5","q10","q25","q90","q95","q99"), 
variable.name = "Quantile", value.name = "Fst")


```

```{r summary_win_zfst}
summary_z_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(z_win_fst_score),
    median = median(z_win_fst_score),
    mean = mean(z_win_fst_score),
    q1 = quantile(z_win_fst_score, 0.01),
    q5 = quantile(z_win_fst_score, 0.05),
    q10 = quantile(z_win_fst_score, 0.10),
    q15 = quantile(z_win_fst_score, 0.15),
    q20 = quantile(z_win_fst_score, 0.20),
    q25 = quantile(z_win_fst_score, 0.25),
    q30 = quantile(z_win_fst_score, 0.30),
    q35 = quantile(z_win_fst_score, 0.35),
    q40 = quantile(z_win_fst_score, 0.40),
    q45 = quantile(z_win_fst_score, 0.45),
    q50 = quantile(z_win_fst_score, 0.50),
    q55 = quantile(z_win_fst_score, 0.55),
    q60 = quantile(z_win_fst_score, 0.60),
    q65 = quantile(z_win_fst_score, 0.65),
    
    q75 = quantile(z_win_fst_score, 0.75),
    q80 = quantile(z_win_fst_score, 0.80),
    q85 = quantile(z_win_fst_score, 0.85),
    q90 = quantile(z_win_fst_score, 0.90),
    q95 = quantile(z_win_fst_score, 0.95),
    q99 = quantile(z_win_fst_score, 0.99),
    max = max(z_win_fst_score)
    ) 

summary_z_win_fst %>% kable(digits = 2, caption = "z(Mean Window Fst)") %>% kable_styling(full_width = F)

q_zfst <- summary_z_win_fst %>% 
  dplyr::select(contrast, paste0("q",c(1,5,10,25,90,95,99))) %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q1","q5","q10","q25","q90","q95","q99"), 
                 variable.name = "Quantile", value.name = "zFst")


#saveRDS(summary_z_win_fst, here("00_dashboard","summary_z_win_fst.rds"))
```

### Windowed Fst Distribution
```{r, fig.width=10,fig.height=7}

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = MEAN_FST)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("Mean Window Fst") +
    geom_vline(q_fst, mapping = aes(xintercept = Fst, color = Quantile), linetype = "dashed")

p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = MEAN_FST)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("Mean Window Fst") +
    geom_hline(q_fst, mapping = aes(yintercept = Fst, color = Quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed Fst Distribution (violin-boxplot)
```{r,  fig.width=8,fig.height=10}
ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = MEAN_FST, y = NA), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = MEAN_FST, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw() +
  ylab(NULL) +
  xlab("Mean Window Fst") +
  labs(
    title = "Windowed Fst Distribution"
  ) +
  theme_bw(base_size = 15) +
  theme(
    plot.title = element_text(hjust = 0.5),
  #  strip.text = element_text(size = 7),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    strip.text.y = element_text(size = 10)
  ) +
  facet_wrap(~contrast, ncol = 1, strip.position = "right") +
  #scale_y_discrete(limits = rev(levels(win_fst_sel_vs_ctrl_prep$contrast))) +
  geom_vline(q_fst, mapping = aes(xintercept = Fst, color = Quantile), linetype = "dashed") 
  
```

### Windowed z-Fst Distribution
```{r, fig.width=10,fig.height=7}
p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = z_win_fst_score)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("z(Mean Window Fst)") +
    geom_vline(q_zfst, mapping = aes(xintercept = zFst, color = Quantile), linetype = "dashed")


p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = z_win_fst_score)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("z(Mean Window Fst)") +
    geom_hline(q_zfst, mapping = aes(yintercept = zFst, color = Quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed z-Fst Distribution (violin-boxplot)
```{r for_dashboard_and_manuscript_figure4B, fig.width=8,fig.height=10}
#png(here("00_dashboard/for_manuscript/figure_4B.png"), res = 300, height = 2900, width = 2300, units = "px")

ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = z_win_fst_score, y = NA), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = z_win_fst_score, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw(
    base_size = 20
  ) +
  ylab(NULL) +
  xlab("zFst") +
  labs(
    title = "Windowed z-Fst Distribution"
  ) +
  theme_bw(base_size = 15) +
  theme(
    plot.title = element_text(hjust = 0.5),
  #  strip.text = element_text(size = 7),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    strip.text.y = element_text(size = 10)
  )  +
  geom_vline(q_zfst, mapping = aes(xintercept = zFst, color = Quantile), linetype = "dashed") +
  facet_wrap(~contrast, ncol = 1, strip.position = "right") 

#dev.off()
```

### Genomewide Manhattan
```{r prepare_data_mahattans, include=F}
manhattan_dat <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  group_by(contrast, CHROM) %>% 
  arrange(contrast, CHROM, BIN_START) %>% 
  mutate(
    genomic_index = NA, 
    chr_index = 1:n(), 
    center = BIN_START + (BIN_END - BIN_START)/2,
    center_mb = center/1e6,
    chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    )

manhattan_dat$genomic_index <- 1:nrow(manhattan_dat)

ticks <- manhattan_dat %>% 
  group_by(contrast, CHROM) %>% 
  summarise(chr_median = median(genomic_index)) 

```

```{r create_export_genomewide_manhattan, eval = F}
# run this once to create genomewide manhattan and export

# Manhattan of win-fst scores
manhattan_dat %>% 
  #.[sample(nrow(.), 500),] %>%
  ggplot(data = ., mapping = aes(x = genomic_index, y = MEAN_FST, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey","red","blue")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("Mean Window Fst") +
    ggtitle("Windowed Fst - Selected vs Control") +
    geom_hline(filter(q_fst, Quantile %in% c("q99","q95")), mapping = aes(yintercept = Fst, color = Quantile), linetype = "dashed") +
    ggsave(
      filename = "manhattan_genomewide_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
    
# Manhattan of z-win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = z_win_fst_score, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("z(Mean Window Fst)") +
    geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
    geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
    ggtitle("z(Windowed Fst) - Selected vs Control") +
  ggsave(
      filename = "manhattan_genomewide_z_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
    
    
```

```{r import_genomewide_manhattan, out.width = "70%", fig.align = "center"}
# import genomewide manhattan created above
knitr::include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst.png"))

knitr::include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst.png"))

```

```{r create_chr_manhattans, eval = F}

# for fst scores
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst")
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}

# for zfst scores
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_z_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = z_win_fst_score)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst") +
      geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
      geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}

```

### Genomewide mean per SNP Fst
```{r visualize_pairwise_global_fst}
#library(pheatmap)

vis_pairwise_fst <- function(contrast_data){

#contrast_data = global_fst

  pops <- contrast_data$pop1 %>% as.character() %>% c(contrast_data$pop2) %>% unique()

  ma <- matrix(rep(NA, length(pops)*length(pops)), nrow = 6, ncol = 6)
  
  rownames(ma) <- pops
  colnames(ma) <- pops

  for(i in 1:nrow(contrast_data)){
    
    col_i <- contrast_data$pop1[i]
    row_i <- contrast_data$pop2[i]
    
    ma[row_i,col_i] <- contrast_data[i,3] # 3rd col should contain values
    ma[col_i,row_i] <- contrast_data[i,3]
    }

  min_fst <- min(ma, na.rm = T)-0.01
  max_fst <- max(ma, na.rm = T)+0.01
  mid_fst <- min_fst + (max_fst - min_fst)/2
  
  p <- ggcorrplot(ma, lab=T, type = "lower") +
        scale_fill_gradient2(
          limit = c(min_fst,max_fst), low = "blue", high = "red", mid = "white", midpoint = mid_fst
          )
  return(p)
  }

#png(here("00_dashboard/for_manuscript/figure_4A.png"), res = 300, height = 2000, width = 2000, units = "px")
vis_pairwise_fst(global_fst)
#dev.off()

```



Diversity
====================================

Column {.tabset}
-----------------

```{r get_polymorphic_snps_per_population, eval = FALSE}

# Get polymorphic sites and their locations
alt_af_polym <- lapply(
  # get alternative allele frequencies
  list.files(here("batches123_08_LDD_SFS/output"), pattern = "ALTfrq2", full.names = TRUE), 
  function(fl){
    #--fl=list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE)[6]
    # print file name to keep track
    print(fl)
    # load alt frq file
    vroom(fl, col_types = c(CHROM = "c")) %>% 
      # fix columns
      separate(col = POS, sep = " ", into = c("POS", "FRQ"), convert = TRUE) %>% 
      # remove fixed sites
      filter(FRQ > 0 & FRQ < 1)
    }
  )


names(alt_af_polym) <- list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2") %>% 
  str_remove(".ALTfrq2") %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.")


sapply(alt_af_polym, nrow)

alt_af_polym$DU6 %>% head()

alt_af_polym$DU6 %>% tail()

alt_af_polym$DU6$CHROM %>% is.na() %>% any()

```

```{r get_fixed_per_population, eval = FALSE}
# Get polymorphic sites and their locations
alt_af_fixed <- lapply(
  # get alternative allele frequencies
  list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE), 
  function(fl){
    #--fl=list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE)[6]
    # print file name to keep track
    print(fl)
    # load alt frq file
    vroom(fl, col_types = c(CHROM = "c")) %>% 
      # fix columns
      separate(col = POS, sep = " ", into = c("POS", "FRQ"), convert = TRUE) %>% 
      # keep fixed alleles
      filter(FRQ == 1)
    }
  )


names(alt_af_fixed) <- list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2") %>% 
  str_remove(".ALTfrq2") %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.")


alt_af_fixed$DU6$CHROM %>% is.na() %>% any()

```

```{r import_prepare_pi_dat, include=FALSE} 
# define files
fls <- list.files(here("batches123_07_selection_analysis/02_pi_ratio/output"), pattern = ".windowed.pi$", full.names = TRUE)

# import and merge data, ignore n variants
pi_dat <- lapply(fls, function(fl){
  
  pop <- fl %>% 
    basename() %>% 
    str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.") %>% 
    str_remove(".windowed.pi")
  
  d <- fl %>% 
    vroom() %>% 
    #ignore n variants - greedy - to get any region of high diversity, even if there's only one SNP
    dplyr::select(-N_VARIANTS)
  
  names(d)[names(d) == "PI"] <- paste0("PI_",pop)
  
  return(d)
  
  }) %>% 
  purrr::reduce(full_join, by = c("CHROM", "BIN_START","BIN_END"))

# impute NAs with zero
pi_dat[is.na(pi_dat)] <- 0

# reshape data and add zscores

pi_dat <- pi_dat %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END"), value.name = "PI", variable.name = "pop") %>% 
  mutate(pop = str_remove(pop, "PI_")) %>% 
  mutate(is_x = CHROM == "X") %>% 
  group_by(pop, is_x) %>% 
  mutate(z_pi = scale(PI)[,1]) 

head(pi_dat)
```

```{r label_outliers, include = FALSE}

# RTD=regions of highest diversity --> top1%

pi_dat_outliers_labeled <- pi_dat %>% 
  mutate(is_top_1_pct = (z_pi > quantile(z_pi, 0.99)),
         is_top_5_pct = (z_pi > quantile(z_pi, 0.95)))

```


### Intro

* Exploring polymorphic SNPs, in order to characterize the diversity of each population.

* What are the effects for polymorphic SNPs?

* Is there a difference between effects for polymorphic SNPs and fixed SNPs?

* Where are the regions of high diversity located in the genomes of each population?

* What genes appear in regions of high diversity?

### impact_effect_counts
```{r add_snp_effects_tabulate_and_export_table, eval = FALSE}

alt_af_polym_impact_effect_tab <- lapply(alt_af_polym, function(d){
  
  #d=alt_af_polym[[2]] #!!
  #d %>% head() #!!

  d %>% 
    inner_join(
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
        vroom(col_types = c(CHROM = "c")) %>% 
        dplyr::rename(
          effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT"
          ), 
      by = c("CHROM","POS")
    ) %>% 
    group_by(impact, effect) %>% 
    summarise(n=n())
  }
  ) %>% 
  bind_rows(.id="pop") %>% 
  reshape2::dcast(impact + effect ~ pop, value.var = "n")


alt_af_polym_impact_effect_tab %>% 
  write.csv(here("00_dashboard/data/alt_af_polym_impact_effect_tab.csv"))

```

```{r display_alt_af_polym_impact_effect_tab}

read.csv(here("00_dashboard/data/alt_af_polym_impact_effect_tab.csv")) %>% 
  dplyr::select(-X) %>% 
  datatable(caption = "Impact and Effects Polymorphic SNPs", options = list(pageLength = 30))

```

### Correlation_effects_fixed_vs_polymorphic
```{r add_fixed_snp_effects_tabulate_and_export_table, eval = FALSE}
alt_af_fixed_impact_effect_tab <- lapply(alt_af_fixed, function(d){
  
  #d=alt_af_polym[[2]] #!!
  #d %>% head() #!!

  d %>% 
    inner_join(
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
        vroom(col_types = c(CHROM = "c")) %>% 
        dplyr::rename(
          effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT"
          ), 
      by = c("CHROM","POS")
    ) %>% 
    group_by(impact, effect) %>% 
    summarise(n=n())
  }
  ) %>% 
  bind_rows(.id="pop") %>% 
  reshape2::dcast(impact + effect ~ pop, value.var = "n")
```

```{r calculate_pct_effect_and_export_cor_plots, eval = FALSE}

get_pct <- function(d){
  
  tmp <- d %>% 
    dplyr::select(DUK,DUC,DU6,DU6P,DUhLB,FZTDU) %>% 
    lapply(function(x){
  
      #x=alt_af_fixed_impact_effect_tab$DUK
      
      total_cts <- sum(x, na.rm = TRUE)
      
      pct <- (x/total_cts)*100
  
    }) %>% 
    bind_cols()
    
  alt_af_fixed_impact_effect_tab_pct <- bind_cols(
    
    d %>% dplyr::select(impact, effect),
    tmp
  ) 
  
  return(alt_af_fixed_impact_effect_tab_pct)
}

alt_af_polym_impact_effect_pct <- get_pct(alt_af_polym_impact_effect_tab)
alt_af_fixed_impact_effect_tab_pct <- get_pct(alt_af_fixed_impact_effect_tab)


make_export_cor_plot <- function(pop){
  #pop="DUK"
  
  x <- alt_af_polym_impact_effect_pct %>%.[,c("impact", "effect",pop)]
  names(x)[names(x) == pop] <- "pct_polymorphic"
  
  y <- alt_af_fixed_impact_effect_tab_pct %>%.[,c("impact", "effect",pop)]
  names(y)[names(y) == pop] <- "pct_fixed"

  d <- inner_join(x, y, by = c("impact","effect")) %>% 
    mutate(nrows = 1:nrow(.), lab_set = nrows%%2)

  p <- ggplot(d, aes(x = pct_polymorphic, y = pct_fixed)) +
    geom_point() +
    geom_abline(color = "red", slope = 1) +
    theme_bw(base_size = 15) +
    ggrepel::geom_label_repel(data = filter(d, lab_set == 0),  aes(label = effect, fill = impact) , 
                              xlim = c(50,NA), direction = "y") +
    ggrepel::geom_label_repel(data = filter(d, lab_set == 1),  aes(label = effect, fill = impact) , 
                              xlim = c(25,NA), direction = "y") +
    ggtitle(pop) +
    theme(plot.title = element_text(hjust = 0.5))
  
  fl_nm <- paste0("scatter_plot_cor_pct_effects_fixed_vs_polym_",pop,".png")
  
  png(here("00_dashboard/figures", fl_nm), res = 300, width = 3500, height = 3500, units = "px")
  print(p)
  dev.off()
  }


# export each plot 
make_export_cor_plot("DUK")
make_export_cor_plot("DUC")
make_export_cor_plot("DU6")
make_export_cor_plot("DU6P")
make_export_cor_plot("DUhLB")
make_export_cor_plot("FZTDU")
```

```{r display_cor_pct_effect_plots_fixed_vs_polym, out.width="50%", fig.align="center"}

include_graphics(
  list.files(here("00_dashboard/figures"), full.names = TRUE) %>% .[grep("scatter_plot_cor_pct_effects_fixed_vs_polym_", .)]
  )

```

### Genes
```{r tabulate_genes_with_polymorphic_SNPs_and_export, eval=FALSE}

alt_af_polym_gn_cts <- lapply(alt_af_polym, function(d){

  gr <- d %>% 
	  mutate(tmp = POS) %>% 
	  dplyr::rename(seqnames = CHROM, start = tmp, end = POS) %>% 
	  dplyr::select(seqnames, start, end) %>% 
	  makeGRangesFromDataFrame()

  subsetByOverlaps(mmu_gene_set_gr, gr, minoverlap = 1) %>% 
    as.data.frame() %>% 
    group_by(mcols.gene_biotype) %>% 
    summarise(n = n()) %>% 
    janitor::adorn_totals()
  }) %>% 
  bind_rows(.id="pop") %>% 
  dplyr::rename(gene_biotype = mcols.gene_biotype) %>% 
  reshape2::dcast(gene_biotype ~ pop, value.var = "n")

write.csv(alt_af_polym_gn_cts, here("00_dashboard/data/alt_af_polym_gn_cts.csv"))

```

```{r display_genes_with_polymorphic_SNPs_counts}
read.csv(here("00_dashboard/data/alt_af_polym_gn_cts.csv")) %>% 
  dplyr::select(-X) %>% 
  datatable(caption = "Genes Types with Polymorphic SNPs", options = list(pageLength = 100))

```

### Pi_distribution
```{r pi_distribution_histogram, eval = FALSE}

png(here("00_dashboard/figures/histogram_pi_distribution.png"), res = 300, units = "px", height = 2000, width = 2000)

pi_dat %>% 
  ungroup() %>% 
  dplyr::select(-is_x) %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop"), variable.name = "pi_class") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  ggplot(aes(x = value)) +
    geom_histogram() +
    theme_bw(base_size = 15) +
    facet_grid(pop ~ pi_class, scale = "free") +
    ggtitle("Most regions display low diversity") +
    xlab(NULL)

dev.off()

```

```{r zpi_quantiles, include=FALSE}

z_pi_quantiles <- pi_dat %>% 
  summarise(q95 = quantile(z_pi, 0.95),
            q99 = quantile(z_pi, 0.99)) %>% 
  arrange(is_x, pop) %>% 
  mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::select(-is_x) %>% 
  ungroup() %>% 
  reshape2::melt(id.vars = c("pop","chr_class"), variable.name = "quantile", value_name = "quantile_value") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU")))
  
z_pi_quantiles
```

```{r pi_distribution_boxplot_violin, eval=FALSE}

png(here("00_dashboard/figures/violin_boxplot_pi_distribution.png"), res = 300, units = "px",height = 2900, width = 2300)

pi_dat %>% 
  #ungroup() %>% 
  #dplyr::select(-is_x, -PI) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop"), variable.name = "pi_class") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
  ggplot(data = .) + 
    geom_violin(aes(x = z_pi, y = NA), alpha = 0.4, fill = "red", color = "red") +
    geom_boxplot(aes(x = z_pi, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
    ylab(NULL) +
    xlab(NULL) +
    labs(
      title = "Windowed Pi/z_Pi Distribution"
    ) +
    theme_bw(base_size = 15) +
    theme(
      plot.title = element_text(hjust = 0.5),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      strip.text.y = element_text(size = 10)
    ) +
    facet_grid(pop~chr_class, scale = "free") +
    geom_vline(data = z_pi_quantiles, aes(xintercept = value, color = quantile), linetype = "dashed")
    #facet_wrap(~pop, ncol = 1, strip.position = "left")

dev.off()
```

```{r display_pi_distribution_plots, out.width="50%", fig.align="center"}

include_graphics(c(here("00_dashboard/figures/histogram_pi_distribution.png") , 
                   here("00_dashboard/figures/violin_boxplot_pi_distribution.png")))

```

```{r display_extremely_diverse_windows_duk_duc}

x_chr_zpi_above_20 <- pi_dat_outliers_labeled %>% 
  filter(z_pi > 20) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = x_chr_zpi_above_20, subject =  mmu_gene_set_gr, minoverlap = 1)

bind_cols(
  x_chr_zpi_above_20[queryHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(window_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(window_locus, pop, z_pi),
  mmu_gene_set_gr[subjectHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol,  mcols.gene_biotype)
  ) %>% 
  kable(caption = "Genes extreme regions X chr - DUK DUC") %>% 
  kable_styling(full_width = F)

```




### Genomewide_Manhattan
```{r make_and_export_zPi_manhattan, eval = FALSE}

make_z_pi_manhattan <- function(mouse_line){

  #mouse_line="FZTDU"
  
  manhattan_pi_dat <- pi_dat %>% 
    filter(pop == mouse_line) %>% 
    mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
    #--mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
    group_by(pop, CHROM) %>% 
    arrange(pop, CHROM, BIN_START) %>% 
    mutate(
      genomic_index = NA, 
      chr_index = 1:n(), 
      center = BIN_START + (BIN_END - BIN_START)/2,
      center_mb = center/1e6,
      chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    ) %>% 
    mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
    ungroup() %>% 
    group_by(pop, is_x) %>% 
    mutate(is_top_5_pct = (z_pi > quantile(z_pi, 0.95)),
           is_top_1_pct = (z_pi > quantile(z_pi, 0.99))) 
  
  manhattan_pi_dat$genomic_index <- 1:nrow(manhattan_pi_dat)
  
  
  chr_offset_un <- 1000
  len_out <- length(unique(manhattan_pi_dat$CHROM))-1
  chr_offset <- c(0, seq(chr_offset_un, chr_offset_un*len_out, length.out = len_out))
  names(chr_offset) <- unique(manhattan_pi_dat$CHROM)
  
  
  manhattan_pi_dat <- lapply(unique(manhattan_pi_dat$CHROM), function(chr){
    #chr_offset[chr] * 100
    manhattan_pi_dat %>% 
      filter(CHROM == chr) %>% 
      mutate(genomic_index_offset = genomic_index + chr_offset[chr]) 
    }) %>% 
    bind_rows()
  
  ticks <- manhattan_pi_dat %>% 
    group_by(pop, CHROM) %>% 
    summarise(chr_min = min(genomic_index_offset), 
              chr_max = max(genomic_index_offset),
              chr_sep = chr_max + chr_offset_un/2,
              chr_center = chr_min + (chr_max-chr_min)/2) 
  
  
  #d <- manhattan_pi_dat %>% .[sample(1:nrow(.), 1000),]
  d <- manhattan_pi_dat
  
  p <- d %>% 
    filter(!is_top_5_pct & !is_top_1_pct) %>% 
    ggplot(data = ., aes(x = genomic_index_offset, y = z_pi)) +
      geom_point(alpha = 0.5, color = "darkgrey", size = 1) +
    
      geom_point(filter(d, is_top_5_pct & !is_top_1_pct),
                 mapping = aes(x = genomic_index_offset, y = z_pi), 
                 alpha = 0.5, color = "blue", size = 1) +
    
      geom_point(filter(d, is_top_1_pct),
                 mapping = aes(x = genomic_index_offset, y = z_pi), 
                 alpha = 0.5, color = "red", size = 1) +
  
      theme_bw(base_size = 15) +
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank()
      ) +
      scale_x_continuous(breaks = ticks$chr_center, labels = ticks$CHROM, expand = c(0.01,0.01)) + 
      scale_y_continuous(expand = c(0.05,0.05), labels = scales::number_format(accuracy = 1)) +
      geom_vline(filter(ticks, CHROM != "X"), mapping = aes(xintercept = chr_sep), color = "black", linetype = "dashed")  +
      xlab(NULL) +
      ylab("Window zPi") +
      facet_wrap(~pop, strip.position = "right")

    return(p)
  
}


png(here("00_dashboard/figures/manhattan_pi/manhattan_genomewide_zPi.png"), units = "px", res = 300, height = 3000, width = 5000)
grid.arrange(make_z_pi_manhattan("DUK"), make_z_pi_manhattan("DUC"),
             make_z_pi_manhattan("DU6"), make_z_pi_manhattan("DU6P"),
             make_z_pi_manhattan("DUhLB"), make_z_pi_manhattan("FZTDU"),
             ncol = 1)
dev.off()


```

```{r display_genomewide_zPi_manhattan, out.width="50%", fig.align="center"}
include_graphics(here("00_dashboard/figures/manhattan_pi/manhattan_genomewide_zPi.png"))
```

### Chr_Manhattan
```{r make_and_export_zPi_chr_manhattan, eval = FALSE}

make_z_pi_manhattan_chr<- function(chr){
  #--mouse_line="FZTDU"
  #--chr=1
  
  manhattan_pi_dat <- pi_dat %>% 
    filter(CHROM == chr) %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
    mutate(
      center = BIN_START + (BIN_END - BIN_START)/2,
      center_mb = center/1e6
    ) %>% 
    group_by(is_x) %>% 
    mutate(is_top_5_pct = (z_pi > quantile(z_pi, 0.95)),
           is_top_1_pct = (z_pi > quantile(z_pi, 0.99))) 
  
  #d <- manhattan_pi_dat %>% .[sample(1:nrow(.), 1000),]
  d <- manhattan_pi_dat
  
  p <- d %>% 
    filter(!is_top_5_pct & !is_top_1_pct) %>% 
    ggplot(data = ., aes(x = center_mb, y = z_pi)) +
      geom_point(alpha = 0.5, color = "darkgrey", size = 1) +
    
      geom_point(filter(d, is_top_5_pct & !is_top_1_pct),
                 mapping = aes(x = center_mb, y = z_pi), 
                 alpha = 0.5, color = "blue", size = 1) +
    
      geom_point(filter(d, is_top_1_pct),
                 mapping = aes(x = center_mb, y = z_pi), 
                 alpha = 0.5, color = "red", size = 1) +
  
      theme_bw(base_size = 12) +
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank()
      ) +
      scale_y_continuous(expand = c(0.05,0.05), labels = scales::number_format(accuracy = 1)) +
      xlab("Mb") +
      ylab("Window zPi") +
      facet_grid(pop~CHROM)

    return(p)
}


lapply(c(1:19,"X"), function(chr){
  #--chr=1
  fl_nm <- paste0("manhattan_genomewide_zPi_chr", chr,".png")
  png(here("00_dashboard/figures/manhattan_pi", fl_nm), units = "px", res = 300, height = 2000, width = 2500)
  make_z_pi_manhattan_chr(chr) %>% print()
  dev.off()
})

```

```{r display_genomewide_zPi_chr_manhattan, out.width="50%", fig.align="center"}

fls <- sapply(c(1:19,"X"), function(x){
  fl_nm <- paste0("manhattan_genomewide_zPi_chr", x, ".png")
  fl_nm_full <- here("00_dashboard/figures/manhattan_pi", fl_nm)
  return(fl_nm_full)
})

include_graphics(fls)
```



### Genes_RHD
```{r find_genes_in_diversity_outliers_top_1_pct, include = FALSE} 

outliers_gr_top_1_pct <- pi_dat_outliers_labeled %>% 
  ungroup() %>% 
  dplyr::select(CHROM, BIN_START, BIN_END,is_top_1_pct, pop) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop")) %>% 
  filter(is_top_1_pct) %>% 
  dplyr::select(-is_top_1_pct) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov_top_1_pct <- findOverlaps(query = outliers_gr_top_1_pct, subject = mmu_gene_set_gr, minoverlap = 1)


outliers_and_genes_1_pct <- bind_cols(
  outliers_gr_top_1_pct[queryHits(ov_top_1_pct)] %>% 
    as.data.frame() %>% 
    dplyr::select(seqnames, start, end, pop),
  
  mmu_gene_set_gr[subjectHits(ov_top_1_pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol, mcols.gene_biotype) %>% 
    dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, biotype = mcols.gene_biotype) 
  
) 
```

```{r display_gene_counts_in_highest_diverse_regions_top_1_pct}
outliers_and_genes_1_pct %>% 
  ungroup() %>% 
  #filter(outlier_class == "is_top_1_pct") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  group_by(pop, biotype) %>% 
  summarise(n = n()) %>% 
  reshape2::dcast(biotype ~ pop, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  kable(caption = "Gene Types in top 1% most diverse regions") %>% 
  kable_styling(full_width = F)
  #datatable(caption = "Gene Types in top 1% most diverse regions", options = list(pageLength = 100))
```

```{r find_genes_in_diversity_outliers_top_5_pct, include = FALSE}

outliers_gr_top_5_pct <- pi_dat_outliers_labeled %>% 
  ungroup() %>% 
  dplyr::select(CHROM, BIN_START, BIN_END,is_top_5_pct, pop) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop")) %>% 
  filter(is_top_5_pct) %>% 
  dplyr::select(-is_top_5_pct) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov_top_5_pct <- findOverlaps(query = outliers_gr_top_5_pct, subject = mmu_gene_set_gr, minoverlap = 1)


outliers_and_genes_5_pct <- bind_cols(
  outliers_gr_top_5_pct[queryHits(ov_top_5_pct)] %>% 
    as.data.frame() %>% 
    dplyr::select(seqnames, start, end, pop),
  
  mmu_gene_set_gr[subjectHits(ov_top_5_pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol, mcols.gene_biotype) %>% 
    dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, biotype = mcols.gene_biotype) 
  
) 
```

```{r display_gene_counts_in_highest_diverse_regions_top_5_pct}
outliers_and_genes_5_pct %>% 
  ungroup() %>% 
  #filter(outlier_class == "is_top_5_pct") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  group_by(pop, biotype) %>% 
  summarise(n = n()) %>% 
  reshape2::dcast(biotype ~ pop, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  kable(caption = "Gene Types in top 5% most diverse regions") %>% 
  kable_styling(full_width = F)
  #datatable(caption = "Gene Types in top 1% most diverse regions", options = list(pageLength = 100))
```


### RHD_ORA_top1pct 
```{r RHD_GOBP_top1pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP") 
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
)  #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
)
```

```{r RHD_KEGG_top1pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)
```

```{r RHD_GOBP_import_results_top1pct, include=FALSE}

RHD_GOBP_res_top1pct <- here("00_dashboard/data/WebgestaltR_RHD_top1pct/GOBP") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) 


RHD_GOBP_res_top1pct[[4]] # replace this empty element for row binding
RHD_GOBP_res_top1pct[[4]] <- RHD_GOBP_res_top1pct[[1]]
RHD_GOBP_res_top1pct[[4]] <- lapply(RHD_GOBP_res_top1pct[[4]], function(x){
  x <- NA
  }) %>% 
  as.data.frame() %>% 
  mutate(pop = "DUhLB")

RHD_GOBP_res_top1pct <- RHD_GOBP_res_top1pct %>% bind_rows() 


#RHD_GOBP_res %>% datatable(caption = "Enriched GOBP terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))
```

```{r RHD_KEGG_import_results_top1pct, include=FALSE}

RHD_KEGG_res_top1pct <- here("00_dashboard/data/WebgestaltR_RHD_top1pct/KEGG") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) %>% 
  bind_rows()

#RHD_KEGG_res %>% bind_rows() %>% datatable(caption = "Enriched KEGG terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))

```

```{r make_and_export_RHD_ORA_sig_heatmaps_top1pct, eval = FALSE} 
RHD_GOBP_res_sig_top1pct <- RHD_GOBP_res_top1pct %>% 
  mutate(is_sig = (FDR <= 0.1 ) %>% as.integer()) %>% 
  filter(!is.na(description)) %>% 
  reshape2::dcast(description ~ pop, value.var = "is_sig") %>% 
  column_to_rownames(., var = "description")

RHD_GOBP_res_sig_top1pct[is.na(RHD_GOBP_res_sig_top1pct)] <- 0

png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct.png"), res = 300, height = 3000, width = 3000, units = "px")
pheatmap(RHD_GOBP_res_sig_top1pct, main = "RHD_GOBP_res_sig_top1pct")
dev.off()

RHD_KEGG_res_sig_top1pct <- RHD_KEGG_res_top1pct %>% 
  mutate(is_sig = (FDR <= 0.1 ) %>% as.integer()) %>% 
  filter(!is.na(description)) %>% 
  reshape2::dcast(description ~ pop, value.var = "is_sig") %>% 
  column_to_rownames(., var = "description") 

RHD_KEGG_res_sig_top1pct[is.na(RHD_KEGG_res_sig_top1pct)] <- 0

png(here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct.png"), res = 300, height = 3000, width = 3000, units = "px")
pheatmap(RHD_KEGG_res_sig_top1pct, main = "RHD_KEGG_res_sig_top1pct")
dev.off()
```

```{r RHD_GOBP_FDR1, eval=FALSE}

RHD_GOBP_res_sig_top1pct_terms <- RHD_GOBP_res_sig_top1pct %>% rownames()

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
)

setdiff(
  RHD_GOBP_res_sig_top1pct_terms,
  here(
    "00_dashboard/data/WebGestaltR_RHD_lenient/GOBP/Project_WebGestaltR_FZTDU/geneontology_Biological_Process_sig_results.csv"
    ) %>% 
    vroom() %>%
    .$description
  )



```

```{r make_heatmap_enrichmentRatio_GOBP, eval = FALSE}

dir_nm <- here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")


prep_ora_dat <- function(pop){
  #pop="DUK"

  nm <- paste0("Project_WebGestaltR_",pop)

  d <- file.path(dir_nm, nm, "geneontology_Biological_Process_sig_results.csv") %>% 
    vroom() %>% 
    dplyr::select(description, enrichmentRatio) %>% 
    filter(description %in% RHD_GOBP_res_sig_top1pct_terms)
  
  names(d)[names(d) == "enrichmentRatio"] <- pop
  
  d
  
}

png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct_enrichmentRatio.png"), 
    res = 300, height = 3000, width = 3000, units = "px")


full_join(prep_ora_dat("DUK"),prep_ora_dat("DUC"),by = "description") %>% 
  full_join(prep_ora_dat("DU6"), by = "description") %>% 
  full_join(prep_ora_dat("DU6P"), by = "description") %>% 
  full_join(prep_ora_dat("DUhLB"), by = "description") %>% 
  full_join(prep_ora_dat("FZTDU"), by = "description") %>% 
  reshape2::melt(id.vars = "description") %>% 
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  reshape2::dcast(description ~ variable, value.var = "value") %>% 
  column_to_rownames(var = "description") %>% 
  #as.matrix() %>% 
  pheatmap(., main = "enrichmentRatio - GOBP")
  
dev.off()



```

```{r RHD_KEGG_FDR1, eval=FALSE}

RHD_KEGG_res_sig_top1pct_terms <- RHD_KEGG_res_sig_top1pct %>% rownames()

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
)

```


```{r make_heatmap_enrichmentRatio_KEGG, eval = FALSE}

dir_nm <- here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")


prep_ora_dat <- function(pop){
  #pop="DUK"

  nm <- paste0("Project_WebGestaltR_",pop)

  d <- file.path(dir_nm, nm, "pathway_KEGG_sig_results.csv") %>% 
    vroom() %>% 
    dplyr::select(description, enrichmentRatio) %>% 
    filter(description %in% RHD_KEGG_res_sig_top1pct_terms)
  
  names(d)[names(d) == "enrichmentRatio"] <- pop
  
  d
  
}

png(here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct_enrichmentRatio.png"), 
    res = 300, height = 3000, width = 3000, units = "px")


full_join(prep_ora_dat("DUK"),prep_ora_dat("DUC"),by = "description") %>% 
  full_join(prep_ora_dat("DU6"), by = "description") %>% 
  full_join(prep_ora_dat("DU6P"), by = "description") %>% 
  full_join(prep_ora_dat("DUhLB"), by = "description") %>% 
  full_join(prep_ora_dat("FZTDU"), by = "description") %>% 
  reshape2::melt(id.vars = "description") %>% 
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  reshape2::dcast(description ~ variable, value.var = "value") %>% 
  column_to_rownames(var = "description") %>% 
  #as.matrix() %>% 
  pheatmap(., main = "enrichmentRatio - KEGG")
  
dev.off()



```



```{r display_RHD_ORA_sig_heatmaps_top1pct, out.width="50%", fig.align="center"}

include_graphics(
  c(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct.png") , 
    here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct.png"),
    here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct_enrichmentRatio.png"),
    here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct_enrichmentRatio.png"))
  )

```

#### **Description**

Overrepresentation analysis of RDD genes. 

First two figures represent GOBP and KEGG results. In there, terms are either 1 (signficant) or 0 (not-significant).

Figures 3 and 4 show enrichment scores accross all significant terms shown in figures 1 and 2. Enrichment scores were collected by re-running ORA setting FDR=1 and then subsetting by the score already detected as sifnificant (FDR <= 0.1).



### RHD_ORA_top5pct
```{r RHD_GOBP_top5pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
)
```

```{r RHD_KEGG_top5pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

```

```{r RHD_GOBP_import_results_top5pct, include=FALSE}

RHD_GOBP_res_top5pct <- here("00_dashboard/data/WebgestaltR_RHD_top5pct/GOBP") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) %>% 
  bind_rows() 


#RHD_GOBP_res %>% datatable(caption = "Enriched GOBP terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))
```

```{r RHD_KEGG_import_results_top5pct, include=FALSE}

RHD_KEGG_res_top5pct <- here("00_dashboard/data/WebgestaltR_RHD_top5pct/KEGG") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  })

RHD_KEGG_res_top5pct[[3]] # replace this empty element for row binding
RHD_KEGG_res_top5pct[[3]] <- RHD_KEGG_res_top5pct[[1]]
RHD_KEGG_res_top5pct[[3]] <- lapply(RHD_KEGG_res_top5pct[[3]], function(x){
  x <- NA
  }) %>% 
  as.data.frame() %>% 
  mutate(pop = "DUC")

RHD_KEGG_res_top5pct <- RHD_KEGG_res_top5pct %>% bind_rows() 

#RHD_KEGG_res %>% bind_rows() %>% datatable(caption = "Enriched KEGG terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))

```

```{r find_common_GOBP_terms, include=FALSE}

RHD_GOBP_res_top5pct_common <- RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUK"] %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUC"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DU6"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DU6P"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUhLB"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "FZTDU"]) 

```

```{r find_common_KEGG_terms, eval = FALSE}
RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUK"] %>% 
  #intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUC"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DU6"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DU6P"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUhLB"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "FZTDU"]) %>% 
  length() # no intersection
```

```{r heatmap_common_GOBP, eval=FALSE}
png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top5pct_common.png"), res = 300, height = 1500, width = 2000, units = "px")

RHD_GOBP_res_top5pct %>% 
  filter(description %in% RHD_GOBP_res_top5pct_common) %>% 
  filter(FDR < 0.1) %>% 
  reshape2::dcast(description~pop, value.var = "overlap") %>% 
  column_to_rownames(., var = "description") %>% 
  as.matrix() %>% 
  pheatmap(., main = "RHD_GOBP_res_top5pct_common (n genes)", display_numbers = TRUE, number_format = "%.0f")

dev.off()
```

```{r display_heatmap_common_GOBP, out.width="50%", fig.align="center"}

include_graphics(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top5pct_common.png"))
```

###--End




ROH
=========================

Column {.tabset}
------------------

```{r ROH_import_regions, include=FALSE}

roh_dat <- vroom(here("batches123_09_ROH/output/ROH_regions.table")) %>% 
  dplyr::rename(RG = `# RG`, sample = `[2]Sample`, chr = `[3]Chromosome`, start = `[4]Start`, end = `[5]End`, length_bp = `[6]Length (bp)`,
                n_markers = `[7]Number of markers`, qual_avg_fwd_bwd_phred = `[8]Quality (average fwd-bwd phred score)`) %>% 
  filter(RG != "# RG") %>% 
  mutate(sample = str_remove(sample, "-L1")) %>% 
  inner_join(
    
    # include an ordered-naming for later when plotting all samples together (i.e. boxplot)
    sample_info %>% 
      mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6", "DU6P", "DUhLB", "FZTDU"))) %>% 
      arrange(Linie, sample_id) %>% 
      mutate(   sample_order = paste0(    "sample",   seq(101,nrow(.)+100, 1)   )    ) ,
    
    by = c("sample" = "sample_id")) %>% 
  dplyr::rename(pop = Linie) %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6", "DU6P", "DUhLB", "FZTDU")))


dim(roh_dat) #350878      11
```

### Intro

* Runs of homozygosity (ROHs) were calculated with bcftools 1.9 (roh)


### Summary
```{r summarise_roh_length_cohort}

roh_dat %>% 
  group_by(pop) %>% 
  summarise(min = min(length_bp),q25 = quantile(length_bp, 0.25), median = median(length_bp), 
            mean = mean(length_bp), q75 = quantile(length_bp, 0.75), max = max(length_bp),
            sd = sd(length_bp)) %>% 
  kable(format.args = list(big.mark = ","), caption = "Length (bp) Summary by Population", digits = 0) %>% 
  kable_styling(full_width = F)
```

### hist_length
```{r hist_check_roh_length_bp_distribution, fig.width=10, fig.height=10}

ggplot(roh_dat, aes(x = length_bp)) +
  geom_histogram() +
  facet_wrap(~pop) +
  theme_bw(base_size = 12) +
  ggtitle("ROH length distribution by population") +
  theme(plot.title = element_text(hjust = 0.5))
```

### boxplot_length_by_pop
```{r boxplot_check_roh_length_bp_distribution, fig.width=10, fig.height=10}

ggplot(roh_dat, aes(y = pop, x = length_bp, fill = pop)) +
  geom_boxplot() +
  theme_bw(base_size = 12) +
  ggtitle("ROH length distribution by population") +
  theme(plot.title = element_text(hjust = 0.5))
```

### boxplot_length_by_sample
```{r boxplot_check_roh_length_bp_distribution_by_samp, fig.width=10, fig.height=10}
roh_dat %>% 
  mutate(pop = factor(pop, levels = rev(c("DUK","DUC","DU6", "DU6P", "DUhLB", "FZTDU")))) %>% 
  ggplot(aes(y = sample_order, x = length_bp/1000, fill = pop)) +
    geom_boxplot(outlier.size = 0.1, outlier.colour = "red") +
    theme_bw(base_size = 15) +
    ggtitle("ROH length distribution by sample") +
    theme(plot.title = element_text(hjust = 0.5),axis.text.y = element_blank()) +
    ylab("samples") +
    xlab("length (Kbp)")

```


### genome_fractions
```{r per_sample_avg_genome_prop_as_roh}

fractions <- bind_rows(
  
  # compute total fractions
  roh_dat %>% 
    group_by(pop,sample) %>% 
    summarise(per_sample_genome_fraction = sum(length_bp)/genome_length) %>% 
    ungroup() %>% 
    group_by(pop) %>% 
    summarise(mean_per_sample_genome_fraction = mean(per_sample_genome_fraction),
              sd_per_sample_genome_fraction = sd(per_sample_genome_fraction)) %>% 
    mutate(data_set = "total"),

  # compute fractions only for ROHs > 2Mbp
  roh_dat %>% 
    filter(length_bp > 1e6) %>% 
    group_by(pop,sample) %>% 
    summarise(per_sample_genome_fraction = sum(length_bp)/genome_length) %>% 
    ungroup() %>% 
    group_by(pop) %>% 
    summarise(mean_per_sample_genome_fraction = mean(per_sample_genome_fraction),
              sd_per_sample_genome_fraction = sd(per_sample_genome_fraction)) %>% 
    mutate(data_set = "at_least_1Mbp"),

  # compute fractions only for ROHs > 2Mbp
  roh_dat %>% 
    filter(length_bp > 2e6) %>% 
    group_by(pop,sample) %>% 
    summarise(per_sample_genome_fraction = sum(length_bp)/genome_length) %>% 
    ungroup() %>% 
    group_by(pop) %>% 
    summarise(mean_per_sample_genome_fraction = mean(per_sample_genome_fraction),
              sd_per_sample_genome_fraction = sd(per_sample_genome_fraction)) %>% 
    mutate(data_set = "at_least_2Mbp"),
  
  # compute fractions only for ROHs > 3Mbp
  roh_dat %>% 
    filter(length_bp > 3e6) %>% 
    group_by(pop,sample) %>% 
    summarise(per_sample_genome_fraction = sum(length_bp)/genome_length) %>% 
    ungroup() %>% 
    group_by(pop) %>% 
    summarise(mean_per_sample_genome_fraction = mean(per_sample_genome_fraction),
              sd_per_sample_genome_fraction = sd(per_sample_genome_fraction)) %>% 
    mutate(data_set = "at_least_3Mbp")
  
  ) %>% 
  mutate(data_set = factor(data_set, levels = c("total","at_least_1Mbp","at_least_2Mbp","at_least_3Mbp"))) %>% 
  mutate(errors_lower = mean_per_sample_genome_fraction - sd_per_sample_genome_fraction,
         errors_upper = mean_per_sample_genome_fraction + sd_per_sample_genome_fraction)


ggplot(fractions, aes(x = pop, y = mean_per_sample_genome_fraction, fill = data_set)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(aes(ymin = errors_lower, ymax = errors_upper), position =  position_dodge2(width = 0.5, padding = 0.5)) +
  theme_bw(base_size = 12) +
  xlab(NULL) +
  ylab("genome fraction") +
  ggtitle("Mean per-sample genome fractions as ROHs") +
  scale_fill_brewer(palette="Dark2") +
  scale_y_continuous(breaks = seq(0,1,0.05), expand = c(0.01,0.01)) +
  theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5))

```


REDs {data-navmenu="REDs"}
====================================

Column {.tabset}
----------------

### Overview

* Regions of Extreme Differentiation (REDs)

* Corresponding to the top 1% Fst windows of each contrast.

* Genes overlapping those windows were identified

### Mb

```{r get_REDs_and_RED_genes}
get_REDs <- function(contr){
  #contr="DU6_vs_FZTDU"
  
  #-----------------------
  # prepare q99 intervals
  #-----------------------
  # subset data q99
  ss1 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr1 <- GRanges(seqnames = ss1$CHROM, IRanges(start = ss1$BIN_START, end = ss1$BIN_END))
  
  # merge adjacent overlapping ranges
  gr_red1 <- GenomicRanges::reduce(gr1, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  #-----------------------------------------
  # prepare <99th - 95th quantile intervals
  #-----------------------------------------
  # subset data <99th - 95th
  ss2 <- win_fst_sel_vs_ctrl_prep %>% 
    filter(contrast == contr & is_95th_quantile & !is_99th_quantile) %>% 
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)
  
  # convert to granges
  gr2 <- GRanges(seqnames = ss2$CHROM, IRanges(start = ss2$BIN_START, end = ss2$BIN_END))

  # merge adjacent overlapping ranges
  gr_red2 <- GenomicRanges::reduce(gr2, min.gapwidth=1L) #do not merge beyond 1bp gap
  
  # Only keep <99th - 95th regions adjacent-overlapping q99 regions
  gr_red2_ov <- subsetByOverlaps(gr_red2, gr_red1, maxgap = 1)
  
  #-----------------------------------------------------------
  # combine both quantile sets and merge
  #-----------------------------------------------------------
  # Combine q99 and overlapping/adjacent <99th - 95th regions
  res <- c(gr_red1, gr_red2_ov) %>% sort()
    
  # merge adjacent/overlapping regions
  res <- reduce(res, min.gapwidth=1L)
  return(res)
  }

# apply function for each contrast
REDs <- lapply(unique(win_fst_sel_vs_ctrl_prep$contrast), function(contr) get_REDs(contr = contr))
names(REDs) <- unique(win_fst_sel_vs_ctrl_prep$contrast) # add contrast name
```

```{r display_REDs_Mb_tables}
# Summarize Mb per contrast and chromosome
tab <- lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% 
    as.data.frame() %>% 
    group_by(seqnames) %>% 
    summarise(Mb = sum(width_Mb)) %>% 
    mutate(seqnames = factor(seqnames, levels = c(1:19,"X"))) %>% 
    arrange(seqnames)
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(seqnames ~ contrast, value.var = "Mb") %>% 
  janitor::adorn_totals("row") %>% 
  dplyr::rename(CHROM = seqnames) %>% 
  dplyr::select(CHROM, DUK_vs_FZTDU, DUC_vs_FZTDU, DU6_vs_FZTDU, DU6P_vs_FZTDU, DUhLB_vs_FZTDU, FERT_vs_FZTDU)

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_17.csv"))

# Summarize Mb per contrast
lapply(REDs, function(x){
  x$width_Mb <- width(x)/1e6
  x %>% as.data.frame() %>% summarise(Mb = sum(width_Mb))
}) %>% bind_rows(.id = "contrast") %>% 
  reshape2::dcast(. ~ contrast, value.var = "Mb") %>% 
  .[,-1] %>% 
  kable(caption = "Mb Regions of Extreme Differentiation per contrast") %>% 
  kable_styling(full_width = F)

tab %>% 
  #kable(caption = "Mb Regions of Extreme Differentiation per chromosome") %>% 
  #kable_styling(full_width = F)
  datatable(caption = "Mb Regions of Extreme Differentiation per chromosome", 
            options = list(pageLength = 21))


```

### Genes
```{r count_the_number_of_SNPs_in_REDs}
# count total number of overlaps
lapply(1:length(REDs), function(i){ 
  
  # get gr for contrast
  gr <- REDs[[i]]
  
  # count number of overlaps between SNPs and REDs.
  cts <- countOverlaps(query = vcf_final_snp_ids_gr, subject = gr, minoverlap = 1) %>% 
          # remove SNPs wihout overlas
          .[-grep( 0, .)] %>% 
          # count how many SNPs overlap REDs
          length()
  data.frame(contrast = names(REDs)[i], Nr_RED_SNPs = cts)
  }
  ) %>% 
  bind_rows() %>% 
  mutate(contrast = factor(contrast, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU"))) %>% 
  arrange(contrast) %>% 
  kable(digits = 2, caption = "Number of SNPs in REDs",format.args = list(big.mark = ",")) %>% kable_styling(full_width = F)


# count overlaps by chromosome
lapply(1:length(REDs), function(i){ 

  # get gr for contrast
  gr <- REDs[[i]]

  ov <- findOverlaps(query = vcf_final_snp_ids_gr, subject = gr, minoverlap = 1)
  
  vcf_final_snp_ids_gr[queryHits(ov)] %>% 
    as.data.frame() %>% 
    group_by(seqnames) %>% 
    summarise(n_snps = n()) %>% 
    mutate(seqnames = factor(seqnames, levels = c(1:19,"X")), contrast = names(REDs)[i]) %>% 
    arrange(seqnames) %>% 
    dplyr::select(contrast, everything()) %>% 
    dplyr::rename(chr = seqnames)

    }
  ) %>% 
  bind_rows() %>% 
  mutate(contrast = factor(contrast, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU"))) %>% 
  arrange(contrast) %>% 
  reshape2::dcast(chr ~ contrast,value.var = "n_snps") %>% 
  kable(digits = 2, caption = "Number of SNPs in REDs by chr",format.args = list(big.mark = ",")) %>% kable_styling(full_width = F)

```

```{r RED_genes_counts}

RED_genes <- lapply(REDs, function(x) subsetByOverlaps(mmu_gene_set_gr, x, maxgap = 1))

# Number of features per contrast
RED_genes %>% 
  lapply(length) %>% 
  bind_rows() %>% 
  kable(caption = "All features in Mmu ensembl gene set") %>% 
  kable_styling(full_width = F)

# Number of protein_coding per contrast
lapply(RED_genes, function(x) x[x$mcols.gene_biotype == "protein_coding"]) %>% 
  lapply(length) %>% 
  bind_rows() %>% 
  kable(caption = "Protein coding Genes in Mmu ensembl gene set") %>% 
  kable_styling(full_width = F)

# break down by biotype
tab <- lapply(RED_genes, function(x){
  as.data.frame(x) %>% 
  group_by(mcols.gene_biotype) %>% 
  summarise(n = n())
  }) %>% 
  bind_rows(.id = "contrast") %>% 
  reshape2::dcast(mcols.gene_biotype ~ contrast, value.var = "n") %>% 
  dplyr::rename(gene_biotype = mcols.gene_biotype) 

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_5.csv"))
tab %>% 
  kable(caption = "Features by biotype in Mmu ensembl gene set") %>% 
  kable_styling(full_width = F)


```

```{r add_GOBP_and_summarize_terms_and_export, eval = FALSE}

# biomart query done on Monday 26/10/2020
#mart <- useDataset(dataset = "mmusculus_gene_ensembl", mart = useMart("ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org"))


RED_genes2 <- lapply(1:length(RED_genes), function(i){
  
  #i=1
  
  x <- RED_genes[[i]]
  
  # do BM quary
  mart_res <- getBM(attributes = c("ensembl_gene_id","mgi_symbol","name_1006","namespace_1003"), 
                  filters = "ensembl_gene_id", 
                  values = x$mcols.gene_id,
                  mart = mart) %>% 
              filter(namespace_1003 == "biological_process") 

  # collapse GOBP terms
  mart_res$ensembl_gene_id %>% 
    unique() %>% 
    lapply(function(y){
      
      gobps <- mart_res %>% 
        filter(ensembl_gene_id == y) %>%
        dplyr::select(name_1006) %>% 
        unlist() %>% 
        paste(collapse = "; ")
      
      data.frame(
        ensembl_gene_id = y,
        mgi_symbol = mart_res$mgi_symbol[mart_res$ensembl_gene_id == y] %>% unique(), 
        GOBP = gobps
        )
    }) %>% 
    bind_rows() %>% 
    full_join(
      x %>% as.data.frame() %>% dplyr::rename(ensembl_gene_id = mcols.gene_id, gene_biotype = mcols.gene_biotype),
      by = "ensembl_gene_id"
    ) %>% 
    mutate(contrast = names(RED_genes)[i] ) %>% 
    dplyr::select(contrast, ensembl_gene_id,mgi_symbol,seqnames, start, end, width, strand, gene_biotype, GOBP) 
})


names(RED_genes2) <- names(RED_genes)

# export for later use offline (biomart query not needed)

RED_genes2$DUK_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_DUK.csv"))
RED_genes2$DUC_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_DUC.csv"))
RED_genes2$DU6_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_DU6.csv"))
RED_genes2$DU6P_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_DU6P.csv"))
RED_genes2$DUhLB_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_DUhLB.csv"))
RED_genes2$FERT_vs_FZTDU %>% write.csv(here("00_dashboard/data/RED_genes/RED_genes_FERT.csv"))

```

### DUK
```{r display_DUK_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_DUK.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUC
```{r display_DUC_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_DUC.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6
```{r display_DU6_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_DU6.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6P
```{r display_DU6P_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_DU6P.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUhLB
```{r display_DUhLB_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_DUhLB.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### FERT
```{r display_FERT_RED_genes_information}
here("00_dashboard/data/RED_genes/RED_genes_FERT.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```



### Gene-mean-Fst
```{r Export_table_of_RED_genes_with_mean_fst_for_manual_exploration, eval = FALSE}

get_gene_mean_fst <- function(cont_tab, fst_fl){

  #cont_tab=RED_genes$DUK_vs_FZTDU
  #fst_fl="DUK_vs_FZTDU.weir.fst"

  # import per SNP Fst scores and 
  fst_tab <- vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fst_fl)) %>% 
    #fix chrX that is changed to NA (not sure why)
    mutate(CHROM = ifelse(is.na(CHROM), "X", CHROM))
  
  # turn into granges
  fst_gr <- GRanges(seqnames = fst_tab$CHROM, IRanges(start = fst_tab$POS, end = fst_tab$POS)) # locus
  fst_gr$WEIR_AND_COCKERHAM_FST <- fst_tab$WEIR_AND_COCKERHAM_FST # fst scores
  
  # calculate gene-mean Fst and export
  gn_mean_df <- lapply(1:length(cont_tab), function(i){
  
  #lapply(1:10, function(i){
    
    gn <- cont_tab[i] # get gene entry
  
    ov <- findOverlaps(query = gn, subject = fst_gr) # find snps in gene
    
    gn_snp_fst <- fst_gr[subjectHits(ov)]$WEIR_AND_COCKERHAM_FST # add per snp fst
    
    # combine gene information and mean-gene fst
    data.frame(gene_locus = paste0(as.character(seqnames(gn)), ":", start(gn), "-", end(gn) ),
               width_bp = width(gn),
               gene_id = gn$mcols.gene_id, 
               gene_biotype = gn$mcols.gene_biotype, 
               n_snp = length(gn_snp_fst), 
               mean_fst = mean(gn_snp_fst, na.rm=TRUE))
    }) %>% 
    bind_rows()
  
  # prepare name of file
  fl_nm <- paste0("RED_gene_mean_fst_", str_remove(fst_fl,".weir.fst"),".csv")
  
  # export file
  write.csv(gn_mean_df, here("00_dashboard/data",fl_nm))
}

#rm(cont_tab, fst_tab, fst_gr, gn, gn_mean_df, fl_nm)

get_gene_mean_fst(cont_tab=RED_genes$DUK_vs_FZTDU, fst_fl="DUK_vs_FZTDU.weir.fst")
get_gene_mean_fst(cont_tab=RED_genes$DUC_vs_FZTDU, fst_fl="DUC_vs_FZTDU.weir.fst")
get_gene_mean_fst(cont_tab=RED_genes$DU6_vs_FZTDU, fst_fl="DU6_vs_FZTDU.weir.fst")
get_gene_mean_fst(cont_tab=RED_genes$DU6P_vs_FZTDU, fst_fl="DU6P_vs_FZTDU.weir.fst")
get_gene_mean_fst(cont_tab=RED_genes$DUhLB_vs_FZTDU, fst_fl="DUhLB_vs_FZTDU.weir.fst")
get_gene_mean_fst(cont_tab=RED_genes$FERT_vs_FZTDU, fst_fl="FERT_vs_FZTDU.weir.fst")

```

```{r matrix_mean_fst_all_RED_genes_at_each_contrast, eval=FALSE}

# generate_mean_REDgene_fst_matrix

# combine all red genes of all contrast 
RED_genes_all <- c(RED_genes$DUK_vs_FZTDU, 
  RED_genes$DUC_vs_FZTDU, 
  RED_genes$DU6_vs_FZTDU, 
  RED_genes$DU6P_vs_FZTDU, 
  RED_genes$DUhLB_vs_FZTDU, 
  RED_genes$FERT_vs_FZTDU) %>% 
  unique() # total = 3751 genes

# make a function to calculate the mean fst of each gene for a given contrast
get_gene_mean_fst2 <- function(cont_tab, fst_fl){

  #cont_tab=RED_genes_all[1:10]
  #fst_fl="DUK_vs_FZTDU.weir.fst"

  # import per SNP Fst scores and 
  fst_tab <- vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fst_fl)) %>% 
    #fix chrX that is changed to NA (not sure why)
    mutate(CHROM = ifelse(is.na(CHROM), "X", CHROM))
  
  # turn into granges
  fst_gr <- GRanges(seqnames = fst_tab$CHROM, IRanges(start = fst_tab$POS, end = fst_tab$POS)) # locus
  fst_gr$WEIR_AND_COCKERHAM_FST <- fst_tab$WEIR_AND_COCKERHAM_FST # fst scores
  
  # calculate gene-mean Fst
  gn_mean_df <- lapply(1:length(cont_tab), function(i){
  
    gn <- cont_tab[i] # get gene entry
  
    ov <- findOverlaps(query = gn, subject = fst_gr) # find snps in gene
    
    gn_snp_fst <- fst_gr[subjectHits(ov)]$WEIR_AND_COCKERHAM_FST # add per snp fst
    
    # combine gene information and mean-gene fst into a data frame
    res <- data.frame(gene_id = gn$mcols.gene_id, 
               mean_fst = mean(gn_snp_fst, na.rm=TRUE))
    }) %>% 
    # stack entries into one data frame
    bind_rows()
  
    # specify contrast information
    names(gn_mean_df)[names(gn_mean_df) == "mean_fst"] <- paste0(str_remove(fst_fl,".weir.fst"),"_mean_fst")
    
    return(gn_mean_df)
}

# calculate mean fst for each RED gene at each contrast
RED_genes_all_mean_fst <- list(
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="DUK_vs_FZTDU.weir.fst"),
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="DUC_vs_FZTDU.weir.fst"),
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="DU6_vs_FZTDU.weir.fst"),
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="DU6P_vs_FZTDU.weir.fst"),
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="DUhLB_vs_FZTDU.weir.fst"),
  get_gene_mean_fst2(cont_tab=RED_genes_all, fst_fl="FERT_vs_FZTDU.weir.fst")
  ) %>% 
  purrr::reduce(inner_join, by = "gene_id") #dim = 3751    7

# export this table to avoid running this part again (it takes too long to complete)
write.csv(RED_genes_all_mean_fst, here("00_dashboard/data/RED_genes/RED_gene_mean_fst_all_contrasts.csv"))


```

```{r display_Fst_RED_genes}
read.csv(
  here("00_dashboard/data/RED_genes/RED_gene_mean_fst_all_contrasts.csv")
  ) %>% 
  dplyr::select(-X) %>% 
  datatable(
    #options = list(pageLength = 300), 
    caption = "Mean Fst of SNPs in RED Genes", 
    rownames = FALSE, 
    filter = "top"
    ) %>% 
  formatRound(
    paste0(c("DUK","DUC","DU6P","DU6","DUhLB","FERT"),"_vs_FZTDU_mean_fst"),
    2
    )
```

### Gene-NrSNPs
```{r get_number_of_snps_for_RED_genes, eval = FALSE}
# combine all red genes of all contrast 
RED_genes_all <- c(RED_genes$DUK_vs_FZTDU, 
  RED_genes$DUC_vs_FZTDU, 
  RED_genes$DU6_vs_FZTDU, 
  RED_genes$DU6P_vs_FZTDU, 
  RED_genes$DUhLB_vs_FZTDU, 
  RED_genes$FERT_vs_FZTDU) %>% 
  unique() # total = 3751 genes

# make a function to calculate the mean fst of each gene for a given contrast
get_gene_n_snps <- function(cont_tab, fst_fl){ 
  
  # this function counts the number of SNPs that are not NAs for Fst calculation.
  # Those are the "active" SNPs used to calculte the gene's mean Fst

  #cont_tab=RED_genes_all[1:10]
  #fst_fl="DUK_vs_FZTDU.weir.fst"

  # import per SNP Fst scores and 
  fst_tab <- vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fst_fl)) %>% 
    #fix chrX that is changed to NA (not sure why)
    mutate(CHROM = ifelse(is.na(CHROM), "X", CHROM))
  
  # turn into granges
  fst_gr <- GRanges(seqnames = fst_tab$CHROM, IRanges(start = fst_tab$POS, end = fst_tab$POS)) # locus
  fst_gr$WEIR_AND_COCKERHAM_FST <- fst_tab$WEIR_AND_COCKERHAM_FST # fst scores
  
  # calculate gene-mean Fst
  gn_nsnp_df <- lapply(1:length(cont_tab), function(i){
  
    gn <- cont_tab[i] # get gene entry
  
    ov <- findOverlaps(query = gn, subject = fst_gr) # find snps in gene
    
    gn_snp_fst <- fst_gr[subjectHits(ov)]$WEIR_AND_COCKERHAM_FST # add per snp fst
    
    # remove entries with NAs and count number of SNPs "not NAs"
    n_snps <- gn_snp_fst[!is.na(gn_snp_fst)] %>% length()
    
    # combine gene information and snp number into a data frame
    res <- data.frame(gene_id = gn$mcols.gene_id, 
                      snp_nr = n_snps)
    }) %>% 
    # stack entries into one data frame
    bind_rows()
  
    # specify contrast information
    names(gn_nsnp_df)[names(gn_nsnp_df) == "snp_nr"] <- paste0(str_remove(fst_fl,".weir.fst"),"_n_snps")
    
    return(gn_nsnp_df)
}

    
# calculate mean fst for each RED gene at each contrast
RED_genes_all_nsnps <- list(
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="DUK_vs_FZTDU.weir.fst"),
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="DUC_vs_FZTDU.weir.fst"),
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="DU6_vs_FZTDU.weir.fst"),
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="DU6P_vs_FZTDU.weir.fst"),
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="DUhLB_vs_FZTDU.weir.fst"),
  get_gene_n_snps(cont_tab=RED_genes_all, fst_fl="FERT_vs_FZTDU.weir.fst")
  ) %>% 
  purrr::reduce(inner_join, by = "gene_id") #dim = 3751    7

# export this table to avoid running this part again (it takes too long to complete)
write.csv(RED_genes_all_nsnps, here("00_dashboard/data/RED_genes/RED_gene_nsnp_all_contrasts.csv"))
```

```{r display_nsnps_RED_genes}
read.csv(
  here("00_dashboard/data/RED_genes/RED_gene_nsnp_all_contrasts.csv")
  ) %>% 
  dplyr::select(-X) %>% 
  datatable(
    #options = list(pageLength = 300), 
    caption = "Number of SNPs in RED Genes (used for mean gene calculation)", 
    rownames = FALSE, 
    filter = "top"
    )

```

### Gene-SNPeff
```{r add_SNPeff_gene_table_for_RED_genes}
lapply(1:length(RED_genes), function(i){
  gn_ids <- RED_genes[[i]]$mcols.gene_id %>% unique() %>% str_trim()
  # import snpeff gene table
  vroom(here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.summary.genes.txt"),skip = 1) %>% 
    # subset snpeff table by RED-genes
    filter(GeneId %in% gn_ids) %>% 
    mutate(contrast = names(RED_genes)[i]) %>% 
    dplyr::select(contrast, everything()) %>% 
    dplyr::rename(GeneName = `#GeneName`) 
}) %>% 
  bind_rows() %>% 
  datatable(
    options = list(pageLength = 20), 
    caption = "SNPeff summary annotations for SNPs in RED Genes", 
    rownames = FALSE,
    filter = "top"
    )
  
```

### Vis-Distinct-genes
```{r find_distinctly_differentiated_RED_genes}
# import matrix-like table with RED-gene mean fst
ma <- read.csv(here("00_dashboard/data/RED_genes/RED_gene_mean_fst_all_contrasts.csv")) %>% dplyr::select(-X)

# Add rownames as gene ids
rownames(ma) <- ma$gene_id
ma <- ma %>% dplyr::select(-gene_id) %>% as.matrix()

# find genes distinctly differentiated
gns <- list(
  duk = get_distinct_genes(ma = ma,cont1 = "DUK_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("DUK_vs_FZTDU_mean_fst","FERT_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1),
  duc = get_distinct_genes(ma = ma,cont1 = "DUC_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("DUC_vs_FZTDU_mean_fst","FERT_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1),
  du6 = get_distinct_genes(ma = ma,cont1 = "DU6_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("DU6_vs_FZTDU_mean_fst","FERT_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1),
  du6p = get_distinct_genes(ma = ma,cont1 = "DU6P_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("DU6P_vs_FZTDU_mean_fst","FERT_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1),
  duhlb = get_distinct_genes(ma = ma,cont1 = "DUhLB_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("DUhLB_vs_FZTDU_mean_fst","FERT_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1),
  # to get distinct FERT (DUK and DUC pooled) genes, compare to DU6, DU6P and DuhLB contrasts only.
  fert = get_distinct_genes(ma = ma, cont1 = "FERT_vs_FZTDU_mean_fst", 
                   rest_cont = colnames(ma)[!colnames(ma) %in% c("FERT_vs_FZTDU_mean_fst", "DUK_vs_FZTDU_mean_fst","DUC_vs_FZTDU_mean_fst")],
                   upper_bound = 0.7,
                   lower_bound = 0.1)
) %>% 
  lapply(function(x) rownames(x)) %>% unlist() %>% unique() #173 genes

# export for sharing with colleagues
#ma[gns,] %>% write.csv(here("00_dashboard/data/RED_genes/RED_gene_mean_fst_distinct.csv"))

# subset matrix for genes with high distinctiveness
ma_ss <- ma[gns,] #173   6

# fixnames
colnames(ma_ss) <- colnames(ma_ss) %>% str_remove("_mean_fst")
```

```{r vis_distinctly_differentiated_RED_genes, fig.width=10, fig.height=8}
ma_ss %>% 
  # remove FERT line (sticking to the Jenni's idea that fert lines should be evaluated in the context of convergent selection)
  #.[,colnames(.) != "FERT_vs_FZTDU"] %>% 
  pheatmap(cluster_cols = F, cluster_rows = F, display_numbers = F, angle_col = 45,
           main = "Distinctive RED-genes \n(mean Fst per gene)")

# remove objects with generic names
rm(gns, ma)
```

### Distinct-genes-Mean-Fst
```{r datatable_for_distinct_genes_fsts}
RED_gene_mean_fst_distinct <- ma_ss %>%
      round(2) %>% 
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>% 
      inner_join(mmu_gene_set, by = "gene_id") %>% 
      dplyr::select(seqname, start, end, gene_id, gene_symbol, gene_biotype, everything()) %>% 
      dplyr::rename(chr=seqname) %>% 
      # add a label to indicate for which contrast a gene is distinct
      mutate(
        distinct_for = ifelse(DUK_vs_FZTDU >= 0.7 & DUC_vs_FZTDU <= 0.1, "DUK_vs_FZTDU", NA),
        distinct_for = ifelse(DUC_vs_FZTDU >= 0.7 & DUK_vs_FZTDU <= 0.1, "DUC_vs_FZTDU", distinct_for),
        distinct_for = ifelse(DU6_vs_FZTDU >= 0.7, "DU6_vs_FZTDU", distinct_for),
        distinct_for = ifelse(DU6P_vs_FZTDU >= 0.7, "DU6P_vs_FZTDU", distinct_for),
        distinct_for = ifelse(DUhLB_vs_FZTDU >= 0.7, "DUhLB_vs_FZTDU", distinct_for),
        distinct_for = ifelse(FERT_vs_FZTDU >= 0.7, "FERT_vs_FZTDU", distinct_for)
      ) %>% 
      dplyr::select(distinct_for, everything())


RED_gene_mean_fst_distinct %>% 
  datatable(
    options = list(pageLength = 300), 
    caption = "Mean Fst distinct RED Genes (Fst (contrast of interest) >=0.7 & Fst (rest) < 0.1)", 
    rownames = FALSE, filter = "top"
    )

# export for sharing with colleagues
# RED_gene_mean_fst_distinct %>% write.csv(here("00_dashboard/data/RED_genes/RED_gene_mean_fst_distinct2.csv"))

```

### UpSetR (excl. FERT)
```{r upsetr_RED_genes_exclFERT, fig.width=10}
RED_genes_ids <- RED_genes[1:5] %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_genes_sets <- lapply(RED_genes, function(x){
  (RED_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_genes_sets) <- RED_genes_ids$mcols.gene_id

names(RED_genes_sets) <- names(RED_genes)

upset(RED_genes_sets, nintersects = NA, empty.intersections = T)
grid.text("Intersections RED-gene sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
```

### UpSetR (incl FERT)
```{r upsetr_RED_genes_inclFERT, fig.width=10}
RED_genes_ids <- RED_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_genes_sets <- lapply(RED_genes, function(x){
  (RED_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_genes_sets) <- RED_genes_ids$mcols.gene_id

names(RED_genes_sets) <- names(RED_genes)

#png(here("00_dashboard/for_manuscript/supplementary_figure_2.png"), res = 300, height = 2000, width = 4000, units = "px")
upset(RED_genes_sets, nintersects = NA, nsets = 6, empty.intersections = T)
grid.text("Intersections RED-gene sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()

```

### UpSetR (protein coding, excl FERT)
```{r upsetr_RED_genes_protein_coding_exclFERT, fig.width=10}
RED_pc_genes_ids <- RED_genes[1:5] %>% 
  lapply(function(x) as.data.frame(x) %>% filter(mcols.gene_biotype == "protein_coding")  ) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_pc_genes_sets <- lapply(RED_genes, function(x){
  (RED_pc_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_pc_genes_sets) <- RED_pc_genes_ids$mcols.gene_id

names(RED_pc_genes_sets) <- names(RED_genes)

upset(RED_pc_genes_sets, nintersects = NA, empty.intersections = T)
grid.text("Intersections RED-gene (protein coding) sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))

```

### UpSetR (protein coding, incl FERT)
```{r upsetr_RED_genes_protein_coding_inclFERT, fig.width=10}
RED_pc_genes_ids <- RED_genes %>% 
  lapply(function(x) as.data.frame(x) %>% filter(mcols.gene_biotype == "protein_coding")  ) %>% 
  bind_rows(.id = "contrast") %>% 
  dplyr::select(mcols.gene_id) %>% 
  unique()

RED_pc_genes_sets <- lapply(RED_genes, function(x){
  (RED_pc_genes_ids$mcols.gene_id %in% x$mcols.gene_id) %>% as.integer()
  }) %>% 
  bind_cols() %>% 
  as.data.frame() 

rownames(RED_pc_genes_sets) <- RED_pc_genes_ids$mcols.gene_id

names(RED_pc_genes_sets) <- names(RED_genes)

#png(here("00_dashboard/for_manuscript/supplementary_figure_3.png"), res = 300, height = 2000, width = 4000, units = "px")
upset(RED_pc_genes_sets, nintersects = NA, nsets = 6, empty.intersections = T)
grid.text("Intersections RED-gene (protein coding) sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()

```


RED-ORA {data-navmenu="REDs"}
====================================
```{r run_WebGestaltR_gobp_once_and_export, eval = F}
#--------
# GOBP
#--------

run_WebGestaltR(
  interestGene=RED_genes$DUK_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DUC_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
) # no sig!

run_WebGestaltR(
  interestGene=RED_genes$DU6_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DU6P_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$DUhLB_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

run_WebGestaltR(
  interestGene=RED_genes$FERT_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR/GOBP")
)

#-------
# KEGG
#-------
run_WebGestaltR(
  interestGene=RED_genes$DUK_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig!

run_WebGestaltR(
  interestGene=RED_genes$DUC_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig

run_WebGestaltR(
  interestGene=RED_genes$DU6_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
) # no sig

run_WebGestaltR(
  interestGene=RED_genes$DU6P_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

run_WebGestaltR(
  interestGene=RED_genes$DUhLB_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

run_WebGestaltR(
  interestGene=RED_genes$FERT_vs_FZTDU$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR/KEGG")
)

```


Column {.tabset}
-----------------

### gobp_DUK_vs_FZTDU
```{r}
gobp_DUK_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)

```

### gobp_DU6_vs_FZTDU
```{r}
gobp_DU6_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_DU6P_vs_FZTDU
```{r}
gobp_DU6P_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_DUhLB_vs_FZTDU
```{r}
gobp_DUhLB_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### gobp_FERT_vs_FZTDU
```{r}
gobp_FERT_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_DU6P_vs_FZTDU
```{r}
kegg_DU6P_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_DUhLB_vs_FZTDU
```{r}
kegg_DUhLB_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

### kegg_FERT_vs_FZTDU
```{r}
kegg_FERT_vs_FZTDU %>% 
  dplyr::select(geneSet, description, size, overlap, pValue, FDR, enrichmentRatio) %>% 
  datatable(rownames = F, options = list(pageLength = 150)) %>% 
  formatRound(columns=c("pValue", "FDR", "enrichmentRatio"), digits=3)
```

```{r supp_table_6to13, eval=F}
gobp_DUK_vs_FZTDU %>% 
  mutate(contrast = "DUK_vs_FZTDU", db = "gobp") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_6.csv"))
gobp_DU6_vs_FZTDU %>% 
  mutate(contrast = "DU6_vs_FZTDU", db = "gobp")  %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_7.csv"))
gobp_DU6P_vs_FZTDU %>% 
  mutate(contrast = "DU6P_vs_FZTDU", db = "gobp") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_8.csv"))
gobp_DUhLB_vs_FZTDU %>% 
  mutate(contrast = "DUhLB_vs_FZTDU", db = "gobp") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_9.csv"))
gobp_FERT_vs_FZTDU %>% 
  mutate(contrast = "FERT_vs_FZTDU", db = "gobp") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_10.csv"))


kegg_DU6P_vs_FZTDU %>% 
  mutate(contrast = "DU6P_vs_FZTDU", db = "kegg") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_11.csv"))
kegg_DUhLB_vs_FZTDU %>% 
  mutate(contrast = "DUhLB_vs_FZTDU", db = "kegg") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_12.csv"))
kegg_FERT_vs_FZTDU %>% 
  mutate(contrast = "FERT_vs_FZTDU", db = "kegg") %>% 
  dplyr::select(contrast, db, everything()) %>% 
  write.csv(here("00_dashboard/for_manuscript/supplementary_table_13.csv"))
```

### Top Results
```{r, fig.width=15, fig.height=10}
gobp_sig_top <- list(
  DUK_vs_FZTDU = gobp_DUK_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DU6_vs_FZTDU = gobp_DU6_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DU6P_vs_FZTDU = gobp_DU6P_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  DUhLB_vs_FZTDU = gobp_DUhLB_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio)),
  FERT_vs_FZTDU = gobp_FERT_vs_FZTDU %>% arrange(FDR) %>% .[1:5,] %>% arrange(desc(enrichmentRatio))
  ) %>% 
    bind_rows(.id = "contrast") %>% 
    #dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio) %>% 
    #remove NAs introduced when there are no 5 sig entries
    filter(!is.na(description)) %>% 
    mutate(db = "GOBP")

kegg_sig <- list(
  DU6P_vs_FZTDU = kegg_DU6P_vs_FZTDU,
  DUhLB_vs_FZTDU = kegg_DUhLB_vs_FZTDU,
  FERT_vs_FZTDU = kegg_FERT_vs_FZTDU
  ) %>% 
    bind_rows(.id = "contrast") %>% 
    #dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio)  %>% 
    mutate(db = "KEGG")

#png(here("00_dashboard/for_manuscript/figure_5.png"), res = 300, height = 3000, width = 4000, units = "px")

d <- bind_rows(gobp_sig_top, kegg_sig) %>% 
  mutate(
    contrast = factor(contrast, levels = c("DUK_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
  ) %>% 
  arrange(contrast, enrichmentRatio) %>% 
  mutate(
    description = factor(description, levels = .$description)#,
    #contrast = factor(contrast, levels = c("DUK_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    ) %>% 
  dplyr::select(contrast, geneSet, description, FDR, enrichmentRatio)

ggplot(data = d, aes(x = enrichmentRatio, y = description, fill = contrast)) +
  geom_bar(stat = "identity") +
  geom_text(aes(x = 1.5, y = description, label = geneSet)) +
  ylab(NULL) +
  theme_bw(base_size = 15) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) +
  labs(
    title = "Top Enriched Terms and Pathways"
  ) +
  scale_y_discrete(limits = rev(levels(d$description)))

#dev.off()

```




RED-candidates? {data-navmenu="REDs"}
================================================
```{r get_RED_genes_with_HIGH_impact_snps}

RED_genes_HIGH <- lapply(RED_genes, function(contr){
  contr %>% 
    as.data.frame() %>% 
    # intersect with high impact SNPs
    inner_join(
      snpeff_snps_HIGH,
      by = c("mcols.gene_id"="gene_ensembl")
    ) %>% 
    # prepare snp and gene loci columns
    mutate(
      gene_locus = paste0(seqnames,":",start,"-",end),
      snp_locus = paste0(CHROM,":",POS)
      ) %>% 
    # keep more informative columns
    dplyr::select(snp_locus, impact, effect, gene, mcols.gene_id, mcols.gene_biotype, gene_locus, ends_with("_Fst")) %>% 
    # rename columns for clarity
    dplyr::rename(
      snp_impact = impact, 
      snp_effect = effect, 
      gene_name = gene, 
      gene_ensembl = mcols.gene_id,
      gene_biotype = mcols.gene_biotype
      ) 
  })  %>% 
  bind_rows(.id = "contrast") %>% 
  # set levels contrasts
  mutate(
    contrast = factor(
      contrast, 
      levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU")
      )
    )
``` 

```{r matrices_heatmaps}
#---------------------------------------------------------
# matrix for heatmap all RED-genes with a high-impact SNP
#---------------------------------------------------------

# remove duplicate entries due to multiple snp-gene_isoform hits
ss <- RED_genes_HIGH %>% dplyr::select(gene_name, snp_locus, ends_with("_Fst")) %>% unique()

ma <- ss %>% dplyr::select(ends_with("_Fst")) %>% as.matrix()
rownames(ma) <- paste0(ss$gene_name, " | snp_locus:",ss$snp_locus)

# set NAs to smallest observed Fst (NA = all samples with same genotype for SNP, thus Fst is lowest)
# needed for clustering
ma[is.na(ma)] <- min(ma, na.rm=T)

#ma <- ma[,c("DUK_vs_FZTDU_Fst","DUC_vs_FZTDU_Fst","FERT_vs_FZTDU_Fst","DU6_vs_FZTDU_Fst", "DU6P_vs_FZTDU_Fst","DUhLB_vs_FZTDU_Fst")]
ma <- ma[,c("DUK_vs_FZTDU_Fst","DUC_vs_FZTDU_Fst","FERT_vs_FZTDU_Fst","DU6_vs_FZTDU_Fst", "DU6P_vs_FZTDU_Fst","DUhLB_vs_FZTDU_Fst")]

colnames(ma) <- str_remove(colnames(ma), "_Fst")

#------------------------------------------------------------------------------------
# matrix excluding high-impact SNPs (in RED-genes) with all contrast with Fst < 0.3
#------------------------------------------------------------------------------------

ma2 <- ma[apply(ma,1,function(x) any(x > 0.3)),]


```


Column {.tabset}
-----------------

### Overview

* Genes in regions of extreme differentiation were identified (RED-genes).

* Then, genes carrying a high-impact mutation (as per SNPeff) were identified --> candidate genes

* Thus candidate genes = RED-genes with a high impact mutation.

* The candidate-gene mean Fst was calculated using all SNPs inside the gene, including high-impact and other effects.

* To understand the results, candidate-genes were intersected, visualized in heatmaps (high-impact-SNP Fst and mean gene Fst). Genes were serched for in significantly enriched terms, etc.


### Tables

```{r}

# tabulate high-snp-genes intersecting with RED-genes
RED_genes_HIGH %>% 
  dplyr::select(contrast, gene_ensembl) %>% 
  unique() %>% 
  group_by(contrast) %>% 
  summarise(gene_number = n()) %>% 
  reshape2::dcast(.~contrast, value.var = "gene_number") %>% 
  .[,-1] %>% 
  kable(caption = "How many RED-genes carrying high-impact SNPs are there?") %>% 
  kable_styling(full_width = F)


```

```{r}

RED_genes_HIGH %>% 
  group_by(contrast, snp_effect) %>% 
  summarise(n=n()) %>% 
  reshape2::dcast(snp_effect ~ contrast, value.var = "n")%>% 
  kable(caption = "What effect correspond to those high-impact SNPs?") %>% 
  kable_styling(full_width = F)

```

### Gene intersection
```{r, fig.width=10}
RED_genes_HIGH_ids <- RED_genes_HIGH %>% 
  dplyr::select(contrast, gene_ensembl) %>% 
  dplyr::select(gene_ensembl) %>% 
  unique() %>% 
  unlist() %>% 
  as.character()

RED_genes_HIGH_sets <- lapply(
  unique(RED_genes_HIGH$contrast), function(contr){
    
    gns <- RED_genes_HIGH %>% 
      filter(contrast == contr) %>% 
      dplyr::select(gene_ensembl) %>% 
      unlist() %>% 
      as.character()

    (RED_genes_HIGH_ids %in% gns) %>% as.integer()
    
  }
) %>% 
  bind_cols() %>% 
  as.data.frame()

names(RED_genes_HIGH_sets) <- unique(RED_genes_HIGH$contrast)
rownames(RED_genes_HIGH_sets) <- RED_genes_HIGH_ids

#png(here("00_dashboard/for_manuscript/supplementary_figure_4.png"), res = 300, height = 2000, width = 3000, units = "px")
upset(RED_genes_HIGH_sets, nsets = 6)
grid.text("Intersections RED-genes with HIGH-impact-SNPs",x = 0.65, y=0.95, gp=gpar(fontsize=15))
#dev.off()
```

### Fst Heatmap
```{r fst_heatmap, fig.width=10, fig.height=15}

#pheatmap(ma,na_col = "darkgrey",cluster_cols = F, cluster_rows = F, cellwidth = 20, cellheight = 10, main = "Fst scores per (HIGH-impact)SNP in RED-genes")

pheatmap(ma,na_col = "darkgrey",cluster_cols = F, cluster_rows = T, main = "Fst scores per (HIGH-impact)SNP in RED-genes")

```

### Fst Heatmap (any Fst > 0.3)
```{r fst_heatmap_low_scores_removed, fig.width=10, fig.height=13}

#png(here("00_dashboard/for_manuscript/supplementary_figure_5.png"), res = 300, height = 3500, width = 2500, units = "px")

pheatmap(ma2,na_col = "darkgrey",cluster_cols = T, cluster_rows = T, display_numbers = T, angle_col = 45,main = "Fst scores (>0.3 any) per (HIGH-impact)SNP in RED-genes")

#dev.off()


```

```{r for_manuscript_nr_snps_genes_after_filtering_by_fst, eval = F}

# Get nr of high-impact SNPs after removing low Fst contrasts (all < 0.3)
rownames(ma2) %>% str_split("[|]") %>% sapply(function(x) x[2]) %>% str_trim() %>% unique() %>% length()

# Get nr of RED genes after removing low Fst contrasts (all < 0.3)
rownames(ma2) %>% str_split("[|]") %>% sapply(function(x) x[1]) %>% str_trim() %>% unique() %>% length()

# Get number of SNPs with min 0.7 in at least one contrast
rownames(ma2)[apply(ma2,1, function(x) any(x >= 0.7))] %>% str_split("[|]") %>% sapply(function(x) x[2]) %>% str_trim() %>% unique() %>% length()

rownames(ma2)[apply(ma2,1, function(x) any(x >= 0.7))] %>% str_split("[|]") %>% sapply(function(x) x[1]) %>% str_trim() %>% unique() %>% length()
```

### Genes in sig. GOBP-KEGG
```{r genes_in_sig_GOBP_KEGG}
names(RED_genes_HIGH) <- str_remove(names(RED_genes_HIGH), "_Fst")

tab <- RED_genes_HIGH %>% 
  # filter RED-genes with high-impact SNP
  filter(
    # filter by loci: high-impact SNPs "any contrast Fst > 0.2"
    snp_locus %in% (rownames(ma2) %>% lapply(function(x) str_split(x,"snp_locus:") %>% unlist() %>% .[2]))
  ) %>% 
  # select relevant columns
  dplyr::select(contrast, snp_locus, gene_ensembl, gene_name, snp_effect, snp_impact, contains("_vs_")) %>% 
  # intersect with genes in significant pathway terms
  inner_join(
    sig_gobp_kegg_one_gene_per_row,
    by = c("contrast","gene_ensembl")
  ) %>% 
  dplyr::select(contrast, snp_locus, gene_ensembl, gene_name, snp_effect, snp_impact, db, description, contains("_vs_"))

#write.csv(tab, here("00_dashboard/for_manuscript/supplementary_table_15.csv"))

tab %>% 
  kable(caption = "Number of RED-genes containing HIGH-impact SNPs in Significant GOBP or KEGG") %>% 
  kable_styling(full_width = F) 

```

### Manhattan plots 
```{r make_figures_once, eval = F}
# run this once to export

#chr=3

for(chr in unique(manhattan_dat$CHROM)){ #loop on each chromosome from same manhattan fst data above
  
  # chr subset RED-genes with high impact SNPs
  ss_genes <- RED_genes_HIGH %>% 
    # get gene coordinates
    separate(gene_locus,into = c("CHROM","tmp"),sep = ":") %>% 
    separate(tmp, into = c("start","end"), sep = "-") %>% 
    filter(CHROM == chr) %>% 
    dplyr::select(gene_name, CHROM, start, end) %>% 
    unique() 
  
  cond <- nrow(ss_genes) >= 1 # set condition: proceed it RED-genes have a high impact SNP for this chr

  if(cond){
    # make a granges of RED-genes
    ss_genes_gr <- GRanges(
      seqnames = ss_genes$CHROM,
      IRanges(
        start = as.numeric(ss_genes$start), 
        end = as.numeric(ss_genes$end)
        )
      )
    ss_genes_gr$gene_name <- ss_genes$gene_name # add gene name as mcols

    # calculate gene means using prepared function
    mean_fst_genes <- bind_rows(
      get_mean_fst_gene_ranges(ss_genes_gr, "DUK_vs_FZTDU"),
      get_mean_fst_gene_ranges(ss_genes_gr, "DUC_vs_FZTDU"),
      get_mean_fst_gene_ranges(ss_genes_gr, "DU6_vs_FZTDU"),
      get_mean_fst_gene_ranges(ss_genes_gr, "DU6P_vs_FZTDU"),
      get_mean_fst_gene_ranges(ss_genes_gr, "DUhLB_vs_FZTDU"),
      get_mean_fst_gene_ranges(ss_genes_gr, "FERT_vs_FZTDU")
    ) %>% 
      # convert gene position (gene center) to Mb
      mutate(center_mb = ( start + (end - start)/2 )/1e6 ) %>% 
      dplyr::rename(CHROM = seqnames, MEAN_FST = mean_gene_fst) %>% 
      dplyr::select(contrast, CHROM, center_mb, MEAN_FST, gene_name) %>% 
      mutate(CHROM = paste0("Chromosome ", CHROM))
  
    # subset manhattan data
    ss_win_fst_dat <-  manhattan_dat %>% 
      filter(CHROM == chr) %>% 
      ungroup() %>% 
      mutate(CHROM = paste0("Chromosome ", CHROM))
  
    # prepare figure name for exporting
    fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,"_RED_genes_HIGH_labs.png")
  
    # make plot
    ss_win_fst_dat %>% 
      # plot fst windowed data
      ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
        geom_point(alpha = 0.5) +
        # add mean gene Fst data and labels
        geom_point(data = mean_fst_genes, aes(x = center_mb, y = MEAN_FST), color = "red", size = 2) +
        geom_label_repel(data = mean_fst_genes, aes(x = center_mb, y = MEAN_FST, label = gene_name), alpha = 0.7) +
        # work out the appeareance
        theme_bw(base_size = 15) +
        scale_color_manual(values = "darkgrey") +
        scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
        facet_grid(contrast ~ CHROM) +
        xlab("Mb") +
        ylab("Mean Window Fst") +
        # save plot
        ggsave(
          filename = fl_nm, 
          device = "png",
          path = here("00_dashboard/figures"),
          width = 15,
          height = 10,
          units = "in",
          dpi = 300
        )
  }else{ # if condition is note met print a usefull message
    print(
      paste0("There were no RED-genes with high impact SNPs in chromosome ", chr)
    )  
    }
}

```

```{r message_to_redirect_to_figure_files}
data.frame(
  file = list.files(here("00_dashboard/figures"), pattern = "RED_genes_HIGH_labs.png") %>% 
  sapply(function(x) paste0("00_dashboard/figures/",x))
) %>% 
  kable(row.names = FALSE, caption = "Inspect figures manualy:")
```

### Gene-mean Fst heatmap
```{r get_per_candidate_gene_mean_fst, eval = F}
# get chrs with RED-genes with high impact SNPs
chrs <- RED_genes_HIGH %>% 
  mutate(
    CHROM = str_split(snp_locus,":") %>% sapply(function(x) x[1])
    ) %>% 
  dplyr::select(CHROM) %>% 
  unique() %>% 
  unlist() %>% 
  as.numeric() %>% 
  sort()

RED_genes_HIGH_mean_fst <- lapply(chrs, function(chr){
    # chr subset RED-genes with high impact SNPs
  ss_genes <- RED_genes_HIGH %>% 
    # get gene coordinates
    separate(gene_locus,into = c("CHROM","tmp"),sep = ":") %>% 
    separate(tmp, into = c("start","end"), sep = "-") %>% 
    filter(CHROM == chr) %>% 
    dplyr::select(gene_name, CHROM, start, end) %>% 
    unique() 
  
  # make a granges of RED-genes
  ss_genes_gr <- GRanges(
    seqnames = ss_genes$CHROM,
    IRanges(
      start = as.numeric(ss_genes$start), 
      end = as.numeric(ss_genes$end)
      )
    )
  ss_genes_gr$gene_name <- ss_genes$gene_name # add gene name as mcols

  # calculate gene means using prepared function
  mean_fst_genes <- bind_rows(
    get_mean_fst_gene_ranges(ss_genes_gr, "DUK_vs_FZTDU"),
    get_mean_fst_gene_ranges(ss_genes_gr, "DUC_vs_FZTDU"),
    get_mean_fst_gene_ranges(ss_genes_gr, "DU6_vs_FZTDU"),
    get_mean_fst_gene_ranges(ss_genes_gr, "DU6P_vs_FZTDU"),
    get_mean_fst_gene_ranges(ss_genes_gr, "DUhLB_vs_FZTDU"),
    get_mean_fst_gene_ranges(ss_genes_gr, "FERT_vs_FZTDU")
  ) 

})

names(RED_genes_HIGH_mean_fst) <- paste0("chr_",chrs)

# extract number of SNPs
ns <- lapply(RED_genes_HIGH_mean_fst, function(x){
  reshape2::dcast(x, gene_name ~ contrast, value.var = "n")
  }) %>% 
  bind_rows() %>% 
  .[2:ncol(.)] %>% 
  apply(1,unique)

  
lapply(RED_genes_HIGH_mean_fst, function(x){
  reshape2::dcast(x, gene_name ~ contrast, value.var = "mean_gene_fst")
  }) %>% 
  bind_rows() %>% 
  mutate(n_snp = ns) %>% 
  dplyr::select(gene_name, n_snp, everything()) %>% 
  # export to avoid running the above again (takes too long!)
  write.csv(here("00_dashboard/data/mean_fst_candidate_genes.csv"))
```

```{r heatmap_candiate_gene_mean_fst, fig.width=10, fig.height=13}
# import data created in previous chunk
ma_mean_fst_genes <- read.csv(here("00_dashboard/data/mean_fst_candidate_genes.csv")) %>% dplyr::select(-X, -n_snp)

rownames(ma_mean_fst_genes) <- ma_mean_fst_genes$gene_name

ma_mean_fst_genes$gene_name <- NULL

ma_mean_fst_genes <- as.matrix(ma_mean_fst_genes)

# replace NaN with min Fst observed
ma_mean_fst_genes[is.na(ma_mean_fst_genes)] <- min(ma_mean_fst_genes, na.rm=T)


pheatmap(ma_mean_fst_genes, cluster_cols = T, cluster_rows = T, display_numbers = T, angle_col = 45,main = "Candidate genes mean Fst scores")
```

### Distinctive
```{r Distinctive_candidate_REDgenes_per_contrast, fig.width=10, fig.height=8}

most_distinct_diff_genes <- list(
  duk = get_distinct_genes(ma_mean_fst_genes, "DUK_vs_FZTDU", c("DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU"), 0.7, 0.5),
  duc = get_distinct_genes(ma_mean_fst_genes, "DUC_vs_FZTDU", c("DUK_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU"), 0.7, 0.5),
  fert = get_distinct_genes(ma_mean_fst_genes, "FERT_vs_FZTDU", c("DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU"), 0.7, 0.5),
  du6 = get_distinct_genes(ma_mean_fst_genes,"DU6_vs_FZTDU", c("DUC_vs_FZTDU","DUK_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU"), 0.7, 0.5),
  du6p = get_distinct_genes(ma_mean_fst_genes,"DU6P_vs_FZTDU", c("DUC_vs_FZTDU","DU6_vs_FZTDU","DUK_vs_FZTDU","DUhLB_vs_FZTDU"), 0.7, 0.5),
  duhlb = get_distinct_genes(ma_mean_fst_genes,"DUhLB_vs_FZTDU", c("DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUK_vs_FZTDU"), 0.7, 0.5)
)
  

  
gns <- most_distinct_diff_genes %>% lapply(function(x) rownames(x))  %>% unlist() %>% unique()

mat_distinct <- ma_mean_fst_genes[gns,] %>% 
  as.data.frame() %>% 
  mutate(gene_name = gns) %>% 
  inner_join(
    RED_genes_HIGH %>% 
      filter(gene_name %in% gns) %>% 
      dplyr::select(gene_locus, gene_ensembl, gene_name) %>% 
      unique(),
    by = "gene_name"
  )

rownames(mat_distinct) <- paste0(mat_distinct$gene_name, " / ", mat_distinct$gene_locus)

#png(here("00_dashboard/for_manuscript/supplementary_figure_7.png"), res = 300, height = 2000, width = 2000, units = "px")
mat_distinct %>% 
  dplyr::select(contains("_vs_")) %>% 
  as.matrix() %>% 
  .[,c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU")] %>% 
  pheatmap(cluster_cols = F, cluster_rows = F, display_numbers = T, angle_col = 90,
           main = "Distinctive candidate RED-genes \n(mean Fst per gene)")
#dev.off()

```

### High Impact SNP Fst 
```{r, details_snp_effects_high_impact}
tab <- RED_genes_HIGH %>% 
  mutate(
    CHROM = str_split(snp_locus, ":") %>% sapply(function(x) x[1]),
    snp_pos = str_split(snp_locus, ":") %>% sapply(function(x) x[2]),
    gene_start = str_split(gene_locus, ":") %>% sapply(function(x) x[2] %>% str_split("-") %>% sapply(function(y) y[1])),
    gene_end = str_split(gene_locus, ":") %>% sapply(function(x) x[2] %>% str_split("-") %>% sapply(function(y) y[2]))
    ) %>% 
  dplyr::select(CHROM, snp_pos, snp_effect, snp_impact, gene_name, gene_ensembl, contains("_vs_")) %>% 
  unique() 

tab %>% 
  datatable(options = list(pageLength = 100), caption = "High Impact SNP effect information and associated gene") %>% 
  formatRound(
    paste0(c("DUK","DUC","DU6P","DU6","DUhLB","FERT"),"_vs_FZTDU"),
    2
    )

#write.csv(tab,here("00_dashboard/data/High_Impact_SNP_Details.csv"))
```

### Mean Fst Candidate Genes
```{r, details_candidate_genes}
tab2 <- read.csv(here("00_dashboard/data/mean_fst_candidate_genes.csv")) %>% 
  inner_join(
    dplyr::select(tab, gene_name, gene_ensembl, CHROM, snp_pos, snp_effect, snp_impact) %>% unique(), 
    by = "gene_name"
    ) %>% 
  dplyr::select(CHROM, snp_pos, snp_effect, snp_impact, gene_ensembl, gene_name, n_snp, everything(), -X) %>% 
  unique() 

tab2 %>%
  datatable(options = list(pageLength = 100), 
            caption = "Mean Fst Candidate Genes (genes with a high-impact SNP)") %>% 
  formatRound(
    paste0(c("DUK","DUC","DU6P","DU6","DUhLB","FERT"),"_vs_FZTDU"),
    2
    )

#write.csv(tab2,here("00_dashboard/data/Mean_Fst_Candidate_Genes_Full.csv"))

rm(tab, tab2)
```

### High Impact SNP GT counts
```{r}

by_cols <- c("CHROM","SNP_POS","REF","ALT","snp_effect","snp_impact","gene_name","gene_ensembl")

tabulate_genotypes_per_contrast("DUK") %>% 
  inner_join(
    tabulate_genotypes_per_contrast("DUC"), by = by_cols
  ) %>%  
  inner_join(
    tabulate_genotypes_per_contrast(c("DUK","DUC")), by = by_cols #FERT
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast("DU6"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast("DU6P"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast("DUhLB"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast("FZTDU"), by = by_cols
  ) %>% 
  datatable(options = list(pageLength = 100),
            caption = "Genotype counts per population for HIGH impact SNPs")

```

### Gene Details

---

*Candidate Genes DUK_vs_FZTDU*

* **Mirg  - miRNA containing gene**

  * SNP
    - Position: chr12:109733279
    - Fst: SNP Fst, DUK_vs_FZTDU: 0.975635 (all other contrast with very low snp-fst (~0), except for duhlb_vs_fztdu (~0.32) and FERT (~0.36).
    - Genotypes: There are 20 hom-ref (A) genotypes (5(./.);20(A/A)), whereas FZTDU has 21 hom-alt (G) genotypes (3(./.);1(A/G);21(G/G))
    - Effect: the type of high-impact effect associated to the SNP is indicated as "splice_acceptor_variant&intron_variant"
    
  * Gene
    - Gene Mean Fst: Mean gene-fst is ~0.97 (all other contrasts with very low snp-fst (~0), except for duhlb_vs_fztdu (~0.35)). Mean calculated over 160 SNPs in the gene.
    - GOBP:
    - Expression:
    - Affected transcript:
    - Type:

  * Summary
    - The Mirg gene contains a high-impact mutation affecting gene splicing. 
    - However, Mirg does not encode a protein but rather a number of mirnas.
    - Mirg is known to be important for mouse embryogenesis.
    - Alternative splicing of this gene can affect mirna biogenesis. However, the SNP position does not overlap with mirnas. 
    - The two transcripts affected (Mirg-212 & Mirg-208) contain several mirnas

  
  * Literature
    - Spatiotemporal expression pattern of Mirg, an imprinted non-coding gene, during mouse embryogenesis
      - https://pubmed.ncbi.nlm.nih.gov/22033866/
      
      - "Recent research has revealed that the maternal non-coding RNA genes (Gtl2, Rian and Mirg) from the Dlk1-Dio3 imprinted cluster are closely related to the full development potential of the induced pluripotent stem cells (iPSCs)."
      
      - "we demonstrated that Mirg non-coding RNA exhibited sustained expression throughout mouse embryogenesis from E8.5 to E18.5. Strong expression was detected in the central nervous system (E9.5-E15.5) and various skeletal muscles (E13.5 and E15.5), and the subcellular localization appeared to be in the nuclei. The pituitary and adrenal gland also showed high expression of Mirg, but, unlike the skeletal muscles and the neural circuitry, the signals were not concentrated in the nuclei. In the major internal organs, Mirg maintained low expression during embryogenesis (E12.5-E18.5) whereas in the liver and the developing lung, Mirg was expressed with a gradually decreasing trend and a gradually raising trend, respectively. These findings indicate that temporal regulation of Mirg expression may be required during specific stages and in specific tissues during embryonic development."
      
      - "The Dlk1-Gtl2 imprinted cluster, approximately 1 Mb long,
      is located on mouse distal chromosome 12 and human
      chromosome 14q32 (da Rocha et al. 2008). In this region,
      all of the maternally expressed genes (MEGs) are noncoding, including multiple long non-coding RNA (lncRNA)
      genes, Gtl2 (Miyoshi et al. 2000), Rian (Hatada et al. 2001),
      Mirg (Seitz et al. 2003),"
      
      - "These data indicate that Gtl2, Rian and Mirg genes have
      significant contribution to mouse embryonic development."
      
      - "C:\DATA\BackUpWork\03_SAW\Literature\han2011_mirg"
      
    - A Large Imprinted microRNA Gene Cluster at the Mouse Dlk1-Gtl2 Domain
      - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC515320/

        - "Although a few of these clustered miRNA genes are embedded within introns or exons of Mirg, a maternally expressed gene lacking a conserved protein-coding potential (Seitz et al. 2003), the genomic organization of most of them is still unclear (Figs. ​(Figs.2B,2B, ​,5).5)."

    - https://pubmed.ncbi.nlm.nih.gov/12796779/
    	- Imprinted microRNA genes transcribed antisense to a reciprocally imprinted retrotransposon-like gene
    
    		- "Here we show that the imprinted mouse distal chromosome 12 locus encodes two miRNA genes expressed from the maternally inherited chromosome and antisense to a retrotransposon-like gene (Rtl1) expressed only from the paternal allele" 

    - https://www.sciencedirect.com/science/article/pii/S221112471730493X
    	- The Rodent-Specific MicroRNA Cluster within the Sfmbt2 Gene Is Imprinted and Essential for Placental Development
    
    		- "The miR-379/miR-410 miRNA cluster located in the Mirg gene within the Dlk1-Dio3 imprint region is a maternally expressed cluster of genes observed throughout eutherian mammals (Seitz et al., 2004). The function of the miR-379/miR-410 cluster was investigated using conventional knockout mice with a deletion of the corresponding 59-kb region of the Mirg gene (Labialle et al., 2014). The mutants showed partially penetrant postnatal death, which was caused by defects in the maintenance of energy homeostasis."
    
    - https://www.embopress.org/doi/full/10.15252/embj.201387038
    	- The miR‐379/miR‐410 cluster at the imprinted Dlk1‐Dio3 domain controls neonatal metabolic adaptation
    
    		- The functional significance of the corresponding imprinted human genes at 14q32 is supported by human syndromes of respiratory insufficiency, altered thoracic development, mental retardation, obesity, growth retardation, and precocious puberty (Kagami et al, 2008). The Dlk1‐Dio3 genomic interval has also been linked genetically to diabetes (Wallace et al, 2010; Kameswaran et al, 2014). Whether the miR‐379/miR‐410 (C14MC) cluster contributes to one (or several) of these physiological processes or pathological contexts remains an open question.

    - https://www.sciencedirect.com/science/article/pii/S1097276513003705

    	- For example, the splice-site-overlapping mmu-miR-412 is transcribed as a part of a long noncoding RNA gene (Mercer et al., 2008), Mirg, which contains nine additional miRNA precursors. 

    - https://www.sciencedirect.com/science/article/pii/S088875430500282X?via%3Dihub
    
    	- "High-resolution map and imprinting analysis of the Gtl2–Dnchc1 domain on mouse chromosome 12"
    
    		- RT-PCR experiments support the predicted presence of a maternally expressed intergenic transcript(s) encompassing Gtl2, Rian, and Mirg. 
    
    		- The identified maternally expressed transcripts Gtl2, Rtl1-as, Rian, and Mirg are likely to represent nontranslated transcripts that are all transcribed in the same orientation. Three of the maternally expressed transcripts encompass the precursors for small RNAs: the Rian transcript contains precursors for snoRNAs, the Rtl1-as transcript and Mirg are associated with miRNAs [7], [11], [12].
    
    		- Two were previously described, encompassing the snoRNAs in Rian/MEG8 and miRNAs in and around the Mirg gene [11]. 
    
    		- Mirg expression decreases progressively during postnatal development: in newborn mice, expression is most pronounced in brain, limbs, and tongue and becomes limited to brain in adult mice
    
    		- The biological importance of the (imprinted) expression of the Gtl2, Rian, and Mirg genes in adult brain remains unclear. Most of the complex phenotypes associated with genes in this imprinting cluster have been analyzed in mice carrying uniparental disomies of chromosome 12. These mice show placental and skeletal defects and distinct muscle morphologies and die before birth [3], [5]. Thus, possible brain-specific functions of these imprinted genes in postnatal tissues remain to be analyzed using alternative models.
    
    		- "The identified maternally expressed transcripts Gtl2, Rtl1-as, Rian, and Mirg are likely to represent nontranslated transcripts that are all transcribed in the same orientation. Three of the maternally expressed transcripts encompass the precursors for small RNAs: the Rian transcript contains precursors for snoRNAs, the Rtl1-as transcript and Mirg are associated with miRNAs [7], [11], [12]."
    
    		- "Mirg is predominantly expressed from the maternal allele in all tissues analyzed, confirming the maternal expression of Mirg "
  
* **Nos1ap ("nitric oxide synthase 1 (neuronal) adaptor protein")**
    * Distinct for the contrast DUK_vs_FZTDU 

    * SNP
      -  Position: 1	170,556,702
      - Fst (DUK_vs_FZTDU) = 1. Followed by DUC_vs_FZTDU (Fst = 0.48). All other contrasts NAs.
      - 5(./.);20(A/A) vs 4(./.);21(G/G) --> duk_homref vs fztdu_homalt
      - Splice donor variant
    
    * Gene
      - Gene mean Fst = 0.89 (2712 SNPs). All other contrasts Fst<0.1, except DUC_vs_FZTDU (~0.42)
      - related to cardiac muscle contraction and neuronal function.
      - The transcript affected is Nos1ap-203 (retained intron)
      	- http://www.ensembl.org/info/genome/genebuild/biotypes.html#:~:text=Retained%20intron%3A%20An%20alternatively%20spliced,does%20not%20overlap%20any%20exons.
      	- "Retained intron: An alternatively spliced transcript believed to contain intronic sequence relative to other, coding, transcripts of the same gene"
    
    * Interpretation
      - Most of the literature is concerned with the role of NOS1AP to cariac and neural function.
      - However, because of its involvment in the production of NO, it can be linked to reproductive functions.
      - interestingly the gene has been found to be expressed in mouse placenta.
      - The mutation affects the last exon of the transcript Nos1ap-203.
      - Nos1ap-203 is described as "retained intron" but I am not sure what intron is retained.
    
    * Literature
    
      - https://pubmed.ncbi.nlm.nih.gov/29319606/
        - Disruption of nNOS-NOS1AP protein-protein interactions suppresses neuropathic pain in mice (2018)
        - "disrupting nNOS-NOS1AP protein-protein interactions attenuates mechanistically distinct forms of neuropathic pain"
      
      - https://pubmed.ncbi.nlm.nih.gov/27237129/
        - Research progress in NOS1AP in neurological and psychiatric diseases (2016)

        - "An increasing body of evidence is pointing to the significant roles of NOS1AP in excitotoxic neuronal damage, traumatic nervous system injury, bipolar disorder, and schizophrenia."
        - "The review will promote the further investigation of NOS1AP in brain disorders and the development of drugs targeting the NOS1AP PTB domain or PDZ-binding motif in the future."

      - ...many papers related to schizofrenia and cardiac contraction

      - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5747323/
        - NOS1AP genetic variation is associated with impaired healing of diabetic foot ulcers and diminished response to healing of circulating stem/progenitor cells (2017)
        - "The NOS1AP gene produces a protein called capon that was initially shown to interact with NOS1, regulating nitric oxide (NO) production at post-synaptic sites in neurons.(3) NO, a gaseous free radical, is a messenger molecule with diverse functions throughout the body. NO has been shown to be a cell signaling molecule with varied effects on angiogenesis, stem progenitor cell (SPC) mobilization, and wound repair.(4, 5)"

      - https://pubmed.ncbi.nlm.nih.gov/10559838/#:~:text=Nitric%20oxide%20(NO)%20plays%20a,lordosis%20reflex%20in%20female%20rats.
        - The role of nitric oxide in reproduction (1999)

        - "Nitric oxide (NO) plays a crucial role in reproduction at every level in the organism. In the brain, it activates the release of luteinizing hormone-releasing hormone (LHRH). The axons of the LHRH neurons project to the mating centers in the brain stem and by afferent pathways evoke the lordosis reflex in female rats. In males, there is activation of NOergic terminals that release NO in the corpora cavernosa penis to induce erection by generation of cyclic guanosine monophosphate (cGMP). NO also activates the release of LHRH which reaches the pituitary and activates the release of gonadotropins by activating neural NO synthase (nNOS) in the pituitary gland. In the gonad, NO plays an important role in inducing ovulation and in causing luteolysis, whereas in the reproductive tract, it relaxes uterine muscle via cGMP and constricts it via prostaglandins (PG)."

      - https://en.wikipedia.org/wiki/Nitric_oxide_synthase
  
        - The neuronal isoform is involved in the development of nervous system. It functions as a retrograde neurotransmitter important in long term potentiation and hence is likely to be important in memory and learning. nNOS has many other physiological functions, including regulation of cardiac function and peristalsis and sexual arousal in males and females. An alternatively spliced form of nNOS is a major muscle protein that produces signals in response to calcium release from the SR. nNOS in the heart protects against cardiac arrhythmia induced by myocardial infarction.[5]
  
      - http://www.ensembl.org/Mus_musculus/Gene/ExpressionAtlas?db=core;g=ENSMUSG00000038473;r=1:170311151-170618344
        - Expressed in placenta of pregnant mice

    - https://www.sciencedirect.com/science/article/pii/S088875430500282X?via%3Dihub
    
    	- "High-resolution map and imprinting analysis of the Gtl2–Dnchc1 domain on mouse chromosome 12"
    
    		- RT-PCR experiments support the predicted presence of a maternally expressed intergenic transcript(s) encompassing Gtl2, Rian, and Mirg. 
    
    		- The identified maternally expressed transcripts Gtl2, Rtl1-as, Rian, and Mirg are likely to represent nontranslated transcripts that are all transcribed in the same orientation. Three of the maternally expressed transcripts encompass the precursors for small RNAs: the Rian transcript contains precursors for snoRNAs, the Rtl1-as transcript and Mirg are associated with miRNAs [7], [11], [12].
    
    		- Two were previously described, encompassing the snoRNAs in Rian/MEG8 and miRNAs in and around the Mirg gene [11]. 
    
    		- Mirg expression decreases progressively during postnatal development: in newborn mice, expression is most pronounced in brain, limbs, and tongue and becomes limited to brain in adult mice
    
    		- The biological importance of the (imprinted) expression of the Gtl2, Rian, and Mirg genes in adult brain remains unclear. Most of the complex phenotypes associated with genes in this imprinting cluster have been analyzed in mice carrying uniparental disomies of chromosome 12. These mice show placental and skeletal defects and distinct muscle morphologies and die before birth [3], [5]. Thus, possible brain-specific functions of these imprinted genes in postnatal tissues remain to be analyzed using alternative models.
    
    		- "The identified maternally expressed transcripts Gtl2, Rtl1-as, Rian, and Mirg are likely to represent nontranslated transcripts that are all transcribed in the same orientation. Three of the maternally expressed transcripts encompass the precursors for small RNAs: the Rian transcript contains precursors for snoRNAs, the Rtl1-as transcript and Mirg are associated with miRNAs [7], [11], [12]."
    
    		- "Mirg is predominantly expressed from the maternal allele in all tissues analyzed, confirming the maternal expression of Mirg "
  
* **Nos1ap ("nitric oxide synthase 1 (neuronal) adaptor protein")**
    * Distinct for the contrast DUK_vs_FZTDU 

    * SNP
      -  Position: 1	170,556,702
      - Fst (DUK_vs_FZTDU) = 1. Followed by DUC_vs_FZTDU (Fst = 0.48). All other contrasts NAs.
      - 5(./.);20(A/A) vs 4(./.);21(G/G) --> duk_homref vs fztdu_homalt
      - Splice donor variant
    
    * Gene
      - Gene mean Fst = 0.89 (2712 SNPs). All other contrasts Fst<0.1, except DUC_vs_FZTDU (~0.42)
      - related to cardiac muscle contraction and neuronal function.
      - The transcript affected is Nos1ap-203 (retained intron)
      	- http://www.ensembl.org/info/genome/genebuild/biotypes.html#:~:text=Retained%20intron%3A%20An%20alternatively%20spliced,does%20not%20overlap%20any%20exons.
      	- "Retained intron: An alternatively spliced transcript believed to contain intronic sequence relative to other, coding, transcripts of the same gene"
    
    * Interpretation
      - Most of the literature is concerned with the role of NOS1AP to cariac and neural function.
      - However, because of its involvment in the production of NO, it can be linked to reproductive functions.
      - interestingly the gene has been found to be expressed in mouse placenta.
      - The mutation affects the last exon of the transcript Nos1ap-203.
      - Nos1ap-203 is described as "retained intron" but I am not sure what intron is retained.
    
    * Literature
    
      - https://pubmed.ncbi.nlm.nih.gov/29319606/
        - Disruption of nNOS-NOS1AP protein-protein interactions suppresses neuropathic pain in mice (2018)
        - "disrupting nNOS-NOS1AP protein-protein interactions attenuates mechanistically distinct forms of neuropathic pain"
      
      - https://pubmed.ncbi.nlm.nih.gov/27237129/
        - Research progress in NOS1AP in neurological and psychiatric diseases (2016)

        - "An increasing body of evidence is pointing to the significant roles of NOS1AP in excitotoxic neuronal damage, traumatic nervous system injury, bipolar disorder, and schizophrenia."
        - "The review will promote the further investigation of NOS1AP in brain disorders and the development of drugs targeting the NOS1AP PTB domain or PDZ-binding motif in the future."

      - ...many papers related to schizofrenia and cardiac contraction

      - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5747323/
        - NOS1AP genetic variation is associated with impaired healing of diabetic foot ulcers and diminished response to healing of circulating stem/progenitor cells (2017)
        - "The NOS1AP gene produces a protein called capon that was initially shown to interact with NOS1, regulating nitric oxide (NO) production at post-synaptic sites in neurons.(3) NO, a gaseous free radical, is a messenger molecule with diverse functions throughout the body. NO has been shown to be a cell signaling molecule with varied effects on angiogenesis, stem progenitor cell (SPC) mobilization, and wound repair.(4, 5)"

      - https://pubmed.ncbi.nlm.nih.gov/10559838/#:~:text=Nitric%20oxide%20(NO)%20plays%20a,lordosis%20reflex%20in%20female%20rats.
        - The role of nitric oxide in reproduction (1999)

        - "Nitric oxide (NO) plays a crucial role in reproduction at every level in the organism. In the brain, it activates the release of luteinizing hormone-releasing hormone (LHRH). The axons of the LHRH neurons project to the mating centers in the brain stem and by afferent pathways evoke the lordosis reflex in female rats. In males, there is activation of NOergic terminals that release NO in the corpora cavernosa penis to induce erection by generation of cyclic guanosine monophosphate (cGMP). NO also activates the release of LHRH which reaches the pituitary and activates the release of gonadotropins by activating neural NO synthase (nNOS) in the pituitary gland. In the gonad, NO plays an important role in inducing ovulation and in causing luteolysis, whereas in the reproductive tract, it relaxes uterine muscle via cGMP and constricts it via prostaglandins (PG)."

      - https://en.wikipedia.org/wiki/Nitric_oxide_synthase
  
        - The neuronal isoform is involved in the development of nervous system. It functions as a retrograde neurotransmitter important in long term potentiation and hence is likely to be important in memory and learning. nNOS has many other physiological functions, including regulation of cardiac function and peristalsis and sexual arousal in males and females. An alternatively spliced form of nNOS is a major muscle protein that produces signals in response to calcium release from the SR. nNOS in the heart protects against cardiac arrhythmia induced by myocardial infarction.[5]
  
      - http://www.ensembl.org/Mus_musculus/Gene/ExpressionAtlas?db=core;g=ENSMUSG00000038473;r=1:170311151-170618344
        - Expressed in placenta of pregnant mice

---


*Candidate Genes DUC_vs_FZTDU*

* **Lmntd1 - "lamin tail domain containing 1"**

  * SNP
    - Position: 6	145,433,764
    - start_lost
    - DUC_vs_FZTDU Fst = 0.98. Followed by DU6P (Fst = 0.4), then rest are NAs.
    - 5(./.);1(A/G);19(G/G) vs 2(./.);23(A/A) --> ref=A;alt=G
  
  * Gene
    - Gene mean = 0.8 over 3097 SNPs. Other contrast between 0.27 and 0.03.
    - GOBP: cell population proliferation
    - Transcripts affected: Lmntd1-202 & Lmntd1-206
    - Expressed in testis. 
  
  * Interpreation
    - Lmntd1 is expressed in mouse, with higher expression in testis. 
    - The function of this gene relates to proliferation of cells (see GOBP).
    - The start-lost mutation affects two transcripts. Could this have an impact in testicular function?
  
  
  * Literature
  
    - https://link.springer.com/article/10.1186/s12915-020-00826-z
      - Large-scale discovery of male reproductive tract-specific genes through analysis of RNA-seq datasets
    
      - "Of the 291 genes identified as human testis-specific through one or more of the human germ cell datasets, 252 genes have not been previously identified, 201 of which have one or more equivalent mouse orthologs with 139 of these genes having not been knocked out in mouse. Of these 139 novel genes with no mouse model, 8 encode enzymes (GLT6D1, PRSS48, SATL1, SULT6B1, TMPRSS7, TPTE2, TRIML2, and TTLL8), 1 encodes an epigenetic protein (TAF1L), 6 encode GPCRs (GPR156, TAS2R13, TAS2R30, TAS2R46, TAS2R50, VN1R2), 1 encodes a kinase (CDKL4), 33 encode oGPCRs (such as OR2D3, OR3A2, OR52E5, OR8G5, OR10J1, and OR14A2), 6 encode transcription factors (NKX2-4, ZNF99, ZNF648, ZNF705B, ZNF705G, and ZNF709), 3 encode transporters (ABCC12, SLC35G3, SLC35G5), and 81 encode proteins of unknown drug target type (such as AC008073.3, AC113554.1, AKNAD1, AL049634.2, DNLZ, ERVW-1, ETDA, FBXW10, FER1L5, LMNTD1, LRRC72, NMS, PRR23A, PRR23B, PRR23C, PXT1, RFPL4B, and SSX4B)."

    - https://link.springer.com/article/10.1186/s12862-019-1462-8
      - An evolutionary approach to recover genes predominantly expressed in the testes of the zebrafish, chicken and mouse
    
      - "Among the 68 mouse genes, four have no orthologs in chicken (RhoA, Shcbp1l, Lmna and Lmntd1)."
      - "In total, 10 mouse genes have no zebrafish orthologs: Shcbp1l (but its paralog Shcbp1 has one) and Lmntd1 as observed in chicken, plus Boll, Lrrc46, Dmrtb1, Myh15, Phf7, Tcp10a, Tcp10b, Tcp10c. "
      
      
* **Rassf8 - "Ras association (RalGDS/AF-6) domain family (N-terminal) member 8" - ENSMUSG00000030259**

  * SNP
    - Position: 6	145820060 (A/T)
    - stop_gained
    - DUC_vs_FZTDU Fst = 0.83. Followed by DU6P (0.34). Rest 0.00-.26.
    - 1(./.);1(A/T);23(T/T) vs 25(A/A)

  * Gene
    - Gene mean Fst = 0.98 over 707 SNPs. Other genes ranged 0.34-0.00
    - Low expression in epididimus, testis, uterus and ovary.
    - Transcripts affected: Rassf8-202, Rassf8-203

  * Interpreation
    - Rassf8 is a tumor supresor gene that inhibits cell growth.
    - It is expressed in reproductive tissues, including testis.
    - In testis it plays a role in gametogenesis through the Wnt/ β-catenin pathway.

  * Literature
  
    - https://www.genecards.org/cgi-bin/carddisp.pl?gene=RASSF8
      - "This gene encodes a member of the Ras-assocation domain family (RASSF) of tumor suppressor proteins. This gene is essential for maintaining adherens junction function in epithelial cells and has a role in epithelial cell migration."
    
    - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6429012/
      - "MicroRNA-322 Regulates Self-renewal of Mouse Spermatogonial Stem Cells through Rassf8"(2019)
    
      - "In this study, we report that miR-322 is a novel intrinsic RNA molecule that regulates self-renewal and differentiation of SSCs via the WNT/β-catenin signaling by targeting Rassf8"
    
      - "Rassf8 is a member of the RAS association domain family (RASSF) involved in a myriad of biological processes including cell death, proliferation, and microtubule stability32. RASSF8 inhibits cell growth and regulates the Wnt signaling pathway33, while the Wnt pathway is involved in various cellular processes such as proliferation, differentiation, apoptosis, fate determination, and migration34. "
    
      - "Rassf8 is found both in the nucleus and cell membrane33. Rassf8 is related to the cell cycle, accelerates apoptosis, and restrains migration and invasion52. Overexpression of Rassf8 induces G0/G1 arrest, apoptosis, and downregulation of Cyclin D1 downstream of the WNT/β-catenin pathway."
    
      - "A recent study reported that Rassf8 colocalizes with the adherens junction component β-catenin, binds to E-cadherin, and then regulates the Wnt/ β-catenin pathway33. The Wnt/β-catenin pathway is a critical player in the regulation of mouse and human spermatogonia"
      
* **Hpf1 - "histone PARylation factor 1" - ENSMUSG00000038005**

	* SNP 
		- 8	60905579 (C/A)
		- stop_gained
		- DUC_vs_FZTDU Fst = 0.98. Negiligeble differentiation in the other contrasts.
		- 4(./.);20(A/A);1(C/A) vs 1(./.);24(C/C)

	* Gene
		- Distinct for DUC_vs_FZTDU (Fst = 0.99, over 87 SNPs). All other contrasts ~0.
		- nonsense mediated decay for transcript Hpf1-205.
		- cellular response to DNA damage stimulus, ADP-ribosylation
		- high/mid expression in ovary and testis. Also mid expression in uterus, epididimus, placenta.

	* Interpretation
		- Hpf1 is involved in DNA damage repair and genome stability.
		- Its function is fairly new.
		- It is expressed in mouse reproductive tissues, also in zebra fish where is needed for normal embryogenesis.

	* Literature

		- https://www.genecards.org/cgi-bin/carddisp.pl?gene=HPF1
			- "Promotes histone serine ADP-ribosylation in response to DNA damage, limiting DNA damage-induced PARP1 hyper-automodification, and ensuring genome stability (PubMed:27067600, PubMed:28190768). Serine ADP-ribosylation of proteins constitutes the primary form of ADP-ribosylation of proteins in response to DNA damage (PubMed:29480802). HPF1 also promotes tyrosine ADP-ribosylation, probably by conferring tyrosine specificity on PARP1 (PubMed:30257210"

		- https://link.springer.com/article/10.1007/s00427-018-0608-9 (2018)
			- "Recently, human HPF1/C4orf27 was identified as a novel interacting component of PARP1 complex that plays an important role in PARP1-dependent ADP-ribosylation signaling in the DNA damage response and the maintenance of genome stability (Gibbs-Seymour et al. 2016). However, the function of hpf1 in animal is not studied yet. Whether hpf1 can regulate genome stability during early embryogenesis remains unknown."

			- "To our knowledge, this work analyzed the expression and down-regulation of zebrafish orthologue of human HPF1/C4orf27 for the first time, and the results showed hpf1 is specifically expressed in adult gonad tissues and is required for normal zebrafish embryogenesis, providing more insights for the mechanism of genome stability and early embryogenesis."
			
---

*Candidate Genes FERT_vs_FZTDU*

  * **Olfr664 - "olfactory receptor 664" - ENSMUSG00000051885**
  
  	* SNP
  		- 7:104733444 (A/G)
  		- stop_lost
  		- stop_lost&splice_region_variant
  		- Fst (FERT_vs_FZTDU) = 0.7 (DUK_vs_FZTDU=1, DUC_vs_FZTDU=0.52). Rest all NAs.
  		- 10(./.);3(A/A);13(A/G);24(G/G) vs 5(./.);20(A/A)
  
  	* Gene
  		- Fst gene mean (FERT_vs_FZTDU) = 0.71 over only 6 SNPs. (DUK_vs_FZTDU=1, DUC_vs_FZTDU=0.53).
  		- Mutation affects two transcripts (Olfr664-201, Olfr664-202), which are all. Only one exon for each.
  		- Olfr664-201 is a polymorphic pseudogene and Olfr664-202 codes for a protein.
  		- GOBP-summary: smell perception, G protein-coupled receptor signaling pathway
  		- expressed in olfactory epithelium
  
  	* Interpreation
  		- the mutation "erases" a stop codon (stop_lost) in Olfr664. 
  		- The gene is highly differentiated in FERT but only 6 SNPs were available for mean calculation.
  		- Literature is scarce.
  		- GOBP points out towards the expected function (olfaction).
  
  	* Literature
  		- "Gene expression profiles of vitrified in vivo derived 8‐cell stage mouse embryos detected by high density oligonucleotide microarrays"

  * **Usp17lb - "ubiquitin specific peptidase 17-like B" - ENSMUSG00000062369**
  
  	* SNP
  		- 7:104840875 G/A
  		- stop_gained
  		- This mutation has an Fst=NA --> 6(./.);44(G/G) vs 1(./.);24(G/G)
  		- The mutation is not relevant due to its null differentiation.

  	* Gene

  	* Interpretation
  		- although the mutation has a high impact (stop_gained) there is null differentiation for the SNP relative to the founder population.
  		- SO IGNORE!
  
  	* Literature
  	
  * **1110036E04Rik - ENSMUSG00000097036**
  
  	* This identifier is not in the current EnsEMBL database --> not possible to show mutation
  
  	* RIKEN cDNA 1110036E04 gene http://www.informatics.jax.org/marker/MGI:1915947
  
  	* https://www.ncbi.nlm.nih.gov/gene/68697
  
  	* Highly expressed in testis adult.
  
  	* https://www.alliancegenome.org/gene/MGI:1915947 #--> to have an idea of the transcript
  
  	* SNP
  		- 9	64053187 - splice_acceptor_variant&intron_variant - A	C - Fst = 0.91 	
  		- 9	64053188 - splice_acceptor_variant&intron_variant - G	T - Fst = 0.91
  
  
  	* Interpretation
  		- Two consecutive mutations affecting the splice region.
  		- The gene is no longer supported by ensembl. 
  
  	* Literature
  
  	- "Time course gene expression data in colon of mice after exposure to food-grade E171"
  		- https://www.sciencedirect.com/science/article/pii/S2352340917306637
  	
---

*Candidate Genes DU6_vs_FZTDU*

* **Rbck1 - ignored!**
  * SNP
    - Position: 2	152324431	G	A
    - Fst: 0.02 ---> high impact SNP is not differentiated ... ignored!
    - Genotypes:
    - Effect:
    
  * Gene
    - Gene Mean Fst:
    - Number of SNPs in gene:
    - Expresion:
    - GOBP:	
    - Affected transcript: 
    
  * Summary
    - ... 
    
  * Literature
    - ...


* **Dmap1 - ignored!**
  * SNP
    - Position:
    - Fst: 0.15 --> too low, ignore!
    - Genotypes:
    - Effect:
    
  * Gene
    - Gene Mean Fst:
    - Number of SNPs in gene:
    - Expresion:
    - GOBP:	
    - Affected transcript: 
    
  * Summary
    - ... 
    
  * Literature
    - ...

* **Cpne4 - copine IV - ENSMUSG00000032564**
  * SNP
    - Position: 9	104554416	A	G
    - Fst: 1.00 (followed by DU6P with 0.37, rest all negligeble)
    - Genotypes: 9(./.);16(A/A) vs 4(./.);21(G/G)
    - Effect: splice_acceptor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.84 (followd by DUhLB=0.43, DU6P=0.27, DUC=0.21, DUK=0.11)
    - Number of SNPs in gene: 2369
    - Expresion: highest in brain and olfatory bulb
    - GOBP:	cellular response to calcium ion	
    - Affected transcript: Cpne4-203
    - type:protein coding
    - GeneCards:"This gene belongs to the highly conserved copine family. It encodes a calcium-dependent, phospholipid-binding protein, which may be involved in membrane trafficking, mitogenesis and development. Alternative splicing results in multiple transcript variants. "; "CPNE4 (Copine 4) is a Protein Coding gene. Diseases associated with CPNE4 include Baraitser-Winter Syndrome and Motion Sickness. An important paralog of this gene is CPNE7."; "Probable calcium-dependent phospholipid-binding protein that may play a role in calcium-mediated intracellular processes. CPNE4_HUMAN,Q96A23"
    

  * Summary
    - Cpne4 is involved in neuronal function and possibly muscle glucose metabolism.
    - The high-impact SNP is fully differentiated and affects gene splicing.
    
  * Literature
    - https://pubmed.ncbi.nlm.nih.gov/30866042/
      - "Differential expression and subcellular localization of Copines in mouse retina" 2019
      - "We find that Cpne5, 6, and 9 are expressed in the ganglion cell layer (GCL) and inner nuclear layer (INL) in both amacrine cells and RGCs. Cpne4 expression is restricted to one amacrine cell population of the INL, but is specifically expressed in RGCs in the GCL. Cpne4 expression in RGCs is regulated by Brn3b both cell autonomously (in Brn3b+ RGCs) and cell nonautonomously (in Brn3b- RGCs)."
      - "Given their expression profile and previously proven role as Ca2+ sensors, they may participate in the morphogenetic processes that shape RGC dendrite and axon formation at early postnatal ages."
    
    - https://pubmed.ncbi.nlm.nih.gov/22467039/
      - "Gene expression analysis to identify molecular correlates of pre- and post-conditioning derived neuroprotection" 2012
      - "The five most upregulated genes within the neuroprotective clusters were Tagln, Nes, Ptrf, Vim and Adamts9, and the five most downregulated genes were Slc7a3, Bex1, Brunol4, Nrxn3 and Cpne4. Pathway analysis revealed that the intracellular and second messenger signalling pathways in addition to cell death were predominantly associated with downregulated pre- and post-conditioning associated genes, suggesting that modulation of cell death and signal transduction pathways plays a role in the neuroprotection. A high degree of similarity in the pathways associated with the differentially expressed genes in the pre- and post-conditioning treatments suggests that similar molecular mechanisms may mediate their neuroprotective effects."
    
    - https://pubmed.ncbi.nlm.nih.gov/23202439/
      - "Genome-wide scan for copy number variation association with age at onset of Alzheimer's disease" 2013 
      - " Two CNV events are intragenic causing a deletion in CPNE4. In addition, to further study the mutational load at the 10 known susceptibility loci, CNVs overlapping with these loci were also catalogued. We identified rare small events overlapping CR1 and BIN1 in AD and normal controls with opposite CNV dosage. The CR1 events are consistent with previous reports. Larger scale studies with deeper genotyping specifically addressing CNV are needed to evaluate the significance of these findings."
    
    - https://pubmed.ncbi.nlm.nih.gov/32366026/
      - "Genome-Wide Association Study of Muscle Glycogen in Jingxing Yellow Chicken" 2020
      - " Two significant SNPs were found in introns 12 and 13 of copine 4 (CPNE4) on chromosome 2. The wild-type and mutant individuals had significant differences in glycogen metabolism at two loci (p < 0.01 for both). Individuals carrying two mutations had increased muscle glycogen content. "
      - "There was a significant difference between wild-type and mutant individuals (p < 0.01). These mutations likely affecting two genes (CPNE4 and NKD1) may affect glycogen storage in a pleiotropic manner. Gene annotation indicates that CPNE4 and NKD1 may affect the process of glucose metabolism. Our findings contribute to understanding the genetic regulation of muscle glycogen metabolism and provide theoretical support."

---

*Candidate Genes DU6P_vs_FZTDU*

* **1700012B07Rik - ENSMUSG00000020617 - RIKEN cDNA 1700012B07 gene**
  * SNP
    - Position: 11	109814010	A	G
    - Fst: 0.84 (rest all < 0.18)
    - Genotypes: 5(./.);20(G/G) vs	4(./.);21(A/A)
    - Effect: splice_donor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.85 (rest all < 0.17)
    - Number of SNPs in gene: 251
    - Expresion: mostly expressed in testis 
    - GOBP:	"biological_process" ... not very useful ensembl!
    - Affected transcript: 1700012B07Rik-203
    - type: protein coding
    - GeneCards: NA
    
  * Summary
    - Highly differentiated SNP (Fst = 0.84) affecting the splicing of 1700012B07Rik. 
    - It is a protein coding gene but its function is not well known.
    
  * Literature
    - NA

* **Arsg - ENSMUSG00000020604 - arylsulfatase G**
  * SNP
    - Position: 11	109543817  T	C
    - Fst: 0.77 (rest <= 0.29)
    - Genotypes: 3(./.);22(C/C) vs 3(./.);2(C/C);6(T/C);14(T/T)
    - Effect: start_lost
    
  * Gene
    - Gene Mean Fst: 0.76 (rest <= 0.26)
    - Number of SNPs in gene: 825
    - Expresion: ubiquitous, incl muscle. 
    - GOBP:	sulfur compound metabolic process	
    - Affected transcript: Arsg-202
    - type: Protein coding
    - GeneCards:"The protein encoded by this gene belongs to the sulfatase enzyme family. Sulfatases hydrolyze sulfate esters from sulfated steroids, carbohydrates, proteoglycans, and glycolipids. They are involved in hormone biosynthesis, modulation of cell signaling, and degradation of macromolecules. This protein displays arylsulfatase activity at acidic pH, as is typical of lysosomal sulfatases, and has been shown to localize in the lysosomes. Alternatively spliced transcript variants have been found for this gene"; "ARSG (Arylsulfatase G) is a Protein Coding gene. Diseases associated with ARSG include Usher Syndrome, Type Iv and Usher Syndrome. Among its related pathways are Lysosome and Gamma carboxylation, hypusine formation and arylsulfatase activation. Gene Ontology (GO) annotations related to this gene include sulfuric ester hydrolase activity and arylsulfatase activity. An important paralog of this gene is ARSA."; "Displays arylsulfatase activity at acidic pH with pseudosubstrates, such as p-nitrocatechol sulfate and also, but with lower activity, p-nitrophenyl sulfate and 4-methylumbelliferyl sulfate."
    
  * Summary
    - The mutation is mid-high differentiated. Affecting translation by modifying the start codon (start lost).
    - The gene is an enzyme (sulfatase) involved in hormone biosynthesis, modulation of cell signaling, and degradation of macromolecules. 
    - This gene is found in the literature mostly associated to a disease called dystonia.
    
  * Literature
    - https://pubmed.ncbi.nlm.nih.gov/30643666/
      - "Risk Factor Genes in Patients with Dystonia: A Comprehensive Review"
      - "A total of 43 published studies regarding TOR1A, BDNF, DRD5, APOE, ARSG, NALC, OR4X2, COL4A1, TH, DDC, DBH, MAO, COMT, DAT, GCH1, PRKRA, MR-1, SGCE, ATP1A3, TAF1, THAP1, GNAL, DRD2, HLA-DRB, CBS, MTHFR, and MS genes, were included in the current review."
      
    - https://pubmed.ncbi.nlm.nih.gov/25825126/
      - "Accumulation of rare variants in the arylsulfatase G (ARSG) gene in task-specific dystonia" 2014
      - "In conclusion, we did not detect any conclusive mutation in ARSG. However, we showed an association with rs61999318 in patients with writer's cramp that contributed to an overall enrichment for rare, protein-changing variants in these patients. Thus, our data provide further support for a role of ARSG variants in task-specific dystonia, especially writer's cramp."
      
    - https://pubmed.ncbi.nlm.nih.gov/20679209/
      - "A canine Arylsulfatase G (ARSG) mutation leading to a sulfatase deficiency is associated with neuronal ceroid lipofuscinosis" 2010
      - "We eventually identified a worldwide breed-specific variant in exon 2 of the Arylsulfatase G (ARSG) gene, which causes a p.R99H substitution in the vicinity of the catalytic domain of the enzyme. In transfected cells or leukocytes from affected dogs, the missense change leads to a 75% decrease in sulfatase activity, providing a functional confirmation that the variant might be the NCL-causing mutation. Our results uncover a protein involved in neuronal homeostasis, identify a family of candidate genes to be screened in patients with Kufs' disease, and suggest that a deficiency in sulfatase is part of the NCL pathogenesis."
      
    - https://pubmed.ncbi.nlm.nih.gov/31731261/
      - "Whole genome sequencing for the genetic diagnosis of heterogenous dystonia phenotypes" 2019
      - "We found an association between the known risk variant ARSG rs11655081 and dystonia (p = 0.003)."
    
    - https://pubmed.ncbi.nlm.nih.gov/26975023/
      - "Degeneration of Photoreceptor Cells in Arylsulfatase G-Deficient Mice" 2016
      - "his is the first study demonstrating that ARSG deficiency results in progressive photoreceptor degeneration and dysregulation of various lysosomal proteins"
      
    - https://pubmed.ncbi.nlm.nih.gov/19960030/
      - "Genetic variations in ATP2B1, CSK, ARSG and CSMD1 loci are related to blood pressure and/or hypertension in two Korean cohorts" 2010
      - "Despite the difficulty of obtaining replication results for a complex trait genetic association between blood pressure and hypertension, we were able to identify consistent genetic factors in both the Korean cohorts in ATP2B1, CSK, ARSG and CSMD1 genes."

* **Gm11684 - ENSMUSG00000085734 - predicted gene 11684 **
  * SNP
    - Position: 11	109768595 G	A	
    - Fst: 0.86 (rest <= 0.16)
    - Genotypes: 4(./.);21(A/A) vs 2(./.);1(A/A);4(G/A);18(G/G)
    - Effect: splice_donor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.85 (rest <= 0.16)
    - Number of SNPs in gene: 37
    - Expresion: testis and medial nasal prominence
    - GOBP:	NA
    - Affected transcript: Gm11684-201
    - type: Antisense
    - GeneCards: NA
    
  * Summary
    - SNP affecting transcription (splice_donor_variant&intron_variant) with high differentiation (0.86)
    - The predicted-antisense gene is highly differentiated as well (mean Fst 0.85 over 37 SNPs).
    - There is no information for this gene in the literature.
    
  * Literature
    - NA

* **Map2k6 - mitogen-activated protein kinase kinase 6 - ENSMUSG00000020623**
  * SNP
    - Position: 11	110524602	A	C
    - Fst: 0.86
    - Genotypes: 1(./.);24(C/C) vs 4(./.);15(A/A);6(A/C)
    - Effect: splice_acceptor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.76 (rest <= 0.35)
    - Number of SNPs in gene: 1107
    - Expresion: uniquitous.
    - GOBP:	MAPK cascade	, activation of MAPK activity	, positive regulation of protein phosphorylation	, response to ischemia	, protein phosphorylation	, apoptotic process	, phosphorylation, peptidyl-tyrosine phosphorylation	, ovulation cycle process	, positive regulation of prostaglandin secretion	, response to drug	, positive regulation of apoptotic process	, positive regulation of nitric-oxide synthase biosynthetic process	, cardiac muscle contraction	, cellular response to sorbitol	, negative regulation of cold-induced thermogenesis	
    - Affected transcript: Map2k6-202  
    - type: Protein coding
    - GeneCards:
      - "This gene encodes a member of the dual specificity protein kinase family, which functions as a mitogen-activated protein (MAP) kinase kinase. MAP kinases, also known as extracellular signal-regulated kinases (ERKs), act as an integration point for multiple biochemical signals. This protein phosphorylates and activates p38 MAP kinase in response to inflammatory cytokines or environmental stress. As an essential component of p38 MAP kinase mediated signal transduction pathway, this gene is involved in many cellular processes such as stress induced cell cycle arrest, transcription activation and apoptosis. [provided by RefSeq, Jul 2008]" ; 
      - "MAP2K6 (Mitogen-Activated Protein Kinase Kinase 6) is a Protein Coding gene. Diseases associated with MAP2K6 include Cardiomyopathy, Familial Hypertrophic, 25 and Ischemia. Among its related pathways are Regulation of TP53 Activity through Acetylation and Nucleotide-binding domain, leucine rich repeat containing receptor (NLR) signaling pathways. Gene Ontology (GO) annotations related to this gene include transferase activity, transferring phosphorus-containing groups and protein tyrosine kinase activity. An important paralog of this gene is MAP2K3."; 
      - "Dual specificity protein kinase which acts as an essential component of the MAP kinase signal transduction pathway. With MAP3K3/MKK3, catalyzes the concomitant phosphorylation of a threonine and a tyrosine residue in the MAP kinases p38 MAPK11, MAPK12, MAPK13 and MAPK14 and plays an important role in the regulation of cellular responses to cytokines and all kinds of stresses. Especially, MAP2K3/MKK3 and MAP2K6/MKK6 are both essential for the activation of MAPK11 and MAPK13 induced by environmental stress, whereas MAP2K6/MKK6 is the major MAPK11 activator in response to TNF. MAP2K6/MKK6 also phosphorylates and activates PAK6. The p38 MAP kinase signal transduction pathway leads to direct activation of transcription factors. Nuclear targets of p38 MAP kinase include the transcription factors ATF2 and ELK1. Within the p38 MAPK signal transduction pathway, MAP3K6/MKK6 mediates phosphorylation of STAT4 through MAPK14 activation, and is therefore required for STAT4 activation and STAT4-regulated gene expression in response to IL-12 stimulation. The pathway is also crucial for IL-6-induced SOCS3 expression and down-regulation of IL-6-mediated gene induction; and for IFNG-dependent gene transcription. Has a role in osteoclast differentiation through NF-kappa-B transactivation by TNFSF11, and in endochondral ossification and since SOX9 is another likely downstream target of the p38 MAPK pathway. MAP2K6/MKK6 mediates apoptotic cell death in thymocytes. Acts also as a regulator for melanocytes dendricity, through the modulation of Rho family GTPases. "
    
  * Summary
    - SNP with high differentiation (0.86) affecting gene splicing.
    - The gene is mid-highly differentiated (0.76). It is a protein coding gene with uniquitous expression.
    - The protein is involved in many function as part of MAPK signaling pathway.
    - However, mutations of this gene have been reported to be associated with carcass type in bovine.
    
  * Literature (Map2k6 muscle)
    - https://pubmed.ncbi.nlm.nih.gov/21617946/
      - "Association of bovine carcass phenotypes with genes in an adaptive thermogenesis pathway" 2011
      - "Carcass weight (CW) was found to be associated with MAP2K6 and UCP2 genes; back fat thickness (BFT) was found to be associated with PPARGC1A, MAP2K6, and UCP2 genes; marbling score (MS) was found to be associated with PPARGC1A and MAP2K6 genes"
      - "Further analyses found significant associations of interactions between PPARGC1A and MAP2K6 genes with CW and MS. Especially, interactive genetic associations were identified between c.424 and 222 G>A in PPARGC1A and c.17-10118 T>G in MAP2K6 and between c.228+28619 A>G in PPARGC1A and c.17-10118 T>G in MAP2K6, and they were both detected for CW and MS at a significant level (P < 0.05). The current study suggested that the individual and interactive associations of PPARGC1A, MAP2K6, and UCP2 genes with carcass phenotypes might be resulted from the pathway with fat and energy metabolism through the adaptive thermogenesis."

---

*Candidate Genes DUhLB_vs_FZTDU*

* **Nisch - nischarin - ENSMUSG00000021910**

  * SNP
    - Position: 14:31186720 (C/T)
    - Fst: 0.9, followed by DU6P_vs_FZTDU (0.17), rest all NAs
    - Genotypes: 5(C/T);20(T/T)	vs 25(C/C)
    - Effect: stop_gained
    
  * Gene
    - Gene Mean Fst: 0.89
    - GOBP: glucose metabolic process, apoptotic process, regulation of blood pressure,	Rac protein signal transduction,	actin cytoskeleton organization,	negative regulation of cell migration,	regulation of synaptic transmission, GABAergic	norepinephrine secretion	
    - Affected transcript: the mutation affects exons in three transcripts (Nisch-208, Nisch-207, Nisch-212). The transcript Nisch-207 is labeled as "nonsense mediated decay" (nonsense mutation = stop gained)
  
  * Summary
    - Nish contains a stop-gained point mutation in high differentiation relative to FZTDU (Fst=0.9).
    - The gene is highly differentiated as well (mean Fst = 0.89 over 56 genes)
    - Nish affects glucose metabolism and body size. 
    
  * Literature
    - https://www.genecards.org/cgi-bin/carddisp.pl?gene=NISCH
      - This gene encodes a nonadrenergic imidazoline-1 receptor protein that localizes to the cytosol and anchors to the inner layer of the plasma membrane. The orthologous mouse protein has been shown to influence cytoskeletal organization and cell migration by binding to alpha-5-beta-1 integrin. In humans, this protein has been shown to bind to the adapter insulin receptor substrate 4 (IRS4) to mediate translocation of alpha-5 integrin from the cell membrane to endosomes. Expression of this protein was reduced in human breast cancers while its overexpression reduced tumor growth and metastasis; possibly by limiting the expression of alpha-5 integrin. **In human cardiac tissue, this gene was found to affect cell growth and death while in neural tissue it affected neuronal growth and differentiation**. Alternative splicing results in multiple transcript variants encoding differerent isoforms. Some isoforms lack the expected C-terminal domains of a functional imidazoline receptor. [provided by RefSeq, Jan 2013]
  
    - https://pubmed.ncbi.nlm.nih.gov/30546133/
      - "Development of insulin resistance in Nischarin mutant female mice" 2019
        - "Our previous studies have shown that Nisch mutant male mice increased glucose tolerance in chow-fed conditions"
        - "We found that **insulin signaling was impaired** in major insulin-sensitive tissues of Nisch mutant female mice. When mice were fed with HFD for 15 weeks, the Nisch mutant female mice not only developed severe **insulin resistance and decreased glucose tolerance** compared with wild-type control mice, but also **accumulated more white fat, had larger adipocytes and developed severe hepatic steatosis** than wild-type control mice"
        
    - https://pubmed.ncbi.nlm.nih.gov/28842496/  
      - "Nischarin inhibition alters energy metabolism by activating AMP-activated protein kinase" 2017
        - "Nisch-mutant embryos exhibited delayed development, characterized by **small size and attenuated weight gain**. We uncovered the reason for this phenotype by showing that **Nisch binds to and inhibits the activity of AMP-activated protein kinase (AMPK), which regulates energy homeostasis by suppressing anabolic and activating catabolic processes**. The Nisch mutations enhanced AMPK activation and inhibited mechanistic target of rapamycin signaling in mouse embryonic fibroblasts as well as in muscle and liver tissues of mutant mice. Nisch-mutant mice also exhibited **increased rates of glucose oxidation with increased energy expenditure, despite reduced overall food intake**. Moreover, the Nisch-mutant mice had **reduced expression of liver markers of gluconeogenesis associated with increased glucose tolerance. As a result, these mice displayed decreased growth and body weight.**"
    
    - https://pubmed.ncbi.nlm.nih.gov/20935629/
      - "Meta-analysis identifies 13 new loci associated with waist-hip ratio and reveals sexual dimorphism in the genetic basis of fat distribution" 2011
        - "Waist-hip ratio (WHR) is a measure of body fat distribution and a predictor of metabolic consequences independent of overall adiposity. WHR is heritable, but few genetic variants influencing this trait have been identified. We conducted a meta-analysis of 32 genome-wide association studies for WHR adjusted for body mass index (comprising up to 77,167 participants), following up 16 loci in an additional 29 studies (comprising up to 113,636 subjects). We identified 13 new loci in or near RSPO3, VEGFA, TBX15-WARS2, NFE2L3, GRB14, DNM3-PIGC, ITPR2-SSPN, LY86, HOXC13, ADAMTS9, ZNRF3-KREMEN1, NISCH-STAB1 and CPEB4 (P = 1.9 × 10⁻⁹ to P = 1.8 × 10⁻⁴⁰) and the known signal at LYPLAL1. Seven of these loci exhibited marked sexual dimorphism, all with a stronger effect on WHR in women than men (P for sex difference = 1.9 × 10⁻³ to P = 1.2 × 10⁻¹³). These findings provide evidence for multiple loci that modulate body fat distribution independent of overall adiposity and reveal strong gene-by-sex interactions."

    - https://pubmed.ncbi.nlm.nih.gov/29996974/
      - "Activation of imidazoline receptor I 2, and improved pancreatic β-cell function in human islets" 2018
        - "RNA-sequencing data revealed that NISCH is highly expressed in fat tissues, islets, liver and muscles, with eight detectable splice variants of transcripts in islets"

* **Gm26789 - ENSMUSG00000096991 - "predicted gene, 26789 "**
  * SNP
    - Position: 10	63292683	T/C
    - Fst: 0.89 (rest all NAs)
    - Genotypes: 1(./.);19(C/C);5(T/C) vs	1(./.);24(T/T)
    - Effect: splice_acceptor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.89 (over 9 SNPs)
    - GOBP: NA
    - Expression: NA
    - Affected transcript: Gm26789-201 
    - Processed transcript -> Long non-coding RNA (lncRNA) -> Antisense: Transcripts that overlap the genomic span (i.e. exon or introns) of a protein-coding locus on the opposite strand.
    - Expressed in most tissues (http://www.informatics.jax.org/marker/MGI:5477283)

  * Summary
    - little information available for this gene
    
  * Literature
    - NA
    
* **Gm7075	ENSMUSG00000048185 "predicted gene 7075"**
  * SNP
    - Position: 10	63421672	G/T
    - Fst: 0.90 (rest all NAs)
    - Genotypes: 5(G/T);20(T/T)	25(G/G)
    - Effect: stop_gained
    
  * Gene
    - Gene Mean Fst: 0.90 (over 4 SNPs)
    - GOBP: ubiquitin-dependent protein catabolic process, protein ubiquitination	
    - Expression: incl. alimentary, endocrine, cardiovascular, muscle,  nervous systems. http://www.informatics.jax.org/marker/MGI:3646561
    - Affected transcript: Gm7075-201 ... apparently the only transcript.
    - Type: Protein coding 

  * Summary
    - A predicted gene with not much information
    - Widespread expression throughout systems.
    
  * Literature  
    - NA
    
*  **9630013K17Rik - ENSMUSG00000086359 - "RIKEN cDNA 9630013K17 gene"**   
  * SNP
    - Position: 11	63966373	G	/ T
    - Fst: 0.96 (rest max 0.04)
    - Genotypes: 3(./.);22(T/T)	vs 23(G/G);2(G/T)
    - Effect: splice_acceptor_variant&intron_variant
    
  * Gene
    - Gene Mean Fst: 0.96	over 13 SNPs (rest max 0.24 in DUK ... not that low for DUK)
    - GOBP: NA
    - Expression: uniquitous expression http://www.informatics.jax.org/marker/MGI:1926133
    - Affected transcript:
    - Type: Processed transcript -> Long non-coding RNA (lncRNA) -> Antisense: Transcripts that overlap the genomic span (i.e. exon or introns) of a protein-coding locus on the opposite strand.

  * Summary
    - Again very little information for this ubiquitous non coding gene.
  
  * Literature
    - "Lipotoxic Injury Differentially Regulates Brain Microvascular Gene Expression in Male Mice" 2020 
      - google scholar "… Gm24771 11.03 Gm12603 * 2.59 Gm22504 8.42 D5Ertd605e 2.54 Snord14a* 7.54
9330102E08Rik 2.51 Gm23527 5.66 9230105E05Rik 2.43 Gm26148 4.58 9630013K17Rik
2.43 Gm25376 4.27 4933432I03Rik 2.36 Gm25432 3.18 4930565D16Rik 2.35" --> need to read more in detail, text doesnt show gene.






RED-genes_x_DEGs {data-navmenu="REDs"}
================================================

Column {.tabset}
------------------

### Summary 
```{r find_and_display_REDs_x_DEGs} 
#FL1=DUK
#FL2=DUC

# Load data

## FL1 (DUK) - testis
FL1_testis_DEGs <- read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/DEGs testis FL1.csv"),
  header = FALSE
) %>% 
  dplyr::select("V11") %>% 
  filter(V11 != "Gene Symbol" & V11 != "") 

## FL2 (DUC) - testis
FL2_testis_DEGs <- read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/Michaelis_DEGs testis FL2.csv"),
  header = FALSE
) %>% 
  dplyr::select("V11") %>% 
  filter(V11 != "Gene Symbol" & V11 != "")

## FL1 (DUK) - ovary
FL1_ovary_DEGs <- read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/Vanselow_FL1 ovary_12864_2008_1500_MOESM1_ESM.csv"),
  header = TRUE
) 

# extract gene ids
FL1_testis_DEGs_gene_symbols <- FL1_testis_DEGs %>%
  apply(1, function(row_i){
  str_split(row_i, ";") %>% unlist()
    }) %>% 
  unlist() %>% 
  str_trim() #just in case

FL2_testis_DEGs_gene_symbols <- FL2_testis_DEGs %>%
  apply(1, function(row_i){
  str_split(row_i, ";") %>% unlist()
    }) %>% 
  unlist() %>% 
  str_trim() #just in case

FL1_ovary_DEGs_gene_symbols <- FL1_ovary_DEGs$Gene.Symbol %>% str_trim()



# Find DEGs in list of REDs
lapply(1:length(RED_genes), function(i){
  gn_ids <- RED_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols) %>% # find overlap with FL1_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RED_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RED genes x DEGs (FL1=DUK testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RED_genes), function(i){
  gn_ids <- RED_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols) %>% # find overlap with FL2_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RED_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RED genes x DEGs (FL2=DUC testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RED_genes), function(i){
  gn_ids <- RED_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols) %>% # find overlap with FL1_ovary_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RED_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RED genes x DEGs (FL1=DUK ovary)") %>% 
  kable_styling(full_width = F)
```


```{r export_tables_for_jenni, eval=FALSE}
lapply(1:length(RED_genes), function(i){
  RED_genes[[i]] %>% 
    as.data.frame() %>% 
    inner_join(
      mmu_gene_set,
      by = c("mcols.gene_id" = "gene_id", "seqnames"="seqname","start","end")
    ) %>% 
    mutate(
      gene_symbol = str_trim(gene_symbol), #just in case
      contrast = names(RED_genes)[i]
      ) %>% 
    dplyr::select(contrast, everything()) %>% 
    dplyr::select(-mcols.gene_biotype, -strand, -width) %>% 
    dplyr::rename(gene_id = mcols.gene_id) %>% 
    filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols) # find overlap with FL1_testis_DEGs
  }) %>% 
  bind_rows() %>% 
  write.csv(here("00_dashboard/data/FL1_testis_DEGs_overlap.csv"))


lapply(1:length(RED_genes), function(i){
  RED_genes[[i]] %>% 
    as.data.frame() %>% 
    inner_join(
      mmu_gene_set,
      by = c("mcols.gene_id" = "gene_id", "seqnames"="seqname","start","end")
    ) %>% 
    mutate(
      gene_symbol = str_trim(gene_symbol), #just in case
      contrast = names(RED_genes)[i]
      ) %>% 
    dplyr::select(contrast, everything()) %>% 
    dplyr::select(-mcols.gene_biotype, -strand, -width) %>% 
    dplyr::rename(gene_id = mcols.gene_id) %>% 
    filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols) # find overlap with FL2_testis_DEGs
  }) %>% 
  bind_rows() %>% 
  write.csv(here("00_dashboard/data/FL2_testis_DEGs_overlap.csv"))


lapply(1:length(RED_genes), function(i){
  RED_genes[[i]] %>% 
    as.data.frame() %>% 
    inner_join(
      mmu_gene_set,
      by = c("mcols.gene_id" = "gene_id", "seqnames"="seqname","start","end")
    ) %>% 
    mutate(
      gene_symbol = str_trim(gene_symbol), #just in case
      contrast = names(RED_genes)[i]
      ) %>% 
    dplyr::select(contrast, everything()) %>% 
    dplyr::select(-mcols.gene_biotype, -strand, -width) %>% 
    dplyr::rename(gene_id = mcols.gene_id) %>% 
    filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols) # find overlap with FL1_ovary_DEGs
  }) %>% 
  bind_rows() %>% 
  write.csv(here("00_dashboard/data/FL1_ovary_DEGs_overlap.csv"))
```

### FL1_testis_DEGs
```{r}
read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/DEGs testis FL1.csv"),
  header = TRUE,
  skip=1
) %>% 
  dplyr::rename(
   fc_1st_set = FC..1st.set.,
   fc_2nd_set = FC..2nd.set.
  ) %>% 
  dplyr::select(Gene.Symbol, fc_1st_set, fc_2nd_set) %>% 
  filter(Gene.Symbol != "") %>% 
  datatable(caption = "Fold Changes of All Original DEGs", rownames = FALSE, 
            options = list(
              pageLength = 100, 
              columnDefs = list(list(className = 'dt-center', targets = "_all"))
              )
            )

```

### FL2_testis_DEGs
```{r}
read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/Michaelis_DEGs testis FL2.csv"),
  header = TRUE,
  skip=1
) %>% 
  dplyr::rename(
   fc_1st_set = FC.1st.array..set,
   fc_2nd_set = FC.2nd.array..set
  ) %>% 
  dplyr::select(Gene.Symbol, fc_1st_set, fc_2nd_set) %>% 
  filter(Gene.Symbol != "") %>% 
  datatable(caption = "Fold Changes of All Original DEGs", rownames = FALSE, 
            options = list(
              pageLength = 100, 
              columnDefs = list(list(className = 'dt-center', targets = "_all"))
              )
            )
```

### FL1_ovary_DEGs
```{r}
read.csv(
  here("00_dashboard/data/Microarray_data_FL1_FL2/Vanselow_FL1 ovary_12864_2008_1500_MOESM1_ESM.csv"),
  header = TRUE
) %>% 
  dplyr::select(Gene.Symbol, FC) %>% 
  filter(Gene.Symbol != "") %>% 
  datatable(caption = "Fold Changes of All Original DEGs", rownames = FALSE, 
            options = list(
              pageLength = 100, 
              columnDefs = list(list(className = 'dt-center', targets = "_all"))
              )
            )
```


RED-reg_SNPs  {data-navmenu="REDs"}
============================================


```{r find_RED_SNPs_in_regulatory_regions}

# find 
reg_SNPs_in_REDs <- lapply(1:length(REDs), function(i){ 

  #i=1
  
  # get gr for contrast
  gr <- REDs[[i]]
  
  # get SNPs in REDs
  SNPs_in_REDs <- subsetByOverlaps(x = vcf_final_snp_ids_gr, ranges = gr, minoverlap = 1) 
  
  # Find overlaps between SNPs in REDs and regulatory or motif features
  ov_rf <- findOverlaps(query = SNPs_in_REDs, subject = regulatory_features, minoverlap = 1)

  # subset overlapping snps
  SNPs_in_REDs_in_rf <- SNPs_in_REDs[queryHits(ov_rf)]

  # add feature information
  mcols(SNPs_in_REDs_in_rf) <- regulatory_features[subjectHits(ov_rf)] %>% 
    as.data.frame() %>% 
    dplyr::rename(
      feat_seqname = seqnames, feat_start=start, feat_end=end, feat_width=width, feat_strand = strand,
      feat_source = mcols.source, feat_feature=mcols.feature, feat_attribute=mcols.attribute
      ) 
  
  # import per SNP Fst scores and convert to granges
  tmp <- vroom(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output", paste0(names(REDs)[i],".weir.fst")),
    col_types = c(CHROM = "c")
    )
  
  # add fst information to SNPs-in-REDs-in-regulatory-regions
  SNPs_in_REDs_in_rf <- SNPs_in_REDs_in_rf %>% 
    # turn back to data frame
    as.data.frame() %>% 
    # inner join with fst data
    inner_join(tmp, by = c("seqnames" = "CHROM", "start" = "POS")) %>% 
    # turn back into granges
    makeGRangesFromDataFrame(keep.extra.columns=TRUE)

  return(SNPs_in_REDs_in_rf)  
  
  }
) 


names(reg_SNPs_in_REDs) <- names(REDs)
```

Column {data-width=300}
------------------------

### Overview

* SNPs found in REDs (RED-SNPs).

* Regulatory regions (ensembl regulatory build) containing RED-SNPs. 

* Downstream genes to the above.


Column  {data-width=700, .tabset}
-----------------------------------

### Number of regSNPs

```{r display_nr_hits}

lapply(reg_SNPs_in_REDs, length) %>% as.data.frame() %>% 
  #mutate(contrast = factor(contrast, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU"))) %>% 
  dplyr::select(paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU")) %>% 
  kable(digits = 2, caption = "Number of regSNPs in REDs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


lapply(reg_SNPs_in_REDs, 
       function(x){
         x %>% 
           as.data.frame() %>% 
           dplyr::select(feat_seqname, feat_start, feat_end) %>% 
           unique() %>% 
           nrow()
         }
       ) %>% 
  as.data.frame() %>% 
  dplyr::select(paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU")) %>% 
  kable(digits = 2, caption = "Number of regulatory regions containing RED-SNPs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


```


### Downstream_genes
```{r find_downstream_genes_for_regSNPs_regions, eval = FALSE}

#load bioconductor gene set
mm10_ensGene <- annotateTranscripts(TxDb.Mmusculus.UCSC.mm10.ensGene) 

# for every set of regSNPs find the nearest downstream gene

downstream_genes <- lapply(1:length(reg_SNPs_in_REDs), function(i){

  print(names(reg_SNPs_in_REDs)[i])
  reg_snps <- reg_SNPs_in_REDs[[i]]
  
  # extract regulatory feature information 
  reg_range <- GRanges(seqnames = paste0("chr",reg_snps$feat_seqname), 
                     IRanges(start = reg_snps$feat_start, end = reg_snps$feat_end))
  
  # add feature width
  reg_range$feature_width <- reg_snps$feat_width


  # extract feature type information and make it a column
  reg_range$feature_type <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("feature_type",x)] %>% str_remove("feature_type=") )

  # extract regulatory ensembl id and make it a column
  reg_range$feature_id <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("ID=regulatory_region:",x)] %>% str_remove("ID=regulatory_region:") )

  # many entries are repeated, because >1 SNP was found in a regulatory region
  reg_range_uniq <- reg_range %>% unique()
  
  # search for nearest gene downstream
  tab <- matchGenes(reg_range_uniq, mm10_ensGene, skipExons=TRUE, type = "fiveprime") %>% 
    # add biotype and gene symbol  
    left_join(
      mmu_gene_set %>% dplyr::select(gene_id, gene_symbol, gene_biotype), 
      by = c("Geneid"="gene_id")
      ) 

  # add gene information to matching regulatory region
  reg_range_uniq$matching_gene_id <- tab$Geneid
  reg_range_uniq$gene_symbol <- tab$gene_symbol
  reg_range_uniq$gene_biotype <- tab$gene_biotype
  reg_range_uniq$gene_strand <- tab$strand
  reg_range_uniq$distance_to_gene <- tab$distance

return(reg_range_uniq)
})

# name elements in the list
names(downstream_genes) <- names(reg_SNPs_in_REDs)


# export for later use
lapply(names(downstream_genes), function(nm){
  nm_fl <- paste0("RED_reg_features_",nm,".csv")
  downstream_genes[[nm]] %>% write.csv(here("00_dashboard/data/regulatory_RED_SNPs",nm_fl))
})

```



```{r display_feature_tabulation_tyoe_of_feature_by_contrast}

lapply(1:length(reg_SNPs_in_REDs), function(i){
  nm <- names(reg_SNPs_in_REDs)[i]
  nm_fl <- paste0("RED_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RED_SNPs",nm_fl)) %>% 
    group_by(feature_type) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm)
}) %>% 
  bind_rows() %>% 
  reshape2::dcast(feature_type ~ contrast, value.var = "n") %>% 
  dplyr::select(feature_type, paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU")) %>% 
  kable(digits = 2, 
        caption = "Number of regulatory regions (with >=1 RED-SNP) by contrast",
        format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)

```

```{r display_feature_tabulation_tyoe_of_feature_by_genebiotype}
#lapply(1:length(reg_SNPs_in_REDs), function(i){
#  nm <- names(reg_SNPs_in_REDs)[i]

contr_nms <- paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU")

tabs <- lapply(contr_nms, function(nm){
  nm_fl <- paste0("RED_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RED_SNPs",nm_fl)) %>% 
    group_by(feature_type, gene_biotype) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm) %>% 
    reshape2::dcast(gene_biotype ~ feature_type, value.var = "n") %>% 
    filter(!is.na(gene_biotype)) 
}) 


print_kable <- function(tab, contr_nm){
  ttl <- paste0("Number of regulatory regions (with >=1 RED-SNP) by gene biotype (", contr_nm, ")")

  tab %>% 
    kable(digits = 2, caption = ttl, format.args = list(big.mark = ",")) %>% 
    kable_styling(full_width = F)
}

print_kable(tabs[[1]], contr_nms[1])
print_kable(tabs[[2]], contr_nms[2])
print_kable(tabs[[3]], contr_nms[3])
print_kable(tabs[[4]], contr_nms[4])
print_kable(tabs[[5]], contr_nms[5])
print_kable(tabs[[6]], contr_nms[6])

```



RDDs {data-navmenu="RDDs"}
====================================

Column {.tabset}
-----------------

### Overview

* Regions of Distinct Differentiation (RDDs)

* RDDs = Windows in the top 5% of a given contrast that are found in the bottom 10% of all other contrasts.

* Distinct genes were those overlapping RDDs (RDD-genes).

* Functional annotation of RDD genes was done using biomart to collect associated biological processes (GOBP).

* For each gene SNPs were identified and their effects summarized.

* Likewise, for every SNP in a RDD-gene, their Fst scores and effects are displayed.

### Summaries 

```{r define_contrasting_quantiles}
quantiles <- c("q10","q95") 
```

```{r define_function_to_find_distinct_REDs} 

summary_z_win_fst <- readRDS( here("00_dashboard","summary_z_win_fst.rds") )

find_distinct_REDs <- function(contr, rest, q_low, q_high){
  
  #contr="DUK_vs_FZTDU"
  #rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" )
  #q_low = "q10"
  #q_high = "q90"
  
  # for each contrast in rest, find bottom 
  rest_bottom_windows <- lapply(rest, function(x){
    
    # set threshold for bottom scores
    q_low_threshold <- summary_z_win_fst[summary_z_win_fst$contrast == x, q_low] %>% as.numeric()
    
    # get rest-contrast individual data
    win_fst_sel_vs_ctrl_prep %>% filter(contrast %in% x) %>% 
      # get bottom scores using zFst scores
      filter(z_win_fst_score < q_low_threshold) %>% 
      ungroup() %>% 
      # select columns for joining
      dplyr::select(CHROM, BIN_START, BIN_END)
    }) %>% 
    # inner join all contrasts in rest
    purrr::reduce(inner_join, by = c("CHROM","BIN_START","BIN_END"))
  
  # set upper score
  q_high_threshold <- summary_z_win_fst[summary_z_win_fst$contrast == contr, q_high] %>% as.numeric()
  
  # get top windows in contrast of interest
  contr_extreme_windows <- win_fst_sel_vs_ctrl_prep %>% 
    filter(
      # get contrast of interest data
      contrast == contr &
        # keep top zFst scores
        z_win_fst_score > q_high_threshold
    ) %>% 
    # select columns for merging
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)

  # get distinct windows
  distinct_windows <- inner_join(contr_extreme_windows, 
                                 rest_bottom_windows, 
                                 by = c("CHROM","BIN_START","BIN_END")) %>% 
    # add fst and number of variants information
    inner_join(
      win_fst_sel_vs_ctrl_prep %>% 
        filter(contrast == contr) %>% 
        dplyr::select(contrast, CHROM, BIN_START, BIN_END, N_VARIANTS, MEAN_FST),
      by = c("CHROM","BIN_START","BIN_END")
      ) %>% 
    mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
    arrange(CHROM)
  
  return(distinct_windows)
}


distinct_windows_list <- list(
  DUK = find_distinct_REDs(
    contr = "DUK_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DUC = find_distinct_REDs(
    contr = "DUC_vs_FZTDU", rest = paste0( c("DUK","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DU6 = find_distinct_REDs(
    contr = "DU6_vs_FZTDU", rest = paste0( c("DUC","DUK","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DU6P = find_distinct_REDs(
    contr = "DU6P_vs_FZTDU", rest = paste0( c("DUC","DU6","DUK","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DUhLB = find_distinct_REDs(
    contr = "DUhLB_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUK"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  FERT = find_distinct_REDs(
    contr = "FERT_vs_FZTDU", rest = paste0( c("DU6","DU6P","DUhLB"), "_vs_FZTDU" ),  quantiles[1], quantiles[2]
    )
)

```

```{r display_summary_distinct_regions}

summarise_distinct_win <- function(line_sel, rest){

#args
#line_sel = "DUK"
#rest = c("DUC","DU6","DU6P","DUhLB")

  #{
  x <- distinct_windows_list[[line_sel]]

  xx <- lapply(rest, function(y){
    #y="DUC"
    win_fst_sel_vs_ctrl_prep %>% 
      filter(contrast == paste0(y,"_vs_FZTDU")) %>% 
      inner_join(
        x %>% dplyr::select(-MEAN_FST), 
        by = c("CHROM","BIN_START","BIN_END")) %>% 
      .$MEAN_FST 
  }) %>% 
    unlist() %>% 
    range() %>%
    round(2) %>% 
    paste0(collapse = " - ")
    

  data.frame(
    contrast = x$contrast %>% unique(),
    n_windows = nrow(x),
    chrs = unique(x$CHROM) %>% paste(collapse = ","),
    n_win_by_chr = x %>% 
      group_by(CHROM) %>% 
      summarise(n=n()) %>%
      ungroup() %>% 
      mutate(tmp =  paste0("chr",CHROM,"(",n,")")  ) %>% 
      dplyr::select(tmp) %>% 
      unlist() %>% 
      paste(collapse = "; "),
    n_vars_range = range(x$N_VARIANTS) %>% paste(collapse = " - "),
    fst_range = range(x$MEAN_FST) %>% round(2) %>% paste(collapse = " - "),
    fst_range_rest = xx,
    rest = paste(rest, collapse = ", ")
    ) 
  }

bind_rows(
  summarise_distinct_win("DUK", c("DUC","DU6","DU6P","DUhLB")),
  summarise_distinct_win("DUC", c("DUK","DU6","DU6P","DUhLB")),
  summarise_distinct_win("DU6", c("DUC","DUK","DU6P","DUhLB")),
  summarise_distinct_win("DU6P", c("DUC","DU6","DUK","DUhLB")),
  summarise_distinct_win("DUhLB", c("DUC","DU6","DU6P","DUK")),
  summarise_distinct_win("FERT", c("DU6","DU6P","DUhLB"))
  ) %>% 
  kable(digits = 2, caption = paste0("Summary Distinct Windows (",quantiles[1]," vs ", quantiles[2], ")" )) %>% 
  kable_styling(full_width = F)


```

```{r get_SNPs_in_RDDs_and_display}
snps_in_RDDs <- lapply(1:length(distinct_windows_list), function(i){
  #i=1

  x <- distinct_windows_list[[i]] %>% 
    dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
    makeGRangesFromDataFrame()
  
  snps <- subsetByOverlaps(x = vcf_final_snp_ids_gr, ranges = x, minoverlap = 1) %>% unique()
  
  return(snps)
})

names(snps_in_RDDs) <- names(distinct_windows_list)

data.frame(
    contrast = names(snps_in_RDDs),
    Nr_RDD_SNPs = sapply(snps_in_RDDs, length)
  ) %>% 
  kable(digits = 2, caption = "Number of SNPs in RDDs", row.names = FALSE) %>% 
  kable_styling(full_width = F)

```

```{r find_genes_in_distinct_windows}

RDD_genes <- lapply(distinct_windows_list, function(x){
  
  #x=distinct_windows_list$DUK
  
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))

  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()

})

```

```{r find_SNPs_in_RDD_genes_and_export, eval = FALSE}  

# this will not be shown in the dashboard

snps_in_RDD_genes <- lapply(RDD_genes, function(x)
  
  subsetByOverlaps(vcf_final_snp_ids_gr, x, minoverlap = 1) %>% unique()
  
)

snps_in_RDD_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows() %>% 
  mutate(interval = paste0(seqnames,":",start,"-",end)) %>% 
  dplyr::select(interval) %>% 
  unique() %>% 
  write.table(
    here("00_dashboard/data/RDD_genes","snps_in_RDD_genes.intervals"), 
    col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE
  )

```

```{r GT_counts_SNPs_in_RDD_genes, eval = FALSE}

SNPs_in_RDD_genes_GT_tab <- vroom(
    here(
      "batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered_SNPs_in_RDD_genes_GT.table"
      )
  ) 

names(SNPs_in_RDD_genes_GT_tab) <- str_remove(names(SNPs_in_RDD_genes_GT_tab),".GT") %>% str_remove("-L1")

tabulate_genotypes_per_contrast2 <- function(pop){ 
  
  #pop <- "DUK"

  # Get sample ids for each line
  samps_pop <- sample_info %>% filter(Linie %in% pop) %>% .$sample_id
  
  # tabulate genotypes of selected line
  sel_GT_cts <- SNPs_in_RDD_genes_GT_tab %>% 
    dplyr::select(all_of(samps_pop)) %>% 
    apply(1, function(x){
      tb <- table(x)
      tb_nms <- names(tb)
      paste0(tb, "(", tb_nms, ")") %>% paste(collapse = ";")
    }) 

  # add locus information
  ss <- SNPs_in_RDD_genes_GT_tab %>% dplyr::select(CHROM, POS)
  
  # add column with GT counts
  new_col <- paste0(paste(pop, collapse = "_"),"_GT_counts")
  ss$tmp <- sel_GT_cts
  names(ss)[names(ss) == "tmp"] <- new_col

  return(ss)  

}

by_cols <- c("CHROM","POS")

tabulate_genotypes_per_contrast2("DUK") %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DUC"), by = by_cols
  ) %>%  
  inner_join(
    tabulate_genotypes_per_contrast2(c("DUK","DUC")), by = by_cols #FERT
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DU6"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DU6P"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DUhLB"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("FZTDU"), by = by_cols
  ) %>% 
  left_join(
    vroom(
      here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids2"), 
      col_names = FALSE 
    ) %>% 
      dplyr::rename(CHROM = X1, POS = X3, REF = X5, ALT = X7) %>% 
      dplyr::select(CHROM, POS, REF, ALT),
    by = by_cols
  ) %>% 
  dplyr::select(CHROM, POS, REF, ALT, everything()) %>% 
  #export for later use
  write.csv(
    here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")
  )
```

```{r display_gene_summary_table}
lapply(RDD_genes, function(x){
  x %>% 
    as.data.frame() %>% 
    dplyr::select(mcols.gene_id, mcols.gene_biotype) %>% 
    unique() %>% 
    group_by( mcols.gene_biotype) %>% 
    summarise(n=n()) 
  }) %>% 
  bind_rows(.id="contrast") %>% 
  reshape2::dcast(mcols.gene_biotype ~ contrast, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  dplyr::rename(gene_biotype=mcols.gene_biotype) %>% 
  dplyr::select(gene_biotype, DUK,DUC,DU6,DU6P,DUhLB,FERT) %>% 
  kable(digits = 2, caption = paste0("Summary Distinct Genes (",quantiles[1]," vs ", quantiles[2], ")" )) %>% 
  kable_styling(full_width = F)
```

```{r add_gene_ontology, eval = FALSE}

# biomart query done on Tuesday 20/10/2020
#mart <- useDataset(dataset = "mmusculus_gene_ensembl", mart = useMart("ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org"))

#x = RDD_genes$FERT

RDD_genes2 <- lapply(1:length(RDD_genes), function(i){ 
  
  #i=1
  
  x <- RDD_genes[[i]]
  
  # do BM quary
  mart_res <- getBM(attributes = c("ensembl_gene_id","mgi_symbol","name_1006","namespace_1003"), 
                  filters = "ensembl_gene_id", 
                  values = x$mcols.gene_id,
                  mart = mart) %>% 
              filter(namespace_1003 == "biological_process") 

  # collapse GOBP terms
  mart_res$ensembl_gene_id %>% 
    unique() %>% 
    lapply(function(y){
      
      gobps <- mart_res %>% 
        filter(ensembl_gene_id == y) %>%
        dplyr::select(name_1006) %>% 
        unlist() %>% 
        paste(collapse = "; ")
      
      data.frame(
        ensembl_gene_id = y,
        mgi_symbol = mart_res$mgi_symbol[mart_res$ensembl_gene_id == y] %>% unique(), 
        GOBP = gobps
        )
    }) %>% 
    bind_rows() %>% 
    full_join(
      x %>% as.data.frame() %>% dplyr::rename(ensembl_gene_id = mcols.gene_id, gene_biotype = mcols.gene_biotype),
      by = "ensembl_gene_id"
    ) %>% 
    mutate(contrast = paste0(names(RDD_genes)[i], "_vs_FZTDU") ) %>% 
    dplyr::select(contrast, ensembl_gene_id,mgi_symbol,seqnames, start, end, width, strand, gene_biotype, GOBP) 
})


names(RDD_genes2) <- names(RDD_genes)

# export for later use offline (biomart query not needed)

RDD_genes2$DUK %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUK.csv"))
RDD_genes2$DUC %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUC.csv"))
RDD_genes2$DU6 %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DU6.csv"))
RDD_genes2$DU6P %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DU6P.csv"))
RDD_genes2$DUhLB %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB.csv"))
RDD_genes2$FERT %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_FERT.csv"))

```

### DUK_genes 
```{r} 
here("00_dashboard/data/RDD_genes/RDD_genes_DUK.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUC_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DUC.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6P_genes
```{r} 
here("00_dashboard/data/RDD_genes/RDD_genes_DU6P.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUhLB_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### FERT_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_FERT.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### RDD-SNPs_effect_counts
```{r} 

display_snp_effects_per_transcript <- function(pop){
  
  csv_fl <- paste0("RDD_genes_",pop,".csv")
  
  here("00_dashboard/data/RDD_genes",csv_fl) %>% 
  read.csv() %>% 
  dplyr::select(contrast, ensembl_gene_id, mgi_symbol, gene_biotype) %>% 
  left_join(
    vroom(here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.summary.genes.txt"),skip = 1),
    by = c("ensembl_gene_id" = "GeneId")
  ) %>% 
  unique() 
  }

lapply(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), display_snp_effects_per_transcript) %>% 
  bind_rows() %>% 
  datatable(options = list(paging = FALSE), caption = "SNP effect counts (SNPs associated to RDD genes (all contrasts) as per SNPeff)")

```

### RDDs_x_REDs_Venn 
```{r visualize_RED_RDD_intersections, fig.width=5, fig.height=8} 

RED_genes <- readRDS(here("00_dashboard/RED_genes.rds"))

display_venn_RED_RDD_genes <- function(red_contr, rdd_contr){
  
  #...
  #red_contr="DUK_vs_FZTDU"
  #rdd_contr="DUK"
  #...
  
  d <- data.frame(
    ids = unique(RED_genes[[red_contr]]$mcols.gene_id, RDD_genes[[rdd_contr]]$mcols.gene_id)
    ) %>% 
    mutate(
      RED_genes = ids %in% RED_genes[[red_contr]]$mcols.gene_id,
      RDD_genes = ids %in% RDD_genes[[rdd_contr]]$mcols.gene_id
    )
  
  vc <- vennCounts(
    d %>% dplyr::select(-ids)
    )
  
  vennDiagram(vc, circle.col = c("red","blue"), mar = rep(0.1,4), cex = 0.7)
  title(red_contr, line = 0, adj = 0.5)
  
  }

par(mfrow = c(3, 2))

display_venn_RED_RDD_genes("DUK_vs_FZTDU","DUK")
display_venn_RED_RDD_genes("DUC_vs_FZTDU","DUC")
display_venn_RED_RDD_genes("DU6_vs_FZTDU","DU6")
display_venn_RED_RDD_genes("DU6P_vs_FZTDU","DU6P")
display_venn_RED_RDD_genes("DUhLB_vs_FZTDU","DUhLB")
display_venn_RED_RDD_genes("FERT_vs_FZTDU","FERT")


```

### RDDgenes_loci 

```{r prepare_data-manhattans, eval = FALSE}
d_man <- win_fst_sel_vs_ctrl_prep %>% 
  dplyr::select(contrast, CHROM, BIN_START, BIN_END, MEAN_FST) %>%
  reshape2::dcast(CHROM + BIN_START + BIN_END ~ contrast, value.var = "MEAN_FST") %>% 
  mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
         center_window_Mb = center_window/1000000) %>% 
  # relevel chromosomes and colors for chromosomes
  mutate(CHROM = factor(CHROM, levels = c(1:19, "X"))) %>% 
  # sort by chromosomes and window position
  arrange(CHROM, center_window_Mb) %>% 
  mutate(genomic_index = 1:nrow(.)) %>% 
  mutate(chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")) %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","center_window","center_window_Mb","genomic_index","chr_color"), 
                 variable.name = "contrast", value.name = "MEAN_FST")

# define ticks for manhattan
ticks <- d_man %>% 
  group_by(CHROM) %>% 
  summarise(chr_median = median(genomic_index), chr_end = max(genomic_index)) 

```

```{r function_make_manhattans, eval = FALSE}
plot_manhattan <- function(pop){

#pop="DUK" #arg
  
  set.seed(456)
  
  #>>>>>>>>>>>>>>>>>>>>>
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
  )
  
  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
  ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 
  
  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )
  
  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()
  
  #>>>>>>>>>>>>>>>>>>>>>
  
  # prepare plot data 
  plot_dat <- d_man %>% 
    # subset by contrast and by min Fst
    filter(contrast == contr, MEAN_FST > 0.6) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )  
  
  # make plot
  p <- ggplot(# main data points ("background" -- black and grey dots)
        data = filter(plot_dat, is.na(gene_symbol)), 
        aes(x = genomic_index, y = MEAN_FST),
        alpha = 0.5, 
        show.legend = FALSE) +
    geom_point(aes(color = chr_color), show.legend = FALSE) +
    # red points for RDD gene windows
    geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
    # RDD gene labels
    geom_label_repel(
      data =  filter(plot_dat, !is.na(gene_symbol)),
      aes(label = gene_symbol),
      segment.color = "red", 
      arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
      fill = alpha(c("white"),0.5)
      ) +
    # Make it look nicer
    theme_bw(base_size = 15) +
    scale_color_manual(values = c("black", "darkgrey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0.01,0.01)) +
    scale_y_continuous(expand = c(0.01,0.01)) +
    xlab("chromosome") +
    ylab(
      bquote(
        Fst[(.(paste0(pop,"_vs_FZTDU")))] > 0.6
        )
      )
   return(p)
}

```

```{r make_manhattan_plots_and_export, eval = FALSE}

fig_path <- "00_dashboard/figures/manhattan_win_fst_rdd_genes_labs"

png(here(fig_path, "duk_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUK"); dev.off() 
png(here(fig_path, "duc_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUC"); dev.off() 
png(here(fig_path, "du6_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DU6"); dev.off() 
png(here(fig_path, "du6p_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DU6P"); dev.off() 
png(here(fig_path, "duhlb_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUhLB"); dev.off() 
png(here(fig_path, "fert_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="FERT"); dev.off() 

```

```{r function_make_flat_manhattans, eval = FALSE}
plot_manhattan_flat <- function(pop){ 

#pop="DUK" #arg
  
  set.seed(45)

  #>>>>>>>>>>>>>>>>>>>>>
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
  )
  
  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
  ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 
  
  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )
  
  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()
  
  #>>>>>>>>>>>>>>>>>>>>>  
  
  # prepare plot data 
  plot_dat <- d_man %>% 
    # subset by contrast and by min Fst
    filter(contrast == contr) %>% 
    # ...
    mutate(y = 0) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )  
  
  # make plot
  p <- ggplot(data =  plot_dat, aes(x = genomic_index, y = y, label = gene_symbol)) +
    geom_point(aes(color = chr_color), show.legend = FALSE, shape = 15) +
    # RDD gene labels
    geom_label_repel(
      data =  filter(plot_dat, !is.na(gene_symbol)),
      aes(label = gene_symbol),
      segment.color = "red", 
      segment.alpha = 0.5,
      arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
      fill = alpha(c("white"),0.5),
      nudge_y = 0.3
      #direction = "x"
    ) +
    # Make it look nicer
    #theme_bw(base_size = 15) +
    scale_color_manual(values = c("black", "darkgrey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM) +
    scale_y_continuous(limits = c(0, 1), expand = c(0.01,0)) +
    xlab(NULL) +
    labs(caption = contr) +
    theme(
      axis.line.y  = element_blank(),
      axis.ticks.y = element_blank(),
      axis.text.y  = element_blank(),
      axis.title.y = element_blank(),
      axis.line.x = element_blank(),
      axis.text.x = element_text(size = 20),
      axis.ticks.x = element_blank(),
      panel.background = element_blank()
    )  
  
   return(p)
}

```

```{r make_flat_manhattan_plots_and_export, eval = FALSE}

png(here(fig_path, "duk_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUK"); dev.off() 
png(here(fig_path, "duc_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUC"); dev.off() 
png(here(fig_path, "du6_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DU6"); dev.off() 
png(here(fig_path, "du6p_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DU6P"); dev.off() 
png(here(fig_path, "duhlb_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUhLB"); dev.off() 
png(here(fig_path, "fert_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="FERT"); dev.off() 

```

```{r display_RDDgenes_loci, out.width = "90%", fig.align = "center"} 

fig_path <- here("00_dashboard/figures/manhattan_win_fst_rdd_genes_labs")

include_graphics(list.files(fig_path, full.names = TRUE) %>% .[grep("_win_fst_rdd_genes_labs_flat.png",.)])

```

```{r function_make_vertical_flat_manhattans_labels_interspaced, eval = FALSE}
plot_manhattan_vertical_flat_interspaced <- function(pop){ 
  
  #pop="DUhLB" #arg
  
  #>>>>>>>>>>>>>>>>>>>>>
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
  )
  
  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
  ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 
  
  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )
  
  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()
  
  # interspace gene labels    
  gene_labs_final$lab_set <- NA
  gene_labs_final$lab_set[seq(1,nrow(gene_labs_final), 2)] <- "1"
  gene_labs_final$lab_set[seq(2,nrow(gene_labs_final), 2)] <- "2"
  
  # prepare plot data 
  plot_dat <- d_man %>% 
    # subset by contrast and by min Fst
    filter(contrast == contr) %>% 
    # ...
    mutate(x = 5) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol, lab_set),
      by = "genomic_index"
    )  %>% 
  mutate(genomic_index = rev(genomic_index))
  
  seg_dat <- plot_dat %>% 
    group_by(CHROM) %>% 
    summarise(ymin = min(genomic_index), ymax = max(genomic_index)) %>% 
    mutate(x = 5, xmax = 10, text_pos = ymin + ((ymax-ymin)/2))
  
  seg_dat$chr_color <- NA
  seg_dat$chr_color[seq(1,nrow(seg_dat), 2)] <- "1"
  seg_dat$chr_color[seq(2,nrow(seg_dat), 2)] <- "2"
  
  
  set.seed(45)
  
  p <- ggplot(data =  plot_dat, aes(x = x, y =  genomic_index)) +
    
    geom_segment(data = seg_dat,  aes(x = x, xend = x, y = ymin, yend = ymax, color = chr_color), 
                 size = 25, show.legend = FALSE) +
    
    geom_text(data = seg_dat, aes(x=x, y = text_pos, label = CHROM), color = "white", size = 15) +
  
    scale_color_manual(values = c("black", "darkgrey")) +
    
    scale_y_continuous(expand = c(0,0)) +
    
    theme_void() +
    
    geom_label_repel(data =  filter(plot_dat, !is.na(gene_symbol), lab_set == "1"),aes(label = gene_symbol),
                     segment.color = "red",segment.alpha = 0.5, segment.size = 1, 
                     nudge_x = 0.05, size = 15, direction = "y", fill = alpha("white",0.5)) +
    
    geom_label_repel(data =  filter(plot_dat, !is.na(gene_symbol), lab_set == "2"),aes(label = gene_symbol),
                     segment.color = "red",segment.alpha = 0.5, segment.size = 1, 
                     nudge_x = -0.05, size = 15, direction = "y", fill = alpha("white",0.5)) 
    
  return(p)
}
```

```{r, make_vertical_flat_manhattans_labels_interspaced_and_export, eval = FALSE}

png(here(fig_path, "duk_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("DUK"); dev.off() 

png(here(fig_path, "duc_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("DUC"); dev.off() 

png(here(fig_path, "du6_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("DU6"); dev.off() 

png(here(fig_path, "du6p_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("DU6P"); dev.off() 

png(here(fig_path, "fert_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("FERT"); dev.off() 

png(here(fig_path, "duhlb_win_fst_rdd_genes_labs_vert_flat.png"),  res = 300, height = 8000, width = 6000); plot_manhattan_vertical_flat_interspaced("DUhLB"); dev.off() 


```

```{r function_zoom_in_manhattan, eval=FALSE}
plot_manhattan_zoom <- function(pop, chr, from_mb, to_mb, nudge_y, height_png, width_png, export_plot=TRUE){
  
  set.seed(478)
  
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")

  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
    )

  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
    ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 

  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )

  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()

  # prepare plot data 
  plot_dat <- d_man %>% 
    filter(CHROM == chr & center_window_Mb > from_mb & center_window_Mb < to_mb) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )
  
  # subset data for gene labels existing
  lab_dat <- plot_dat %>% 
    filter(!is.na(gene_symbol) & contrast == contr) 

  # plot according to number of gene labels available
  if(nrow(lab_dat) > 1){

  # interspace gene labels    
  lab_dat$lab_set <- NA
  lab_dat$lab_set[seq(1,nrow(lab_dat), 2)] <- "1"
  lab_dat$lab_set[seq(2,nrow(lab_dat), 2)] <- "2"
  
    # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "1"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1
      ) +
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "2"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = -nudge_y,
        vjust = 1
      ) + 
      # Make it look nicer
      theme_bw(base_size = 15) +
      theme(panel.background = element_blank()) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2), breaks = seq(0,1,0.2)) +
      xlab("Mb") 
  }else{
          # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  lab_dat,aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1
      ) +
      # Make it look nicer
      theme_bw(base_size = 15) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2), breaks = seq(0,1,0.2)) +
      xlab("Mb") 
  }
  
  if(export_plot){
    # export plot
    fl_nm <- paste0(pop, "_win_fst_rdd_genes_labs_chr",chr,"_from",from_mb,"mb_to", to_mb, "mb.png")
    png(here(fig_path, fl_nm), res = 300, height = height_png, width = width_png)
    print(p)
    dev.off()
  }else{
    print(p)
  }
  

}

```

```{r make_and_export_goi_zoom_labs_manhattan, eval = FALSE}

# DUK 
plot_manhattan_zoom(pop="DUK", chr=1, from_mb=24.9, to_mb=25.05, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=2, from_mb=28.25, to_mb=29.5, nudge_y=0.1, height_png = 1500, width_png = 4000) 
plot_manhattan_zoom(pop="DUK", chr=6, from_mb=27.25, to_mb=28.1, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=6, from_mb=72.1, to_mb=72.5, nudge_y=0.1, height_png = 1500, width_png = 4000) 
plot_manhattan_zoom(pop="DUK", chr=6, from_mb=115.5, to_mb=115.7, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=9, from_mb=29.1, to_mb=29.4, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=12, from_mb=109.5, to_mb=110.75, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=15, from_mb=98.5, to_mb=98.7, nudge_y=0.1, height_png = 1500, width_png = 4000)
plot_manhattan_zoom(pop="DUK", chr=16, from_mb=25, to_mb=27, nudge_y=0.1,height_png = 1500, width_png = 4000)

# DUC
plot_manhattan_zoom(pop="DUC", chr=3, from_mb=79.5, to_mb=80, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUC", chr=7, from_mb=39.5, to_mb=40.75, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUC", chr=8, from_mb=61.3, to_mb=61.7, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUC", chr=8, from_mb=121.5, to_mb=122.25, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUC", chr=9, from_mb=7, to_mb=9.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUC", chr=18, from_mb=76.75, to_mb=77.3, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)

# DU6
plot_manhattan_zoom(pop="DU6", chr=2, from_mb=4, to_mb=6, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6", chr=3, from_mb=35.5, to_mb=36.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6", chr=4, from_mb=129.25, to_mb=129.6, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6", chr=11, from_mb=100.5, to_mb=100.9, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6", chr=17, from_mb=14.5, to_mb=15.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6", chr=18, from_mb=60.5, to_mb=61, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)

#DU6P
plot_manhattan_zoom(pop="DU6P", chr=1, from_mb=149.1, to_mb=149.6, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6P", chr=4, from_mb=3.5, to_mb=5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6P", chr=8, from_mb=119, to_mb=123, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6P", chr=11, from_mb=109, to_mb=110.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DU6P", chr=19, from_mb=55, to_mb=57.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)

#DUhLB
plot_manhattan_zoom(pop="DUhLB", chr=2, from_mb=10.25, to_mb=13.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=2, from_mb=162.5, to_mb=163.25, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=6, from_mb=64.5, to_mb=66.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=6, from_mb=128.35, to_mb=128.75, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=7, from_mb=74.3, to_mb=74.55, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=11, from_mb=60.75, to_mb=62, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=12, from_mb=5, to_mb=8, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=14, from_mb=38, to_mb=41, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="DUhLB", chr=17, from_mb=31, to_mb=31.75, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)

# FERT
plot_manhattan_zoom(pop="FERT", chr=6, from_mb=70.5, to_mb=71.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="FERT", chr=10, from_mb=6.8, to_mb=7.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="FERT", chr=10, from_mb=15.5, to_mb=16, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="FERT", chr=13, from_mb=19.8, to_mb=20.25, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
plot_manhattan_zoom(pop="FERT", chr=13, from_mb=99, to_mb=99.5, nudge_y=0.1, height_png = 1500, width_png = 4000, export_plot = TRUE)
```

```{r function_zoom_in_manhattan_nonsyn_in_genes, eval = FALSE}
# twick above function to make labels larger
plot_manhattan_zoom2 <- function(pop, chr, from_mb, to_mb, nudge_y, height_png, width_png, lab_size, export_plot=TRUE){
  
  set.seed(478)
  
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")

  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
    )

  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
    ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 

  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )

  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()

  # prepare plot data 
  plot_dat <- d_man %>% 
    filter(CHROM == chr & center_window_Mb > from_mb & center_window_Mb < to_mb) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )
  
  # subset data for gene labels existing
  lab_dat <- plot_dat %>% 
    filter(!is.na(gene_symbol) & contrast == contr) 

  # plot according to number of gene labels available
  if(nrow(lab_dat) > 1){

  # interspace gene labels    
  lab_dat$lab_set <- NA
  lab_dat$lab_set[seq(1,nrow(lab_dat), 2)] <- "1"
  lab_dat$lab_set[seq(2,nrow(lab_dat), 2)] <- "2"
  
    # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "1"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1,
        size = lab_size
      ) +
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "2"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = -nudge_y,
        vjust = 1,
        size = lab_size
      ) + 
      # Make it look nicer
      theme_bw(base_size = 15) +
      theme(panel.background = element_blank()) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2), breaks = seq(0,1,0.2)) +
      xlab("Mb") 
  }else{
          # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  lab_dat,aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1,
        size = lab_size
      ) +
      # Make it look nicer
      theme_bw(base_size = 15) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2), breaks = seq(0,1,0.2)) +
      xlab("Mb") 
  }
  
  if(export_plot){
    # export plot
    fl_nm <- paste0(pop, "_win_fst_rdd_genes_nonsyn_labs_chr",chr,"_from",from_mb,"mb_to", to_mb, "mb.png")
    png(here(fig_path, fl_nm), res = 300, height = height_png, width = width_png)
    print(p)
    dev.off()
  }else{
    print(p)
  }
  

}

```

```{r make_goi_with_nonsyn_snp_zoom_labs_manhattan, eval=FALSE}

#duk
plot_manhattan_zoom2(pop="DUK", chr=2, from_mb=28.25, to_mb=29.5, nudge_y=0.2, height_png = 1000, width_png = 5000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUK", chr=6, from_mb=72, to_mb=73, nudge_y=0.2, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUK", chr=6, from_mb=115.5, to_mb=116, nudge_y=0.2, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUK", chr=16, from_mb=25, to_mb=27, nudge_y=0.2, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)

#duc
plot_manhattan_zoom2(pop="DUC", chr=8, from_mb=61.25, to_mb=61.75, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUC", chr=9, from_mb=8, to_mb=9, nudge_y=0.2, height_png = 1000, width_png = 5000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUC", chr=18, from_mb=76.8, to_mb=76.95, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)

#du6
plot_manhattan_zoom2(pop="DU6", chr=2, from_mb=4.25, to_mb=5, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6", chr=3, from_mb=35.6, to_mb=35.9, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6", chr=4, from_mb=129.25, to_mb=129.6, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6", chr=11, from_mb=100.6, to_mb=100.9, nudge_y=0.2, height_png = 1000, width_png = 2500, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6", chr=17, from_mb=14.7, to_mb=15, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)

#du6p
plot_manhattan_zoom2(pop="DU6P", chr=4, from_mb=4.7, to_mb=5, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6P", chr=8, from_mb=119, to_mb=123, nudge_y=0.2, height_png = 1200, width_png = 3500, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6P", chr=11, from_mb=110, to_mb=110.25, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DU6P", chr=19, from_mb=56.25, to_mb=57, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)

#fert
plot_manhattan_zoom2(pop="FERT", chr=6, from_mb=70.6, to_mb=71.2, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="FERT", chr=10, from_mb=6.9, to_mb=7.4, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="FERT", chr=13, from_mb=99, to_mb=99.5, nudge_y=0.2, height_png = 1000, width_png = 2000, lab_size = 5, export_plot = T)

# duhlb
plot_manhattan_zoom2(pop="DUhLB", chr=2, from_mb=11.05, to_mb=11.9, nudge_y=0.2, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUhLB", chr=6, from_mb=128.5, to_mb=128.65, nudge_y=0.2, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)
plot_manhattan_zoom2(pop="DUhLB", chr=11, from_mb=61, to_mb=61.5, nudge_y=0.1, height_png = 1000, width_png = 3000, lab_size = 5, export_plot = T)



```

###--EndOfSection

RDD-genes_DEGs {data-navmenu="RDDs"} 
================================================

Column {.tabset}
-----------------

### Summaries 
```{r find_and_display_RDDs_x_DEGs_DUK_DUC} 

FL1_testis_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL1_testis_DEGs_gene_symbols.rds"))
FL2_testis_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL2_testis_DEGs_gene_symbols.rds"))
FL1_ovary_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL1_ovary_DEGs_gene_symbols.rds"))

#FL1=DUK
#FL2=DUC

# Find DEGs in list of RDDs
lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols) %>% # find overlap with FL1_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  dplyr::rename(DEGs_found_in = contrast, n_DEGs_found = n_common_genes) %>% 
  filter(n_DEGs_found > 0) %>% 
  kable(caption = "RDD genes x DEGs (FL1=DUK testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols) %>% # find overlap with FL2_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  dplyr::rename(DEGs_found_in = contrast, n_DEGs_found = n_common_genes) %>% 
  filter(n_DEGs_found > 0) %>% 
  kable(caption = "RDD genes x DEGs (FL2=DUC testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols) %>% # find overlap with FL1_ovary_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  dplyr::rename(DEGs_found_in = contrast, n_DEGs_found = n_common_genes) %>% 
  kable(caption = "RDD genes x DEGs (FL1=DUK ovary)") %>% 
  kable_styling(full_width = F)
```

```{r find_and_display_RDDs_x_DEGs_DuhLB}
RDD_genes_DUhLB_BRE <- here("00_dashboard/data/DEGs_Julia/RDD_genes_DUhLB_BRE_for_R.csv") %>% read.table(sep = ",", header = TRUE) %>% 
  dplyr::select(-X, -X.1, -X.2, -X.3) %>% 
  reshape2::melt(id.vars = c("contrast","ensembl_gene_id","mgi_symbol","seqnames", "start","end", "width","strand", "gene_biotype", "GOBP")) %>% 
  separate(col = variable, into = c("tmp","tissue"), sep = "_") %>% 
  reshape2::dcast(
    contrast + ensembl_gene_id + mgi_symbol + seqnames + start + end + width + strand + gene_biotype + GOBP + tissue ~ tmp, value.var = "value" 
    ) 

lapply(1:length(RDD_genes), function(i){

  d <- inner_join(
    RDD_genes[[i]] %>% 
      as.data.frame() %>% 
      dplyr::select(mcols.gene_id, mcols.gene_symbol),
    RDD_genes_DUhLB_BRE %>% 
      filter(FDR != "") %>% 
      dplyr::select(ensembl_gene_id, tissue), 
    by = c("mcols.gene_id" ="ensembl_gene_id")
    )
  
  if(nrow(d) > 0){
    d %>% 
      group_by(tissue) %>% 
      summarise(n_common_genes = n(), gene_symbol = paste(mcols.gene_symbol, collapse = " ; ")) %>% 
      mutate(contrast = names(RDD_genes)[i]) %>% 
      dplyr::select(contrast, everything())
  }else{
    data.frame(contrast = names(RDD_genes)[i], tissue = "", n_common_genes = 0, gene_symbol = "")
  }
}) %>% 
  bind_rows() %>% 
  dplyr::rename(DEGs_found_in = contrast, n_DEGs_found = n_common_genes) %>% 
  filter(n_DEGs_found > 0) %>% 
  kable(caption = "RDD genes x DEGs (DUhLB)") %>% 
  kable_styling(full_width = F)


```

```{r find_and_display_RDDs_x_DEGs_DU6}

YoungTitan_vs_YoungControl_liver <- here("00_dashboard/data/DEGs_Shahaf/res.Sample_YoungTitan_vs_YoungControl.csv") %>% 
  read.csv(stringsAsFactors = FALSE)

RDD_genes_in_DU6_liver <- RDD_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "pop") %>% 
  inner_join(YoungTitan_vs_YoungControl_liver, by = c("mcols.gene_id" = "X"))

RDD_genes_in_DU6_liver %>% 
  filter(padj <= 0.1) %>%
  group_by(pop) %>% 
  summarise(tissue = "liver", n_common_genes = n(), gene_symbol = paste0(mcols.gene_symbol, collapse = " ; ") ) %>% 
  dplyr::rename(DEGs_found_in = pop, n_DEGs_found = n_common_genes) %>% 
  filter(n_DEGs_found > 0) %>% 
  kable(caption = "RDD genes x DEGs (DU6)") %>% 
  kable_styling(full_width = F)

```

```{r find_and_display_RDDs_x_DEGs_DU6P}
# young 
YoungRock_vs_YoungControl_liver <- here("00_dashboard/data/DEGs_Shahaf/res.Sample_YoungRock_vs_YoungControl.csv") %>% 
  read.csv(stringsAsFactors = FALSE)

RDD_genes_in_DU6Pyoung_liver <- RDD_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "pop") %>% 
  inner_join(YoungRock_vs_YoungControl_liver, by = c("mcols.gene_id" = "X"))

RDD_genes_in_DU6Pyoung_liver %>% 
  filter(padj <= 0.1) %>%
  group_by(pop) %>% 
  summarise(tissue = "liver", n_common_genes = n(), gene_symbol = paste0(mcols.gene_symbol, collapse = " ; ") ) %>% 
  dplyr::rename(DEGs_found_in = pop, n_DEGs_found = n_common_genes) %>% 
  kable(caption = "RDD genes x DEGs (DU6P - young)") %>% 
  kable_styling(full_width = F)


# old
OldRock_vs_OldControl_liver <- here("00_dashboard/data/DEGs_Shahaf/res.Sample_OldRock_vs_OldControl.csv") %>% 
  read.csv(stringsAsFactors = FALSE)

RDD_genes_in_DU6Pold_liver <- RDD_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows(.id = "pop") %>% 
  inner_join(OldRock_vs_OldControl_liver, by = c("mcols.gene_id" = "X"))

RDD_genes_in_DU6Pold_liver %>% 
  filter(padj <= 0.1) %>%
  group_by(pop) %>% 
  summarise(tissue = "liver", n_common_genes = n(), gene_symbol = paste0(mcols.gene_symbol, collapse = " ; ") ) %>% 
  dplyr::rename(DEGs_found_in = pop, n_DEGs_found = n_common_genes) %>% 
  kable(caption = "RDD genes x DEGs (DU6P - old)") %>% 
  kable_styling(full_width = F)

```

```{r shahaf_request_find_orexin, eval = FALSE}

YoungTitan_vs_YoungControl_liver %>% filter(X == "ENSMUSG00000045471")

YoungTitan_vs_YoungControl_liver %>% filter(symbol == "Hcrt")

gr_mof <- mmu_gene_set_gr %>% 
  as.data.frame() %>%
  filter(mcols.gene_id == "ENSMUSG00000045471") %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

win_fst_sel_vs_ctrl_prep_gr <- win_fst_sel_vs_ctrl_prep %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = gr_mof, subject = win_fst_sel_vs_ctrl_prep_gr, minoverlap = 1)

win_fst_sel_vs_ctrl_prep_gr[subjectHits(ov)] %>% 
  as.data.frame() %>% 
  #dplyr::select(seqnames, start, end, contrast, MEAN_FST) %>% 
  dplyr::rename(chrom = seqnames, win_start = start, win_end = end) %>% 
  group_by(contrast) %>% 
  summarise(win_Fst = mean(MEAN_FST)) %>% 
  arrange(desc(win_Fst)) %>% 
  filter(contrast != "FERT_vs_FZTDU")


#---------------------
YoungTitan_vs_YoungControl_liver %>% filter(symbol == "Hcrtr2") # this is the one he was most excited about

gr_mof <- mmu_gene_set_gr %>% 
  as.data.frame() %>%
  filter(mcols.gene_id == "ENSMUSG00000032360") %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

win_fst_sel_vs_ctrl_prep_gr <- win_fst_sel_vs_ctrl_prep %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = gr_mof, subject = win_fst_sel_vs_ctrl_prep_gr, minoverlap = 1)

win_fst_sel_vs_ctrl_prep_gr[subjectHits(ov)] %>% 
  as.data.frame() %>% 
  #dplyr::select(seqnames, start, end, contrast, MEAN_FST) %>% 
  dplyr::rename(chrom = seqnames, win_start = start, win_end = end) %>% 
  group_by(contrast) %>% 
  summarise(win_Fst = mean(MEAN_FST)) %>% 
  arrange(desc(win_Fst)) %>% 
  filter(contrast != "FERT_vs_FZTDU")

#----------------------------




```

```{r shahaf_request_find_MOF, eval = FALSE}

#MOF = kat8 = ENSMUSG00000030801

gr_mof <- mmu_gene_set_gr %>% 
  as.data.frame() %>%
  filter(mcols.gene_id == "ENSMUSG00000030801") %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

win_fst_sel_vs_ctrl_prep_gr <- win_fst_sel_vs_ctrl_prep %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = gr_mof, subject = win_fst_sel_vs_ctrl_prep_gr, minoverlap = 1)

win_fst_sel_vs_ctrl_prep_gr[subjectHits(ov)] %>% 
  as.data.frame() %>% 
  #dplyr::select(seqnames, start, end, contrast, MEAN_FST) %>% 
  dplyr::rename(chrom = seqnames, win_start = start, win_end = end) %>% 
  group_by(contrast) %>% 
  summarise(win_Fst = mean(MEAN_FST), n_snps = sum(N_VARIANTS)) %>% 
  arrange(desc(win_Fst)) %>% 
  filter(contrast != "FERT_vs_FZTDU")


subsetByOverlaps(vcf_final_snp_ids_gr, gr_mof, minoverlap = 1) %>% 
  as.data.frame() %>% 
  inner_join(
    here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% vroom(col_types = c(CHROM = "c")),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  inner_join(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUK_vs_FZTDU.weir.fst") %>%  
      vroom(col_types = c(CHROM = "c")) %>% 
      dplyr::rename(DUK_vs_FZTDU=WEIR_AND_COCKERHAM_FST),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  inner_join(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUC_vs_FZTDU.weir.fst") %>%  
      vroom(col_types = c(CHROM = "c")) %>% 
      dplyr::rename(DUC_vs_FZTDU=WEIR_AND_COCKERHAM_FST),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  inner_join(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6_vs_FZTDU.weir.fst") %>%  
      vroom(col_types = c(CHROM = "c")) %>% 
      dplyr::rename(DU6_vs_FZTDU=WEIR_AND_COCKERHAM_FST),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  inner_join(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DU6P_vs_FZTDU.weir.fst") %>%  
      vroom(col_types = c(CHROM = "c")) %>% 
      dplyr::rename(DU6P_vs_FZTDU=WEIR_AND_COCKERHAM_FST),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  inner_join(
    here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/DUhLB_vs_FZTDU.weir.fst") %>%  
      vroom(col_types = c(CHROM = "c")) %>% 
      dplyr::rename(DUhLB_vs_FZTDU=WEIR_AND_COCKERHAM_FST),
    by = c("seqnames" = "CHROM", "start"="POS")
  ) %>% 
  dplyr::select(-strand, -width, -end) %>%
  dplyr::rename(chr = seqnames, pos = start) 

```

### RDD-Genes_x_DEGs-DuhLB
```{r RDD-Genes_x_DEGs-DuhLB}

lapply(1:length(RDD_genes), function(i){

  d <- inner_join(
    RDD_genes[[i]] %>% 
      as.data.frame() %>% 
      dplyr::select(mcols.gene_id, mcols.gene_symbol),
    RDD_genes_DUhLB_BRE %>% 
      filter(FDR != ""),
      #dplyr::select(ensembl_gene_id, tissue), 
    by = c("mcols.gene_id" ="ensembl_gene_id")
    )
  d %>% 
      #group_by(tissue) %>% 
      #summarise(n_common_genes = n(), gene_symbol = paste(mcols.gene_symbol, collapse = " ; ")) %>% 
      dplyr::select(contrast, everything())
}) %>% 
  bind_rows() %>% 
  dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol) %>% 
  dplyr::select(-mgi_symbol, -seqnames, -start, -end, -strand, -width, -GOBP) %>% 
  dplyr::select(contrast, gene_id, gene_symbol, tissue, FDR, logFC, everything()) %>% 
  arrange(tissue) %>% 
  datatable(
    options = list(pageLength = 100), 
    caption = "RDD genes x DEGs (DUhLB)", 
    rownames = FALSE
    )
```

### RDD-Genes_x_DEGs-DU6
```{r RDD-Genes_x_DEGs-DU6}

RDD_genes_in_DU6_liver %>% 
  filter(padj <= 0.1) %>% 
  mutate(tissue = "liver", padj = round(padj, 3), log2FoldChange = round(log2FoldChange, 3), baseMean = round(baseMean,3)) %>% 
  dplyr::select(pop, mcols.gene_id, mcols.gene_symbol, tissue, baseMean, log2FoldChange, padj, mcols.gene_biotype) %>% 
  dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, gene_biotype=mcols.gene_biotype) %>% 
  datatable(
    options = list(pageLength = 100), 
    caption = "RDD genes x DEGs (DU6)", 
    rownames = FALSE
    )

```

```{r tables_for_shahaf, eval=FALSE}

#------------------------------------------

# get rdds at different quantiles

## make a function
find_distinct_REDs_list <- function(q_low, q_high){
  
  qs <- paste0("q", c(q_low, q_high))
  
  res <- list(
    DUK = find_distinct_REDs(
      contr = "DUK_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), qs[1], qs[2]
    ),
    DUC = find_distinct_REDs(
      contr = "DUC_vs_FZTDU", rest = paste0( c("DUK","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), qs[1], qs[2]
    ),
    DU6 = find_distinct_REDs(
      contr = "DU6_vs_FZTDU", rest = paste0( c("DUC","DUK","DU6P","DUhLB"), "_vs_FZTDU" ), qs[1], qs[2]
    ),
    DU6P = find_distinct_REDs(
      contr = "DU6P_vs_FZTDU", rest = paste0( c("DUC","DU6","DUK","DUhLB"), "_vs_FZTDU" ), qs[1], qs[2]
    ),
    DUhLB = find_distinct_REDs(
      contr = "DUhLB_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUK"), "_vs_FZTDU" ), qs[1], qs[2]
    ),
    FERT = find_distinct_REDs(
      contr = "FERT_vs_FZTDU", rest = paste0( c("DU6","DU6P","DUhLB"), "_vs_FZTDU" ),  qs[1], qs[2]
    )
  ) 
  
  return(res)
}


## q10vsq90
q10vsq90 <- find_distinct_REDs_list(10,90)

RDD_genes_q10vsq90 <- lapply(q10vsq90, function(x){
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()
  })

## q15vsq85
q15vsq85 <- find_distinct_REDs_list(15,85)
RDD_genes_q15vsq85 <- lapply(q15vsq85, function(x){
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()
})

## q20vsq80
q20vsq80 <- find_distinct_REDs_list(20,80)
RDD_genes_q20vsq80 <- lapply(q20vsq80, function(x){
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()
})

## q25vsq75
q25vsq75 <- find_distinct_REDs_list(25,75)
RDD_genes_q25vsq75 <- lapply(q25vsq75, function(x){
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()
})

#-----------------------------------add diff gene exprs info and export for shahaf
add_diff_gene_exprs_info <- function(d){
  d %>% 
    lapply(as.data.frame) %>% 
    bind_rows(.id = "pop") %>% 
    full_join(YoungTitan_vs_YoungControl_liver, by = c("mcols.gene_id" = "X")) %>%  
    #filter(padj <= 0.1) %>% 
    dplyr::select(pop, mcols.gene_id, mcols.gene_symbol, mcols.gene_biotype, baseMean, log2FoldChange, padj) %>% 
    dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, gene_biotype = mcols.gene_biotype) %>% 
    mutate(padj = round(padj, 3), log2FoldChange = round(log2FoldChange,3), baseMean = round(baseMean, 3))
}


add_diff_gene_exprs_info(RDD_genes_q10vsq90) %>% write.csv(here("00_dashboard/data/DEGs_Shahaf", "RDD_genes_q10vsq90.csv"))
add_diff_gene_exprs_info(RDD_genes_q15vsq85) %>% write.csv(here("00_dashboard/data/DEGs_Shahaf", "RDD_genes_q15vsq85.csv"))
add_diff_gene_exprs_info(RDD_genes_q20vsq80) %>% write.csv(here("00_dashboard/data/DEGs_Shahaf", "RDD_genes_q20vsq80.csv"))
add_diff_gene_exprs_info(RDD_genes_q25vsq75) %>% write.csv(here("00_dashboard/data/DEGs_Shahaf", "RDD_genes_q25vsq75.csv"))

#-------------------------------------

summarise_overlap_RDD_genes_DEGs <- function(d){
  d %>% 
    lapply(as.data.frame) %>% 
    bind_rows(.id = "pop") %>% 
    inner_join(YoungTitan_vs_YoungControl_liver, by = c("mcols.gene_id" = "X")) %>%  
    filter(padj <= 0.1)%>%
    group_by(pop) %>% 
    summarise(tissue = "liver", n_common_genes = n(), gene_symbol = paste0(mcols.gene_symbol, collapse = " ; ") )
}


summarise_overlap_RDD_genes_DEGs(RDD_genes_q10vsq90) %>% kable(caption = "(q10vsq90)genes x liverDEGs(DU6vsFZTDU)") %>% kable_styling(full_width = F) 
summarise_overlap_RDD_genes_DEGs(RDD_genes_q15vsq85) %>% kable(caption = "(q15vsq85)genes x liverDEGs(DU6vsFZTDU)") %>% kable_styling(full_width = F)
summarise_overlap_RDD_genes_DEGs(RDD_genes_q20vsq80) %>% kable(caption = "(q20vsq80)genes x liverDEGs(DU6vsFZTDU)") %>% kable_styling(full_width = F)
summarise_overlap_RDD_genes_DEGs(RDD_genes_q25vsq75) %>% kable(caption = "(q25vsq75)genes x liverDEGs(DU6vsFZTDU)") %>% kable_styling(full_width = F)

```

### RDD-Genes_x_DEGs-DU6P_young
```{r RDD-Genes_x_DEGs-DU6P_young}

RDD_genes_in_DU6Pyoung_liver %>% 
  filter(padj <= 0.1) %>% 
  mutate(tissue = "liver", padj = round(padj, 3), log2FoldChange = round(log2FoldChange, 3), baseMean = round(baseMean,3)) %>% 
  dplyr::select(pop, mcols.gene_id, mcols.gene_symbol, tissue, baseMean, log2FoldChange, padj, mcols.gene_biotype) %>% 
  dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, gene_biotype=mcols.gene_biotype) %>% 
  datatable(
    options = list(pageLength = 100), 
    caption = "RDD genes x DEGs (DU6P-young)", 
    rownames = FALSE
    )

```

### RDD-Genes_x_DEGs-DU6P_old
```{r RDD-Genes_x_DEGs-DU6P_old}

RDD_genes_in_DU6Pold_liver %>% 
  filter(padj <= 0.1) %>% 
  mutate(tissue = "liver", padj = round(padj, 3), log2FoldChange = round(log2FoldChange, 3), baseMean = round(baseMean,3)) %>% 
  dplyr::select(pop, mcols.gene_id, mcols.gene_symbol, tissue, baseMean, log2FoldChange, padj, mcols.gene_biotype) %>% 
  dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, gene_biotype=mcols.gene_biotype) %>% 
  datatable(
    options = list(pageLength = 100), 
    caption = "RDD genes x DEGs (DU6P-old)", 
    rownames = FALSE
    )

```


### MA plots

```{r make_and_export_ma_plots_du6, eval = FALSE}
set.seed(123)

png(here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6_in_LiverDEGsYoungDU6vsYoungCTRL.png"), res = 300, units = "px", width = 2500, height = 2000)
ggplot() +
  geom_point(
    data = YoungTitan_vs_YoungControl_liver %>% filter(padj > 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "black", alpha = 0.4
    ) +  
  geom_point(
    data = YoungTitan_vs_YoungControl_liver %>% filter(padj <= 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "red", alpha = 0.7
    ) +
  geom_label_repel(
    data = RDD_genes_in_DU6_liver %>% filter(pop == "DU6") %>% filter(padj < 0.1),
    aes(x = log10(baseMean), y = log2FoldChange, label = mcols.gene_symbol),
    box.padding = 2
    ) +
  theme_bw(base_size = 12) +
  scale_color_manual(values = c("black","red")) +
  labs(title = "DU6-RDD-genes found as DEGs (Young_DU6 vs Young_Control - liver)")

dev.off()
```

```{r make_and_export_ma_plots_du6p_young, eval = FALSE}
set.seed(123)

png(here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6P_in_LiverDEGsYoungDU6PvsYoungCTRL.png"), res = 300, units = "px", width = 2500, height = 2000)
ggplot() +
  geom_point(
    data = YoungRock_vs_YoungControl_liver %>% filter(padj > 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "black", alpha = 0.4
    ) +  
  geom_point(
    data = YoungRock_vs_YoungControl_liver %>% filter(padj <= 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "red", alpha = 0.7
    ) +
  geom_label_repel(
    data = RDD_genes_in_DU6Pyoung_liver %>% filter(pop == "DU6P") %>% filter(padj < 0.1),
    aes(x = log10(baseMean), y = log2FoldChange, label = mcols.gene_symbol),
    box.padding = 2
    ) +
  theme_bw(base_size = 12) +
  scale_color_manual(values = c("black","red")) +
  labs(title = "DU6P-RDD-genes found as DEGs (Young_DU6P vs Young_Control - liver)")

dev.off()
```

```{r make_and_export_ma_plots_du6p_old, eval = FALSE}
set.seed(123)

png(here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6P_in_LiverDEGsOldDU6PvsOldCTRL.png"), res = 300, units = "px", width = 2500, height = 2000)
ggplot() +
  geom_point(
    data = OldRock_vs_OldControl_liver %>% filter(padj > 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "black", alpha = 0.4
    ) +  
  geom_point(
    data = OldRock_vs_OldControl_liver %>% filter(padj <= 0.1),
    aes(x = log10(baseMean), y = log2FoldChange),
    color = "red", alpha = 0.7
    ) +
  geom_label_repel(
    data = RDD_genes_in_DU6Pold_liver %>% filter(pop == "DU6P") %>% filter(padj < 0.1),
    aes(x = log10(baseMean), y = log2FoldChange, label = mcols.gene_symbol),
    box.padding = 2
    ) +
  theme_bw(base_size = 12) +
  scale_color_manual(values = c("black","red")) +
  labs(title = "DU6P-RDD-genes found as DEGs (Old_DU6P vs Old_Control - liver)")
dev.off()
```

```{r import_and_display, out.width="50%", fig.align="center"}
include_graphics(
  c(
    here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6_in_LiverDEGsYoungDU6vsYoungCTRL.png"),
    here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6P_in_LiverDEGsYoungDU6PvsYoungCTRL.png"),
    here("00_dashboard/figures/MA_plots/MAplot_RDDgenesDU6P_in_LiverDEGsOldDU6PvsOldCTRL.png")
  )
)

```

###--EndOfSection

RDD-SNP-effects {data-navmenu="RDDs"}
================================================ 

Column {.tabset}
-----------------

### Overview

* Find SNPs in RDD-genes, then add SNPeff annotations and Fst scores.

* Note: this was not done for RED-genes!

```{r make_multiple_query_function_and_export, eval=FALSE}

# Function to get SNPs in RDD-genes and extract their effects from SNPeff table output.

  display_snp_effects_per_gene <- function(pop){
  
    # define data set  
    x <- RDD_genes[[pop]]
    
    # define chromosomes of interest (the ones with RDD genes)
    chrs <- x %>% seqnames() %>% unique() %>% as.character()
    
    snp_effect_fst <- 
      # import snp annotations data set
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
      vroom() %>% 
      # retain chromosomes of interest to reduce size
      filter(CHROM %in% chrs) %>% 
      # prepare columns for granges
      mutate(start = POS, end = POS) %>% 
      dplyr::rename(
        seqnames = CHROM, effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT", gene_symbol = "ANN[*].GENE", gene_id = "ANN[*].GENEID"
        ) %>% 
      dplyr::select(seqnames, start, end, REF, ALT, effect, impact, gene_symbol, gene_id) %>% 
      # kepp unique snp-gene-effect associations (some snps have the same effect in different gene-transcripts)
      unique() %>% 
      # add snp fst information
      left_join(
        vroom(
          here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",paste0(pop,"_vs_FZTDU.weir.fst"))
          ),
        by = c("seqnames"="CHROM", "start"="POS")
      ) %>% 
      # filter by RDD_genes
      filter(gene_id %in% x$mcols.gene_id) %>% 
      # prepare table for display
      mutate(snp_locus = paste0(seqnames, ":", start,"-",end)) %>% 
      dplyr::select(gene_id, gene_symbol, snp_locus, effect, impact, WEIR_AND_COCKERHAM_FST) %>% 
      unique() %>% 
      mutate(contrast = names(RDD_genes)[names(RDD_genes) == pop]) %>%
      dplyr::select(contrast, gene_id, gene_symbol, everything()) 
    
    # export queries (avoid running this every time dashboard is knitted)
    csv_nm <- paste0("RDD_genes_" , names(RDD_genes)[names(RDD_genes) == pop], "_SNPeff.csv")
    write.csv(snp_effect_fst, here("00_dashboard/data/RDD_genes", csv_nm))
  }

display_snp_effects_per_gene("DUK") 
display_snp_effects_per_gene("DUC") 
display_snp_effects_per_gene("DU6") 
display_snp_effects_per_gene("DU6P") 
display_snp_effects_per_gene("DUhLB") 
display_snp_effects_per_gene("FERT") 

```

### DUK
```{r find_snp_in_RDD_genes_add_effects_DUK}
here("00_dashboard/data/RDD_genes/RDD_genes_DUK_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>%
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUK_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))
```

### DUC
```{r find_snp_in_RDD_genes_add_effects_DUC}
here("00_dashboard/data/RDD_genes/RDD_genes_DUC_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUC_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DU6
```{r find_snp_in_RDD_genes_add_effects_DU6}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DU6_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DU6P
```{r find_snp_in_RDD_genes_add_effects_DU6P}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6P_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DU6P_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DUhLB
```{r find_snp_in_RDD_genes_add_effects_DUhLB}
here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUhLB_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### FERT
```{r find_snp_in_RDD_genes_add_effects_FERT}
here("00_dashboard/data/RDD_genes/RDD_genes_FERT_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUK_DUC_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>%
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### AF-heatmap
```{r make_heatmap_allele_frq_nonsyn_SNPs_in_RDD_genes, eval = FALSE}

get_frqs <- function(chr, pos, ref_allele, alt_allele){
  
  hom_ref_gt <- paste(ref_allele,ref_allele,sep="/")
  hom_alt_gt <- paste(alt_allele,alt_allele,sep="/")
  het_gt <- paste(ref_allele,alt_allele,sep="/")  
  
  read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
    dplyr::select(-X) %>% 
    filter(CHROM == chr & POS == pos) %>% 
    reshape2::melt(id.vars = c("CHROM", "POS", "REF", "ALT")) %>% 
    mutate(
      
      hom_ref = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% 
                         .[grep(hom_ref_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_ref_gt,")"), TRUE))) %>% as.numeric(),
      
      hom_alt = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% .[grep(hom_alt_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_alt_gt,")"), TRUE))) %>% as.numeric(),
      
      het = sapply(value, function(x) str_split(x,";") %>% 
                     unlist() %>% 
                     .[grep(het_gt,.)] %>% 
                     str_remove(fixed(paste0("(",het_gt,")"), TRUE))) %>% as.numeric(),
      
      miss = sapply(value, function(x) str_split(x,";") %>% 
                      unlist() %>% 
                      .[grep("./.",., fixed = TRUE)] %>% 
                      str_remove(fixed("(./.)", TRUE))) %>% as.numeric(),
      
      hom_ref = ifelse(is.na(hom_ref), 0, hom_ref),
      hom_alt = ifelse(is.na(hom_alt), 0, hom_alt),
      het = ifelse(is.na(het), 0, het),
      
      ref_frq = (hom_ref*2 + het*1)/((hom_alt+hom_ref+het)*2),
      alt_frq = (hom_alt*2 + het*1)/((hom_alt+hom_ref+het)*2),
      
      variable = str_remove(variable, "_GT_counts"),
      variable = ifelse(variable == "DUK_DUC", "FERT", variable)
    ) %>% 
    dplyr::rename(pop=variable) %>% 
    reshape2::dcast(CHROM + POS ~ pop, value.var = "alt_frq")
}





tab_alt_frq_gn_eff <- lapply(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), function(pop){
  
  # import SNP effects for population
  d_gt <- here(paste0("00_dashboard/data/RDD_genes/RDD_genes_",pop,"_SNPeff.csv")) %>% read.csv() %>% dplyr::select(-X) %>% 
    
    # keep only high and moderate impacts
    filter(impact %in% c("HIGH","MODERATE")) %>% 
    
    # add GT counts
    inner_join(
      read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
        mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)),
      by = "snp_locus"
    ) %>% 
    
    # reorder columns
    dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) 
  
  # add allele frq with function created above
  d_alt_frq_gn_eff <- lapply(1:nrow(d_gt), function(i){
    get_frqs(chr = d_gt[i,"CHROM"], pos = d_gt[i,"POS"], ref_allele = d_gt[i,"REF"], alt_allele = d_gt[i,"ALT"])  
  }) %>% 
    bind_rows() %>% 
    inner_join(
      d_gt %>% dplyr::select(CHROM, POS, gene_symbol, effect, impact), 
      by = c("CHROM","POS")
    ) %>% 
    mutate(target_population = pop) #%>% 
    #dplyr::select(target_population, everything())
  
  return(d_alt_frq_gn_eff)
  
  }) 

names(tab_alt_frq_gn_eff) <- c("DUK","DUC","DU6","DU6P","DUhLB","FERT")


make_export_pheatmap <- function(d, width = 1800, height = 2000){
  
  # prepare data
  dd <- d %>% mutate(CHROM =as.numeric(CHROM), POS = as.numeric(POS)) %>% arrange(CHROM, POS)
  pmat <- dd %>% dplyr::select("DUK","DUC","DU6","DU6P","DUhLB","FERT","FZTDU")
  rownames(pmat) <- paste0(dd$CHROM, ":",dd$POS," (", dd$gene_symbol,")")
  pop <- dd$target_population %>% unique()
  
  # make pheatmap
  ttl <- paste0("AF Non-Syn SNPs in RDD Genes (", pop,")")
  p <- pheatmap::pheatmap(pmat, cluster_cols = FALSE, cluster_rows = FALSE, main = ttl, display_numbers = TRUE)
  
  # export plot
  fl_nm <- paste0("af_nonsyn_heatmap_", pop, ".png")
  
  png(here("00_dashboard/figures", fl_nm), res = 300, width = width, height = height)
  print(p)
  dev.off()

}


make_export_pheatmap(tab_alt_frq_gn_eff$DUK, 1800, 1500)
make_export_pheatmap(tab_alt_frq_gn_eff$DUC, 1800, 1000)
make_export_pheatmap(tab_alt_frq_gn_eff$DU6, 1800, 1100)
make_export_pheatmap(tab_alt_frq_gn_eff$DU6P, 1800, 1800)
make_export_pheatmap(tab_alt_frq_gn_eff$DUhLB, 1800, 2000)
make_export_pheatmap(tab_alt_frq_gn_eff$FERT, 1800, 800)


```

```{r display_heatmap_allele_frq_nonsyn_SNPs_in_RDD_genes, out.width = "50%", fig.align = "center"}


fig_path <- here("00_dashboard/figures")

include_graphics(
  list.files(fig_path, full.names = TRUE) %>% .[grep("af_nonsyn_heatmap_", .)]
  )


```


RDD-ORA {data-navmenu="RDDs"}
====================================

Column {.tabset}
-----------------

### Overview

- Mostly null results.

- Significant enrichments were found as follows:

```{r eval=FALSE, q10_v_q95_GOBP_run_once_and_export} 
run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
)

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!
```

```{r eval = FALSE, q10_v_q95_KEGG_run_once_and_export}

run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
)

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
)

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!
```

```{r q10_v_q95_GOBP_queries_only_protein_coding_genes, eval = FALSE}
run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id[RDD_genes$DUK$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id[RDD_genes$DUC$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) # sig results!

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id[RDD_genes$DU6$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id[RDD_genes$DU6P$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id[RDD_genes$DUhLB$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id[RDD_genes$FERT$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP_pt_coding")
) #no sig
```

```{r q10_v_q95_KEGG_queries_only_protein_coding_genes, eval = FALSE}
run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id[RDD_genes$DUK$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) # sig results!

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id[RDD_genes$DUC$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id[RDD_genes$DU6$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) #no sig

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id[RDD_genes$DU6P$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) # no sig

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id[RDD_genes$DUhLB$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) # sig results

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id[RDD_genes$FERT$mcols.gene_biotype == "protein_coding"],
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG_pt_coding")
) # no sig
```

```{r relaxed_RDDs_function, eval=FALSE}

# make a function to find regions of distinct differentiation
# then find genes overlapping those regions
# report both results (regions and genes).

get_RDDs_and_RDDgenes <- function(q_low, q_high){
  
  # Regions
  rdds <- list(
    DUK = find_distinct_REDs(
      contr = "DUK_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), q_low, q_high
      ),
    DUC = find_distinct_REDs(
      contr = "DUC_vs_FZTDU", rest = paste0( c("DUK","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), q_low, q_high
      ),
    DU6 = find_distinct_REDs(
      contr = "DU6_vs_FZTDU", rest = paste0( c("DUC","DUK","DU6P","DUhLB"), "_vs_FZTDU" ), q_low, q_high
      ),
    DU6P = find_distinct_REDs(
      contr = "DU6P_vs_FZTDU", rest = paste0( c("DUC","DU6","DUK","DUhLB"), "_vs_FZTDU" ), q_low, q_high
      ),
    DUhLB = find_distinct_REDs(
      contr = "DUhLB_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUK"), "_vs_FZTDU" ), q_low, q_high
      ),
    FERT = find_distinct_REDs(
      contr = "FERT_vs_FZTDU", rest = paste0( c("DU6","DU6P","DUhLB"), "_vs_FZTDU" ),  q_low, q_high
      )
  )

  # Genes
  rdd_genes <- lapply(rdds, function(x){
  
    wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  
    gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()
  
  })
  
  res <- list(rdds = rdds, rdd_genes = rdd_genes)
  
  return(res)

}

```

```{r relaxed_apply_RDDs_function, eval=FALSE}

rdd_q10_q90 <- get_RDDs_and_RDDgenes(q_low = "q10", q_high = "q90")

rdd_q15_q85 <- get_RDDs_and_RDDgenes(q_low = "q15", q_high = "q85")

rdd_q20_q80 <- get_RDDs_and_RDDgenes(q_low = "q20", q_high = "q80")

rdd_q25_q75 <- get_RDDs_and_RDDgenes(q_low = "q25", q_high = "q75")

```

```{r relaxed_function_run_webgestalt_per_contrast, eval = FALSE}

WebGestaltR_GOBP_wrapper <- function(pop, rdd_gene_dat, q_low, q_high){

  #pop="DU6"
  #rdd_gene_dat = rdd_q10_q90 
  #q_low = "q10"
  #q_high = "q90"

    gns <- rdd_gene_dat$rdd_genes[[pop]]$mcols.gene_id # prep gene list
    proj_nm <- paste0("WebGestaltR_", pop,"_vs_FZTDU") # prep name of results dir
    out_dir <- paste0("WebGestaltR_RDD_",q_low,"_",q_high) # prep name of output dir
    
    print(paste0("# ", pop,"_vs_FZTDU. Gene list length:", length(gns)))

    run_WebGestaltR(
      interestGene=gns,
      projectName=proj_nm,
      enrichDatabase="geneontology_Biological_Process",
      outputDirectory=here("00_dashboard/data", out_dir,"GOBP")
    )
    
}


WebGestaltR_KEGG_wrapper <- function(pop, rdd_gene_dat, q_low, q_high){
  
    gns <- rdd_gene_dat$rdd_genes[[pop]]$mcols.gene_id # prep gene list
    proj_nm <- paste0("WebGestaltR_", pop,"_vs_FZTDU") # prep name of results dir
    out_dir <- paste0("WebGestaltR_RDD_",q_low,"_",q_high) # prep name of output dir
    
    print(paste0("# ", pop,"_vs_FZTDU. Gene list length:", length(gns)))

    run_WebGestaltR(
      interestGene=gns,
      projectName=proj_nm,
      enrichDatabase="pathway_KEGG",
      outputDirectory=here("00_dashboard/data", out_dir,"KEGG")
    )
}

```

```{r relaxed_run_WebGestaltR_wrapper, eval = FALSE}
# rdd_q10_q90
WebGestaltR_GOBP_wrapper(pop = "DUK", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") #notsig - 86 genes
WebGestaltR_GOBP_wrapper(pop = "DUC", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # sig - 43 genes
WebGestaltR_GOBP_wrapper(pop = "DU6", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") #notsig - 38 genes
WebGestaltR_GOBP_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") #notsig - 36 genes
WebGestaltR_GOBP_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") #notsig - 83 genes
WebGestaltR_GOBP_wrapper(pop = "FERT", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") #notsig - 10 genes

WebGestaltR_KEGG_wrapper(pop = "DUK", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90")# sig - 86 genes
WebGestaltR_KEGG_wrapper(pop = "DUC", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # notsig - 43 genes
WebGestaltR_KEGG_wrapper(pop = "DU6", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # notsig - 38 genes
WebGestaltR_KEGG_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # notsig - 36 genes
WebGestaltR_KEGG_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # notsig - 83 genes
WebGestaltR_KEGG_wrapper(pop = "FERT", rdd_gene_dat = rdd_q10_q90, q_low = "q10", q_high = "q90") # notsig - 10 genes


# rdd_q15_q85
WebGestaltR_GOBP_wrapper(pop = "DUK", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 127 genes
WebGestaltR_GOBP_wrapper(pop = "DUC", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 68 genes
WebGestaltR_GOBP_wrapper(pop = "DU6", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 61 genes
WebGestaltR_GOBP_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # sig - 101 genes
WebGestaltR_GOBP_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 189 genes
WebGestaltR_GOBP_wrapper(pop = "FERT", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 103


WebGestaltR_KEGG_wrapper(pop = "DUK", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # sig - 127 genes
WebGestaltR_KEGG_wrapper(pop = "DUC", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 68 genes
WebGestaltR_KEGG_wrapper(pop = "DU6", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 61 genes
WebGestaltR_KEGG_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # sig - 101 genes
WebGestaltR_KEGG_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 189 genes
WebGestaltR_KEGG_wrapper(pop = "FERT", rdd_gene_dat = rdd_q15_q85, q_low = "q15", q_high = "q85") # notsig - 103 genes


# rdd_q20_q80
WebGestaltR_GOBP_wrapper(pop = "DUK", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 193 genes
WebGestaltR_GOBP_wrapper(pop = "DUC", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 154 genes
WebGestaltR_GOBP_wrapper(pop = "DU6", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # sig - 123 genes
WebGestaltR_GOBP_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # sig - 177 genes
WebGestaltR_GOBP_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 302 genes
WebGestaltR_GOBP_wrapper(pop = "FERT", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 198 genes


WebGestaltR_KEGG_wrapper(pop = "DUK", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 193 genes
WebGestaltR_KEGG_wrapper(pop = "DUC", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 154 genes
WebGestaltR_KEGG_wrapper(pop = "DU6", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 123 genes
WebGestaltR_KEGG_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 177 genes
WebGestaltR_KEGG_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # sig - 302 genes
WebGestaltR_KEGG_wrapper(pop = "FERT", rdd_gene_dat = rdd_q20_q80, q_low = "q20", q_high = "q80") # notsig - 198


# rdd_q25_q75
WebGestaltR_GOBP_wrapper(pop = "DUK", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 298
WebGestaltR_GOBP_wrapper(pop = "DUC", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 274
WebGestaltR_GOBP_wrapper(pop = "DU6", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # sig - 286
WebGestaltR_GOBP_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # sig - 248
WebGestaltR_GOBP_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 448
WebGestaltR_GOBP_wrapper(pop = "FERT", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # sig - 315


WebGestaltR_KEGG_wrapper(pop = "DUK", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # sig - 298
WebGestaltR_KEGG_wrapper(pop = "DUC", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 274
WebGestaltR_KEGG_wrapper(pop = "DU6", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 286
WebGestaltR_KEGG_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 248
WebGestaltR_KEGG_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # notsig - 448
WebGestaltR_KEGG_wrapper(pop = "FERT", rdd_gene_dat = rdd_q25_q75, q_low = "q25", q_high = "q75") # sig - 315

```

```{r corroborate_origal_query, eval = FALSE}
# results reproduced!

rdd_q10_q95 <- get_RDDs_and_RDDgenes(q_low = "q10", q_high = "q95")

WebGestaltR_GOBP_wrapper(pop = "DUK", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #notsig - 76
WebGestaltR_GOBP_wrapper(pop = "DUC", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #sig - 37
WebGestaltR_GOBP_wrapper(pop = "DU6", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #notsig - 28
WebGestaltR_GOBP_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # notsig - 35
WebGestaltR_GOBP_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # notsig - 49
WebGestaltR_GOBP_wrapper(pop = "FERT", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # notsig - 9

WebGestaltR_KEGG_wrapper(pop = "DUK", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #sig - 76
WebGestaltR_KEGG_wrapper(pop = "DUC", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #notsig - 37
WebGestaltR_KEGG_wrapper(pop = "DU6", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") #notsig - 28
WebGestaltR_KEGG_wrapper(pop = "DU6P", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # notsig - 35
WebGestaltR_KEGG_wrapper(pop = "DUhLB", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # sig - 49
WebGestaltR_KEGG_wrapper(pop = "FERT", rdd_gene_dat = rdd_q10_q95, q_low = "q10", q_high = "q95") # notsig - 9

```

```{r export_RDD_gene_data, eval = FALSE}
# original approach
rdd_q10_q95$rdd_genes %>% lapply(as.data.frame) %>% bind_rows(.id = "pop") %>% write.csv(here("00_dashboard/data/WebGestaltR_RDD_q10_q95","rdd_genes_q10_q95.csv"))
# relaxed thresholds
rdd_q10_q90$rdd_genes %>% lapply(as.data.frame) %>% bind_rows(.id = "pop") %>% write.csv(here("00_dashboard/data/WebGestaltR_RDD_q10_q90","rdd_genes_q10_q90.csv"))
rdd_q15_q85$rdd_genes %>% lapply(as.data.frame) %>% bind_rows(.id = "pop") %>% write.csv(here("00_dashboard/data/WebGestaltR_RDD_q15_q85","rdd_genes_q15_q85.csv"))
rdd_q20_q80$rdd_genes %>% lapply(as.data.frame) %>% bind_rows(.id = "pop") %>% write.csv(here("00_dashboard/data/WebGestaltR_RDD_q20_q80","rdd_genes_q20_q80.csv"))
rdd_q25_q75$rdd_genes %>% lapply(as.data.frame) %>% bind_rows(.id = "pop") %>% write.csv(here("00_dashboard/data/WebGestaltR_RDD_q25_q75","rdd_genes_q25_q75.csv"))
```

```{r summary_table_significance}

find_if_sig <- function(q_low, q_high, db){
  # define dir where to look for results
  proj_dir <- paste0("WebGestaltR_RDD_q",q_low,"_q",q_high)  
  # Define paths to results in a db
  fls <- list.files(here("00_dashboard/data", proj_dir, db), pattern = "Project_WebGestaltR", full.names = TRUE)
  # names for list
  contr <- fls %>% sapply(basename) %>% str_remove("Project_WebGestaltR_")
  # count if tables are empty or if they have results
  check_sig <- fls %>% sapply(function(d){
    n_lines <- list.files(d, pattern = "_sig_results.csv", full.names = TRUE) %>% readLines() %>% length()
    #ifelse(n_lines > 1, "SIG", "not_sig")
    
    return(n_lines - 1) # substract colnames
    
    })
  # name list
  names(check_sig) <- contr
  
  # sort list according to contrasts
  check_sig <- check_sig[match(paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU"), names(check_sig))]
  
  q_thresholds <- paste0("q",q_low,"vs_q", q_high)
  names(q_thresholds) <- "q_thresholds"
  names(db) <- "db"
  res <- c(db, q_thresholds, check_sig)

  return(res)
}



bind_rows(
  find_if_sig(10,95,"GOBP"), 
  find_if_sig(10,90,"GOBP"), 
  find_if_sig(15,85,"GOBP"),
  find_if_sig(20,80,"GOBP"),
  find_if_sig(25,75,"GOBP")
  ) %>% 
  kable(caption = "Number of significant (FDR < 0.1) terms (GOBP) by gene lists", row.names = FALSE) %>% kable_styling(full_width = F)


bind_rows(
  find_if_sig(10,95,"KEGG"), 
  find_if_sig(10,90,"KEGG"), 
  find_if_sig(15,85,"KEGG"),
  find_if_sig(20,80,"KEGG"),
  find_if_sig(25,75,"KEGG")
  ) %>% 
  kable(caption = "Number of significant (FDR < 0.1) terms (KEGG) by gene lists", row.names = FALSE) %>% kable_styling(full_width = F)
```

```{r function_import_sig_terms}

import_if_sig <- function(q_low, q_high, db){
  # define dir where to look for results
  proj_dir <- paste0("WebGestaltR_RDD_q",q_low,"_q",q_high)  
  # Define paths to results in a db
  fls <- list.files(here("00_dashboard/data", proj_dir, db), pattern = "Project_WebGestaltR", full.names = TRUE)
  # count if tables are empty or if they have results
  get_sig_tab <- fls %>% sapply(function(d){
    
    sig_tab <- list.files(d, pattern = "_sig_results.csv", full.names = TRUE)
    n_lines <- sig_tab %>% readLines() %>% length()
    #ifelse(n_lines > 1, "SIG", "not_sig")
    
    if(n_lines > 1){
      read.csv(sig_tab) %>% 
        dplyr::select(-X) %>% 
        mutate(q_thresholds = paste0("q",q_low,"_vs_q", q_high), contrast = d %>% basename() %>% str_remove("Project_WebGestaltR_")) %>% 
        dplyr::select(contrast, q_thresholds, everything())
    }
    
    })
  
  names(get_sig_tab) <- basename(fls) %>% str_remove("Project_WebGestaltR_")
  
  get_sig_tab[ sapply(get_sig_tab, function(x) !is.null(x) ) ] %>% bind_rows() # remove empty elements and bind rows
}

```

```{r function_to_vis_nr_gene_in_sig_terms}

vis_overlap_heatmap_sig_terms <- function(dat, db){
  
  #dat=gobp_sig_terms
  
  ma <- dat %>% 
    mutate(tmp = paste0(contrast," (",q_thresholds,")")) %>% 
    reshape2::dcast(description ~ tmp, value.var = "overlap") %>% 
    tibble::column_to_rownames(var = "description") %>% 
    as.matrix()
    
  ma[is.na(ma)] <- 0
  
  #coldat <- data.frame( contrast = colnames(ma) %>% str_split("[+]") %>% sapply(function(x) x[1]) )
  #rownames(coldat) <- colnames(ma) %>% str_split("[+]") %>% sapply(function(x) x[2]) 
  
  pheatmap::pheatmap(ma, display_numbers = TRUE, number_format = "%.0f", 
                     main = paste0("Significant terms (", db,") and number of overlapping genes"))
  
}

```

```{r import_sig_terms}

gobp_sig_terms <- bind_rows(
  import_if_sig(10,95,"GOBP"), 
  import_if_sig(10,90,"GOBP"), 
  import_if_sig(15,85,"GOBP"),
  import_if_sig(20,80,"GOBP"),
  import_if_sig(25,75,"GOBP")
  )

kegg_sig_terms <- bind_rows(
  import_if_sig(10,95,"KEGG"), 
  import_if_sig(10,90,"KEGG"), 
  import_if_sig(15,85,"KEGG"),
  import_if_sig(20,80,"KEGG"),
  import_if_sig(25,75,"KEGG")
  )

```

### GOBP_summary
```{r GOBP_summary, fig.dim=c(10,8)}
vis_overlap_heatmap_sig_terms(gobp_sig_terms, "GOBP")
```

### KEGG_summary
```{r KEGG_summary, fig.dim = c(10,8)}
vis_overlap_heatmap_sig_terms(kegg_sig_terms, "KEGG")
```
###--EndOfSection


RDD-regSNPs {data-navmenu="RDDs"}
===================================

```{r find_RDD_SNPs_in_regulatory_regions} 

#distinct_windows_list aka RDDs

# find 
reg_SNPs_in_RDDs <- lapply(1:length(distinct_windows_list), function(i){ 

  #i=1
  
  # get gr for contrast
  x <- distinct_windows_list[[i]]
  gr <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  
  # get SNPs in distinct_windows_lists
  SNPs_in_RDDs <- subsetByOverlaps(x = vcf_final_snp_ids_gr, ranges = gr, minoverlap = 1) 
  
  # Find overlaps between SNPs in RDDs and regulatory or motif features
  ov_rf <- findOverlaps(query = SNPs_in_RDDs, subject = regulatory_features, minoverlap = 1)

  # subset overlapping snps
  SNPs_in_RDDs_in_rf <- SNPs_in_RDDs[queryHits(ov_rf)]

  # add feature information
  mcols(SNPs_in_RDDs_in_rf) <- regulatory_features[subjectHits(ov_rf)] %>% 
    as.data.frame() %>% 
    dplyr::rename(
      feat_seqname = seqnames, feat_start=start, feat_end=end, feat_width=width, feat_strand = strand,
      feat_source = mcols.source, feat_feature=mcols.feature, feat_attribute=mcols.attribute
      ) 
  
  # import per SNP Fst scores and convert to granges
  tmp <- vroom(
    here(
      "batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",
      paste0(names(distinct_windows_list)[i],"_vs_FZTDU.weir.fst")
      ),
    col_types = c(CHROM = "c")
    )
  
  # add fst information to SNPs-in-RDDs-in-regulatory-regions
  SNPs_in_RDDs_in_rf <- SNPs_in_RDDs_in_rf %>% 
    # turn back to data frame
    as.data.frame() %>% 
    # inner join with fst data
    inner_join(tmp, by = c("seqnames" = "CHROM", "start" = "POS")) %>% 
    # turn back into granges
    makeGRangesFromDataFrame(keep.extra.columns=TRUE)

  return(SNPs_in_RDDs_in_rf)  
  
  }
) 

names(reg_SNPs_in_RDDs) <- names(distinct_windows_list)
```

Column {data-width=300}
------------------------

### Overview

* RDDs = Regions of Distinct Genetic Differentiation. 

* Genetic Differentiation = Genetic differentiation of a selected line relative to FZTDU.

* Distinct = a genomic region that is highly differentiated in one population, but has low genetic differentiation in all other populations

* Region = window of 50Kb with at least 10 SNPs. In which Fst was calculated at each SNP and then individual Fst scores were averaged.

* high genetic differentiation = window with a mean Fst in the 95th quantile.

* low genetic differentiation = window with a mean Fst in the 10th quantile.

* Regulatory regions = The Ensembl Regulatory Build. Publication: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0621-5. Data set: ftp://ftp.ensembl.org/pub/release-93/regulation/mus_musculus/mus_musculus.GRCm38.motiffeatures.20180516.gff.gz

* RDD-SNPs were overlapped with regulatory features. Redundant hits (a regulatory feature overlaped by >1 SNP) were uniq'd.

* The first gene downstream to a regulatory features of interest was identified. Genes were found with the function matchGenes() of bioconductor's package bumphunter. Genes set used also from bioconductor (TxDb.Mmusculus.UCSC.mm10.ensGene).


Column  {data-width=700, .tabset}
-----------------------------------

### Number of regSNPs

```{r display_nr_hits_RDD_regSNPs} 

lapply(reg_SNPs_in_RDDs, length) %>% as.data.frame() %>% 
  #mutate(contrast = factor(contrast, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU"))) %>% 
  dplyr::select(c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, caption = "Number of regSNPs in RDDs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


lapply(reg_SNPs_in_RDDs, 
       function(x){
         x %>% 
           as.data.frame() %>% 
           dplyr::select(feat_seqname, feat_start, feat_end) %>% 
           unique() %>% 
           nrow()
         }
       ) %>% 
  as.data.frame() %>% 
  dplyr::select(c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, caption = "Number of regulatory regions containing RDD-SNPs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


```

### Downstream_genes
```{r find_downstream_genes_for_regSNPs_RDD_regions, eval = FALSE}

#load bioconductor gene set
mm10_ensGene <- annotateTranscripts(TxDb.Mmusculus.UCSC.mm10.ensGene) 

# for every set of regSNPs find the nearest downstream gene

downstream_genes <- lapply(1:length(reg_SNPs_in_RDDs), function(i){

  print(names(reg_SNPs_in_RDDs)[i])
  reg_snps <- reg_SNPs_in_RDDs[[i]]
  
  # extract regulatory feature information 
  reg_range <- GRanges(seqnames = paste0("chr",reg_snps$feat_seqname),
                       IRanges(start = reg_snps$feat_start, end = reg_snps$feat_end))
  
  # add feature width
  reg_range$feature_width <- reg_snps$feat_width
  
  # add fst scores
  reg_range$WEIR_AND_COCKERHAM_FST <- reg_snps$WEIR_AND_COCKERHAM_FST

  # extract feature type information and make it a column
  reg_range$feature_type <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("feature_type",x)] %>% str_remove("feature_type=") )

  # extract regulatory ensembl id and make it a column
  reg_range$feature_id <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("ID=regulatory_region:",x)] %>% str_remove("ID=regulatory_region:") )

  # many entries are repeated, because >1 SNP was found in a regulatory region
  # include the number of SNPs in the feature.
  reg_range_uniq <- reg_range %>% 
    as.data.frame() %>% 
    group_by(seqnames, start, end, width, strand, feature_width, feature_type, feature_id) %>% 
    summarise(n_snps_in_feature = n(), mean_fst_snps_in_feature = mean(WEIR_AND_COCKERHAM_FST)) %>% 
    ungroup() %>% 
    dplyr::select(seqnames, start, end, width, feature_width, n_snps_in_feature, mean_fst_snps_in_feature, feature_type, feature_id)
    
  # search for nearest gene downstream
  tab <- matchGenes(reg_range_uniq, mm10_ensGene, skipExons=TRUE, type = "fiveprime") %>% 
    # add biotype and gene symbol  
    left_join(
      mmu_gene_set %>% dplyr::select(gene_id, gene_symbol, gene_biotype), 
      by = c("Geneid"="gene_id")
      ) 

  # add gene information to matching regulatory region
  reg_range_uniq$matching_gene_id <- tab$Geneid
  reg_range_uniq$gene_symbol <- tab$gene_symbol
  reg_range_uniq$gene_biotype <- tab$gene_biotype
  reg_range_uniq$gene_strand <- tab$strand
  reg_range_uniq$distance_to_gene <- tab$distance

 return(reg_range_uniq)
})

# name elements in the list
names(downstream_genes) <- names(reg_SNPs_in_RDDs)


# export for later use
lapply(names(downstream_genes), function(nm){
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  downstream_genes[[nm]] %>% write.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl))
})

```

```{r display_feature_tabulation_type_of_feature_by_contrast} 

lapply(1:length(reg_SNPs_in_RDDs), function(i){
  nm <- names(reg_SNPs_in_RDDs)[i]
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl)) %>% 
    group_by(feature_type) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm)
}) %>% 
  bind_rows() %>% 
  reshape2::dcast(feature_type ~ contrast, value.var = "n") %>% 
  dplyr::select(feature_type, c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, 
        caption = "Number of regulatory regions (with >=1 RDD-SNP) by contrast",
        format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)

```

```{r display_feature_tabulation_type_of_feature_by_gene_biotype}
#lapply(1:length(reg_SNPs_in_RDDs), function(i){
#  nm <- names(reg_SNPs_in_RDDs)[i]

contr_nms <- c("DUK","DUC","DU6","DU6P","DUhLB","FERT")

tabs <- lapply(contr_nms, function(nm){
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl)) %>% 
    group_by(feature_type, gene_biotype) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm) %>% 
    reshape2::dcast(gene_biotype ~ feature_type, value.var = "n") %>% 
    filter(!is.na(gene_biotype)) 
}) 

print_kable2 <- function(tab, contr_nm){
  ttl <- paste0("Number of regulatory regions (with >=1 RDD-SNP) by gene biotype (", contr_nm, ")")

  tab %>% 
    kable(digits = 2, caption = ttl, format.args = list(big.mark = ",")) %>% 
    kable_styling(full_width = F)
}


print_kable2(tabs[[1]], contr_nms[1])
print_kable2(tabs[[2]], contr_nms[2])
print_kable2(tabs[[3]], contr_nms[3])
print_kable2(tabs[[4]], contr_nms[4])
print_kable2(tabs[[5]], contr_nms[5])
print_kable2(tabs[[6]], contr_nms[6])

```

### DEG_hits
```{r find_and_display_hits_BETWEEN_RDD_regSNPs_matching_genes_AND_DEGs}

# are there any hits between RDD-regSNPs-matching-genes and DEGs? 
## Only DUK: Atoh8 (FL1_testis_DEGs_hits)
## and DU6: Cd74 (FL1_testis_DEGs_hits)
RDD_regSNPs_matching_genes_x_DEGs <- list.files(here("00_dashboard/data/regulatory_RDD_SNPs"), full.names = TRUE) %>% 
  lapply(function(x){ 
    
    # import RDD-regSNPs-matching-genes data and convert to genomic ranges
    d <- read.csv(x) %>% 
      dplyr::select(seqnames, start, end, feature_type, feature_id, matching_gene_id, gene_symbol, gene_biotype) %>% 
      mutate(
        pop = basename(x) %>% str_remove(".csv") %>% str_remove("RDD_reg_features_")
      ) 
    
    res <- list(
      FL1_testis_DEGs_hits = d %>% filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols),
      FL2_testis_DEGs_hits = d %>% filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols),
      FL1_ovary_DEGs_hits = d %>% filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols)
      ) %>% 
      bind_rows(.id = "DEG_hit_type")
    
    return(res)
    
    }
  ) %>% 
  bind_rows()


RDD_regSNPs_matching_genes_x_DEGs %>%     
  kable(
    caption = "Hits between RDD-regSNPs-matching-genes and DEGs (only for DU6 and DUK)"
    ) %>% 
  kable_styling(full_width = F)


```

### RDD-genes-hits
```{r find_and_display_hits_BETWEEN_RDD_regSNPs_matching_genes_AND_RDD-genes}

# are there any hits between RDD-regSNPs-matching-genes and RDD-genes? 
## Many of the RDD genes appear downstram RDD-regSNPs ... this is not very informative, 
## ... but is expected, since RDD-regSNPs and RDD-genes were extracted from the same genomic intervals (RDD regions).
lapply(names(RDD_genes), function(nm){
  
  d <- paste0(here("00_dashboard/data/regulatory_RDD_SNPs"), "/RDD_reg_features_", nm, ".csv") %>% 
    read.csv() %>% 
    dplyr::select(-X) 

  data.frame(
    pop = nm,
    n_RDD_genes = RDD_genes[[nm]] %>% length(),
    n_hits = RDD_genes[[nm]] %>% as.data.frame() %>% filter(mcols.gene_id %in% d$matching_gene_id) %>% nrow()
    ) %>% 
    mutate(
      pct_hits = ((n_hits/n_RDD_genes) * 100) %>% round(2)
      ) 
}) %>% 
  bind_rows() %>%     
  kable(
    caption = "Hits between RDD-regSNPs-matching-genes and RDD-genes"
    ) %>% 
  kable_styling(full_width = F)

```

### DUK
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_DUK.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```

### DUC
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_DUC.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```

### DU6
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_DU6.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```

### DU6P
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_DU6P.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```

### DUhLB
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_DUhLB.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```

### FERT
```{r}
here("00_dashboard/data/regulatory_RDD_SNPs", "RDD_reg_features_FERT.csv") %>% 
  read.csv() %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(mean_fst_snps_in_feature = round(mean_fst_snps_in_feature,2)) %>% 
  datatable(options = list(pageLength = 100), caption = "Regulatory Features overlapped by RDD-SNPs")
```




RDD-gene-selected {data-navmenu="RDDs"}
==============================================

```{r set_biomart_for_Gviz, eval=FALSE} 
# biomart query done on Monday 05/11/2020

bm <- useMart(host = "www.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL",
              dataset = "mmusculus_gene_ensembl")

```

```{r macro_Gviz_function, eval = FALSE}
vis_goi_macro <- function(ens_gene_id, pos_snps, bp_expand){
  
  
  # merge (full join) fst windows
  win_fst_merged_gr <- lapply(as.character(unique(win_fst_sel_vs_ctrl_prep$contrast)), 
                              function(i){ 
                                x <- win_fst_sel_vs_ctrl_prep %>% 
                                  filter(contrast == i) %>% 
                                  ungroup() %>% 
                                  dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST)
                                
                                names(x)[names(x) == "MEAN_FST"] <- i

                                return(x)
                              }) %>% 
    purrr::reduce(full_join, by = c("CHROM","BIN_START","BIN_END")) %>% 
    dplyr::rename(seqnames=CHROM, start=BIN_START,end=BIN_END) %>% 
    mutate(
      DUK_vs_FZTDU = ifelse(is.na(DUK_vs_FZTDU), 0, DUK_vs_FZTDU),
      DUC_vs_FZTDU = ifelse(is.na(DUC_vs_FZTDU), 0, DUC_vs_FZTDU),
      DU6_vs_FZTDU = ifelse(is.na(DU6_vs_FZTDU), 0, DU6_vs_FZTDU),
      DU6P_vs_FZTDU = ifelse(is.na(DU6P_vs_FZTDU), 0, DU6P_vs_FZTDU),
      DUhLB_vs_FZTDU = ifelse(is.na(DUhLB_vs_FZTDU), 0, DUhLB_vs_FZTDU),
      FERT_vs_FZTDU = ifelse(is.na(FERT_vs_FZTDU), 0, FERT_vs_FZTDU)
    ) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) 


  # get location of gene of interest
  goi_gr <- mmu_gene_set_gr[mmu_gene_set_gr$mcols.gene_id == ens_gene_id]
  
  # expand location 50K to each side
  goi_gr_expanded <- resize(goi_gr, width = width(goi_gr)+(bp_expand), fix = "center")
  
  # define windows containing the gene
  wins <- subsetByOverlaps(win_fst_merged_gr, 
                           goi_gr_expanded, 
                           minoverlap = 1)
  
  
  # define ideogram for chromosome
  itrack <- IdeogramTrack(genome = "mm10", chromosome = as.character(unique(seqnames(wins))))
  
  # track for genomic positions
  gtrack <- GenomeAxisTrack()
  
  wins_track <- DataTrack(wins, name = "Windowed_Fst")
  
  
  # show snp(s) of interest
  ht_snp <- HighlightTrack(trackList = list(wins_track), 
                           start = pos_snps, 
                           end = pos_snps, 
                           lty = "dashed",
                           col = "grey",
                           inBackground=FALSE,
                           chromosome = as.character(seqnames(goi_gr)))

  # show all RDD regions in window
  rdds <- lapply(1:length(distinct_windows_list), function(i){
    
    x <- distinct_windows_list[[i]]
    
    gr_x <- x %>% 
      dplyr::select(CHROM, BIN_START, BIN_END) %>% 
      dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
      makeGRangesFromDataFrame()
    
    gr_xx <- subsetByOverlaps(gr_x, wins, minoverlap = 1)
    
    # convert gr_xx from a range to a point to avoid adjacent low fst windows
    gr_xxx <- GRanges(
      seqnames = seqnames(gr_xx), 
      IRanges(
        start = start(gr_xx) + floor(width(gr_xx)/2),
        end = start(gr_xx) + floor(width(gr_xx)/2)
      )
    )    
    
    # reduce rdd points
    gr_xxx_red <- GenomicRanges::reduce(gr_xxx, min.gapwidth = 25000) #Ranges separated by a gap of at least min.gapwidth positions are not merged
    
    # add population name
    mcols(gr_xxx_red) <- names(distinct_windows_list)[i]
    
    return(gr_xxx_red)

  }) %>% 
    GRangesList() %>% 
    unlist()  
  
 
  #rdd_track <- AnnotationTrack(start = start(rdds), end = end(rdds), 
  #                             chromosome = as.character(unique(seqnames(rdds))), 
  #                             id = unlist(mcols(rdds)), 
  #                             name = "RDDs")

 
  #plotTracks(list(itrack, gtrack, rdd_track, ht_snp), groups = names(mcols(wins)),type = "a", transcriptAnnotation = "symbol", featureAnnotation = "id", fontsize=10)
  plotTracks(list(itrack, gtrack, ht_snp), groups = names(mcols(wins)),type = "a", transcriptAnnotation = "symbol", featureAnnotation = "id")
  
}

```

```{r micro_Gviz_function, eval = FALSE}
vis_goi_micro <- function(ens_gene_id, pos_snps){ 
    
  # define contrasts for convenience
  contrasts <- paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU") %>% 
    # sort to match groups to data later in plotracks
    sort()
  
  # get goi location
  query_genes <- mmu_gene_set_gr[mmu_gene_set_gr$mcols.gene_id == ens_gene_id]
  
  # get goi per snp fst scores at each contrast
  per_snp_fst_goi <- lapply(contrasts, function(contrast){
    
    # define fst per snp file for contrast
    fl_nm <- paste0(contrast, ".weir.fst")
    
    # subset snps fst scores overlapping with goi
    tab <- subsetByOverlaps(
      # create per-snp fst granges here to avoid storing into mem  
      vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl_nm), col_types = c(CHROM="c")) %>% 
        mutate(start=POS, end=POS) %>% 
        dplyr::rename(seqnames=CHROM) %>% 
        dplyr::select(seqnames, start, end, WEIR_AND_COCKERHAM_FST) %>% 
        makeGRangesFromDataFrame(keep.extra.columns = TRUE),
      query_genes, 
      minoverlap = 1
    ) %>% 
      as.data.frame() 
    
    names(tab)[names(tab)=="WEIR_AND_COCKERHAM_FST"] <- contrast
    
    return(tab)
    
  }) %>% 
    # merge goi snp fst scores at each contrast   
    purrr::reduce(full_join, by = c("seqnames","start","end")) %>% 
    # select columns to work with  
    dplyr::select(seqnames, start, end, ends_with("_vs_FZTDU")) %>% 
    # turn NAs into 0s, they are in practice the same    
    mutate(
      DUK_vs_FZTDU = ifelse(is.na(DUK_vs_FZTDU), 0, DUK_vs_FZTDU),
      DUC_vs_FZTDU = ifelse(is.na(DUC_vs_FZTDU), 0, DUC_vs_FZTDU),
      DU6_vs_FZTDU = ifelse(is.na(DU6_vs_FZTDU), 0, DU6_vs_FZTDU),
      DU6P_vs_FZTDU = ifelse(is.na(DU6P_vs_FZTDU), 0, DU6P_vs_FZTDU),
      DUhLB_vs_FZTDU = ifelse(is.na(DUhLB_vs_FZTDU), 0, DUhLB_vs_FZTDU),
      FERT_vs_FZTDU = ifelse(is.na(FERT_vs_FZTDU), 0, FERT_vs_FZTDU)
    ) %>% 
    # convert to granges  
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) 
  
  # make a track for per snp fst scores
  snp_fst_track <- DataTrack(per_snp_fst_goi, name = "SNP_Fst")
  
  # get transcripts for GOI
  gene_track <- BiomartGeneRegionTrack(genome = "mm10", 
                                       name = "RDD-Genes",
                                       gene = ens_gene_id, 
                                       biomart = bm)
  
  # show snp(s) of interest
  ht_snp <- HighlightTrack(trackList = list(snp_fst_track, gene_track), 
                           start = pos_snps, 
                           end = pos_snps, 
                           lty = "dashed",
                           col = "grey",
                           inBackground=FALSE,
                           chromosome = as.character(seqnames(query_genes)))
  
  # track for genomic positions
  gtrack <- GenomeAxisTrack()
  
  # plot tracks
  plotTracks(list(gtrack, ht_snp), 
             groups = names(mcols(per_snp_fst_goi)),
             type = c("a","p"),
             transcriptAnnotation = "transcript")
  }

```

```{r pie_chart_function, eval = FALSE}
make_frq_pie_chart <- function(chr, pos, rsid, ref_allele, alt_allele){
  
  hom_ref_gt <- paste(ref_allele,ref_allele,sep="/")
  hom_alt_gt <- paste(alt_allele,alt_allele,sep="/")
  het_gt <- paste(ref_allele,alt_allele,sep="/")  
  
  ttl <- paste0(rsid, ", ", chr,":",pos,"-",pos, ", ",ref_allele,"/",alt_allele)

  
  p <- read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
    dplyr::select(-X) %>% 
    filter(CHROM == chr & POS == pos) %>% 
    reshape2::melt(id.vars = c("CHROM", "POS", "REF", "ALT")) %>% 
    mutate(
      
      hom_ref = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% 
                         .[grep(hom_ref_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_ref_gt,")"), TRUE))) %>% as.numeric(),
      
      hom_alt = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% .[grep(hom_alt_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_alt_gt,")"), TRUE))) %>% as.numeric(),
      
      het = sapply(value, function(x) str_split(x,";") %>% 
                     unlist() %>% 
                     .[grep(het_gt,.)] %>% 
                     str_remove(fixed(paste0("(",het_gt,")"), TRUE))) %>% as.numeric(),
      
      miss = sapply(value, function(x) str_split(x,";") %>% 
                      unlist() %>% 
                      .[grep("./.",., fixed = TRUE)] %>% 
                      str_remove(fixed("(./.)", TRUE))) %>% as.numeric(),
      
      hom_ref = ifelse(is.na(hom_ref), 0, hom_ref),
      hom_alt = ifelse(is.na(hom_alt), 0, hom_alt),
      het = ifelse(is.na(het), 0, het),
  
      ref_frq = (hom_ref*2 + het*1)/((hom_alt+hom_ref+het)*2),
      alt_frq = (hom_alt*2 + het*1)/((hom_alt+hom_ref+het)*2),
      
      variable = str_remove(variable, "_GT_counts"),
      variable = ifelse(variable == "DUK_DUC", "FERT", variable)
      ) %>% 
    dplyr::rename(pop=variable) %>% 
    reshape2::melt(id.vars = c("pop","REF","ALT"), measure.vars = c("ref_frq", "alt_frq"), variable.name = "allele", value.name = "frq") %>% 
    mutate(
      allele = ifelse(allele == "ref_frq", paste0("REF:",REF), paste0("ALT:",ALT)),
      allele = factor(allele, levels = c(paste0("REF:",ref_allele), paste0("ALT:",alt_allele)))
      ) %>% 
    arrange(pop) %>% 
    dplyr::select(pop, allele, frq) %>% 
    mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FERT","FZTDU"))) %>% 
   
    ggplot(., aes(x="", y=frq, fill=allele)) +
      geom_bar(stat="identity", width=1, color="white") +
      coord_polar("y", start=0) +
      theme_void() +# remove background, grid, numeric labels
      theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = 5)) +
      scale_color_manual(values = c("lightblue","darkgrey"), aesthetics = "fill") +
      facet_wrap(~pop, nrow = 1) +
      labs(title = ttl)
  print(p)
   
}

```




Column {.tabset} 
-----------------

### RDD-genes 

- Looking at the functional annotation of RDD-genes (GOBP) and results of ORA analysis, the following genes could be interesting:

  - **DUK**: 
    - Trp63
      - Relevant GOBPs: negative regulation of intracellular estrogen receptor signaling pathway.
      - SNP effect impacts are only "LOW" or "MODIFIER". Mostly intron variants with Fst > 0.9.
    - ORA showed as significantly enriched the term **"Phospholipase D signaling pathway"** with the RDD-genes (Raf1, Adcy6, Grm8, Tsc1, Ralgds). 
      - There are hints in the literature that this term could influence reproduction.
        - "Role of phospholipase-D and phosphatidic acid in mediating gonadotropin-releasing hormone-induced inhibition of preantral granulosa cell differentiation" https://academic.oup.com/endo/article/135/3/1205/3035944
        - "Role of phospholipase D in regulation of testicular Leydig cell hyperplasia in Sprague–Dawley rats treated with di(2-ethylhexyl) phthalate" https://link.springer.com/article/10.1007/s00204-010-0618-5
        - "Gonadotropin-releasing Hormone Activates Phospholipase D in Ovarian Granulosa Cell" https://www.jbc.org/content/264/20/11762.full.pdf
      - The gene **Tsc1** appears to be involved in reproduction and **contains 3 moderate-missense mutations with Fst>0.95.**. However, none of these non-synonimous mutations has SIFT values indicative of a deleterious effect on the protein (see table below).
        - *"Collectively, these results indicate that Tsc1 has critical roles in preserving the ovarian reserve of oocytes in addition to maintaining the normal architecture of the oviduct for appropriate transport of preimplantation embryos to the uterus."* PMC3690805
        
  - **DUC**: 
    - Pgr:
      - Relevant GOBPs: steroid hormone mediated signaling pathway; progesterone receptor signaling pathway; ovulation from ovarian follicle; female mating behavior,negative regulation of granulosa cell apoptotic process.
      - **SNP effect MODERATE (missense mutation) with high Fst (0.934468).**
      - This is gene was found enriching (ORA) the GOBP terms "hormone-mediated signaling pathway" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
      - Note that P4 is higher in this line than in FZTDU and DUK (Langhammer et al 2014, PMID:24248751)
      - The missense SNP (9  8901222  rs16808511  G  A) SIFT score (VEP query) was 0.04 for two transcripts and therefore deleterious.
    - Pias2:
      - Relevant GOBPs: androgen receptor signaling pathway; negative regulation of androgen receptor signaling pathway.
      - SNP effects MODIFIERS, mainly intron variants, mostly Fst ~0.9.
      - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway", "regulation of dendrite morphogenesis" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
    
    - Rxfp1
       - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway" after ORA analysis.
       - Also, one of the GOBP terms related to reproduction is "parturition".
       - All effects (~200) are modifiers. Mostly intronic variants.
       - Several of those mutations are Fst=1 and the majority are Fst>0.6.
       
    - Yap1   
      -  Relevant GOBPs: regulation of canonical Wnt signaling pathway; response to progesterone; progesterone receptor signaling pathway; positive regulation of intracellular estrogen receptor signaling pathway.
      - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
      - Mutation effects are either low or (mostly) modifiers. Almost all SNPs are > 0.9.
    - "regulation of dendrite morphogenesis"  
      - This term showed up as  significantly enriched by ORA analysis. In contains 3 genes (Pias2, Skor2 and Trpc6).
      - Although this enriched term is not very informative, the RDD-genes therein still might be of interest:
        - Pias2: see above.
        - Skor2: 
          - "In the present study, we identified four significant SNPs after chromosome-wise multiple test correction. The first SNP rs29972765 is located on chr18:77022376 and its nearest gene is Skor2 (SKI family transcriptional corepressor 2). In addition to Skor2, another five genes appeared to locate on the same LD block: Ier3ip1 (immediate early response 3 interacting protein 1), Hdhd2 (haloacid dehalogenase-like hydrolase domain containing 2), Katnal2 (katanin p60 subunit A-like 2), Pias2 (protein inhibitor of activated STAT 2), and St8sia5 (ST8 alpha-N-acetyl-neuraminide alpha-2,8-sialyltransferase 5). " (PMC5015949)
          - GOBP: " negative regulation of BMP signaling pathway" --> (https://www.sciencedirect.com/science/article/pii/S0012160616306686) "BMP signaling regulates multiple reproductive processes across Metazoa."; "Germ line specification, proliferation, migration, and gametogenesis are often BMP-dependent."; "Some BMP ligands play sex-specific reproductive roles in vertebrates.".
          - SNPs effects are mostly modifiers, but there is one moderate-missense effect. However, this mutations has Fst=NA. The other many (>100) modifier effects are Fst > 0.8.
        - Trpc6: 
          - GOBP: "single fertilization" --> "The union of male and female gametes to form a zygote." (http://www.informatics.jax.org/vocab/gene_ontology/GO:0007338).
          - "In conclusion, expression of TRPC1, 2, 3, 4 and 6 isoforms in uterine epithelium were modulated through the stages of the EC and likely as a result of hormonal changes, or due to different expression levels of the estrogen or progesterone receptors throughout the EC." (https://link.springer.com/article/10.1007%2Fs11033-019-04857-w).
          - For these gene there is one moderate-missense effect and 29 modifier effects. However the missense mutatation shows low differentiation (Fst=0.120401). Modifier mutations (~15) are Fst~0.9 and are intron variants.

  - **FERT**: 
    - ORA returned no significant results.
    - Zfp366
      - response to estrogen; negative regulation of intracellular estrogen receptor signaling pathway.
      - SNP effect impacts only "LOW" or "MODIFIER". Negligeble Fst (0.0130608).
    - Eif2ak3
      - "The Unfolded Protein Response Regulates Uterine Myocyte Antioxidant Responsiveness During Pregnancy" PMID: 27733380
      - three missense_variant with Fst > 0.98    

  - **DuhLB**: 
    - Two significant ORA results were returned: "Histidine metabolism" and "beta-Alanine metabolism"
    - For both terms, two RDD genes were used: Aldh3a2 and Aldh3a1. Of which Aldh3a2 could be interensting (see below).
    - Other genes could be interesting as well due to their glucose and fatty acid metabolism related functions:
    - Srsf6
      - response to insulin
      - negative regulation of type B pancreatic cell apoptotic process
      - just modifiers, two mutations with Fst~0.9
    - Aldh3a2
      - lipid metabolic process; fatty acid metabolic process
      - **3 moderate-missense variants with Fst~0.7, two of them SIFT < 0.05 and one SIFT = 0.07**
    - Pfkfb3
      - fructose metabolic process; fructose 2,6-bisphosphate metabolic process; metabolic process; carbohydrate phosphorylation
      - just a modifier with Fst=NA.
    - Slc37a1
      - glucose-6-phosphate transport ; carbohydrate transport
      - just modifiers and low effects. Almost all mutations Fst > 0.9

  - **DU6**
    - ORA returned no significant results.
    - Not clear evidence after inspection of GOBP terms looking for terms related to lipid metabolism, the methionine cycle, and cytochrome P450 regulation, lipid synthesis and cytochrome P450 detoxification (see https://www.biorxiv.org/content/10.1101/2020.05.11.088625v1.full).
    - ORA returned no significant results.
    - However, these genes could be related to body size:
      - Hcrt
        - GOBP: response to starvation, eating behavior.
        - only modifier effects. ~8 SNPs with Fst~0.9.
      - Kat2a	  
        - GOBP: " intracellular distribution of mitochondria", "positive regulation of gluconeogenesis", "response to nutrient levels".
        - There are unfortunatelly no SNPs within this gene.
      - Frmd4a
        - GOBP: positive regulation of protein secretion; negative regulation of protein secretion
        - **It has one missense mutation with high differentiation (~0.97).**
      - Atp11b  
        - GOBP: phospholipid transport; phospholipid translocation
        - **missense_variant with complete differentiation (Fst = 1)**. Tolerated (SIFT ~ 0.55).
        - "Obesity, adipokines, and C-peptide are associated with distinct plasma phospholipid profiles in adult males, an untargeted lipidomic approach" https://www.nature.com/articles/s41598-017-05785-0
        - "Lipidomics Reveals Associations of Phospholipids With Obesity and Insulin Resistance in Young Adults" https://pubmed.ncbi.nlm.nih.gov/26709969/
  
  - **DU6P**
    - ORA returned no significant results.
    - Vwa2
      - GOBP: insulin receptor signaling pathway
        - protein synthesis via mTOR and downstream elements (https://www.cellsignal.de/contents/science-cst-pathways-cellular-metabolism/insulin-receptor-signaling/pathways-irs)
        - "Insulin rapidly activates protein synthesis ... The rapid activation of protein synthesis by insulin is mediated primarily through phosphoinositide 3-kinase. This involves the activation of PKB (protein kinase B) ... PKB also phosphorylates the TSC1 (tuberous sclerosis complex 1)–TSC2 complex to relieve its inhibitory action on the mTOR (mammalian target of rapamycin). Inhibition of mTOR by rapamycin markedly impairs insulin-activated protein synthesis. mTOR controls translation initiation and elongation."  (https://portlandpress.com/biochemsoctrans/article-abstract/34/2/213/63821/Regulation-of-protein-synthesis-by-insulin?redirectedFrom=fulltext)
        - "Insulin signaling regulates protein synthesis and degradation as well as posttranslational modifications at the tissue level and coordinates proteostasis at the organism level...effects of insulin on amino acid flux in skeletal muscle ... The coordination of amino acid flux between tissues by insulin helps orchestrate proteostasis at the organismal level to maintain normal growth. " https://www.sciencedirect.com/science/article/pii/S1550413117303522
      - moderate missense SNP with low differentiation (Fst ~ 0.13)
    - Afap1l2
      - GOBP: positive regulation of epidermal growth factor receptor signaling pathway

        - "In vivo EGF treatment markedly activates asymmetric divisions of dystrophin-deficient satellite cells in mdx mice, increasing progenitor numbers, enhancing regeneration, and restoring muscle strength. Therefore, activating an EGFR-dependent polarity pathway promotes functional rescue of dystrophin-deficient satellite cells and enhances muscle force generation." PMID: 30713094
 
      - **missense_variant with high differentiation Fst~0.95 **. However, it is predicted to be tolerated (SIFT 0.15 - 0.48)

    - Vti1a
      - GOBP: protein transport; intracellular protein transport;voluntary musculoskeletal movement
      - only low and modifier effects
      - many SNPs (intronic) in complete differentiation and Fst>0.9

    - Casp7
      - GOBP: protein processing; **striated muscle cell differentiation**
      - missense_variant with low Fst ~ 0.25

    - Adrb1
      - GOBP: negative regulation of multicellular organism growth; positive regulation of cell growth involved in cardiac muscle cell development
      - only low and modifier effects.

    - Cbfa2t3
      - GOBP: negative regulation of cell population proliferation; positive regulation of proteasomal ubiquitin-dependent protein catabolic process; negative regulation of glycolytic process
      - only modifier and low effects.

    - Aprt
      - GOBP: cellular response to **insulin stimulus**
      - only modifier effects.

    - Cdt1
      - GOBP: positive regulation of protein-containing complex assembly; cell division
      - only modifier effects.

    - Plag1
      - GOBP: multicellular organism growth; organ growth
      - only modifier effects

    - Impad1
      - GOBP: skeletal system development
      - **missense_variant Fst ~ 0.84**

    - Cdh13
      - GOBP: negative regulation of cell population proliferation; positive regulation of smooth muscle cell proliferation;regulation of epidermal growth factor receptor signaling pathway
      - missense_variant low Fst ~ 0.23

    - Hsbp1
      - GOBP: muscle contraction
      - It contains no SNPs


### DUC(Pgr) 
```{r gviz_around_Pgr, eval = FALSE}

png(here("00_dashboard/figures","Pgr_Gviz_macro.png"), res = 300, width = 3500, height = 1500) #width = 3500, height = 1500

vis_goi_macro(ens_gene_id = "ENSMUSG00000031870", pos_snps = 8901222, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Pgr_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Pgr_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000031870", pos_snps = 8901222) 

dev.off()
```

```{r make_frq_pie_chart, eval = FALSE}  
png(here("00_dashboard/figures","Pgr_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 9, pos = 8901222, rsid = "rs16808511", ref_allele = "G", alt_allele = "A")

dev.off()
```

```{r import_gviz_around_Pgr, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Pgr_Gviz_macro.png"), 
    here("00_dashboard/figures","Pgr_Gviz_micro.png"),
    here("00_dashboard/figures","Pgr_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Pgr} 

read.table(here("00_dashboard/data/VEP/rs16808511_chr9_pos8901222_GA_Pgr.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 



### DUK(Tsc1) 
```{r gviz_around_Tsc1, eval = FALSE} 
png(here("00_dashboard/figures","Tsc1_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000026812", pos_snps = c(28671945, 28687118, 28686868), bp_expand=1000000)

dev.off()
``` 

```{r visualize_per_snp_Fst_Tsc1_and_export_micro, eval = FALSE} 

png(here("00_dashboard/figures","Tsc1_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000026812", pos_snps = c(28671945, 28687118, 28686868)) 

dev.off()
```

```{r make_frq_pie_chart_Tsc1, eval = FALSE}

#2:28671945-28671945 rs27210729 A/G
p1 <- make_frq_pie_chart(chr = 2, pos = 28671945, rsid = "rs27210729", ref_allele = "A", alt_allele = "G")

#2:28687118-28687118 rs27210669 A/G
p2 <- make_frq_pie_chart(chr = 2, pos = 28687118, rsid = "rs27210669", ref_allele = "A", alt_allele = "G")


#2:28686868-28686868 rs27210671 C/A
p3 <- make_frq_pie_chart(chr = 2, pos = 28686868, rsid = "rs27210671", ref_allele = "C", alt_allele = "A")


png(here("00_dashboard/figures","Tsc1_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()
```

```{r import_gviz_around_Tsc1, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Tsc1_Gviz_macro.png"), 
    here("00_dashboard/figures","Tsc1_Gviz_micro.png"),
    here("00_dashboard/figures","Tsc1_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Tsc1}

read.table(here("00_dashboard/data/VEP/Tsc1.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
```

### DuhLB(Aldh3a2)  
```{r gviz_around_Aldh3a2, eval = FALSE} 
#11:61248025-61248025
#11:61224649-61224649
#11:61224676-61224676

png(here("00_dashboard/figures","Aldh3a2_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000010025", pos_snps = c(61248025, 61224649, 61224676), bp_expand=1000000)

dev.off()
``` 

```{r visualize_per_snp_Fst_Aldh3a2_and_export_micro, eval = FALSE} 

png(here("00_dashboard/figures","Aldh3a2_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000010025", pos_snps = c(61248025, 61224649, 61224676)) 

dev.off()
```

```{r make_frq_pie_chart_Aldh3a2, eval = FALSE}  

#11:61248025-61248025 11_61248025_G/T
p1 <- make_frq_pie_chart(chr = 11, pos = 61248025, rsid = "rs1133760426", ref_allele = "G", alt_allele = "T")

#11:61224649-61224649 11_61224649_G/A 
p2 <- make_frq_pie_chart(chr = 11, pos = 61224649, rsid = "rs28203629", ref_allele = "G", alt_allele = "A")

#11:61224676-61224676 11_61224676_C/T
p3 <- make_frq_pie_chart(chr = 11, pos = 61224676, rsid = "rs28203628", ref_allele = "C", alt_allele = "T")


png(here("00_dashboard/figures","Aldh3a2_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()
```

```{r import_gviz_around_Aldh3a2, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Aldh3a2_Gviz_macro.png"), 
    here("00_dashboard/figures","Aldh3a2_Gviz_micro.png"),
    here("00_dashboard/figures","Aldh3a2_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Aldh3a2}

read.table(here("00_dashboard/data/VEP/Aldh3a2.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
```



### DU6P(Afap1l2)
```{r gviz_around_Afap1l2, eval = FALSE} 

#19:56944714-56944714	G	A
#ENSMUSG00000025083

png(here("00_dashboard/figures","Afap1l2_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000025083", pos_snps = 56944714, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Afap1l2_and_export_micro, eval = FALSE} 
png(here("00_dashboard/figures","Afap1l2_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000025083", pos_snps = 56944714) 

dev.off()
```

```{r make_frq_pie_chart_Afap1l2, eval = FALSE}  
png(here("00_dashboard/figures","Afap1l2_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 19, pos = 56944714, rsid = "rs215944784", ref_allele = "G", alt_allele = "A")

dev.off()
```

```{r import_gviz_around_Afap1l2, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Afap1l2_Gviz_macro.png"), 
    here("00_dashboard/figures","Afap1l2_Gviz_micro.png"),
    here("00_dashboard/figures","Afap1l2_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Afap1l2} 

read.table(here("00_dashboard/data/VEP/Afap1l2.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 

### DU6(Atp11b)
```{r gviz_around_Atp11b, eval = FALSE} 

#3:35809538-35809538	A	C
#ENSMUSG00000037400

png(here("00_dashboard/figures","Atp11b_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000037400", pos_snps = 35809538, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Atp11b_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Atp11b_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000037400", pos_snps = 35809538) 

dev.off()
```

```{r make_frq_pie_chart_Atp11b, eval = FALSE}  
png(here("00_dashboard/figures","Atp11b_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 3, pos = 35809538, rsid = "rs30557001", ref_allele = "A", alt_allele = "C")

dev.off()
```

```{r import_gviz_around_Atp11b, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Atp11b_Gviz_macro.png"), 
    here("00_dashboard/figures","Atp11b_Gviz_micro.png"),
    here("00_dashboard/figures","Atp11b_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Atp11b} 

read.table(here("00_dashboard/data/VEP/Atp11b.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 

### FERT(Eif2ak3) 
```{r gviz_around_Eif2ak3, eval = FALSE} 

#6:70844760-70844760	T	C
#6:70844967-70844967	C	A
#6:70893040-70893040	T	A
#ENSMUSG00000031668

png(here("00_dashboard/figures","Eif2ak3_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000031668", pos_snps = c(70844760, 70844967, 70893040), bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Eif2ak3_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Eif2ak3_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000031668", pos_snps = c(70844760, 70844967, 70893040)) 

dev.off()
```

```{r make_frq_pie_chart_Eif2ak3, eval = FALSE}  
p1 <- make_frq_pie_chart(chr = 6, pos = 70844760, rsid = "rs262061785", ref_allele = "T", alt_allele = "C")
p2 <- make_frq_pie_chart(chr = 6, pos = 70844967, rsid = "rs248511738", ref_allele = "C", alt_allele = "A")
p3 <- make_frq_pie_chart(chr = 6, pos = 70893040, rsid = "rs37318214", ref_allele = "T", alt_allele = "A")

png(here("00_dashboard/figures","Eif2ak3_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()

dev.off()
```

```{r import_gviz_around_Eif2ak3, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Eif2ak3_Gviz_macro.png"), 
    here("00_dashboard/figures","Eif2ak3_Gviz_micro.png"),
    here("00_dashboard/figures","Eif2ak3_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Eif2ak3} 

read.table(here("00_dashboard/data/VEP/Eif2ak3.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 


INDELs-Part1 {data-navmenu="INDELs"}
=============================

Column {.tabset}
----------------

### Summaries 
```{r load_data_sets, include=FALSE}

cts_raw_vcf <- here("batches123_02_GenotypeGVCFs_by_chr/output/cohort_biallelicINDELs.indel.hist") %>% 
  vroom() %>% 
  mutate(indel_class = ifelse(LENGTH < 0, "deletion", "insertion"))

cts_final_vcf <- here("batches123_04_final_VCF/output/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.indel.hist") %>% 
  vroom() %>% 
  mutate(indel_class = ifelse(LENGTH < 0, "deletion", "insertion"))

info_anno_raw_vcf <- here("batches123_02_GenotypeGVCFs_by_chr/output/cohort_biallelicINDELs.table") %>% 
  vroom() %>% 
  dplyr::select(-"...9") # not sure what this col is about...

mgp_indels <- here("resource_mgp_sanger_REL-1505-SNPs_Indels/mgp.v5.merged.indels.dbSNP142.normed.PASS.locus") %>% 
  vroom(col_names = FALSE) %>% 
  dplyr::rename(CHROM = X1, POS = X2)

dbsnp_indels <- here("reference_genome_ensembl/mus_musculus_only_indels.recode.loci") %>% 
  vroom(col_names = FALSE) %>% 
  dplyr::rename(CHROM = X1, POS = X2)

final_indel_vcf_metrics_detail <- read.delim(here("batches123_04_final_VCF/output/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)

fls <- list.files(
  here("batches123_05_annotation/output"),
  pattern = ".csv"
  ) %>% 
  .[grep("biallelicINDELs_VQSR99",.)] %>% 
  .[grep("cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.summary.csv",.,invert = TRUE)]
snpeff_data_indel <- fls %>% lapply(function(fl) read.csv(here("batches123_05_annotation/output",fl)))
names(snpeff_data_indel) <- fls %>% 
  str_remove("cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.summary_") %>% 
  str_remove(".csv")
rm(fls)

```

```{r table_nr_indels_binned_by_length}
cts_raw_vcf %>% 
  mutate(bin_len = ifelse(abs(LENGTH) >= 300, ">=300", NA)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 300 & abs(LENGTH) >= 200, "<300 & >=200", bin_len)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 200 & abs(LENGTH) >= 100, "<200 & >=100", bin_len)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 100 & abs(LENGTH) >= 50, "<100 & >=50", bin_len)) %>%
  mutate(bin_len = ifelse(abs(LENGTH) < 50, "<50", bin_len)) %>%
  mutate(bin_len = factor(bin_len, levels = c(">=300", "<300 & >=200", "<200 & >=100", "<100 & >=50", "<50"))) %>% 
  group_by(indel_class, bin_len) %>% 
  summarise(n = sum(COUNT)) %>% 
  reshape2::dcast(bin_len ~ indel_class, value.var = "n") %>% 
  janitor::adorn_totals(where = "col") %>% 
  janitor::adorn_totals(where = "row") %>% 
  kable(format.args = list(big.mark = ","), digits = 2, caption = "Number of INDELs by length (raw_vcf)") %>% 
  kable_styling(full_width = F)

cts_final_vcf %>% 
  mutate(bin_len = ifelse(abs(LENGTH) >= 300, ">=300", NA)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 300 & abs(LENGTH) >= 200, "<300 & >=200", bin_len)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 200 & abs(LENGTH) >= 100, "<200 & >=100", bin_len)) %>% 
  mutate(bin_len = ifelse(abs(LENGTH) < 100 & abs(LENGTH) >= 50, "<100 & >=50", bin_len)) %>%
  mutate(bin_len = ifelse(abs(LENGTH) < 50, "<50", bin_len)) %>%
  mutate(bin_len = factor(bin_len, levels = c(">=300", "<300 & >=200", "<200 & >=100", "<100 & >=50", "<50"))) %>% 
  group_by(indel_class, bin_len) %>% 
  summarise(n = sum(COUNT)) %>% 
  reshape2::dcast(bin_len ~ indel_class, value.var = "n") %>% 
  janitor::adorn_totals(where = "col") %>% 
  janitor::adorn_totals(where = "row") %>% 
  kable(format.args = list(big.mark = ","), digits = 2, caption = "Number of INDELs by length (final_vcf)") %>% 
  kable_styling(full_width = F)

```

### raw_cts

```{r make_plots_cts_raw_vcf, fig.dim=c(20,10)} 

p1 <- ggplot(cts_raw_vcf, aes(x = LENGTH, y = COUNT)) +
  geom_bar(stat = "identity") +
  theme_bw(base_size = 15) +
  ggtitle("INDEL counts by length")


# make_histogram_counts_by_indel_length_zoom

limits <- c(-15,15)
intervals <- 1

p2 <- cts_raw_vcf %>% 
  filter(LENGTH > limits[1] & LENGTH < limits[2]) %>% 
  ggplot(aes(x = LENGTH, y = COUNT)) +
    geom_bar(stat = "identity") +
    scale_x_continuous(breaks = seq(limits[1],limits[2],intervals)) +
    theme_bw(base_size = 15) +
    ggtitle(paste0("INDEL counts by length (", limits[1], " to ", limits[2], ")"))


# make_barplot_n_indels

p3 <- cts_raw_vcf %>% 
  group_by(indel_class) %>% 
  summarise(n_indels = sum(COUNT)) %>% 
  ggplot(aes(x = indel_class, y = n_indels)) +
    geom_bar(stat = "identity") +
    theme_bw(base_size = 15) +
    #theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    ggtitle("INDEL counts by INDEL-class")

# put plots together

grid.arrange(p1,p2,p3, ncol = 3)

```


### vis_raw_annot
```{r check_NAs_per_annotation,eval=FALSE}
info_anno_raw_vcf %>% 
  dplyr::select(-CHROM, -POS) %>% 
  reshape2::melt() %>% 
  group_by(variable) %>% 
  summarise(n = n(), sum_na = sum(is.na(value)))
# MQRankSum & ReadPosRankSum --> both habe comparable NA number. Because they are calculated on het sites and these genomes are highly homozygous.
```


```{r density_plots_raw_annot, fig.dim=c(20,10)}

info_anno_raw_vcf %>% 
  dplyr::select(-CHROM, -POS) %>% 
  mutate(log10FS = log10(FS)) %>% 
  reshape2::melt() %>% 
  ggplot(aes(x = value)) +
    geom_density() +
    facet_wrap(~variable, ncol = 2, scales = "free") +
    theme_bw(base_size = 12) +
    xlab(NULL)

```

### vis_after_hard_filters

```{r define_very_good_indels_by_hard_filtering, include=FALSE}
# https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants

# https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering#2
#"For filtering indels, most annotations related to mapping quality have been removed since there is a conflation with the length of an indel in a read and the degradation in mapping quality that is assigned to the read by the aligner. This covariation is not necessarily indicative of being an error in the same way that it is for SNPs."

truth_set <- info_anno_raw_vcf %>% 
  filter(QD > 5 & FS < 1 & SOR < 3) %>%
  filter( (ReadPosRankSum < 2 & ReadPosRankSum > -2) | is.na(ReadPosRankSum) )

```

```{r export_intervals_for_vqsr, eval = FALSE}
truth_set %>% 
  mutate(tst = paste0(CHROM,":",POS,"-",POS)) %>% 
  dplyr::select(tst) %>% 
  write.table(., here("batches123_03_VariantQualityScoreRecalibration/data", "truth_training_INDELs.intervals"),
              quote = FALSE, row.names = FALSE, col.names = FALSE)
```

```{r vis_after_applying_hard_filters, fig.dim=c(20,10)}

truth_set %>% 
  dplyr::select(-CHROM, -POS) %>% 
  reshape2::melt() %>% 
  ggplot(aes(x = value)) +
    geom_density() +
    facet_wrap(~variable, ncol = 2, scales = "free") +
    theme_bw(base_size = 12) +
    xlab(NULL)

```

### VQSR

#### **Tranche selection**
* Variant quality score recalibration was conducted on the raw indel vcf.

* The resources were the hard-filtered-indel-set (truth/taining set) and the external MGP indel set (training set). For the purpose of metrics indels in dbSNP were used.

* The number of INDELs in the hard-filtered-indel-set was `r nrow(truth_set)`.

* The number of INDELs in the MGP indel set was `r nrow(mgp_indels)`.

* The number of INDELs in dbSNP was `r nrow(dbsnp_indels)`.

* These are the number of SNPs in each tranche:

```{r}
here("batches123_03_VariantQualityScoreRecalibration/output/cohort_biallelicINDELs.tranches") %>% 
  read.csv(skip = 2, header = TRUE) %>% 
  mutate(total_sites = numKnown + numNovel,
         pct_in_dbSNP = (numKnown/total_sites)*100) %>% 
  dplyr::select(targetTruthSensitivity, numKnown, numNovel, total_sites, pct_in_dbSNP, everything()) %>% 
  kable(format.args = list(big.mark = ","), digits = 2) %>% 
  kable_styling(full_width = F)
```

* For indels TiTv ratio is not applicable and therefore it is not possible to produce the tranches plot.

* Therefore, choosing a good balance between sensitivity and especificity is not as straight forward as it was for SNPs.

* By looking at the the VQSLod of lowest scored variant in a tranche (minVQSLod) it is evident that "tranche99.9" is way to low (negative means that odds are in favor of the negative model --> variant is more similar to variants not in the truth/training set).

* The "tranche99" looks sensible, the lowest score variant has a VQSLOD of 3.65 (odds of being in the positive model are ~40 to 1).

* Also note the dramatic drop in minVQSLod going from "tranche99" to "tranche99.9".

* Therefore, this tranche was chosen for site-filtering, corresponding to 2,817,271 INDELs.

* This is a study that also used "tranche99" to site-filter variants: https://academic.oup.com/gbe/article/11/6/1514/5423187#136403651


#### **On the low overlap with dbSNP**

* Only ~46 to ~56% of INDELs are observed in dbSNPs. 

* Compare this figure with SNPs (>90% overlap).

* But this is an expected outcome, because INDELs are much less characterized than SNPs.

* Some literature, concerning this fact:

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2953750/

https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-2-r9

https://genome.cshlp.org/content/21/6/830.full.html






### filtered_cts
```{r make_plots_cts_final_vcf, fig.dim=c(20,10)} 

p1 <- ggplot(cts_final_vcf, aes(x = LENGTH, y = COUNT)) +
  geom_bar(stat = "identity") +
  theme_bw(base_size = 15) +
  ggtitle("INDEL counts by length")


# make_histogram_counts_by_indel_length_zoom

limits <- c(-15,15)
intervals <- 1

p2 <- cts_final_vcf %>% 
  filter(LENGTH > limits[1] & LENGTH < limits[2]) %>% 
  ggplot(aes(x = LENGTH, y = COUNT)) +
    geom_bar(stat = "identity") +
    scale_x_continuous(breaks = seq(limits[1],limits[2],intervals)) +
    theme_bw(base_size = 15) +
    ggtitle(paste0("INDEL counts by length (", limits[1], " to ", limits[2], ")"))


# make_barplot_n_indels

p3 <- cts_final_vcf %>% 
  group_by(indel_class) %>% 
  summarise(n_indels = sum(COUNT)) %>% 
  ggplot(aes(x = indel_class, y = n_indels)) +
    geom_bar(stat = "identity") +
    theme_bw(base_size = 15) +
    #theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    ggtitle("INDEL counts by INDEL-class")

# put plots together

grid.arrange(p1,p2,p3, ncol = 3)

```

### indels_per_sample
```{r plot_indels_per_sample}

p1 <- final_indel_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_INDELS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_INDELS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic INDELs per sample (GT filtered VCF)")

#p1

ggplotly(p1) 

```

### indels_per_pop_tab
```{r indels_per_pop}
indels_per_pop <- here("batches123_04_final_VCF/output") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  .[grep("biallelicINDELs_VQSR99",.)] %>% 
  lapply(function(x){ 
    d <- read.table(here("batches123_04_final_VCF/output",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() %>% 
  mutate(line = factor(line, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(line)

indels_per_pop %>% 
  kable(caption = "Per-Line INDEL summary", digits = 2) %>% 
  kable_styling(full_width = F) %>% 
  footnote(general = "Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT==RR) and all-missing sites (GT==./.).")
```

### indels_per_pop_vis
```{r plot_indels_per_pop, fig.dim=c(7,10)}
indels_per_pop %>% 
  ggplot(aes(x = line, y = TOTAL_INDELS)) +
    geom_bar(stat = "identity") +
      theme_bw(base_size = 12) +
      xlab(NULL) +
      ggtitle("Number of INDELs per population")
```

### Intersections
```{r indel_prep_sets, eval = FALSE}

fls <- here("batches123_04_final_VCF/output/") %>% 
  list.files(pattern = "*.INDELids") %>% 
  .[-grep("cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.INDELids",.)]

indel_ids <- lapply(fls, function(fl){ 
  ln <- str_remove(fl, "cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.") %>% 
    str_remove(".INDELids")
  d <- vroom(here("batches123_04_final_VCF/output",fl), col_names = F) %>% mutate(line = ln, indel_id = paste0(X1,"_",X3))
  return(d)
})

names(indel_ids) <- str_remove(fls, "cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.") %>% 
    str_remove(".INDELids")

all_ids <- indel_ids %>% bind_rows() %>% select(indel_id) %>% unique() %>% unlist()

set_dat <- data.frame(ids = all_ids,
		   duk = (all_ids %chin% indel_ids$DUK$indel_id),
		   duc = (all_ids %chin% indel_ids$DUC$indel_id),
		   du6 = (all_ids %chin% indel_ids$DU6$indel_id),
		   du6p = (all_ids %chin% indel_ids$DU6P$indel_id),
		   duhlb = (all_ids %chin% indel_ids$DUhLB$indel_id),
		   fztdu = (all_ids %chin% indel_ids$FZTDU$indel_id))

set_dat_bool <- data.frame(
  duk = as.integer(set_dat$duk),
	duc = as.integer(set_dat$duc),
	du6 = as.integer(set_dat$du6),
	du6p = as.integer(set_dat$du6p),
	duhlb = as.integer(set_dat$duhlb),
	fztdu = as.integer(set_dat$fztdu)
	)

rownames(set_dat_bool) <- set_dat$ids

png(here("00_dashboard/figures/intersection_indels1.png"), res = 300, height = 3000, width = 3800, units = "px")

upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  text.scale = c(
    2, #intersection size title
    1.5, #intersection size tick labels
    1.5, #set size title
    1, #set size tick labels
    2, #set names
    1.5 #numbers above bars
    ),
  number.angles = 45,
  intersections =  list(
    # full intersection
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    # selected line intersctions
    list("duhlb","du6p","du6","duc","duk"),
    # fertility lines vs fztdu
    list("duk","duc","fztdu"),
    # body size lines vs fztdu
    list("du6","du6p","fztdu"),
    # lines vs fztdu
    list("duk","fztdu"),
    list("duc","fztdu"),
    list("du6","fztdu"),
    list("du6p","fztdu"),
    list("duhlb","fztdu"),
    # phenotype related intersection
    list("duk","duc"),
    list("du6","du6p"),
    #list("duhlb","fztdu"), #done already
    # private SNPs
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    ),
  queries = list(
    # Private SNPs per line (red)
    list(query = intersects, params = list("duk"),color = "red", active = T),
    list(query = intersects, params = list("duc"),color = "red", active = T),
    list(query = intersects, params = list("du6"),color = "red", active = T),
    list(query = intersects, params = list("du6p"),color = "red", active = T),
    list(query = intersects, params = list("duhlb"),color = "red", active = T),
    list(query = intersects, params = list("fztdu"),color = "red", active = T),
    
    # private SNPs of selected lines in FZTDU (blue)
    list(query = intersects, params = list("duk","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duc","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("du6p","fztdu"),color = "blue", active = T),
    list(query = intersects, params = list("duhlb","fztdu"),color = "blue", active = T),
    
    # SNPs found only in related lines (yellow)
    list(query = intersects, params = list("duk","duc"),color = "orange", active = T),
    list(query = intersects, params = list("du6","du6p"),color = "orange", active = T),
    
    # SNPs found only in related lines and FZTDU (green)
    list(query = intersects, params = list("duk","duc","fztdu"),color = "green", active = T),
    list(query = intersects, params = list("du6","du6p","fztdu"),color = "green", active = T),
    
    # SNPs found between 5 lines (purple) 
     list(query = intersects, params = list("duhlb","du6p","du6","duc","duk"),color = "purple", active = T),
    
    # common SNPs ()
    list(query = intersects, params = list("fztdu","duhlb","du6p","du6","duc","duk"),color = "black", active = T)
    )

)

grid.text("Intersections INDEL sets",x = 0.65, y=0.95, gp=gpar(fontsize=15))

dev.off()

```

```{r indel_upsetr_plot_minimal_import, fig.width=10} 

include_graphics(here("00_dashboard/figures/intersection_indels1.png"))

```

### snpeff
```{r indels_import_snpeff_summary_tables}

snpeff_data_indel$summary_table %>% kable(caption = "summary_table") %>% kable_styling(full_width = F)

snpeff_data_indel$cts_by_eff %>% 
  mutate(Type = str_trim(Type)) %>% 
  left_join(impact_snpeff, by = c("Type" = "Effect")) %>% 
  unique() %>% 
  mutate(
    Impact = ifelse(Type == "5_prime_UTR_premature_start_codon_gain_variant",
                  "LOW",
                  Impact)
  ) %>% 
  arrange(desc(Count)) %>% 
  kable(caption = "cts_by_eff") %>% kable_styling(full_width = F)

snpeff_data_indel$change_rate_by_chr %>% kable(caption = "change_rate_by_chr") %>% kable_styling(full_width = F)

snpeff_data_indel$cts_by_geno_region %>% kable(caption = "cts_by_geno_region") %>% kable_styling(full_width = F) 

snpeff_data_indel$vars_by_type %>% kable(caption = "vars_by_type") %>% kable_styling(full_width = F) 

snpeff_data_indel$eff_by_impact %>% kable(caption = "eff_by_impact") %>% kable_styling(full_width = F) 
```

### PCA
```{r indel_pc1_pc2_plot, fig.width = 5, fig.height=7}
pc_pct_indels <- read.csv(here("batches123_06_genetic_structure/data/pc_pct_INDELs.csv")) %>% dplyr::select(-X)

pc_eigenscores_indels <- read.csv(here("batches123_06_genetic_structure/data/pc_eigenscores_INDELs.csv")) %>% 
  dplyr::select(-X) %>% 
  mutate(sample.id = str_remove(sample.id, "-L1")) %>% 
  left_join(sample_info, by = c("sample.id"="sample_id")) %>% 
  dplyr::select(Linie, sample.id, target_cvg, everything())

p1 <- pc_eigenscores_indels %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=Linie)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())


p2 <- pc_eigenscores_indels %>% 
        mutate(gp = paste(Linie, target_cvg, sep = "_")) %>% 
        ggplot(data = ., aes(x=PC1,y=PC2,color=gp)) +
          geom_point() +
          theme_bw() +
          theme(legend.title = element_blank())
  
gridExtra::grid.arrange(p1,p2, ncol = 1) 

rm(p1,p2)
```

### HC
```{r indel_hier_clust, fig.width = 10, fig.height=5} 
dend <- readRDS(here("batches123_06_genetic_structure/data/dendrogram_INDELs.rds")) %>% as.hclust()

# change labels to group to avoid legend
#identical(sample_info[,"sample_id"], str_remove(dend[["labels"]], "-L1")) #check order of labels is the same! yes!
dend[["labels"]] <- sample_info$Linie


colors <- c("red", "blue", "green", "black", "purple","orange")
clus6 <- cutree(dend, 6)
plot(as.phylo(dend), 
     type = "fan", 
     cex = 0.6,
     no.margin = TRUE,
     tip.color = colors[clus6])

# good reference: http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning
```

### ADMIXTURE
```{r indel_barplot_k5, fig.width=20, fig.height=7}
# About sample ordering: https://www.biostars.org/p/221817/

fam_fl <- read.table(here("batches123_06_genetic_structure/data/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.ldpruned.fam"))

make_ancestry_barplot(
  k=5, 
  q_fl=here("batches123_06_genetic_structure/scripts/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.ldpruned.5.Q")
) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) 
```

INDELs-Part2_RDD-genes {data-navmenu="INDELs"}
=======================================

Column {.tabset}
----------------

### INDELS_in_RDD_genes
```{r indel_load_fst_data}
# per indel fst
per_indel_fst <- here("batches123_07_selection_analysis/03_fst_indels/output") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(fl){
    
    vroom(fl, col_types = c("CHROM"="c")) %>% 
      mutate(contrast = basename(fl) %>% str_remove(".weir.fst"))
    
  }) %>% 
  bind_rows() %>% 
  reshape2::dcast(CHROM+POS~contrast, value.var = "WEIR_AND_COCKERHAM_FST") %>% 
  mutate(start = POS) %>% 
  dplyr::rename(seqnames = CHROM, end = POS) %>% 
  inner_join(
    vroom(
      here("batches123_04_final_VCF/output/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.INDELids2"),
      col_types = c("X1"="c"), col_names = FALSE
      ),
    by = c("seqnames"="X1", "start"="X2")
    ) %>% 
  dplyr::rename(REF=X3, ALT=X4) %>% 
  dplyr::select(seqnames, start, end, everything()) %>% 
  mutate(indel_class = ifelse( nchar(REF) < nchar(ALT), "insertion", "deletion" )) %>% 
  mutate(indel_length = nchar(ALT) - nchar(REF)) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

```{r find_overlaps_with_RDD_genes}
# import_RDD_genes (still not sure if I should add indels to main dashboard!)
# genes exported earlier (q95 vs q10)
# ... and transform to granges
rdd_gns_tmp <- c("DUK","DUC","DU6","DU6P","DUhLB","FERT") %>% 
    lapply(function(pop){
      here("00_dashboard/data/RDD_genes", paste0("RDD_genes_", pop, ".csv")) %>% 
        read.csv(stringsAsFactors = FALSE) %>% 
        dplyr::select(-X)
      }) %>% 
    bind_rows() %>% 
    dplyr::select(seqnames, start, end, everything()) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = per_indel_fst, subject = rdd_gns_tmp, minoverlap = 1)


indels_in_rdd_genes <- bind_cols(

  rdd_gns_tmp[subjectHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(distinct_gene_locus = paste0(seqnames,":",start,"-",end)) %>% 
    dplyr::select(contrast, distinct_gene_locus, strand, width,  ensembl_gene_id, mgi_symbol, gene_biotype) %>% 
    dplyr::rename(distinct_gene_for=contrast),
    
  per_indel_fst[queryHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(indel_locus = paste0(seqnames,":",start)) %>% 
    dplyr::select(indel_locus,indel_class, indel_length, contains("_vs_FZTDU"), REF, ALT )
)

```

```{r indels_summarise_rdd_genes}
indels_in_rdd_genes %>% 
  mutate(distinct_gene_for = factor(distinct_gene_for, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB", "FERT"), "_vs_FZTDU"))) %>% 
  group_by(distinct_gene_for) %>% 
  mutate(symbol_geneid = paste0("(", mgi_symbol, "_", ensembl_gene_id, ")")) %>% 
  summarise(n_genes=n_distinct(distinct_gene_locus), 
            gene = paste(unique(symbol_geneid), collapse = ", ")) %>% 
  datatable(options=list(paging = FALSE))
```

### RDD_genes_DUK
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "DUK_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### RDD_genes_DUC
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "DUC_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### RDD_genes_DU6
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "DU6_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### RDD_genes_DU6P
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "DU6P_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### RDD_genes_DUhLB
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "DUhLB_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### RDD_genes_FERT
```{r}
indels_in_rdd_genes %>% filter(distinct_gene_for == "FERT_vs_FZTDU") %>% datatable(filter = 'top', options=list(paging = FALSE))

```

### SNPeff
```{r display_impact_table}

here("batches123_05_annotation/output/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.ann.tab") %>% 
  vroom(col_types = c("CHROM"="c")) %>% 
  dplyr::rename(gene_id = "ANN[*].GENEID", symbol = "ANN[*].GENE", effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT") %>%
  filter(gene_id  %in% unique(indels_in_rdd_genes$ensembl_gene_id)) %>% 
  mutate(indel_locus = paste0(CHROM,":",POS)) %>% 
  dplyr::select(-CHROM,-POS) %>%
  dplyr::select(indel_locus, effect, impact, symbol, gene_id, indel_locus, REF, ALT) %>%
  datatable(options = list(pageLength = 100))

```

### Vis_Fst_Indels
```{r heatmap_fst_high_moderte_indels, fig.height=10, fig.width=10}

high_moderate_rdd_genes_indels <- here(
  "batches123_05_annotation/output/cohort_biallelicINDELs_VQSR99_PASS_withmissingness.filtered.ann.tab"
  ) %>% 
  vroom(col_types = c("CHROM"="c")) %>% 
  dplyr::rename(gene_id = "ANN[*].GENEID", symbol = "ANN[*].GENE", effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT") %>%
  filter(gene_id  %in% unique(indels_in_rdd_genes$ensembl_gene_id)) %>% 
  mutate(indel_locus = paste0(CHROM,":",POS)) %>% 
  dplyr::select(-CHROM,-POS) %>%
  dplyr::select(indel_locus, effect, impact, symbol, gene_id, indel_locus, REF, ALT) %>% 
  filter(impact %in% c("HIGH","MODERATE")) %>% 
  inner_join(
    indels_in_rdd_genes,
    by = "indel_locus"
  ) %>% 
  mutate(tmp = indel_locus) %>% 
  tidyr::separate(col = tmp, sep = ":", into = c("chr","pos")) %>% 
  #arrange(distinct_gene_for, impact, mgi_symbol, chr, pos) %>%
  arrange(as.numeric(chr), as.numeric(pos), distinct_gene_for, mgi_symbol, impact) %>% 
  dplyr::select(distinct_gene_for, mgi_symbol, impact, effect, indel_class, indel_locus, 
                paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU")) %>% 
  mutate(impact_effect = paste0(impact,"_",effect)) %>% 
  mutate(indel_locus_gene = paste0(indel_locus, "(",mgi_symbol, ")")) %>% 
  unique()
  

na_idx <- high_moderate_rdd_genes_indels[,paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU")] %>% 
  apply(1, function(r) all(is.na(r)))

ma <- high_moderate_rdd_genes_indels[!na_idx,] %>% 
  dplyr::select(indel_locus_gene, contains("_vs_")) %>% 
  tibble::column_to_rownames("indel_locus_gene") %>% 
  as.matrix()

rowdata <- high_moderate_rdd_genes_indels[!na_idx,] %>% 
  dplyr::select(distinct_gene_for, impact_effect, indel_locus_gene) %>% 
  tibble::column_to_rownames("indel_locus_gene")

pheatmap::pheatmap(ma, na_col = "#DDDDDD", cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE, number_format = "%.1f",
                   annotation_row = rowdata, cellwidth = 30, cellheight=30, 
                   main = "Fst InDels in RDD-genes")
```










