---
title: "Analysis of WGS DU-mice data"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(here)
library(plotly)
library(DT)
library(venn)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
library(knitr)
```

```{r, datasets, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

final_vcf_metrics_detail <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.table.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)


raw_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "raw")

mid_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "site_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)

final_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(data_set = "gt_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```


Read Alignment
===============


Variant Calling 
================

Column {data-width=300}
------------------------

### **Overview** 
* GATK4 best practices
* Site level filtering: 
  * bi-allelic SNPs
  * 95% truth sensitivity (VQSR)
* Genotype level filtering: 
  * DP: 4 - 77 (mean DP + 4*SD)
  * GQ >= 20
  * Else missing (./.)
* If 1/6 groups with min 20/25 non-miss samples: **7064770 SNPs**
* If 6/6 groups with min 20/25 non-miss samples: **34033 SNPs**


### **Number of Variants**
```{r}
vcfs_main_summary <- bind_rows(
  raw_vcf_metrics_summary,
  mid_vcf_metrics_summary,
  final_vcf_metrics_summary
) %>% 
  dplyr::select(data_set, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV, TOTAL_INDELS, TOTAL_MULTIALLELIC_SNPS) %>% 
  melt(id.vars = "data_set") %>% 
  dcast(variable ~ data_set, fill = "value") %>%
  mutate(raw = as.numeric(raw), 
         site_filtered = as.numeric(site_filtered), 
         gt_filtered = as.numeric(gt_filtered)) %>% 
  dplyr::select(variable, raw, site_filtered, gt_filtered) %>% 
  knitr::kable(format.args = list(big.mark = ","), digits = 2)

vcfs_main_summary
```


Column {data-width=700, .tabset}
------------------------

### **SNPs per sample**
```{r}
p1 <- final_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_SNPS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

#p1

#ggplotly(p1) %>% layout(legend = list(orientation = "h", x = -0.05, y = -0.02))
ggplotly(p1) 
```

### **Sample metrics**
```{r}
final_vcf_metrics_detail %>% 
  dplyr::select(Linie, SAMPLE_ALIAS, target_cvg, everything(),
                -PCT_GQ0_VARIANTS, -TOTAL_GQ0_VARIANTS) %>% 
  datatable(rownames = F, options = list(pageLength = 25))
```

### **Population metrics**
```{r}
here("batches123_04_FinalVCF/metrics") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  lapply(function(x){
    d <- read.table(here("batches123_04_FinalVCF/metrics",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() %>% 
  datatable(rownames = F)
```

> Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT=="RR") and all-missing sites (GT=="./.").

### **Intersections (upsetr)**
```{r prepare data for upsetr, eval = F}
fls <- here("batches123_04_FinalVCF/output/") %>% list.files(pattern = "*.snpIDs")

snp_ids <- lapply(fls, function(fl){
  ln <- str_remove(fl, "cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
    str_remove(".snpIDs")
  d <- vroom(here("batches123_04_FinalVCF/output",fl), col_names = F) %>% mutate(line = ln, snp_id = paste0(X1,"_",X2))
  return(d)
})

names(snp_ids) <- str_remove(fls, "cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
    str_remove(".snpIDs")


all_ids <- snp_ids %>% bind_rows() %>% select(-line, -X1, -X2) %>% unique() %>% unlist()

sets <- data.frame(ids = all_ids,
		   duk = (all_ids %chin% snp_ids$DUK$snp_id),
		   duc = (all_ids %chin% snp_ids$DUC$snp_id),
		   du6 = (all_ids %chin% snp_ids$DU6$snp_id),
		   du6p = (all_ids %chin% snp_ids$DU6P$snp_id),
		   duhlb = (all_ids %chin% snp_ids$DUhLB$snp_id),
		   fztdu = (all_ids %chin% snp_ids$FZTDU$snp_id))

write.csv(sets, here("00_dashboard/data","snp_id_sets.csv"))
```

```{r make upsetr plot, fig.width=10}
set_dat <- vroom(here("00_dashboard/data","snp_id_sets.csv")) %>% as.data.frame()

set_dat_bool <- data.frame(
  duk = as.integer(set_dat$duk),
	duc = as.integer(set_dat$duc),
	du6 = as.integer(set_dat$du6),
	du6p = as.integer(set_dat$du6p),
	duhlb = as.integer(set_dat$duhlb),
	fztdu = as.integer(set_dat$fztdu)
	)

rownames(set_dat_bool) <- set_dat$ids

upset(
  set_dat_bool,
  nsets = 6, 
  sets = c("fztdu","duhlb","du6p","du6","duc","duk"), 
  keep.order = T,
  intersections =  list(
    list("fztdu","duhlb","du6p","du6","duc","duk"),
    list("fztdu","duhlb","du6p","du6","duc"),
    list("fztdu","duhlb","du6p","du6","duk"),
    list("fztdu","duhlb","du6p","duc","duk"),
    list("fztdu","duhlb","du6","duc","duk"),
    list("fztdu","du6p","du6","duc","duk"),
    list("duhlb","du6p","du6","duc","duk"),
    list("duk"),
    list("duc"),
    list("du6"),
    list("du6p"),
    list("duhlb"),
    list("fztdu")
    )
)

```

### **Missingness**

```{r}
include_graphics(here("batches123_04_FinalVCF/figures_tables","miss_cts.png"))

include_graphics(here("batches123_04_FinalVCF/figures_tables","miss_cts_facet.png"))
```

```{r}

n_snps <- mid_vcf_metrics_summary$TOTAL_SNPS

tmp <- read.csv(here("batches123_04_FinalVCF/figures_tables","n_snps_by_min_n_called_per_line.csv")) 

fraction_snps_by_min_n_called_per_line <- lapply(tmp[3:ncol(tmp)], function(x) round(x/n_snps, 2)) %>%
  bind_cols() %>%
  mutate(min_called_samples = tmp$min_called_samples) %>%
  dplyr::select(min_called_samples, everything())

fraction_snps_by_min_n_called_per_line[2:26,] %>% kable(align = "l")

rm(tmp)

```

> Figures produced externally from site-filtered biSNPs with script "./batches123_04_FinalVCF/scripts/15_visualize_missingness.R".

SNP Annotations
===============
```{r}

```


LDD & SFS 
========================

Column
-------------------

### Allele frequency state per population
```{r}

```

### LD Decay

Column
-------------------

### Alternative Allele Frequency Distribution
```{r}

```

### Minor Allele Frequency Distribution
```{r}

```

Diversity
===========================

Population Structure
==============================

Column {.tabset}
----------------

### PC1 - PC2

### PC2 - PC3

### PC3 - PC4

### PC4 - PC5

### PC5 - PC6


Column
---------------

### Hieratchical Clustering

### Admixture 

Admixture detailed
==============================

Column
--------

### K2

### K3

### K4

Column
-------

### K5

### K6

Divergent Signatures
==========================


Directional Signatures
===========================


