---
title: "Analysis of WGS DU-mice data"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(here)
library(plotly)
library(DT)
library(venn)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
```

```{r, datasets, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

final_vcf_metrics_detail <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.table.variant_calling_detail_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(SAMPLE_ALIAS = str_remove(SAMPLE_ALIAS, "-L1")) %>% 
  left_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id")) %>% 
  mutate(Linie = factor(Linie, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  arrange(Linie, target_cvg, SAMPLE_ALIAS)


raw_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "raw")

mid_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% mutate(data_set = "site_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)

final_vcf_metrics_summary <- read.delim(here("batches123_04_FinalVCF/metrics/cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.table.variant_calling_summary_metrics"), stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  mutate(data_set = "gt_filtered") %>% 
  mutate(PCT_DBSNP_INDELS = NA)
```


Read Alignment
===============


Variant Calling 
================

Column {data-width=300}
------------------------

### **Overview** 

* GATK4 best practices
* Filtering:
  * Site level: 
    * bi-allelic SNPs
    * 95% truth sensitivity (VQSR)
  * Genotype level (min 1/6 groups, min 20/25 samples): 
    * min DP: 4
    * min GQ: 20
    * max DP: 77 (mean DP + 4*SD)


### **Number of Variants**
```{r}
vcfs_main_summary <- bind_rows(
  raw_vcf_metrics_summary,
  mid_vcf_metrics_summary,
  final_vcf_metrics_summary
) %>% 
  dplyr::select(data_set, TOTAL_SNPS, NUM_IN_DB_SNP, NOVEL_SNPS, PCT_DBSNP, DBSNP_TITV, NOVEL_TITV, TOTAL_INDELS, TOTAL_MULTIALLELIC_SNPS) %>% 
  melt(id.vars = "data_set") %>% 
  dcast(variable ~ data_set, fill = "value") %>%
  mutate(raw = as.numeric(raw), 
         site_filtered = as.numeric(site_filtered), 
         gt_filtered = as.numeric(gt_filtered)) %>% 
  dplyr::select(variable, raw, site_filtered, gt_filtered) %>% 
  knitr::kable(format.args = list(big.mark = ","), digits = 2)

vcfs_main_summary
```



Column {data-width=700, .tabset}
------------------------

### **SNPs per sample**
```{r}
p1 <- final_vcf_metrics_detail %>% 
  #dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, NUM_IN_DB_SNP, NOVEL_SNPS) %>% 
  dplyr::select(SAMPLE_ALIAS, Linie, target_cvg, TOTAL_SNPS) %>% 
  #melt(measure.vars = c("NUM_IN_DB_SNP","NOVEL_SNPS"), variable.name = "SNP_type", value.name = "SNP_number") %>%
  #mutate(SNP_type = factor(SNP_type, levels = c("NOVEL_SNPS", "NUM_IN_DB_SNP"))) %>% 
  #ggplot(data = ., aes(x = SAMPLE_ALIAS, y = SNP_number, fill = SNP_type)) +
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS, fill = target_cvg)) +
    geom_bar(stat = "identity") +
    #facet_grid(~Linie+target_cvg, scales = "free_x")+
    facet_grid(~Linie, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

#p1

#ggplotly(p1) %>% layout(legend = list(orientation = "h", x = -0.05, y = -0.02))
ggplotly(p1) 
```

### **Sample metrics**
```{r}
final_vcf_metrics_detail %>% 
  dplyr::select(Linie, SAMPLE_ALIAS, target_cvg, everything(),
                -PCT_GQ0_VARIANTS, -TOTAL_GQ0_VARIANTS) %>% 
  datatable(rownames = F, options = list(pageLength = 25))
```

### **Population metrics**
```{r}
here("batches123_04_FinalVCF/metrics") %>% 
  list.files() %>% 
  .[grep("DU",.)] %>% 
  .[grep("summary",.)] %>% 
  lapply(function(x){
    d <- read.table(here("batches123_04_FinalVCF/metrics",x),stringsAsFactors = F,dec = ",",comment.char = "#", header = T)
    l <- str_remove(x,"cohort_biallelicSNPs_VQSR95_PASS_AddedMissingness.recode.filtered.") %>% 
      str_remove(".variant_calling_summary_metrics")
    d %>% mutate(line = l) %>% dplyr::select(line, everything())
  }) %>% 
  bind_rows() %>% 
  datatable(rownames = F)
```

> Population VCFs were produced from cohort final-VCF by extracting samples and removing fixed reference sites (GT=="RR") and all-missing sites (GT=="./.").


SNP Annotations
===============
```{r}

```


LDD & SFS 
========================

Column
-------------------

### Allele frequency state per population
```{r}

```

### LD Decay

Column
-------------------

### Alternative Allele Frequency Distribution
```{r}

```

### Minor Allele Frequency Distribution
```{r}

```

Diversity
===========================

Population Structure
==============================

Column {.tabset}
----------------

### PC1 - PC2

### PC2 - PC3

### PC3 - PC4

### PC4 - PC5

### PC5 - PC6


Column
---------------

### Hieratchical Clustering

### Admixture 

Admixture detailed
==============================

Column
--------

### K2

### K3

### K4

Column
-------

### K5

### K6

Divergent Signatures
==========================


Directional Signatures
===========================


