---
title: "WGS DU-mice"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---


```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F} 
options(scipen = 999)
#library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
#library(reshape2)
#library(tidyr)
library(here)
#library(plotly)
library(DT)
#library(UpSetR)
#library(data.table)
library(vroom)
#library(RColorBrewer)
library(knitr)
#library(ggcorrplot)
#library(tibble)
#library(png)
#library(grid)
#library(gridExtra)
#library(fastqcr)
library(GenomicRanges)
library(kableExtra)
#library(ape)
#library(WebGestaltR)
#library(pheatmap)
library(limma)
library(ggrepel)
#library(bumphunter) #BiocManager::install("bumphunter")
#library("TxDb.Mmusculus.UCSC.mm10.ensGene") #BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(biomaRt)
library(Gviz)
```

```{r sample_info_and_genome_length, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

#http://www.ensembl.org/Mus_musculus/Info/Annotation 
genome_length <- 2730871774 # genome length (Golden Path Length	) 
```

```{r vcf_final_SNP_ids}

x <- vroom(
  here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids"), 
  col_names = FALSE, 
  delim = " ",
  col_types = c(X1 = "c")
  ) %>% 
  dplyr::select(-X2) %>% 
  dplyr::rename(seqname = X1, start = X3) %>% 
  mutate(end = start) 

# turn it into granges
vcf_final_snp_ids_gr <- GRanges(seqnames = x$seqname, IRanges(start = x$start, end = x$end))

rm(x)
```

```{r win_fst_data_sel_vs_ctrl_zscore, include=F}  

win_fst_sel_vs_ctrl <- lapply(
  
  list.files(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
             pattern = "windowed.weir.fst$"), 
  
  function(fl){
    
    read.table(
      here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl),
      header = T,
      stringsAsFactors = F
      )
    }
  )

names(win_fst_sel_vs_ctrl) <- list.files(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
  pattern = "windowed.weir.fst$"
  ) %>% str_remove(".windowed.weir.fst")

# prepare data (i.e. zscore transformation) for visualizations
win_fst_sel_vs_ctrl_prep <- lapply(win_fst_sel_vs_ctrl, function(d){
  d %>% 
    # filter by min number of SNPs in window (min 10 SNPs)
    dplyr::filter(N_VARIANTS >= 10) %>% 
    # separate x chromosome
    mutate(is_x = CHROM == "X") %>% 
    # group by autosomes/chrX
    group_by(is_x) %>% 
    # compute zscores
    mutate(z_win_fst_score = scale(MEAN_FST))   
}) %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    )

# label extreme scores
win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  mutate(
    is_1th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.01) ),
    is_5th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.05) ),
    is_10th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.10) ),
    is_25th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.25) ),
    is_90th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.90) ),
    is_95th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.95) ),
    is_99th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.99) )
      )

#win_fst_sel_vs_ctrl_prep %>% summarise(sum(is_99th_quantile)/n()) # it checks out :)



```

```{r ensembl_gene_set, include=F}
# Load gene set
mmu_gene_set <- read.table(here("reference_genome_ensembl/Mus_musculus.GRCm38.93_gene.gtf"), 
                           header = F, stringsAsFactors = F) 
  
names(mmu_gene_set) <- c("seqname","start","end","strand","gene_id","gene_id2","gene_biotype") # as in readme

mmu_gene_set <-mmu_gene_set %>% 
  filter(seqname %in% c(1:19,"X")) %>% # subset for autosomes and X
  dplyr::rename(gene_symbol = gene_id2)

# Turn gene set into a genomics range object
mmu_gene_set_gr <- GRanges(seqnames = mmu_gene_set$seqname,
                           ranges = IRanges(start = mmu_gene_set$start,
                                            end = mmu_gene_set$end),
                           strand = mmu_gene_set$strand,
                           mcols = dplyr::select(mmu_gene_set, gene_id, gene_symbol, gene_biotype))

```

```{r function_run_WebGestaltR, eval = F}
#---------------------------
# function_run_WebGestaltR
#---------------------------

run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}

```

```{r find_distinct_genes, include=FALSE} 
get_distinct_genes <- function(ma, cont1, rest_cont, upper_bound, lower_bound){

  # get entries of matrix with high fst for target contrast
  cond1 <- ma[,cont1] >= upper_bound
  
  # get entries of matrix in which remaining contrast are low 
  cond2 <-  ma[,rest_cont] %>% apply(1, function(x) all(x < lower_bound) )
  
  # turn NAs as FALSE
  cond1[is.na(cond1)] <- FALSE
  cond2[is.na(cond2)] <- FALSE
  
  # combine both conditions into one indexing vector
  idx <- cbind(cond1,cond2) %>% apply(1, all)
  
  # subset matrix
  ma[idx,] %>% as.data.frame()

}
```

```{r load_regulatory_regions_ensembl, include = FALSE} 

# https://www.ensembl.org/info/website/upload/gff.html --> gff format explained

## Regulatory_Build
x <- vroom(
  here("reference_genome_ensembl/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff"), col_names = FALSE
)

regulatory_features <- GRanges(
  seqnames = x$X1, 
  IRanges(x$X4, x$X5),
  mcols = dplyr::select(x,X2, X3, X9) %>% dplyr::rename(source = X2, feature = X3, attribute = X9)
  )


## motiffeatures
y <- vroom(
  here("reference_genome_ensembl/mus_musculus.GRCm38.motiffeatures.20180516.gff"), col_names = FALSE
)


motif_features <- GRanges(
  seqnames = y$X1, 
  IRanges(y$X4, y$X5),
  mcols = dplyr::select(y,X2, X3, X6, X7, X9) %>% dplyr::rename(source = X2, feature = X3, score = X6, strand = X7, attribute = X9)
  )

rm(x,y)
```

RDDs {data-navmenu="RDDs"}
====================================

Column {.tabset}
-----------------

### Overview

* Regions of Distinct Differentiation (RDDs)

* RDDs = Windows in the top 5% of a given contrast that are found in the bottom 10% of all other contrasts.

* Distinct genes were those overlapping RDDs (RDD-genes).

* Functional annotation of RDD genes was done using biomart to collect associated biological processes (GOBP).

* For each gene SNPs were identified and their effects summarized.

* Likewise, for every SNP in a RDD-gene, their Fst scores and effects are displayed.

### Summaries 

```{r define_contrasting_quantiles}
quantiles <- c("q10","q95") 
```

```{r define_function_to_find_distinct_REDs} 

summary_z_win_fst <- readRDS( here("00_dashboard","summary_z_win_fst.rds") )

find_distinct_REDs <- function(contr, rest, q_low, q_high){
  
  #contr="DUK_vs_FZTDU"
  #rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" )
  #q_low = "q10"
  #q_high = "q90"
  
  # for each contrast in rest, find bottom 25% windows
  rest_bottom_windows <- lapply(rest, function(x){
    
    # set threshold for bottom scores
    q_low_threshold <- summary_z_win_fst[summary_z_win_fst$contrast == x, q_low] %>% as.numeric()
    
    # get rest-contrast individual data
    win_fst_sel_vs_ctrl_prep %>% filter(contrast %in% x) %>% 
      # get bottom scores using zFst scores
      filter(z_win_fst_score < q_low_threshold) %>% 
      ungroup() %>% 
      # select columns for joining
      dplyr::select(CHROM, BIN_START, BIN_END)
    }) %>% 
    # inner join all contrasts in rest
    purrr::reduce(inner_join, by = c("CHROM","BIN_START","BIN_END"))
  
  # set upper score
  q_high_threshold <- summary_z_win_fst[summary_z_win_fst$contrast == contr, q_high] %>% as.numeric()
  
  # get top windows in contrast of interest
  contr_extreme_windows <- win_fst_sel_vs_ctrl_prep %>% 
    filter(
      # get contrast of interest data
      contrast == contr &
        # keep top zFst scores
        z_win_fst_score > q_high_threshold
    ) %>% 
    # select columns for merging
    ungroup() %>% 
    dplyr::select(CHROM, BIN_START, BIN_END)

  # get distinct windows
  distinct_windows <- inner_join(contr_extreme_windows, 
                                 rest_bottom_windows, 
                                 by = c("CHROM","BIN_START","BIN_END")) %>% 
    # add fst and number of variants information
    inner_join(
      win_fst_sel_vs_ctrl_prep %>% 
        filter(contrast == contr) %>% 
        dplyr::select(contrast, CHROM, BIN_START, BIN_END, N_VARIANTS, MEAN_FST),
      by = c("CHROM","BIN_START","BIN_END")
      ) %>% 
    mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
    arrange(CHROM)
  
  return(distinct_windows)
}


distinct_windows_list <- list(
  DUK = find_distinct_REDs(
    contr = "DUK_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DUC = find_distinct_REDs(
    contr = "DUC_vs_FZTDU", rest = paste0( c("DUK","DU6","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DU6 = find_distinct_REDs(
    contr = "DU6_vs_FZTDU", rest = paste0( c("DUC","DUK","DU6P","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DU6P = find_distinct_REDs(
    contr = "DU6P_vs_FZTDU", rest = paste0( c("DUC","DU6","DUK","DUhLB"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  DUhLB = find_distinct_REDs(
    contr = "DUhLB_vs_FZTDU", rest = paste0( c("DUC","DU6","DU6P","DUK"), "_vs_FZTDU" ), quantiles[1], quantiles[2]
    ),
  FERT = find_distinct_REDs(
    contr = "FERT_vs_FZTDU", rest = paste0( c("DU6","DU6P","DUhLB"), "_vs_FZTDU" ),  quantiles[1], quantiles[2]
    )
)

```

```{r display_summary_distinct_regions}

summarise_distinct_win <- function(line_sel, rest){

#args
#line_sel = "DUK"
#rest = c("DUC","DU6","DU6P","DUhLB")

  #{
  x <- distinct_windows_list[[line_sel]]

  xx <- lapply(rest, function(y){
    #y="DUC"
    win_fst_sel_vs_ctrl_prep %>% 
      filter(contrast == paste0(y,"_vs_FZTDU")) %>% 
      inner_join(
        x %>% dplyr::select(-MEAN_FST), 
        by = c("CHROM","BIN_START","BIN_END")) %>% 
      .$MEAN_FST 
  }) %>% 
    unlist() %>% 
    range() %>%
    round(2) %>% 
    paste0(collapse = " - ")
    

  data.frame(
    contrast = x$contrast %>% unique(),
    n_windows = nrow(x),
    chrs = unique(x$CHROM) %>% paste(collapse = ","),
    n_win_by_chr = x %>% 
      group_by(CHROM) %>% 
      summarise(n=n()) %>%
      ungroup() %>% 
      mutate(tmp =  paste0("chr",CHROM,"(",n,")")  ) %>% 
      dplyr::select(tmp) %>% 
      unlist() %>% 
      paste(collapse = "; "),
    n_vars_range = range(x$N_VARIANTS) %>% paste(collapse = " - "),
    fst_range = range(x$MEAN_FST) %>% round(2) %>% paste(collapse = " - "),
    fst_range_rest = xx,
    rest = paste(rest, collapse = ", ")
    ) 
  }

bind_rows(
  summarise_distinct_win("DUK", c("DUC","DU6","DU6P","DUhLB")),
  summarise_distinct_win("DUC", c("DUK","DU6","DU6P","DUhLB")),
  summarise_distinct_win("DU6", c("DUC","DUK","DU6P","DUhLB")),
  summarise_distinct_win("DU6P", c("DUC","DU6","DUK","DUhLB")),
  summarise_distinct_win("DUhLB", c("DUC","DU6","DU6P","DUK")),
  summarise_distinct_win("FERT", c("DU6","DU6P","DUhLB"))
  ) %>% 
  kable(digits = 2, caption = paste0("Summary Distinct Windows (",quantiles[1]," vs ", quantiles[2], ")" )) %>% 
  kable_styling(full_width = F)


```

```{r get_SNPs_in_RDDs_and_display}
snps_in_RDDs <- lapply(1:length(distinct_windows_list), function(i){
  #i=1

  x <- distinct_windows_list[[i]] %>% 
    dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
    makeGRangesFromDataFrame()
  
  snps <- subsetByOverlaps(x = vcf_final_snp_ids_gr, ranges = x, minoverlap = 1) %>% unique()
  
  return(snps)
})

names(snps_in_RDDs) <- names(distinct_windows_list)

data.frame(
    contrast = names(snps_in_RDDs),
    Nr_RDD_SNPs = sapply(snps_in_RDDs, length)
  ) %>% 
  kable(digits = 2, caption = "Number of SNPs in RDDs", row.names = FALSE) %>% 
  kable_styling(full_width = F)

```

```{r find_genes_in_distinct_windows}

RDD_genes <- lapply(distinct_windows_list, function(x){
  
  #x=distinct_windows_list$DUK
  
  wins <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))

  gns_grs <- subsetByOverlaps(mmu_gene_set_gr, wins, minoverlap = 1) %>% unique()

})

```

```{r find_SNPs_in_RDD_genes_and_export, eval = FALSE}  

# this will not be shown in the dashboard

snps_in_RDD_genes <- lapply(RDD_genes, function(x)
  
  subsetByOverlaps(vcf_final_snp_ids_gr, x, minoverlap = 1) %>% unique()
  
)

snps_in_RDD_genes %>% 
  lapply(as.data.frame) %>% 
  bind_rows() %>% 
  mutate(interval = paste0(seqnames,":",start,"-",end)) %>% 
  dplyr::select(interval) %>% 
  unique() %>% 
  write.table(
    here("00_dashboard/data/RDD_genes","snps_in_RDD_genes.intervals"), 
    col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE
  )

```

```{r GT_counts_SNPs_in_RDD_genes, eval = FALSE}

SNPs_in_RDD_genes_GT_tab <- vroom(
    here(
      "batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered_SNPs_in_RDD_genes_GT.table"
      )
  ) 

names(SNPs_in_RDD_genes_GT_tab) <- str_remove(names(SNPs_in_RDD_genes_GT_tab),".GT") %>% str_remove("-L1")

tabulate_genotypes_per_contrast2 <- function(pop){ 
  
  #pop <- "DUK"

  # Get sample ids for each line
  samps_pop <- sample_info %>% filter(Linie %in% pop) %>% .$sample_id
  
  # tabulate genotypes of selected line
  sel_GT_cts <- SNPs_in_RDD_genes_GT_tab %>% 
    dplyr::select(all_of(samps_pop)) %>% 
    apply(1, function(x){
      tb <- table(x)
      tb_nms <- names(tb)
      paste0(tb, "(", tb_nms, ")") %>% paste(collapse = ";")
    }) 

  # add locus information
  ss <- SNPs_in_RDD_genes_GT_tab %>% dplyr::select(CHROM, POS)
  
  # add column with GT counts
  new_col <- paste0(paste(pop, collapse = "_"),"_GT_counts")
  ss$tmp <- sel_GT_cts
  names(ss)[names(ss) == "tmp"] <- new_col

  return(ss)  

}

by_cols <- c("CHROM","POS")

tabulate_genotypes_per_contrast2("DUK") %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DUC"), by = by_cols
  ) %>%  
  inner_join(
    tabulate_genotypes_per_contrast2(c("DUK","DUC")), by = by_cols #FERT
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DU6"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DU6P"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("DUhLB"), by = by_cols
  ) %>% 
  inner_join(
    tabulate_genotypes_per_contrast2("FZTDU"), by = by_cols
  ) %>% 
  left_join(
    vroom(
      here("batches123_04_final_VCF/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.SNPids2"), 
      col_names = FALSE 
    ) %>% 
      dplyr::rename(CHROM = X1, POS = X3, REF = X5, ALT = X7) %>% 
      dplyr::select(CHROM, POS, REF, ALT),
    by = by_cols
  ) %>% 
  dplyr::select(CHROM, POS, REF, ALT, everything()) %>% 
  #export for later use
  write.csv(
    here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")
  )
```

```{r display_gene_summary_table}
lapply(RDD_genes, function(x){
  x %>% 
    as.data.frame() %>% 
    dplyr::select(mcols.gene_id, mcols.gene_biotype) %>% 
    unique() %>% 
    group_by( mcols.gene_biotype) %>% 
    summarise(n=n()) 
  }) %>% 
  bind_rows(.id="contrast") %>% 
  reshape2::dcast(mcols.gene_biotype ~ contrast, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  dplyr::rename(gene_biotype=mcols.gene_biotype) %>% 
  dplyr::select(gene_biotype, DUK,DUC,DU6,DU6P,DUhLB,FERT) %>% 
  kable(digits = 2, caption = paste0("Summary Distinct Genes (",quantiles[1]," vs ", quantiles[2], ")" )) %>% 
  kable_styling(full_width = F)
```

```{r add_gene_ontology, eval = FALSE}

# biomart query done on Tuesday 20/10/2020
#mart <- useDataset(dataset = "mmusculus_gene_ensembl", mart = useMart("ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org"))

#x = RDD_genes$FERT

RDD_genes2 <- lapply(1:length(RDD_genes), function(i){ 
  
  #i=1
  
  x <- RDD_genes[[i]]
  
  # do BM quary
  mart_res <- getBM(attributes = c("ensembl_gene_id","mgi_symbol","name_1006","namespace_1003"), 
                  filters = "ensembl_gene_id", 
                  values = x$mcols.gene_id,
                  mart = mart) %>% 
              filter(namespace_1003 == "biological_process") 

  # collapse GOBP terms
  mart_res$ensembl_gene_id %>% 
    unique() %>% 
    lapply(function(y){
      
      gobps <- mart_res %>% 
        filter(ensembl_gene_id == y) %>%
        dplyr::select(name_1006) %>% 
        unlist() %>% 
        paste(collapse = "; ")
      
      data.frame(
        ensembl_gene_id = y,
        mgi_symbol = mart_res$mgi_symbol[mart_res$ensembl_gene_id == y] %>% unique(), 
        GOBP = gobps
        )
    }) %>% 
    bind_rows() %>% 
    full_join(
      x %>% as.data.frame() %>% dplyr::rename(ensembl_gene_id = mcols.gene_id, gene_biotype = mcols.gene_biotype),
      by = "ensembl_gene_id"
    ) %>% 
    mutate(contrast = paste0(names(RDD_genes)[i], "_vs_FZTDU") ) %>% 
    dplyr::select(contrast, ensembl_gene_id,mgi_symbol,seqnames, start, end, width, strand, gene_biotype, GOBP) 
})


names(RDD_genes2) <- names(RDD_genes)

# export for later use offline (biomart query not needed)

RDD_genes2$DUK %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUK.csv"))
RDD_genes2$DUC %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUC.csv"))
RDD_genes2$DU6 %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DU6.csv"))
RDD_genes2$DU6P %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DU6P.csv"))
RDD_genes2$DUhLB %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB.csv"))
RDD_genes2$FERT %>% write.csv(here("00_dashboard/data/RDD_genes/RDD_genes_FERT.csv"))

```

### DUK_genes 
```{r} 
here("00_dashboard/data/RDD_genes/RDD_genes_DUK.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUC_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DUC.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DU6P_genes
```{r} 
here("00_dashboard/data/RDD_genes/RDD_genes_DU6P.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### DUhLB_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### FERT_genes 
```{r}
here("00_dashboard/data/RDD_genes/RDD_genes_FERT.csv") %>% read.csv() %>% dplyr::select(-X) %>% datatable(options=list(paging = FALSE))
```

### RDD-SNPs_effect_counts 
```{r} 

display_snp_effects_per_transcript <- function(pop){
  
  csv_fl <- paste0("RDD_genes_",pop,".csv")
  
  here("00_dashboard/data/RDD_genes",csv_fl) %>% 
  read.csv() %>% 
  dplyr::select(contrast, ensembl_gene_id, mgi_symbol, gene_biotype) %>% 
  left_join(
    vroom(here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.summary.genes.txt"),skip = 1),
    by = c("ensembl_gene_id" = "GeneId")
  ) %>% 
  unique() 
  }

lapply(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), display_snp_effects_per_transcript) %>% 
  bind_rows() %>% 
  datatable(options = list(paging = FALSE), caption = "SNP effect counts (SNPs associated to RDD genes (all contrasts) as per SNPeff)")

```

### RDDs_x_REDs_Venn 
```{r visualize_RED_RDD_intersections, fig.width=5, fig.height=8} 

RED_genes <- readRDS(here("00_dashboard/RED_genes.rds"))

display_venn_RED_RDD_genes <- function(red_contr, rdd_contr){
  
  #...
  #red_contr="DUK_vs_FZTDU"
  #rdd_contr="DUK"
  #...
  
  d <- data.frame(
    ids = unique(RED_genes[[red_contr]]$mcols.gene_id, RDD_genes[[rdd_contr]]$mcols.gene_id)
    ) %>% 
    mutate(
      RED_genes = ids %in% RED_genes[[red_contr]]$mcols.gene_id,
      RDD_genes = ids %in% RDD_genes[[rdd_contr]]$mcols.gene_id
    )
  
  vc <- vennCounts(
    d %>% dplyr::select(-ids)
    )
  
  vennDiagram(vc, circle.col = c("red","blue"), mar = rep(0.1,4), cex = 0.7)
  title(red_contr, line = 0, adj = 0.5)
  
  }

par(mfrow = c(3, 2))

display_venn_RED_RDD_genes("DUK_vs_FZTDU","DUK")
display_venn_RED_RDD_genes("DUC_vs_FZTDU","DUC")
display_venn_RED_RDD_genes("DU6_vs_FZTDU","DU6")
display_venn_RED_RDD_genes("DU6P_vs_FZTDU","DU6P")
display_venn_RED_RDD_genes("DUhLB_vs_FZTDU","DUhLB")
display_venn_RED_RDD_genes("FERT_vs_FZTDU","FERT")


```

### Manhattans_RDDgenes 
```{r make_manhattan_plots_and_export, eval = FALSE} 

d_man <- win_fst_sel_vs_ctrl_prep %>% 
  dplyr::select(contrast, CHROM, BIN_START, BIN_END, MEAN_FST) %>%
  reshape2::dcast(CHROM + BIN_START + BIN_END ~ contrast, value.var = "MEAN_FST") %>% 
  mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
         center_window_Mb = center_window/1000000) %>% 
  # relevel chromosomes and colors for chromosomes
  mutate(CHROM = factor(CHROM, levels = c(1:19, "X"))) %>% 
  # sort by chromosomes and window position
  arrange(CHROM, center_window_Mb) %>% 
  mutate(genomic_index = 1:nrow(.)) %>% 
  mutate(chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")) %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","center_window","center_window_Mb","genomic_index","chr_color"), 
                 variable.name = "contrast", value.name = "MEAN_FST")

# define ticks for manhattan
ticks <- d_man %>% 
  group_by(CHROM) %>% 
  summarise(chr_median = median(genomic_index), chr_end = max(genomic_index)) 



plot_manhattan <- function(pop){

#pop="DUK" #arg
  
  set.seed(456)

  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  # convert contrast data to granges to add RDD gene symbols
  d_man_gr <- d_man %>% 
    filter(contrast == contr) %>% 
    dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  
  # find overlaps between contrast fst data and RDD symbols for pop
  ov <- findOverlaps(query = d_man_gr, subject = RDD_genes[[pop]])
  
  # combine fst contrast data with gene symbols
  ov_dat <- bind_cols(
    d_man_gr[queryHits(ov)] %>% as.data.frame(),
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
  ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol)
  
  # Some genes appear twice because they map to >1 window Fst
  # Get the max Fst per gene to resolve duplicated genes
  # for genes matching more than one window, choose window with highest Fst
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
    ) %>%
    dplyr::select(genomic_index, MEAN_FST, gene_symbol)
  
  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
    }) %>% 
    bind_rows()
  
  # prepare plot data 
  plot_dat <- d_man %>% 
    # subset by contrast and by min Fst
    filter(contrast == contr, MEAN_FST > 0.6) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )  
  
  # make plot
  p <- ggplot(# main data points ("background" -- black and grey dots)
        data = filter(plot_dat, is.na(gene_symbol)), 
        aes(x = genomic_index, y = MEAN_FST),
        alpha = 0.5, 
        show.legend = FALSE) +
    geom_point(aes(color = chr_color), show.legend = FALSE) +
    # red points for RDD gene windows
    geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
    # RDD gene labels
    geom_label_repel(
      data =  filter(plot_dat, !is.na(gene_symbol)),
      aes(label = gene_symbol),
      segment.color = "red", 
      arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
      fill = alpha(c("white"),0.5)
      ) +
    # Make it look nicer
    theme_bw(base_size = 15) +
    scale_color_manual(values = c("black", "darkgrey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0.01,0.01)) +
    scale_y_continuous(expand = c(0.01,0.01)) +
    xlab("chromosome") +
    ylab(
      bquote(
        Fst[(.(paste0(pop,"_vs_FZTDU")))] > 0.6
        )
      )
   return(p)
}


fig_path <- "00_dashboard/figures/manhattan_win_fst_rdd_genes_labs"

png(here(fig_path, "duk_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUK"); dev.off() 
png(here(fig_path, "duc_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUC"); dev.off() 
png(here(fig_path, "du6_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DU6"); dev.off() 
png(here(fig_path, "du6p_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DU6P"); dev.off() 
png(here(fig_path, "duhlb_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="DUhLB"); dev.off() 
png(here(fig_path, "fert_win_fst_rdd_genes_labs.png"), res = 300, height = 2000, width = 6000); plot_manhattan(pop="FERT"); dev.off() 

plot_manhattan_flat <- function(pop){ 

#pop="DUK" #arg
  
  set.seed(45)

  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  # convert contrast data to granges to add RDD gene symbols
  d_man_gr <- d_man %>% 
    filter(contrast == contr) %>% 
    dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  
  # find overlaps between contrast fst data and RDD symbols for pop
  ov <- findOverlaps(query = d_man_gr, subject = RDD_genes[[pop]])
  
  # combine fst contrast data with gene symbols
  ov_dat <- bind_cols(
    d_man_gr[queryHits(ov)] %>% as.data.frame(),
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
  ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol)
  
  # Some genes appear twice because they map to >1 window Fst
  # Get the max Fst per gene to resolve duplicated genes
  # for genes matching more than one window, choose window with highest Fst
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
    ) %>%
    dplyr::select(genomic_index, MEAN_FST, gene_symbol)
  
  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
    }) %>% 
    bind_rows()
  
  # prepare plot data 
  plot_dat <- d_man %>% 
    # subset by contrast and by min Fst
    filter(contrast == contr) %>% 
    # ...
    mutate(y = 0) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )  
  
  # make plot
  p <- ggplot(data =  plot_dat, aes(x = genomic_index, y = y, label = gene_symbol)) +
    geom_point(aes(color = chr_color), show.legend = FALSE, shape = 15) +
    # RDD gene labels
    geom_label_repel(
      data =  filter(plot_dat, !is.na(gene_symbol)),
      aes(label = gene_symbol),
      segment.color = "red", 
      segment.alpha = 0.5,
      arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
      fill = alpha(c("white"),0.5),
      nudge_y = 0.3#,
      #direction = "x"
    ) +
    # Make it look nicer
    #theme_bw(base_size = 15) +
    scale_color_manual(values = c("black", "darkgrey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM) +
    scale_y_continuous(limits = c(0, 1), expand = c(0.01,0)) +
    xlab("chromosome") +
    labs(caption = contr) +
    theme(
      axis.line.y  = element_blank(),
      axis.ticks.y = element_blank(),
      axis.text.y  = element_blank(),
      axis.title.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_blank()
    )  
  
   return(p)
}

png(here(fig_path, "duk_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUK"); dev.off() 
png(here(fig_path, "duc_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUC"); dev.off() 
png(here(fig_path, "du6_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DU6"); dev.off() 
png(here(fig_path, "du6p_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DU6P"); dev.off() 
png(here(fig_path, "duhlb_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="DUhLB"); dev.off() 
png(here(fig_path, "fert_win_fst_rdd_genes_labs_flat.png"), res = 300, height = 2000, width = 6000); plot_manhattan_flat(pop="FERT"); dev.off() 


```

```{r display_manhattan_plots, out.width = "90%", fig.align = "center"} 

fig_path <- here("00_dashboard/figures/manhattan_win_fst_rdd_genes_labs")

include_graphics(
  c(
    list.files(fig_path, full.names = TRUE) %>% .[grep("flat", .)],
    list.files(fig_path, full.names = TRUE) %>% .[-grep("flat", .)]
    )
  )

```

```{r make_goi_zoom_labs_manhattan, eval = FALSE}

plot_manhattan_zoom <- function(pop, chr, from_mb, to_mb, nudge_y){
  # Define contrast name
  contr <- paste0(pop, "_vs_FZTDU")
  
  
  
  # Find overlaps between RDD-windows and RDD-genes
  ov <- findOverlaps(
    
    query = distinct_windows_list[[pop]] %>% 
      dplyr::rename(seqnames=CHROM, start=BIN_START, end=BIN_END) %>% 
      makeGRangesFromDataFrame(keep.extra.columns = TRUE), 
    
    subject = RDD_genes[[pop]], 
    
    minoverlap = 1
    )

  # combine RDD-windows and RDD-gene-symbols
  ov_dat <- bind_cols(
    distinct_windows_list[[pop]][queryHits(ov),],
    RDD_genes[[pop]][subjectHits(ov)] %>% as.data.frame() %>% dplyr::select(mcols.gene_symbol)
    ) %>% 
    dplyr::rename(gene_symbol=mcols.gene_symbol) 

  # temporary data to get the max Fst RDD-window per gene 
  # this is because some genes map to more than one window
  # here only the max Fst RDD window is kept
  gene_labs_tmp <- inner_join(
    ov_dat,
    ov_dat %>% 
      group_by(gene_symbol) %>% 
      summarise(max_per_gene = max(MEAN_FST, na.rm=TRUE)),
    by = c("gene_symbol", c("MEAN_FST"="max_per_gene"))
  ) %>%
    mutate(center_window = (BIN_START + (BIN_END-BIN_START)/2) ,
           center_window_Mb = center_window/1000000) %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, center_window, center_window_Mb, MEAN_FST, gene_symbol) %>% 
    # add genomic index (needed for later)
    inner_join(
      d_man %>% 
        filter(contrast == contr) %>% 
        dplyr::select(CHROM, BIN_START, BIN_END, genomic_index),
      by = c("CHROM", "BIN_START", "BIN_END")
    )

  # Some genes map to the same window ... collapse them into one entryy
  gene_labs_final <- lapply(unique(gene_labs_tmp$genomic_index), function(x){
    d0 <- gene_labs_tmp %>% filter(genomic_index == x) 
    d1 <- d0 %>% dplyr::select(genomic_index) %>% unique()
    d1$gene_symbol <- d0 %>% dplyr::select(gene_symbol) %>% unlist() %>% paste0(collapse = ",")
    return(d1)
  }) %>% 
    bind_rows()

  # prepare plot data 
  plot_dat <- d_man %>% 
    filter(CHROM == chr & center_window_Mb > from_mb & center_window_Mb < to_mb) %>% 
    # add gene symbols
    left_join(
      gene_labs_final %>% dplyr::select(genomic_index, gene_symbol),
      by = "genomic_index"
    )  
  
  lab_dat <- plot_dat %>% 
    filter(!is.na(gene_symbol) & contrast == contr) 
  
  if(nrow(lab_dat) > 1){

  # interspace gene labels    
  lab_dat$lab_set <- NA
  lab_dat$lab_set[seq(1,nrow(lab_dat), 2)] <- "1"
  lab_dat$lab_set[seq(2,nrow(lab_dat), 2)] <- "2"
  
    # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "1"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1
      ) +
      geom_label_repel(
        data =  filter(lab_dat, lab_set == "2"),aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = -nudge_y,
        vjust = 1
      ) +
      # Make it look nicer
      theme_bw(base_size = 15) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2)) +
      xlab("Mb") 
    return(p)
  }else{
          # make plot
    p <- ggplot(# main data points ("background" -- black and grey dots)
      data = plot_dat, #data = filter(plot_dat, is.na(gene_symbol)), 
      aes(x = center_window_Mb, y = MEAN_FST, color = contrast),
      alpha = 0.5, 
      show.legend = TRUE) +
      geom_line() + #geom_point(color = "darkgrey") +
      # red points for RDD gene windows
      geom_point(data =  filter(plot_dat, !is.na(gene_symbol)), color = "red") +
      # RDD gene labels
      geom_label_repel(
        data =  lab_dat,aes(label = gene_symbol),
        segment.color = "black", 
        arrow = arrow(length = unit(0.01, "npc"), type = "closed", ends = "first"),
        fill = alpha(c("white"),0.5),
        show.legend = FALSE,
        color = "black",
        direction = "x",
        nudge_y = nudge_y,
        vjust = 1
      ) +
      # Make it look nicer
      theme_bw(base_size = 15) +
      scale_x_continuous(expand = c(0.01,0.01)) +
      scale_y_continuous(expand = c(0.01,0.01), limits = c(NA, 1.2)) +
      xlab("Mb") 
    return(p)
    }
  }

plot_manhattan_zoom(pop="DUK", chr=1, from_mb=24.9, to_mb=25.05, nudge_y=0.1)
  
plot_manhattan_zoom(pop="DUK", chr=2, from_mb=28.25, to_mb=29.5, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=6, from_mb=27.25, to_mb=28.1, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=6, from_mb=72.1, to_mb=72.5, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=6, from_mb=115.5, to_mb=115.7, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=9, from_mb=29.1, to_mb=29.4, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=12, from_mb=109.5, to_mb=110.75, nudge_y=0.1)#!

plot_manhattan_zoom(pop="DUK", chr=15, from_mb=98.5, to_mb=98.7, nudge_y=0.1)

plot_manhattan_zoom(pop="DUK", chr=16, from_mb=25, to_mb=27, nudge_y=0.1)

```





RDD-genes_DEGs {data-navmenu="RDDs"} 
================================================

```{r find_and_display_RDDs_x_DEGs} 

FL1_testis_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL1_testis_DEGs_gene_symbols.rds"))
FL2_testis_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL2_testis_DEGs_gene_symbols.rds"))
FL1_ovary_DEGs_gene_symbols <- readRDS(here("00_dashboard/FL1_ovary_DEGs_gene_symbols.rds"))

#FL1=DUK
#FL2=DUC

# Find DEGs in list of RDDs
lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols) %>% # find overlap with FL1_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RDD genes x DEGs (FL1=DUK testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols) %>% # find overlap with FL2_testis_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RDD genes x DEGs (FL2=DUC testis)") %>% 
  kable_styling(full_width = F)


lapply(1:length(RDD_genes), function(i){
  gn_ids <- RDD_genes[[i]] %>% 
              as.data.frame() %>% 
              inner_join(
                mmu_gene_set,
                by = c("mcols.gene_id" = "gene_id")
              ) %>% 
              dplyr::select(gene_symbol) %>% 
              mutate(gene_symbol = str_trim(gene_symbol)) %>% #just in case
              filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols) %>% # find overlap with FL1_ovary_DEGs
              unlist()
  names(gn_ids) <- NULL
  gn_ids2 <- paste(gn_ids, collapse = " ; ")
  data.frame(contrast = names(RDD_genes)[i], n_common_genes = length(gn_ids), gene_symbol = gn_ids2)
}) %>% 
  bind_rows() %>% 
  kable(caption = "RDD genes x DEGs (FL1=DUK ovary)") %>% 
  kable_styling(full_width = F)
```



RDD-SNP-effects {data-navmenu="RDDs"}
================================================ 

Column {.tabset}
-----------------

### Overview

* Find SNPs in RDD-genes, then add SNPeff annotations and Fst scores.

* Note: this was not done for RED-genes!

```{r make_multiple_query_function_and_export, eval=FALSE}

# Function to get SNPs in RDD-genes and extract their effects from SNPeff table output.

  display_snp_effects_per_gene <- function(pop){
  
    # define data set  
    x <- RDD_genes[[pop]]
    
    # define chromosomes of interest (the ones with RDD genes)
    chrs <- x %>% seqnames() %>% unique() %>% as.character()
    
    snp_effect_fst <- 
      # import snp annotations data set
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
      vroom() %>% 
      # retain chromosomes of interest to reduce size
      filter(CHROM %in% chrs) %>% 
      # prepare columns for granges
      mutate(start = POS, end = POS) %>% 
      dplyr::rename(
        seqnames = CHROM, effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT", gene_symbol = "ANN[*].GENE", gene_id = "ANN[*].GENEID"
        ) %>% 
      dplyr::select(seqnames, start, end, REF, ALT, effect, impact, gene_symbol, gene_id) %>% 
      # kepp unique snp-gene-effect associations (some snps have the same effect in different gene-transcripts)
      unique() %>% 
      # add snp fst information
      left_join(
        vroom(
          here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",paste0(pop,"_vs_FZTDU.weir.fst"))
          ),
        by = c("seqnames"="CHROM", "start"="POS")
      ) %>% 
      # filter by RDD_genes
      filter(gene_id %in% x$mcols.gene_id) %>% 
      # prepare table for display
      mutate(snp_locus = paste0(seqnames, ":", start,"-",end)) %>% 
      dplyr::select(gene_id, gene_symbol, snp_locus, effect, impact, WEIR_AND_COCKERHAM_FST) %>% 
      unique() %>% 
      mutate(contrast = names(RDD_genes)[names(RDD_genes) == pop]) %>%
      dplyr::select(contrast, gene_id, gene_symbol, everything()) 
    
    # export queries (avoid running this every time dashboard is knitted)
    csv_nm <- paste0("RDD_genes_" , names(RDD_genes)[names(RDD_genes) == pop], "_SNPeff.csv")
    write.csv(snp_effect_fst, here("00_dashboard/data/RDD_genes", csv_nm))
  }

display_snp_effects_per_gene("DUK") 
display_snp_effects_per_gene("DUC") 
display_snp_effects_per_gene("DU6") 
display_snp_effects_per_gene("DU6P") 
display_snp_effects_per_gene("DUhLB") 
display_snp_effects_per_gene("FERT") 

```


### DUK
```{r find_snp_in_RDD_genes_add_effects_DUK}
here("00_dashboard/data/RDD_genes/RDD_genes_DUK_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>%
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUK_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))
```

### DUC 
```{r find_snp_in_RDD_genes_add_effects_DUC}
here("00_dashboard/data/RDD_genes/RDD_genes_DUC_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUC_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DU6
```{r find_snp_in_RDD_genes_add_effects_DU6}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DU6_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DU6P
```{r find_snp_in_RDD_genes_add_effects_DU6P}
here("00_dashboard/data/RDD_genes/RDD_genes_DU6P_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DU6P_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### DUhLB
```{r find_snp_in_RDD_genes_add_effects_DUhLB}
here("00_dashboard/data/RDD_genes/RDD_genes_DUhLB_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUhLB_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>% 
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```

### FERT
```{r find_snp_in_RDD_genes_add_effects_FERT}
here("00_dashboard/data/RDD_genes/RDD_genes_FERT_SNPeff.csv") %>% read.csv() %>% dplyr::select(-X) %>% 
  inner_join(
    read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
      mutate(snp_locus = paste0(CHROM,":",POS,"-",POS)) %>% 
      dplyr::select(snp_locus, REF, ALT, DUK_DUC_GT_counts, FZTDU_GT_counts),
    by = "snp_locus"
  ) %>%
  dplyr::select(snp_locus, REF, ALT, contains("GT_counts"), everything()) %>%
  datatable(caption = "SNP effect (SNPs in RDD genes)", options = list(pageLength = 100))

```



RDD-ORA {data-navmenu="RDDs"}
====================================

### Overview

- Mostly null results.

- Significant enrichments were found for:
  
  * DUC_vs_FZTDU (GOBP): hormone-mediated signaling pathway, regulation of dendrite morphogenesis, intracellular steroid hormone receptor signaling pathway.
  
  * DUK_vs_FZTDU (KEGG): Phospholipase D signaling pathway.
  
  * DUhLB_vs_FZTDU (KEGG): Histidine metabolism, beta-Alanine metabolism.
  
- Except for results using gene list from DUC_vs_FZTDU (incl. Pgr), results are not very informative. 

- For more details see html files stored in './00_dashboard/data/WebGestaltR_RDD/'


```{r eval=FALSE, GOBP_run_once_and_export} 
run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
)

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/GOBP")
) #No significant gene set is identified based on FDR 0.1!
```

```{r eval = FALSE, KEGG_run_once_and_export}

run_WebGestaltR(
  interestGene=RDD_genes$DUK$mcols.gene_id,
  projectName="WebGestaltR_DUK_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
)

run_WebGestaltR(
  interestGene=RDD_genes$DUC$mcols.gene_id,
  projectName="WebGestaltR_DUC_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6$mcols.gene_id,
  projectName="WebGestaltR_DU6_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DU6P$mcols.gene_id,
  projectName="WebGestaltR_DU6P_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=RDD_genes$DUhLB$mcols.gene_id,
  projectName="WebGestaltR_DUhLB_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
)

run_WebGestaltR(
  interestGene=RDD_genes$FERT$mcols.gene_id,
  projectName="WebGestaltR_FERT_vs_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RDD/KEGG")
) #No significant gene set is identified based on FDR 0.1!
```


RDD-regSNPs {data-navmenu="RDDs"}
===================================

```{r find_RDD_SNPs_in_regulatory_regions} 

#distinct_windows_list aka RDDs

# find 
reg_SNPs_in_RDDs <- lapply(1:length(distinct_windows_list), function(i){ 

  #i=1
  
  # get gr for contrast
  x <- distinct_windows_list[[i]]
  gr <- GRanges(seqnames = x$CHROM, IRanges(start = x$BIN_START, end = x$BIN_END))
  
  # get SNPs in distinct_windows_lists
  SNPs_in_RDDs <- subsetByOverlaps(x = vcf_final_snp_ids_gr, ranges = gr, minoverlap = 1) 
  
  # Find overlaps between SNPs in RDDs and regulatory or motif features
  ov_rf <- findOverlaps(query = SNPs_in_RDDs, subject = regulatory_features, minoverlap = 1)

  # subset overlapping snps
  SNPs_in_RDDs_in_rf <- SNPs_in_RDDs[queryHits(ov_rf)]

  # add feature information
  mcols(SNPs_in_RDDs_in_rf) <- regulatory_features[subjectHits(ov_rf)] %>% 
    as.data.frame() %>% 
    dplyr::rename(
      feat_seqname = seqnames, feat_start=start, feat_end=end, feat_width=width, feat_strand = strand,
      feat_source = mcols.source, feat_feature=mcols.feature, feat_attribute=mcols.attribute
      ) 
  
  # import per SNP Fst scores and convert to granges
  tmp <- vroom(
    here(
      "batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",
      paste0(names(distinct_windows_list)[i],"_vs_FZTDU.weir.fst")
      ),
    col_types = c(CHROM = "c")
    )
  
  # add fst information to SNPs-in-RDDs-in-regulatory-regions
  SNPs_in_RDDs_in_rf <- SNPs_in_RDDs_in_rf %>% 
    # turn back to data frame
    as.data.frame() %>% 
    # inner join with fst data
    inner_join(tmp, by = c("seqnames" = "CHROM", "start" = "POS")) %>% 
    # turn back into granges
    makeGRangesFromDataFrame(keep.extra.columns=TRUE)

  return(SNPs_in_RDDs_in_rf)  
  
  }
) 

names(reg_SNPs_in_RDDs) <- names(distinct_windows_list)
```

Column {data-width=300}
------------------------

### Overview

* SNPs found in RDDs (RDD-SNPs).

* Regulatory regions (ensembl regulatory build) containing RDD-SNPs. 

* Downstream genes to the above.

Column  {data-width=700, .tabset}
-----------------------------------

### Number of regSNPs

```{r display_nr_hits_RDD_regSNPs} 

lapply(reg_SNPs_in_RDDs, length) %>% as.data.frame() %>% 
  #mutate(contrast = factor(contrast, levels = paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"_vs_FZTDU"))) %>% 
  dplyr::select(c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, caption = "Number of regSNPs in RDDs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


lapply(reg_SNPs_in_RDDs, 
       function(x){
         x %>% 
           as.data.frame() %>% 
           dplyr::select(feat_seqname, feat_start, feat_end) %>% 
           unique() %>% 
           nrow()
         }
       ) %>% 
  as.data.frame() %>% 
  dplyr::select(c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, caption = "Number of regulatory regions containing RDD-SNPs",format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)


```

### Downstream_genes
```{r find_downstream_genes_for_regSNPs_RDD_regions, eval = FALSE}

#load bioconductor gene set
mm10_ensGene <- annotateTranscripts(TxDb.Mmusculus.UCSC.mm10.ensGene) 

# for every set of regSNPs find the nearest downstream gene

downstream_genes <- lapply(1:length(reg_SNPs_in_RDDs), function(i){

  print(names(reg_SNPs_in_RDDs)[i])
  reg_snps <- reg_SNPs_in_RDDs[[i]]
  
  # extract regulatory feature information 
  reg_range <- GRanges(seqnames = paste0("chr",reg_snps$feat_seqname), 
                     IRanges(start = reg_snps$feat_start, end = reg_snps$feat_end))
  
  # add feature width
  reg_range$feature_width <- reg_snps$feat_width


  # extract feature type information and make it a column
  reg_range$feature_type <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("feature_type",x)] %>% str_remove("feature_type=") )

  # extract regulatory ensembl id and make it a column
  reg_range$feature_id <- reg_snps$feat_attribute %>% str_split(";") %>% 
    sapply(function(x) x[grep("ID=regulatory_region:",x)] %>% str_remove("ID=regulatory_region:") )

  # many entries are repeated, because >1 SNP was found in a regulatory region
  reg_range_uniq <- reg_range %>% unique()
  
  # search for nearest gene downstream
  tab <- matchGenes(reg_range_uniq, mm10_ensGene, skipExons=TRUE, type = "fiveprime") %>% 
    # add biotype and gene symbol  
    left_join(
      mmu_gene_set %>% dplyr::select(gene_id, gene_symbol, gene_biotype), 
      by = c("Geneid"="gene_id")
      ) 

  # add gene information to matching regulatory region
  reg_range_uniq$matching_gene_id <- tab$Geneid
  reg_range_uniq$gene_symbol <- tab$gene_symbol
  reg_range_uniq$gene_biotype <- tab$gene_biotype
  reg_range_uniq$gene_strand <- tab$strand
  reg_range_uniq$distance_to_gene <- tab$distance

return(reg_range_uniq)
})

# name elements in the list
names(downstream_genes) <- names(reg_SNPs_in_RDDs)


# export for later use
lapply(names(downstream_genes), function(nm){
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  downstream_genes[[nm]] %>% write.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl))
})

```

```{r display_feature_tabulation_type_of_feature_by_contrast} 

lapply(1:length(reg_SNPs_in_RDDs), function(i){
  nm <- names(reg_SNPs_in_RDDs)[i]
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl)) %>% 
    group_by(feature_type) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm)
}) %>% 
  bind_rows() %>% 
  reshape2::dcast(feature_type ~ contrast, value.var = "n") %>% 
  dplyr::select(feature_type, c("DUK","DUC","DU6","DU6P","DUhLB","FERT")) %>% 
  kable(digits = 2, 
        caption = "Number of regulatory regions (with >=1 RDD-SNP) by contrast",
        format.args = list(big.mark = ",")) %>% 
  kable_styling(full_width = F)

```

```{r display_feature_tabulation_type_of_feature_by_gene_biotype}
#lapply(1:length(reg_SNPs_in_RDDs), function(i){
#  nm <- names(reg_SNPs_in_RDDs)[i]

contr_nms <- c("DUK","DUC","DU6","DU6P","DUhLB","FERT")

tabs <- lapply(contr_nms, function(nm){
  nm_fl <- paste0("RDD_reg_features_",nm,".csv")
  read.csv(here("00_dashboard/data/regulatory_RDD_SNPs",nm_fl)) %>% 
    group_by(feature_type, gene_biotype) %>% 
    summarise(n=n()) %>% 
    mutate(contrast = nm) %>% 
    reshape2::dcast(gene_biotype ~ feature_type, value.var = "n") %>% 
    filter(!is.na(gene_biotype)) 
}) 

print_kable2 <- function(tab, contr_nm){
  ttl <- paste0("Number of regulatory regions (with >=1 RDD-SNP) by gene biotype (", contr_nm, ")")

  tab %>% 
    kable(digits = 2, caption = ttl, format.args = list(big.mark = ",")) %>% 
    kable_styling(full_width = F)
}


print_kable2(tabs[[1]], contr_nms[1])
print_kable2(tabs[[2]], contr_nms[2])
print_kable2(tabs[[3]], contr_nms[3])
print_kable2(tabs[[4]], contr_nms[4])
print_kable2(tabs[[5]], contr_nms[5])
print_kable2(tabs[[6]], contr_nms[6])

```

### DEG_hits
```{r find_and_display_hits_BETWEEN_RDD_regSNPs_matching_genes_AND_DEGs}

# are there any hits between RDD-regSNPs-matching-genes and DEGs? 
## Only DUK: Atoh8 (FL1_testis_DEGs_hits)
## and DU6: Cd74 (FL1_testis_DEGs_hits)
RDD_regSNPs_matching_genes_x_DEGs <- list.files(here("00_dashboard/data/regulatory_RDD_SNPs"), full.names = TRUE) %>% 
  lapply(function(x){ 
    
    # import RDD-regSNPs-matching-genes data and convert to genomic ranges
    d <- read.csv(x) %>% 
      dplyr::select(seqnames, start, end, feature_type, feature_id, matching_gene_id, gene_symbol, gene_biotype) %>% 
      mutate(
        pop = basename(x) %>% str_remove(".csv") %>% str_remove("RDD_reg_features_")
      ) 
    
    res <- list(
      FL1_testis_DEGs_hits = d %>% filter(gene_symbol %in% FL1_testis_DEGs_gene_symbols),
      FL2_testis_DEGs_hits = d %>% filter(gene_symbol %in% FL2_testis_DEGs_gene_symbols),
      FL1_ovary_DEGs_hits = d %>% filter(gene_symbol %in% FL1_ovary_DEGs_gene_symbols)
      ) %>% 
      bind_rows(.id = "DEG_hit_type")
    
    return(res)
    
    }
  ) %>% 
  bind_rows()


RDD_regSNPs_matching_genes_x_DEGs %>%     
  kable(
    caption = "Hits between RDD-regSNPs-matching-genes and DEGs (only for DU6 and DUK)"
    ) %>% 
  kable_styling(full_width = F)


```

### RDD-genes-hits
```{r find_and_display_hits_BETWEEN_RDD_regSNPs_matching_genes_AND_RDD-genes}

# are there any hits between RDD-regSNPs-matching-genes and RDD-genes? 
## Many of the RDD genes appear downstram RDD-regSNPs ... this is not very informative, 
## ... but is expected, since RDD-regSNPs and RDD-genes were extracted from the same genomic intervals (RDD regions).
lapply(names(RDD_genes), function(nm){
  
  d <- paste0(here("00_dashboard/data/regulatory_RDD_SNPs"), "/RDD_reg_features_", nm, ".csv") %>% 
    read.csv() %>% 
    dplyr::select(-X) 

  data.frame(
    pop = nm,
    n_RDD_genes = RDD_genes[[nm]] %>% length(),
    n_hits = RDD_genes[[nm]] %>% as.data.frame() %>% filter(mcols.gene_id %in% d$matching_gene_id) %>% nrow()
    ) %>% 
    mutate(
      pct_hits = ((n_hits/n_RDD_genes) * 100) %>% round(2)
      ) 
}) %>% 
  bind_rows() %>%     
  kable(
    caption = "Hits between RDD-regSNPs-matching-genes and RDD-genes"
    ) %>% 
  kable_styling(full_width = F)

```




Results Summary
====================================

```{r set_biomart_for_Gviz, eval=FALSE} 
# biomart query done on Monday 05/11/2020

bm <- useMart(host = "www.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL",
              dataset = "mmusculus_gene_ensembl")

```

```{r macro_Gviz_function, eval = FALSE}
vis_goi_macro <- function(ens_gene_id, pos_snps, bp_expand){
  
  
  # merge (full join) fst windows
  win_fst_merged_gr <- lapply(as.character(unique(win_fst_sel_vs_ctrl_prep$contrast)), 
                              function(i){ 
                                x <- win_fst_sel_vs_ctrl_prep %>% 
                                  filter(contrast == i) %>% 
                                  ungroup() %>% 
                                  dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST)
                                
                                names(x)[names(x) == "MEAN_FST"] <- i

                                return(x)
                              }) %>% 
    purrr::reduce(full_join, by = c("CHROM","BIN_START","BIN_END")) %>% 
    dplyr::rename(seqnames=CHROM, start=BIN_START,end=BIN_END) %>% 
    mutate(
      DUK_vs_FZTDU = ifelse(is.na(DUK_vs_FZTDU), 0, DUK_vs_FZTDU),
      DUC_vs_FZTDU = ifelse(is.na(DUC_vs_FZTDU), 0, DUC_vs_FZTDU),
      DU6_vs_FZTDU = ifelse(is.na(DU6_vs_FZTDU), 0, DU6_vs_FZTDU),
      DU6P_vs_FZTDU = ifelse(is.na(DU6P_vs_FZTDU), 0, DU6P_vs_FZTDU),
      DUhLB_vs_FZTDU = ifelse(is.na(DUhLB_vs_FZTDU), 0, DUhLB_vs_FZTDU),
      FERT_vs_FZTDU = ifelse(is.na(FERT_vs_FZTDU), 0, FERT_vs_FZTDU)
    ) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) 


  # get location of gene of interest
  goi_gr <- mmu_gene_set_gr[mmu_gene_set_gr$mcols.gene_id == ens_gene_id]
  
  # expand location 50K to each side
  goi_gr_expanded <- resize(goi_gr, width = width(goi_gr)+(bp_expand), fix = "center")
  
  # define windows containing the gene
  wins <- subsetByOverlaps(win_fst_merged_gr, 
                           goi_gr_expanded, 
                           minoverlap = 1)
  
  
  # define ideogram for chromosome
  itrack <- IdeogramTrack(genome = "mm10", chromosome = as.character(unique(seqnames(wins))))
  
  # track for genomic positions
  gtrack <- GenomeAxisTrack()
  
  wins_track <- DataTrack(wins, name = "Windowed_Fst")
  
  
  # show snp(s) of interest
  ht_snp <- HighlightTrack(trackList = list(wins_track), 
                           start = pos_snps, 
                           end = pos_snps, 
                           lty = "dashed",
                           col = "grey",
                           inBackground=FALSE,
                           chromosome = as.character(seqnames(goi_gr)))

  # show all RDD regions in window
  rdds <- lapply(1:length(distinct_windows_list), function(i){
    
    x <- distinct_windows_list[[i]]
    
    gr_x <- x %>% 
      dplyr::select(CHROM, BIN_START, BIN_END) %>% 
      dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
      makeGRangesFromDataFrame()
    
    gr_xx <- subsetByOverlaps(gr_x, wins, minoverlap = 1)
    
    # convert gr_xx from a range to a point to avoid adjacent low fst windows
    gr_xxx <- GRanges(
      seqnames = seqnames(gr_xx), 
      IRanges(
        start = start(gr_xx) + floor(width(gr_xx)/2),
        end = start(gr_xx) + floor(width(gr_xx)/2)
      )
    )    
    
    # reduce rdd points
    gr_xxx_red <- GenomicRanges::reduce(gr_xxx, min.gapwidth = 25000) #Ranges separated by a gap of at least min.gapwidth positions are not merged
    
    # add population name
    mcols(gr_xxx_red) <- names(distinct_windows_list)[i]
    
    return(gr_xxx_red)

  }) %>% 
    GRangesList() %>% 
    unlist()  
  
 
  #rdd_track <- AnnotationTrack(start = start(rdds), end = end(rdds), 
  #                             chromosome = as.character(unique(seqnames(rdds))), 
  #                             id = unlist(mcols(rdds)), 
  #                             name = "RDDs")

 
  #plotTracks(list(itrack, gtrack, rdd_track, ht_snp), groups = names(mcols(wins)),type = "a", transcriptAnnotation = "symbol", featureAnnotation = "id", fontsize=10)
  plotTracks(list(itrack, gtrack, ht_snp), groups = names(mcols(wins)),type = "a", transcriptAnnotation = "symbol", featureAnnotation = "id")
  
}

```

```{r micro_Gviz_function, eval = FALSE}
vis_goi_micro <- function(ens_gene_id, pos_snps){ 
    
  # define contrasts for convenience
  contrasts <- paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"), "_vs_FZTDU") %>% 
    # sort to match groups to data later in plotracks
    sort()
  
  # get goi location
  query_genes <- mmu_gene_set_gr[mmu_gene_set_gr$mcols.gene_id == ens_gene_id]
  
  # get goi per snp fst scores at each contrast
  per_snp_fst_goi <- lapply(contrasts, function(contrast){
    
    # define fst per snp file for contrast
    fl_nm <- paste0(contrast, ".weir.fst")
    
    # subset snps fst scores overlapping with goi
    tab <- subsetByOverlaps(
      # create per-snp fst granges here to avoid storing into mem  
      vroom(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl_nm), col_types = c(CHROM="c")) %>% 
        mutate(start=POS, end=POS) %>% 
        dplyr::rename(seqnames=CHROM) %>% 
        dplyr::select(seqnames, start, end, WEIR_AND_COCKERHAM_FST) %>% 
        makeGRangesFromDataFrame(keep.extra.columns = TRUE),
      query_genes, 
      minoverlap = 1
    ) %>% 
      as.data.frame() 
    
    names(tab)[names(tab)=="WEIR_AND_COCKERHAM_FST"] <- contrast
    
    return(tab)
    
  }) %>% 
    # merge goi snp fst scores at each contrast   
    purrr::reduce(full_join, by = c("seqnames","start","end")) %>% 
    # select columns to work with  
    dplyr::select(seqnames, start, end, ends_with("_vs_FZTDU")) %>% 
    # turn NAs into 0s, they are in practice the same    
    mutate(
      DUK_vs_FZTDU = ifelse(is.na(DUK_vs_FZTDU), 0, DUK_vs_FZTDU),
      DUC_vs_FZTDU = ifelse(is.na(DUC_vs_FZTDU), 0, DUC_vs_FZTDU),
      DU6_vs_FZTDU = ifelse(is.na(DU6_vs_FZTDU), 0, DU6_vs_FZTDU),
      DU6P_vs_FZTDU = ifelse(is.na(DU6P_vs_FZTDU), 0, DU6P_vs_FZTDU),
      DUhLB_vs_FZTDU = ifelse(is.na(DUhLB_vs_FZTDU), 0, DUhLB_vs_FZTDU),
      FERT_vs_FZTDU = ifelse(is.na(FERT_vs_FZTDU), 0, FERT_vs_FZTDU)
    ) %>% 
    # convert to granges  
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) 
  
  # make a track for per snp fst scores
  snp_fst_track <- DataTrack(per_snp_fst_goi, name = "SNP_Fst")
  
  # get transcripts for GOI
  gene_track <- BiomartGeneRegionTrack(genome = "mm10", 
                                       name = "RDD-Genes",
                                       gene = ens_gene_id, 
                                       biomart = bm)
  
  # show snp(s) of interest
  ht_snp <- HighlightTrack(trackList = list(snp_fst_track, gene_track), 
                           start = pos_snps, 
                           end = pos_snps, 
                           lty = "dashed",
                           col = "grey",
                           inBackground=FALSE,
                           chromosome = as.character(seqnames(query_genes)))
  
  # track for genomic positions
  gtrack <- GenomeAxisTrack()
  
  # plot tracks
  plotTracks(list(gtrack, ht_snp), 
             groups = names(mcols(per_snp_fst_goi)),
             type = c("a","p"),
             transcriptAnnotation = "transcript")
  }

```


```{r pie_chart_function, eval = FALSE}
make_frq_pie_chart <- function(chr, pos, rsid, ref_allele, alt_allele){
  
  hom_ref_gt <- paste(ref_allele,ref_allele,sep="/")
  hom_alt_gt <- paste(alt_allele,alt_allele,sep="/")
  het_gt <- paste(ref_allele,alt_allele,sep="/")  
  
  ttl <- paste0(rsid, ", ", chr,":",pos,"-",pos, ", ",ref_allele,"/",alt_allele)

  
  p <- read.csv(here("00_dashboard/data/RDD_genes/SNPs_in_RDD_genes_GT_counts.tab")) %>% 
    dplyr::select(-X) %>% 
    filter(CHROM == chr & POS == pos) %>% 
    reshape2::melt(id.vars = c("CHROM", "POS", "REF", "ALT")) %>% 
    mutate(
      
      hom_ref = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% 
                         .[grep(hom_ref_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_ref_gt,")"), TRUE))) %>% as.numeric(),
      
      hom_alt = sapply(value, function(x) str_split(x,";") %>% 
                         unlist() %>% .[grep(hom_alt_gt,.)] %>% 
                         str_remove(fixed(paste0("(",hom_alt_gt,")"), TRUE))) %>% as.numeric(),
      
      het = sapply(value, function(x) str_split(x,";") %>% 
                     unlist() %>% 
                     .[grep(het_gt,.)] %>% 
                     str_remove(fixed(paste0("(",het_gt,")"), TRUE))) %>% as.numeric(),
      
      miss = sapply(value, function(x) str_split(x,";") %>% 
                      unlist() %>% 
                      .[grep("./.",., fixed = TRUE)] %>% 
                      str_remove(fixed("(./.)", TRUE))) %>% as.numeric(),
      
      hom_ref = ifelse(is.na(hom_ref), 0, hom_ref),
      hom_alt = ifelse(is.na(hom_alt), 0, hom_alt),
      het = ifelse(is.na(het), 0, het),
  
      ref_frq = (hom_ref*2 + het*1)/((hom_alt+hom_ref+het)*2),
      alt_frq = (hom_alt*2 + het*1)/((hom_alt+hom_ref+het)*2),
      
      variable = str_remove(variable, "_GT_counts"),
      variable = ifelse(variable == "DUK_DUC", "FERT", variable)
      ) %>% 
    dplyr::rename(pop=variable) %>% 
    reshape2::melt(id.vars = c("pop","REF","ALT"), measure.vars = c("ref_frq", "alt_frq"), variable.name = "allele", value.name = "frq") %>% 
    mutate(
      allele = ifelse(allele == "ref_frq", paste0("REF:",REF), paste0("ALT:",ALT)),
      allele = factor(allele, levels = c(paste0("REF:",ref_allele), paste0("ALT:",alt_allele)))
      ) %>% 
    arrange(pop) %>% 
    dplyr::select(pop, allele, frq) %>% 
    mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FERT","FZTDU"))) %>% 
   
    ggplot(., aes(x="", y=frq, fill=allele)) +
      geom_bar(stat="identity", width=1, color="white") +
      coord_polar("y", start=0) +
      theme_void() +# remove background, grid, numeric labels
      theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = 5)) +
      scale_color_manual(values = c("lightblue","darkgrey"), aesthetics = "fill") +
      facet_wrap(~pop, nrow = 1) +
      labs(title = ttl)
  print(p)
   
}

```




Column {.tabset} 
-----------------

### RED-genes

### RDD-genes 

- Looking at the functional annotation of RDD-genes (GOBP) and results of ORA analysis, the following genes could be interesting:

  - **DUK**: 
    - Trp63
      - Relevant GOBPs: negative regulation of intracellular estrogen receptor signaling pathway.
      - SNP effect impacts are only "LOW" or "MODIFIER". Mostly intron variants with Fst > 0.9.
    - ORA showed as significantly enriched the term **"Phospholipase D signaling pathway"** with the RDD-genes (Raf1, Adcy6, Grm8, Tsc1, Ralgds). 
      - There are hints in the literature that this term could influence reproduction.
        - "Role of phospholipase-D and phosphatidic acid in mediating gonadotropin-releasing hormone-induced inhibition of preantral granulosa cell differentiation" https://academic.oup.com/endo/article/135/3/1205/3035944
        - "Role of phospholipase D in regulation of testicular Leydig cell hyperplasia in Sprague–Dawley rats treated with di(2-ethylhexyl) phthalate" https://link.springer.com/article/10.1007/s00204-010-0618-5
        - "Gonadotropin-releasing Hormone Activates Phospholipase D in Ovarian Granulosa Cell" https://www.jbc.org/content/264/20/11762.full.pdf
      - The gene **Tsc1** appears to be involved in reproduction and **contains 3 moderate-missense mutations with Fst>0.95.**. However, none of these non-synonimous mutations has SIFT values indicative of a deleterious effect on the protein (see table below).
        - *"Collectively, these results indicate that Tsc1 has critical roles in preserving the ovarian reserve of oocytes in addition to maintaining the normal architecture of the oviduct for appropriate transport of preimplantation embryos to the uterus."* PMC3690805
        
  - **DUC**: 
    - Pgr:
      - Relevant GOBPs: steroid hormone mediated signaling pathway; progesterone receptor signaling pathway; ovulation from ovarian follicle; female mating behavior,negative regulation of granulosa cell apoptotic process.
      - **SNP effect MODERATE (missense mutation) with high Fst (0.934468).**
      - This is gene was found enriching (ORA) the GOBP terms "hormone-mediated signaling pathway" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
      - Note that P4 is higher in this line than in FZTDU and DUK (Langhammer et al 2014, PMID:24248751)
      - The missense SNP (9  8901222  rs16808511  G  A) SIFT score (VEP query) was 0.04 for two transcripts and therefore deleterious.
    - Pias2:
      - Relevant GOBPs: androgen receptor signaling pathway; negative regulation of androgen receptor signaling pathway.
      - SNP effects MODIFIERS, mainly intron variants, mostly Fst ~0.9.
      - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway", "regulation of dendrite morphogenesis" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
    
    - Rxfp1
       - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway" after ORA analysis.
       - Also, one of the GOBP terms related to reproduction is "parturition".
       - All effects (~200) are modifiers. Mostly intronic variants.
       - Several of those mutations are Fst=1 and the majority are Fst>0.6.
       
    - Yap1   
      -  Relevant GOBPs: regulation of canonical Wnt signaling pathway; response to progesterone; progesterone receptor signaling pathway; positive regulation of intracellular estrogen receptor signaling pathway.
      - This is gene was found enriching the GOBP terms "hormone-mediated signaling pathway" and "intracellular steroid hormone receptor signaling pathway", after ORA analysis.
      - Mutation effects are either low or (mostly) modifiers. Almost all SNPs are > 0.9.
    - "regulation of dendrite morphogenesis"  
      - This term showed up as  significantly enriched by ORA analysis. In contains 3 genes (Pias2, Skor2 and Trpc6).
      - Although this enriched term is not very informative, the RDD-genes therein still might be of interest:
        - Pias2: see above.
        - Skor2: 
          - "In the present study, we identified four significant SNPs after chromosome-wise multiple test correction. The first SNP rs29972765 is located on chr18:77022376 and its nearest gene is Skor2 (SKI family transcriptional corepressor 2). In addition to Skor2, another five genes appeared to locate on the same LD block: Ier3ip1 (immediate early response 3 interacting protein 1), Hdhd2 (haloacid dehalogenase-like hydrolase domain containing 2), Katnal2 (katanin p60 subunit A-like 2), Pias2 (protein inhibitor of activated STAT 2), and St8sia5 (ST8 alpha-N-acetyl-neuraminide alpha-2,8-sialyltransferase 5). " (PMC5015949)
          - GOBP: " negative regulation of BMP signaling pathway" --> (https://www.sciencedirect.com/science/article/pii/S0012160616306686) "BMP signaling regulates multiple reproductive processes across Metazoa."; "Germ line specification, proliferation, migration, and gametogenesis are often BMP-dependent."; "Some BMP ligands play sex-specific reproductive roles in vertebrates.".
          - SNPs effects are mostly modifiers, but there is one moderate-missense effect. However, this mutations has Fst=NA. The other many (>100) modifier effects are Fst > 0.8.
        - Trpc6: 
          - GOBP: "single fertilization" --> "The union of male and female gametes to form a zygote." (http://www.informatics.jax.org/vocab/gene_ontology/GO:0007338).
          - "In conclusion, expression of TRPC1, 2, 3, 4 and 6 isoforms in uterine epithelium were modulated through the stages of the EC and likely as a result of hormonal changes, or due to different expression levels of the estrogen or progesterone receptors throughout the EC." (https://link.springer.com/article/10.1007%2Fs11033-019-04857-w).
          - For these gene there is one moderate-missense effect and 29 modifier effects. However the missense mutatation shows low differentiation (Fst=0.120401). Modifier mutations (~15) are Fst~0.9 and are intron variants.

  - **FERT**: 
    - ORA returned no significant results.
    - Zfp366
      - response to estrogen; negative regulation of intracellular estrogen receptor signaling pathway.
      - SNP effect impacts only "LOW" or "MODIFIER". Negligeble Fst (0.0130608).
    - Eif2ak3
      - "The Unfolded Protein Response Regulates Uterine Myocyte Antioxidant Responsiveness During Pregnancy" PMID: 27733380
      - three missense_variant with Fst > 0.98    

  - **DuhLB**: 
    - Two significant ORA results were returned: "Histidine metabolism" and "beta-Alanine metabolism"
    - For both terms, two RDD genes were used: Aldh3a2 and Aldh3a1. Of which Aldh3a2 could be interensting (see below).
    - Other genes could be interesting as well due to their glucose and fatty acid metabolism related functions:
    - Srsf6
      - response to insulin
      - negative regulation of type B pancreatic cell apoptotic process
      - just modifiers, two mutations with Fst~0.9
    - Aldh3a2
      - lipid metabolic process; fatty acid metabolic process
      - **3 moderate-missense variants with Fst~0.7, two of them SIFT < 0.05 and one SIFT = 0.07**
    - Pfkfb3
      - fructose metabolic process; fructose 2,6-bisphosphate metabolic process; metabolic process; carbohydrate phosphorylation
      - just a modifier with Fst=NA.
    - Slc37a1
      - glucose-6-phosphate transport ; carbohydrate transport
      - just modifiers and low effects. Almost all mutations Fst > 0.9

  - **DU6**
    - ORA returned no significant results.
    - Not clear evidence after inspection of GOBP terms looking for terms related to lipid metabolism, the methionine cycle, and cytochrome P450 regulation, lipid synthesis and cytochrome P450 detoxification (see https://www.biorxiv.org/content/10.1101/2020.05.11.088625v1.full).
    - ORA returned no significant results.
    - However, these genes could be related to body size:
      - Hcrt
        - GOBP: response to starvation, eating behavior.
        - only modifier effects. ~8 SNPs with Fst~0.9.
      - Kat2a	  
        - GOBP: " intracellular distribution of mitochondria", "positive regulation of gluconeogenesis", "response to nutrient levels".
        - There are unfortunatelly no SNPs within this gene.
      - Frmd4a
        - GOBP: positive regulation of protein secretion; negative regulation of protein secretion
        - **It has one missense mutation with high differentiation (~0.97).**
      - Atp11b  
        - GOBP: phospholipid transport; phospholipid translocation
        - **missense_variant with complete differentiation (Fst = 1)**. Tolerated (SIFT ~ 0.55).
        - "Obesity, adipokines, and C-peptide are associated with distinct plasma phospholipid profiles in adult males, an untargeted lipidomic approach" https://www.nature.com/articles/s41598-017-05785-0
        - "Lipidomics Reveals Associations of Phospholipids With Obesity and Insulin Resistance in Young Adults" https://pubmed.ncbi.nlm.nih.gov/26709969/
  
  - **DU6P**
    - ORA returned no significant results.
    - Vwa2
      - GOBP: insulin receptor signaling pathway
        - protein synthesis via mTOR and downstream elements (https://www.cellsignal.de/contents/science-cst-pathways-cellular-metabolism/insulin-receptor-signaling/pathways-irs)
        - "Insulin rapidly activates protein synthesis ... The rapid activation of protein synthesis by insulin is mediated primarily through phosphoinositide 3-kinase. This involves the activation of PKB (protein kinase B) ... PKB also phosphorylates the TSC1 (tuberous sclerosis complex 1)–TSC2 complex to relieve its inhibitory action on the mTOR (mammalian target of rapamycin). Inhibition of mTOR by rapamycin markedly impairs insulin-activated protein synthesis. mTOR controls translation initiation and elongation."  (https://portlandpress.com/biochemsoctrans/article-abstract/34/2/213/63821/Regulation-of-protein-synthesis-by-insulin?redirectedFrom=fulltext)
        - "Insulin signaling regulates protein synthesis and degradation as well as posttranslational modifications at the tissue level and coordinates proteostasis at the organism level...effects of insulin on amino acid flux in skeletal muscle ... The coordination of amino acid flux between tissues by insulin helps orchestrate proteostasis at the organismal level to maintain normal growth. " https://www.sciencedirect.com/science/article/pii/S1550413117303522
      - moderate missense SNP with low differentiation (Fst ~ 0.13)
    - Afap1l2
      - GOBP: positive regulation of epidermal growth factor receptor signaling pathway

        - "In vivo EGF treatment markedly activates asymmetric divisions of dystrophin-deficient satellite cells in mdx mice, increasing progenitor numbers, enhancing regeneration, and restoring muscle strength. Therefore, activating an EGFR-dependent polarity pathway promotes functional rescue of dystrophin-deficient satellite cells and enhances muscle force generation." PMID: 30713094
 
      - **missense_variant with high differentiation Fst~0.95 **. However, it is predicted to be tolerated (SIFT 0.15 - 0.48)

    - Vti1a
      - GOBP: protein transport; intracellular protein transport;voluntary musculoskeletal movement
      - only low and modifier effects
      - many SNPs (intronic) in complete differentiation and Fst>0.9

    - Casp7
      - GOBP: protein processing; **striated muscle cell differentiation**
      - missense_variant with low Fst ~ 0.25

    - Adrb1
      - GOBP: negative regulation of multicellular organism growth; positive regulation of cell growth involved in cardiac muscle cell development
      - only low and modifier effects.

    - Cbfa2t3
      - GOBP: negative regulation of cell population proliferation; positive regulation of proteasomal ubiquitin-dependent protein catabolic process; negative regulation of glycolytic process
      - only modifier and low effects.

    - Aprt
      - GOBP: cellular response to **insulin stimulus**
      - only modifier effects.

    - Cdt1
      - GOBP: positive regulation of protein-containing complex assembly; cell division
      - only modifier effects.

    - Plag1
      - GOBP: multicellular organism growth; organ growth
      - only modifier effects

    - Impad1
      - GOBP: skeletal system development
      - **missense_variant Fst ~ 0.84**

    - Cdh13
      - GOBP: negative regulation of cell population proliferation; positive regulation of smooth muscle cell proliferation;regulation of epidermal growth factor receptor signaling pathway
      - missense_variant low Fst ~ 0.23

    - Hsbp1
      - GOBP: muscle contraction
      - It contains no SNPs


### DUC(Pgr) 
```{r gviz_around_Pgr, eval = FALSE}

png(here("00_dashboard/figures","Pgr_Gviz_macro.png"), res = 300, width = 3500, height = 1500) #width = 3500, height = 1500

vis_goi_macro(ens_gene_id = "ENSMUSG00000031870", pos_snps = 8901222, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Pgr_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Pgr_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000031870", pos_snps = 8901222) 

dev.off()
```

```{r make_frq_pie_chart, eval = FALSE}  
png(here("00_dashboard/figures","Pgr_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 9, pos = 8901222, rsid = "rs16808511", ref_allele = "G", alt_allele = "A")

dev.off()
```

```{r import_gviz_around_Pgr, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Pgr_Gviz_macro.png"), 
    here("00_dashboard/figures","Pgr_Gviz_micro.png"),
    here("00_dashboard/figures","Pgr_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Pgr} 

read.table(here("00_dashboard/data/VEP/rs16808511_chr9_pos8901222_GA_Pgr.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 



### DUK(Tsc1) 
```{r gviz_around_Tsc1, eval = FALSE} 
png(here("00_dashboard/figures","Tsc1_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000026812", pos_snps = c(28671945, 28687118, 28686868), bp_expand=1000000)

dev.off()
``` 

```{r visualize_per_snp_Fst_Tsc1_and_export_micro, eval = FALSE} 

png(here("00_dashboard/figures","Tsc1_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000026812", pos_snps = c(28671945, 28687118, 28686868)) 

dev.off()
```

```{r make_frq_pie_chart_Tsc1, eval = FALSE}

#2:28671945-28671945 rs27210729 A/G
p1 <- make_frq_pie_chart(chr = 2, pos = 28671945, rsid = "rs27210729", ref_allele = "A", alt_allele = "G")

#2:28687118-28687118 rs27210669 A/G
p2 <- make_frq_pie_chart(chr = 2, pos = 28687118, rsid = "rs27210669", ref_allele = "A", alt_allele = "G")


#2:28686868-28686868 rs27210671 C/A
p3 <- make_frq_pie_chart(chr = 2, pos = 28686868, rsid = "rs27210671", ref_allele = "C", alt_allele = "A")


png(here("00_dashboard/figures","Tsc1_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()
```

```{r import_gviz_around_Tsc1, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Tsc1_Gviz_macro.png"), 
    here("00_dashboard/figures","Tsc1_Gviz_micro.png"),
    here("00_dashboard/figures","Tsc1_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Tsc1}

read.table(here("00_dashboard/data/VEP/Tsc1.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
```

### DuhLB(Aldh3a2)  
```{r gviz_around_Aldh3a2, eval = FALSE} 
#11:61248025-61248025
#11:61224649-61224649
#11:61224676-61224676

png(here("00_dashboard/figures","Aldh3a2_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000010025", pos_snps = c(61248025, 61224649, 61224676), bp_expand=1000000)

dev.off()
``` 

```{r visualize_per_snp_Fst_Aldh3a2_and_export_micro, eval = FALSE} 

png(here("00_dashboard/figures","Aldh3a2_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000010025", pos_snps = c(61248025, 61224649, 61224676)) 

dev.off()
```

```{r make_frq_pie_chart_Aldh3a2, eval = FALSE}  

#11:61248025-61248025 11_61248025_G/T
p1 <- make_frq_pie_chart(chr = 11, pos = 61248025, rsid = "rs1133760426", ref_allele = "G", alt_allele = "T")

#11:61224649-61224649 11_61224649_G/A 
p2 <- make_frq_pie_chart(chr = 11, pos = 61224649, rsid = "rs28203629", ref_allele = "G", alt_allele = "A")

#11:61224676-61224676 11_61224676_C/T
p3 <- make_frq_pie_chart(chr = 11, pos = 61224676, rsid = "rs28203628", ref_allele = "C", alt_allele = "T")


png(here("00_dashboard/figures","Aldh3a2_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()
```

```{r import_gviz_around_Aldh3a2, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Aldh3a2_Gviz_macro.png"), 
    here("00_dashboard/figures","Aldh3a2_Gviz_micro.png"),
    here("00_dashboard/figures","Aldh3a2_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Aldh3a2}

read.table(here("00_dashboard/data/VEP/Aldh3a2.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
```



### DU6P(Afap1l2)
```{r gviz_around_Afap1l2, eval = FALSE} 

#19:56944714-56944714	G	A
#ENSMUSG00000025083

png(here("00_dashboard/figures","Afap1l2_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000025083", pos_snps = 56944714, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Afap1l2_and_export_micro, eval = FALSE} 
png(here("00_dashboard/figures","Afap1l2_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000025083", pos_snps = 56944714) 

dev.off()
```

```{r make_frq_pie_chart_Afap1l2, eval = FALSE}  
png(here("00_dashboard/figures","Afap1l2_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 19, pos = 56944714, rsid = "rs215944784", ref_allele = "G", alt_allele = "A")

dev.off()
```

```{r import_gviz_around_Afap1l2, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Afap1l2_Gviz_macro.png"), 
    here("00_dashboard/figures","Afap1l2_Gviz_micro.png"),
    here("00_dashboard/figures","Afap1l2_missense_snp_frq_pie.png")
    )
  )
```

```{r display_VEP_external_query_Afap1l2} 

read.table(here("00_dashboard/data/VEP/Afap1l2.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 

### DU6(Atp11b)
```{r gviz_around_Atp11b, eval = FALSE} 

#3:35809538-35809538	A	C
#ENSMUSG00000037400

png(here("00_dashboard/figures","Atp11b_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000037400", pos_snps = 35809538, bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Atp11b_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Atp11b_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000037400", pos_snps = 35809538) 

dev.off()
```

```{r make_frq_pie_chart_Atp11b, eval = FALSE}  
png(here("00_dashboard/figures","Atp11b_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 1500)

make_frq_pie_chart(chr = 3, pos = 35809538, rsid = "rs30557001", ref_allele = "A", alt_allele = "C")

dev.off()
```

```{r import_gviz_around_Atp11b, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Atp11b_Gviz_macro.png"), 
    here("00_dashboard/figures","Atp11b_Gviz_micro.png"),
    here("00_dashboard/figures","Atp11b_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Atp11b} 

read.table(here("00_dashboard/data/VEP/Atp11b.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 

### FERT(Eif2ak3) 
```{r gviz_around_Eif2ak3, eval = FALSE} 

#6:70844760-70844760	T	C
#6:70844967-70844967	C	A
#6:70893040-70893040	T	A
#ENSMUSG00000031668

png(here("00_dashboard/figures","Eif2ak3_Gviz_macro.png"), res = 300, width = 3500, height = 1500)

vis_goi_macro(ens_gene_id = "ENSMUSG00000031668", pos_snps = c(70844760, 70844967, 70893040), bp_expand=1000000)

dev.off()

```

```{r visualize_per_snp_Fst_Eif2ak3_and_export_micro, eval = FALSE}
png(here("00_dashboard/figures","Eif2ak3_Gviz_micro.png"), res = 300, width = 3500, height = 1500)

vis_goi_micro(ens_gene_id = "ENSMUSG00000031668", pos_snps = c(70844760, 70844967, 70893040)) 

dev.off()
```

```{r make_frq_pie_chart_Eif2ak3, eval = FALSE}  
p1 <- make_frq_pie_chart(chr = 6, pos = 70844760, rsid = "rs262061785", ref_allele = "T", alt_allele = "C")
p2 <- make_frq_pie_chart(chr = 6, pos = 70844967, rsid = "rs248511738", ref_allele = "C", alt_allele = "A")
p3 <- make_frq_pie_chart(chr = 6, pos = 70893040, rsid = "rs37318214", ref_allele = "T", alt_allele = "A")

png(here("00_dashboard/figures","Eif2ak3_missense_snp_frq_pie.png"), res = 300, width = 3500, height = 2000)

gridExtra::grid.arrange(p1,p2,p3, nrow = 3)

dev.off()

dev.off()
```

```{r import_gviz_around_Eif2ak3, fig.align = "center", out.width = "70%"}   
include_graphics(
  c(
    here("00_dashboard/figures","Eif2ak3_Gviz_macro.png"), 
    here("00_dashboard/figures","Eif2ak3_Gviz_micro.png"),
    here("00_dashboard/figures","Eif2ak3_missense_snp_frq_pie.png")
    )
  )
```


```{r display_VEP_external_query_Eif2ak3} 

read.table(here("00_dashboard/data/VEP/Eif2ak3.txt")) %>% 
  dplyr::rename(
   uploaded_variant = V1,
   location = V2,
   allele = V3,
   consequence = V4,
   impact = V5,
   symbol = V6,
   ensemble_gene_id = V7,
   feature_type = V8,
   feature = V9,
   biotype = V10,
   exon = V11,
   cDNA_position = V15,
   CDS_position = V16,
   protein_position = V17,
   aminoacids = V18,
   codons = V19,
   existing_variant = V20,
   feature_strand = V22,
   APRIS = V28,
   SIFT = V29
  ) %>% 
  dplyr::select(-starts_with("V")) %>% 
  kable() %>% 
  kable_styling(full_width = F)
``` 

