---
title: "WGS DU-mice"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---


```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F}
options(scipen = 999)
#library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
#library(reshape2)
library(tidyr)
library(here)
#library(plotly)
#library(DT)
#library(UpSetR)
#library(data.table)
#library(vroom)
#library(RColorBrewer)
#library(knitr)
library(ggcorrplot)
#library(tibble)
#library(png)
#library(grid)
#library(gridExtra)
#library(fastqcr)
#library(GenomicRanges)
library(kableExtra)
#library(ape)
#library(WebGestaltR)
#library(pheatmap)
#library(limma)
#library(ggrepel)
#library(bumphunter) #BiocManager::install("bumphunter")
#library("TxDb.Mmusculus.UCSC.mm10.ensGene") #BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
#library(biomaRt)
#library(Gviz)
```

```{r sample_info_and_genome_length, include=F}
sample_info <- read.csv(here("sample_info/sample_info_batch1_batch2.csv")) %>% 
  dplyr::select(Linie, sample_id) %>% 
  mutate(target_cvg = "30x") %>% 
  bind_rows(
    read.csv(here("sample_info/sample_info_batch3/sample_info.csv")) %>% 
      dplyr::select(Linie,name) %>% 
      dplyr::rename(sample_id = name) %>% 
      mutate(sample_id = str_remove(sample_id, "-S1"), target_cvg = "5x")
  ) %>% 
  mutate(Linie = ifelse(Linie == "HLB", "DUhLB",Linie))

#http://www.ensembl.org/Mus_musculus/Info/Annotation 
genome_length <- 2730871774 # genome length (Golden Path Length	) 
```

```{r win_fst_data_sel_vs_ctrl_zscore, include=F} 

win_fst_sel_vs_ctrl <- lapply(
  
  list.files(here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
             pattern = "windowed.weir.fst$"), 
  
  function(fl){
    
    read.table(
      here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output",fl),
      header = T,
      stringsAsFactors = F
      )
    }
  )

names(win_fst_sel_vs_ctrl) <- list.files(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output"),
  pattern = "windowed.weir.fst$"
  ) %>% str_remove(".windowed.weir.fst")

# prepare data (i.e. zscore transformation) for visualizations
win_fst_sel_vs_ctrl_prep <- lapply(win_fst_sel_vs_ctrl, function(d){
  d %>% 
    # filter by min number of SNPs in window (min 10 SNPs)
    dplyr::filter(N_VARIANTS >= 10) %>% 
    # separate x chromosome
    mutate(is_x = CHROM == "X") %>% 
    # group by autosomes/chrX
    group_by(is_x) %>% 
    # compute zscores
    mutate(z_win_fst_score = scale(MEAN_FST))   
}) %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
    )

# label extreme scores
win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  mutate(
    is_1th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.01) ),
    is_5th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.05) ),
    is_10th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.10) ),
    is_25th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.25) ),
    is_90th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.90) ),
    is_95th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.95) ),
    is_99th_quantile = (z_win_fst_score >= quantile(z_win_fst_score, 0.99) )
      )

#win_fst_sel_vs_ctrl_prep %>% summarise(sum(is_99th_quantile)/n()) # it checks out :)

# fix chr name order
win_fst_sel_vs_ctrl_prep <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  arrange(CHROM, BIN_START)

```

```{r mean_genomewide_fst_pairwise, include=F}

global_fst <- read.table(
  here("batches123_07_selection_analysis/01_win_fst_sel_vs_ctrl/output/global_fst.tab")
  ) %>% 
  dplyr::rename(global_fst = V2) %>% 
  separate(col = V1, sep = "_vs_", into = c("pop1","pop2")) %>% 
  mutate(pop1 = factor(pop1, levels = c("DUK","DUC","DU6","DU6P","DUhLB"))) %>% 
  arrange(pop1)
```


Fst
====================================

Column {.tabset}
-----------------------

### Summary
```{r summary_win_fst}
summary_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  #group_by(contrast, is_x) %>%
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(MEAN_FST),
    median = median(MEAN_FST),
    mean = mean(MEAN_FST),
    q1 = quantile(MEAN_FST, 0.01),
    q5 = quantile(MEAN_FST, 0.05),
    q10 = quantile(MEAN_FST, 0.10),
    q25 = quantile(MEAN_FST, 0.25),
    q90 = quantile(MEAN_FST, 0.90),
    q95 = quantile(MEAN_FST, 0.95),
    q99 = quantile(MEAN_FST, 0.99),
    max = max(MEAN_FST)
    ) #%>% 
  #arrange(is_x)

summary_win_fst %>%
  #mutate(is_x = ifelse(is_x, "X", "Autosome")) %>% 
  #dplyr::rename(chr_type = is_x) %>% 
  kable(digits = 2, caption = "Mean Window Fst") %>% 
  kable_styling(full_width = F) %>% kable_styling(full_width = F)

q_fst <- summary_win_fst %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q1","q5","q10","q25","q90","q95","q99"), 
variable.name = "Quantile", value.name = "Fst")


```

```{r summary_win_zfst}
summary_z_win_fst <- win_fst_sel_vs_ctrl_prep %>% 
  ungroup() %>% 
  group_by(contrast) %>% 
  summarise(
    n_win = n(),
    min = min(z_win_fst_score),
    median = median(z_win_fst_score),
    mean = mean(z_win_fst_score),
    q1 = quantile(z_win_fst_score, 0.01),
    q5 = quantile(z_win_fst_score, 0.05),
    q10 = quantile(z_win_fst_score, 0.10),
    q15 = quantile(z_win_fst_score, 0.15),
    q20 = quantile(z_win_fst_score, 0.20),
    q25 = quantile(z_win_fst_score, 0.25),
    q30 = quantile(z_win_fst_score, 0.30),
    q35 = quantile(z_win_fst_score, 0.35),
    q40 = quantile(z_win_fst_score, 0.40),
    q45 = quantile(z_win_fst_score, 0.45),
    q50 = quantile(z_win_fst_score, 0.50),
    q55 = quantile(z_win_fst_score, 0.55),
    q60 = quantile(z_win_fst_score, 0.60),
    q65 = quantile(z_win_fst_score, 0.65),
    
    q75 = quantile(z_win_fst_score, 0.75),
    q80 = quantile(z_win_fst_score, 0.80),
    q85 = quantile(z_win_fst_score, 0.85),
    q90 = quantile(z_win_fst_score, 0.90),
    q95 = quantile(z_win_fst_score, 0.95),
    q99 = quantile(z_win_fst_score, 0.99),
    max = max(z_win_fst_score)
    ) 

summary_z_win_fst %>% kable(digits = 2, caption = "z(Mean Window Fst)") %>% kable_styling(full_width = F)

q_zfst <- summary_z_win_fst %>% 
  dplyr::select(contrast, paste0("q",c(1,5,10,25,90,95,99))) %>% 
  reshape2::melt(id.vars = c("contrast"), measure.vars = c("q1","q5","q10","q25","q90","q95","q99"), 
                 variable.name = "Quantile", value.name = "zFst")


#saveRDS(summary_z_win_fst, here("00_dashboard","summary_z_win_fst.rds"))
```

### Summary SNPs per window (pending)
```{r, eval = FALSE}

n_var_all <- win_fst_sel_vs_ctrl %>% 
  bind_rows(.id = "contrast") %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
  ) %>% 
  as_tibble() 


n_var_summary_all <- n_var_all %>% 
  group_by(contrast) %>% 
  summarise(n = n(), mean = mean(N_VARIANTS), sd = sd(N_VARIANTS), 
            min = min(N_VARIANTS), median = median(N_VARIANTS), max = max(N_VARIANTS))


n_var_all %>% 
  ggplot(aes(x = N_VARIANTS)) +
    geom_histogram() +
    facet_wrap(~contrast) +
    geom_vline(data = n_var_summary_all, aes(xintercept = mean), color = "red", linetype = "dashed") +
    scale_x_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw()

#---


n_var_all_excl_less_than_10 <- win_fst_sel_vs_ctrl %>% 
  bind_rows(.id = "contrast") %>% 
  filter(N_VARIANTS >= 10) %>% 
  mutate(contrast = factor(
    contrast, 
    levels = c("DUK_vs_FZTDU","DUC_vs_FZTDU","DU6_vs_FZTDU","DU6P_vs_FZTDU","DUhLB_vs_FZTDU","FERT_vs_FZTDU"))
  ) %>% 
  as_tibble() 


n_var_summary__excl_less_than_10 <- n_var_all_excl_less_than_10 %>% 
  group_by(contrast) %>% 
  summarise(n = n(), mean = round(mean(N_VARIANTS), 2), sd = sd(N_VARIANTS), 
            min = min(N_VARIANTS), median = median(N_VARIANTS), max = max(N_VARIANTS))


n_var_all_excl_less_than_10 %>% 
  ggplot(aes(x = N_VARIANTS)) +
    geom_histogram() +
    facet_wrap(~contrast) +
    geom_vline(data = n_var_summary__excl_less_than_10, aes(xintercept = mean), color = "red", linetype = "dashed") +
    scale_x_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw()
```


### Windowed Fst Distribution
```{r, fig.width=10,fig.height=7}

p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = MEAN_FST)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("Mean Window Fst") +
    geom_vline(q_fst, mapping = aes(xintercept = Fst, color = Quantile), linetype = "dashed")

p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = MEAN_FST)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("Mean Window Fst") +
    geom_hline(q_fst, mapping = aes(yintercept = Fst, color = Quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed Fst Distribution (violin-boxplot)
```{r,  fig.width=8,fig.height=10}
ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = MEAN_FST, y = NA), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = MEAN_FST, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw() +
  ylab(NULL) +
  xlab("Mean Window Fst") +
  labs(
    title = "Windowed Fst Distribution"
  ) +
  theme_bw(base_size = 15) +
  theme(
    plot.title = element_text(hjust = 0.5),
  #  strip.text = element_text(size = 7),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    strip.text.y = element_text(size = 10)
  ) +
  facet_wrap(~contrast, ncol = 1, strip.position = "right") +
  #scale_y_discrete(limits = rev(levels(win_fst_sel_vs_ctrl_prep$contrast))) +
  geom_vline(q_fst, mapping = aes(xintercept = Fst, color = Quantile), linetype = "dashed") 
  
```

### Windowed z-Fst Distribution
```{r, fig.width=10,fig.height=7}
p1 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = z_win_fst_score)) +
    geom_histogram(bins = 50) +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    xlab("z(Mean Window Fst)") +
    geom_vline(q_zfst, mapping = aes(xintercept = zFst, color = Quantile), linetype = "dashed")


p2 <- win_fst_sel_vs_ctrl_prep %>% 
  ggplot(aes(x = NA, y = z_win_fst_score)) +
    geom_boxplot(width = 0.2) +
    coord_flip() +
    facet_wrap(~contrast, ncol = 1, strip.position = "right") +
    theme_bw() +
    theme(
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank()
    ) +
    ylab("z(Mean Window Fst)") +
    geom_hline(q_zfst, mapping = aes(yintercept = zFst, color = Quantile), linetype = "dashed")

gridExtra::grid.arrange(p1,p2, nrow = 1)

rm(p1,p2)  
```

### Windowed z-Fst Distribution (violin-boxplot)
```{r for_dashboard_and_manuscript_figure4B, fig.width=8,fig.height=10} 
#png(here("00_dashboard/for_manuscript/figure_4B.png"), res = 300, height = 2900, width = 2300, units = "px")

ggplot(data = win_fst_sel_vs_ctrl_prep) +
  geom_violin(aes(x = z_win_fst_score, y = NA), alpha = 0.4, fill = "red", color = "red") +
  geom_boxplot(aes(x = z_win_fst_score, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
  theme_bw(
    base_size = 20
  ) +
  ylab(NULL) +
  xlab("zFst") +
  labs(
    title = "Windowed z-Fst Distribution"
  ) +
  theme_bw(base_size = 15) +
  theme(
    plot.title = element_text(hjust = 0.5),
  #  strip.text = element_text(size = 7),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    strip.text.y = element_text(size = 10)
  )  +
  geom_vline(q_zfst, mapping = aes(xintercept = zFst, color = Quantile), linetype = "dashed") +
  facet_wrap(~contrast, ncol = 1, strip.position = "right") 

#dev.off()
```

```{r redone_for_manuscript, eval = FALSE}

xx <- q_zfst %>% 
  mutate(contrast2 = str_replace(contrast, "_vs_", "-")) %>% 
  mutate(contrast2 = factor(contrast2, paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"-FZTDU"))) %>% 
  filter(Quantile %in% c("q95","q99")) %>% 
  dplyr::rename(Percentile = Quantile) 

png(here("00_dashboard/for_manuscript/figure_zFst_violin_boxplot.png"), res = 300, height = 1000, width = 1000, units = "px")

win_fst_sel_vs_ctrl_prep %>% 
  mutate(contrast2 = str_replace(contrast, "_vs_", "-")) %>% 
  mutate(contrast2 = factor(contrast2, paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"-FZTDU"))) %>% 
  ggplot(data = .) +
    geom_violin(aes(x = z_win_fst_score, y = NA), alpha = 0.4, fill = "red", color = "red") +
    geom_boxplot(aes(x = z_win_fst_score, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
    theme_bw(
      base_size = 20
    ) +
    ylab(NULL) +
    xlab(bquote(zF[ST])) +
    theme_bw() +
    theme(
      axis.ticks.y = element_blank(),
      axis.text.y = element_blank(),
      axis.text.x = element_text(size = 8), #, vjust = 3
      strip.text.y = element_text(size = 8)
    )  +
    geom_vline(filter(xx, Percentile == "q95"), mapping = aes(xintercept = zFst, color = Percentile), 
               color = "red", linetype = "dashed", alpha = 0.5) +
    geom_vline(filter(xx, Percentile == "q99"), mapping = aes(xintercept = zFst, color = Percentile), 
               color = "blue", linetype = "dashed", alpha = 0.5) +
    facet_wrap(~contrast2, ncol = 3, strip.position = "top") 

dev.off()
```


### Genomewide Manhattan
```{r prepare_data_mahattans, include=F}
manhattan_dat <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
  group_by(contrast, CHROM) %>% 
  arrange(contrast, CHROM, BIN_START) %>% 
  mutate(
    genomic_index = NA, 
    chr_index = 1:n(), 
    center = BIN_START + (BIN_END - BIN_START)/2,
    center_mb = center/1e6,
    chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    )

manhattan_dat$genomic_index <- 1:nrow(manhattan_dat)

ticks <- manhattan_dat %>% 
  group_by(contrast, CHROM) %>% 
  summarise(chr_median = median(genomic_index)) 

```

```{r create_export_genomewide_manhattan, eval = F}
# run this once to create genomewide manhattan and export

# Manhattan of win-fst scores
manhattan_dat %>% 
  #.[sample(nrow(.), 500),] %>%
  ggplot(data = ., mapping = aes(x = genomic_index, y = MEAN_FST, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey","red","blue")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("Mean Window Fst") +
    ggtitle("Windowed Fst - Selected vs Control") +
    geom_hline(filter(q_fst, Quantile %in% c("q99","q95")), mapping = aes(yintercept = Fst, color = Quantile), linetype = "dashed") +
    ggsave(
      filename = "manhattan_genomewide_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      )
    
# Manhattan of z-win-fst scores
manhattan_dat %>% 
  ggplot(data = ., mapping = aes(x = genomic_index, y = z_win_fst_score, color = chr_color)) +
    geom_point(alpha = 0.5) +
    theme_bw() +
    theme(
      legend.position = "none",
      axis.ticks.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    scale_color_manual(values = c("black","grey")) +
    scale_x_continuous(breaks = ticks$chr_median, labels = ticks$CHROM, expand = c(0,0)) + 
    facet_wrap(~contrast, ncol = 1, strip.position = "right", scales = "free_x") +
    xlab(NULL) +
    ylab("z(Mean Window Fst)") +
    geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
    geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
    ggtitle("z(Windowed Fst) - Selected vs Control") +
  ggsave(
      filename = "manhattan_genomewide_z_fst.png", 
      device = "png",
      path = here("00_dashboard/figures"),
      width = 15,
      height = 10,
      units = "in",
      dpi = 300
      ) 
    
    
```

```{r import_genomewide_manhattan, out.width = "70%", fig.align = "center"} 
# import genomewide manhattan created above
knitr::include_graphics(here("00_dashboard/figures","manhattan_genomewide_fst.png"))

knitr::include_graphics(here("00_dashboard/figures","manhattan_genomewide_z_fst.png"))

```

```{r create_chr_manhattans, eval = F} 

# for fst scores
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = MEAN_FST)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst")
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}

# for zfst scores
for(chr in unique(manhattan_dat$CHROM)){

  fl_nm <- paste0("manhattan_genomewide_z_fst_chr",chr,".png")

  manhattan_dat %>% 
    filter(CHROM == chr) %>% 
    ungroup() %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    ggplot(data = ., mapping = aes(x = center_mb, y = z_win_fst_score)) +
      geom_point(alpha = 0.5) +
      theme_bw(base_size = 15) +
      scale_color_manual(values = "darkgrey") +
      scale_x_continuous(breaks = scales::pretty_breaks(n = 15), expand = c(0,0)) + 
      facet_grid(contrast ~ CHROM) +
      xlab("Mb") +
      ylab("Mean Window Fst") +
      geom_hline(filter(q, quantile == "q95"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "blue") +
      geom_hline(filter(q, quantile == "q99"), mapping = aes(yintercept = zFst), linetype = "dashed", color = "red") +
      ggsave(
        filename = fl_nm, 
        device = "png",
        path = here("00_dashboard/figures"),
        width = 15,
        height = 10,
        units = "in",
        dpi = 300
      )
}

```


```{r manhattan_for_manuscript, eval = FALSE} 

data_cum <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  group_by(CHROM) %>% 
  summarise(max_bp = max(center)) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  dplyr::select(CHROM, bp_add)

win_fst_sel_vs_ctrl_prep_manhattan <- win_fst_sel_vs_ctrl_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  inner_join(data_cum, by = "CHROM") %>% 
  mutate(bp_cum = center + bp_add)

axis_set <- win_fst_sel_vs_ctrl_prep_manhattan %>% 
  group_by(CHROM) %>% 
  summarise(center = mean(bp_cum))


xx <- q_fst %>% 
  mutate(contrast2 = str_replace(contrast, "_vs_", "-")) %>% 
  mutate(contrast2 = factor(contrast2, paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"-FZTDU"))) %>% 
  filter(Quantile %in% c("q95","q99")) %>% 
  dplyr::rename(Percentile = Quantile) 

#png(here("00_dashboard/for_manuscript/manhattan_q95_q99_final.png"), res = 300, width = 4000, height = 3000, units = "px")

png(here("00_dashboard/for_manuscript/manhattan_q95_q99_final.png"), res = 300, width = 2000, height = 1600, units = "px")

win_fst_sel_vs_ctrl_prep_manhattan %>% #.[sample(1:nrow(.), 1000),] %>% 
  mutate(contrast2 = str_replace(contrast, "_vs_", "-")) %>% 
  mutate(contrast2 = factor(contrast2, paste0(c("DUK","DUC","DU6","DU6P","DUhLB","FERT"),"-FZTDU"))) %>% 
  ggplot(aes(x = bp_cum, y = MEAN_FST, color = CHROM)) +
    geom_point(alpha = 0.75, size = 0.5) +
    scale_x_continuous(label = axis_set$CHROM, breaks = axis_set$center, expand = c(1/1000,1/1000)) +
    scale_y_continuous(expand = c(1/100,1/100), breaks = seq(0.25,1,0.25)) + #scales::pretty_breaks(n = 5)
    scale_color_manual(values = rep(c("black","darkgrey"), nrow(axis_set)/2) ) +
    ylab(bquote(F[ST])) +
    xlab(NULL) +
    #theme_minimal() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      #axis.text.x = element_text(face = "bold", size = 12), #, vjust = 3
      axis.text.x = element_text(size = 8), #, vjust = 3
      axis.text.y = element_text(size = 8),
      strip.text.y = element_text(size = 8)
    ) +
    geom_hline(filter(xx, Percentile == "q95"), mapping = aes(yintercept = Fst, color = Percentile), 
               color = "red", linetype = "dashed", alpha = 0.5) +
    geom_hline(filter(xx, Percentile == "q99"), mapping = aes(yintercept = Fst, color = Percentile), 
               color = "blue", linetype = "dashed", alpha = 0.5) +
    facet_wrap(~contrast2, nrow = 6, strip.position = "right") 


dev.off()


```


### Genomewide mean per SNP Fst
```{r visualize_pairwise_global_fst} 
#library(pheatmap)

vis_pairwise_fst <- function(contrast_data){

#contrast_data = global_fst

  pops <- contrast_data$pop1 %>% as.character() %>% c(contrast_data$pop2) %>% unique()

  ma <- matrix(rep(NA, length(pops)*length(pops)), nrow = 6, ncol = 6)
  
  rownames(ma) <- pops
  colnames(ma) <- pops

  for(i in 1:nrow(contrast_data)){
    
    col_i <- contrast_data$pop1[i]
    row_i <- contrast_data$pop2[i]
    
    ma[row_i,col_i] <- contrast_data[i,3] # 3rd col should contain values
    ma[col_i,row_i] <- contrast_data[i,3]
    }

  min_fst <- min(ma, na.rm = T)-0.01
  max_fst <- max(ma, na.rm = T)+0.01
  mid_fst <- min_fst + (max_fst - min_fst)/2
  
  colrs <- heat.colors(15)
  
  p <- ggcorrplot(ma, lab=T, type = "lower") +
        #scale_fill_gradient2(limit = c(min_fst,max_fst), low = "blue", high = "red", mid = "white", midpoint = mid_fst)
        scale_fill_gradient2(limit = c(min_fst,max_fst), low = colrs[15], high = colrs[1], mid = colrs[8], midpoint = mid_fst) +
        theme(legend.title = element_blank()) 
  return(p)
  }

#png(here("00_dashboard/for_manuscript/figure_4A.png"), res = 300, height = 2000, width = 2000, units = "px")
#png(here("00_dashboard/for_manuscript/figure_4A_final.png"), res = 300, height = 2000, width = 2000, units = "px")
#png(here("00_dashboard/for_manuscript/pairwise_fst_heatmap.png"), res = 300, height = 1500, width = 1500, units = "px")
vis_pairwise_fst(global_fst)
#dev.off()

```



