---
title: "WGS DU-mice"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
---


```{r results="asis"}
cat("
<style>
caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

```{r, setup, include=F}
options(scipen = 999)
library(flexdashboard)
library(dplyr)
library(stringr)
library(ggplot2)
library(reshape2)
library(tidyr)
library(here)
library(plotly)
library(DT)
library(UpSetR)
library(data.table)
library(vroom)
library(RColorBrewer)
library(knitr)
library(ggcorrplot)
library(tibble)
library(png)
library(grid)
library(gridExtra)
library(fastqcr)
library(GenomicRanges)
library(kableExtra)
library(ape)
library(WebGestaltR)
library(pheatmap)
library(limma)
library(ggrepel)
library(bumphunter) #BiocManager::install("bumphunter")
library("TxDb.Mmusculus.UCSC.mm10.ensGene") #BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(biomaRt)
library(Gviz)
```

```{r ensembl_gene_set, include=F}
# Load gene set
mmu_gene_set <- read.table(here("reference_genome_ensembl/Mus_musculus.GRCm38.93_gene.gtf"), 
                           header = F, stringsAsFactors = F) 
  
names(mmu_gene_set) <- c("seqname","start","end","strand","gene_id","gene_id2","gene_biotype") # as in readme

mmu_gene_set <-mmu_gene_set %>% 
  filter(seqname %in% c(1:19,"X")) %>% # subset for autosomes and X
  dplyr::rename(gene_symbol = gene_id2)

# Turn gene set into a genomics range object
mmu_gene_set_gr <- GRanges(seqnames = mmu_gene_set$seqname,
                           ranges = IRanges(start = mmu_gene_set$start,
                                            end = mmu_gene_set$end),
                           strand = mmu_gene_set$strand,
                           mcols = dplyr::select(mmu_gene_set, gene_id, gene_symbol, gene_biotype))

```

```{r function_run_WebGestaltR, include = F}
run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}


run_WebGestaltR2 <- function( # allow to define fdr threshold
  interestGene, # vector of genes as input
  fdrThr, #fdr between >0 and 1
  topThr = 100, # how many results to write in results table
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "mmusculus",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = fdrThr,
    topThr = topThr,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}
```


Diversity
====================================

Column {.tabset}
-----------------

```{r get_polymorphic_snps_per_population, eval = FALSE}

# Get polymorphic sites and their locations
alt_af_polym <- lapply(
  # get alternative allele frequencies
  list.files(here("batches123_08_LDD_SFS/output"), pattern = "ALTfrq2", full.names = TRUE), 
  function(fl){
    #--fl=list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE)[6]
    # print file name to keep track
    print(fl)
    # load alt frq file
    vroom(fl, col_types = c(CHROM = "c")) %>% 
      # fix columns
      separate(col = POS, sep = " ", into = c("POS", "FRQ"), convert = TRUE) %>% 
      # remove fixed sites
      filter(FRQ > 0 & FRQ < 1)
    }
  )


names(alt_af_polym) <- list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2") %>% 
  str_remove(".ALTfrq2") %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.")


sapply(alt_af_polym, nrow)

alt_af_polym$DU6 %>% head()

alt_af_polym$DU6 %>% tail()

alt_af_polym$DU6$CHROM %>% is.na() %>% any()

```

```{r get_fixed_per_population, eval = FALSE}
# Get polymorphic sites and their locations
alt_af_fixed <- lapply(
  # get alternative allele frequencies
  list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE), 
  function(fl){
    #--fl=list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2", full.names = TRUE)[6]
    # print file name to keep track
    print(fl)
    # load alt frq file
    vroom(fl, col_types = c(CHROM = "c")) %>% 
      # fix columns
      separate(col = POS, sep = " ", into = c("POS", "FRQ"), convert = TRUE) %>% 
      # keep fixed alleles
      filter(FRQ == 1)
    }
  )


names(alt_af_fixed) <- list.files("batches123_08_LDD_SFS/output", pattern = "ALTfrq2") %>% 
  str_remove(".ALTfrq2") %>% 
  str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.")


alt_af_fixed$DU6$CHROM %>% is.na() %>% any()

```

```{r import_prepare_pi_dat, include=FALSE} 
# define files
fls <- list.files(here("batches123_07_selection_analysis/02_pi_ratio/output"), pattern = ".windowed.pi$", full.names = TRUE)

# import and merge data, ignore n variants
pi_dat <- lapply(fls, function(fl){
  
  pop <- fl %>% 
    basename() %>% 
    str_remove("cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.allrecords.") %>% 
    str_remove(".windowed.pi")
  
  d <- fl %>% 
    vroom() %>% 
    #ignore n variants - greedy - to get any region of high diversity, even if there's only one SNP
    dplyr::select(-N_VARIANTS)
  
  names(d)[names(d) == "PI"] <- paste0("PI_",pop)
  
  return(d)
  
  }) %>% 
  purrr::reduce(full_join, by = c("CHROM", "BIN_START","BIN_END"))

# impute NAs with zero
pi_dat[is.na(pi_dat)] <- 0

# reshape data and add zscores

pi_dat <- pi_dat %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END"), value.name = "PI", variable.name = "pop") %>% 
  mutate(pop = str_remove(pop, "PI_")) %>% 
  mutate(is_x = CHROM == "X") %>% 
  group_by(pop, is_x) %>% 
  mutate(z_pi = scale(PI)[,1]) 

head(pi_dat)
```

```{r label_outliers, include = FALSE}

# RTD=regions of highest diversity --> top1%

pi_dat_outliers_labeled <- pi_dat %>% 
  mutate(is_top_1_pct = (z_pi > quantile(z_pi, 0.99)),
         is_top_5_pct = (z_pi > quantile(z_pi, 0.95)))

```


### Intro

* Exploring polymorphic SNPs, in order to characterize the diversity of each population.

* What are the effects for polymorphic SNPs?

* Is there a difference between effects for polymorphic SNPs and fixed SNPs?

* Where are the regions of high diversity located in the genomes of each population?

* What genes appear in regions of high diversity?

### impact_effect_counts
```{r add_snp_effects_tabulate_and_export_table, eval = FALSE}

alt_af_polym_impact_effect_tab <- lapply(alt_af_polym, function(d){
  
  #d=alt_af_polym[[2]] #!!
  #d %>% head() #!!

  d %>% 
    inner_join(
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
        vroom(col_types = c(CHROM = "c")) %>% 
        dplyr::rename(
          effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT"
          ), 
      by = c("CHROM","POS")
    ) %>% 
    group_by(impact, effect) %>% 
    summarise(n=n())
  }
  ) %>% 
  bind_rows(.id="pop") %>% 
  reshape2::dcast(impact + effect ~ pop, value.var = "n")


alt_af_polym_impact_effect_tab %>% 
  write.csv(here("00_dashboard/data/alt_af_polym_impact_effect_tab.csv"))

```

```{r display_alt_af_polym_impact_effect_tab}

read.csv(here("00_dashboard/data/alt_af_polym_impact_effect_tab.csv")) %>% 
  dplyr::select(-X) %>% 
  datatable(caption = "Impact and Effects Polymorphic SNPs", options = list(pageLength = 30))

```

### Correlation_effects_fixed_vs_polymorphic
```{r add_fixed_snp_effects_tabulate_and_export_table, eval = FALSE}
alt_af_fixed_impact_effect_tab <- lapply(alt_af_fixed, function(d){
  
  #d=alt_af_polym[[2]] #!!
  #d %>% head() #!!

  d %>% 
    inner_join(
      here("batches123_05_annotation/output/cohort_biallelicSNPs_VQSR95_PASS_withmissingness.filtered.ann.tab") %>% 
        vroom(col_types = c(CHROM = "c")) %>% 
        dplyr::rename(
          effect = "ANN[*].EFFECT", impact = "ANN[*].IMPACT"
          ), 
      by = c("CHROM","POS")
    ) %>% 
    group_by(impact, effect) %>% 
    summarise(n=n())
  }
  ) %>% 
  bind_rows(.id="pop") %>% 
  reshape2::dcast(impact + effect ~ pop, value.var = "n")
```

```{r calculate_pct_effect_and_export_cor_plots, eval = FALSE}

get_pct <- function(d){
  
  tmp <- d %>% 
    dplyr::select(DUK,DUC,DU6,DU6P,DUhLB,FZTDU) %>% 
    lapply(function(x){
  
      #x=alt_af_fixed_impact_effect_tab$DUK
      
      total_cts <- sum(x, na.rm = TRUE)
      
      pct <- (x/total_cts)*100
  
    }) %>% 
    bind_cols()
    
  alt_af_fixed_impact_effect_tab_pct <- bind_cols(
    
    d %>% dplyr::select(impact, effect),
    tmp
  ) 
  
  return(alt_af_fixed_impact_effect_tab_pct)
}

alt_af_polym_impact_effect_pct <- get_pct(alt_af_polym_impact_effect_tab)
alt_af_fixed_impact_effect_tab_pct <- get_pct(alt_af_fixed_impact_effect_tab)


make_export_cor_plot <- function(pop){
  #pop="DUK"
  
  x <- alt_af_polym_impact_effect_pct %>%.[,c("impact", "effect",pop)]
  names(x)[names(x) == pop] <- "pct_polymorphic"
  
  y <- alt_af_fixed_impact_effect_tab_pct %>%.[,c("impact", "effect",pop)]
  names(y)[names(y) == pop] <- "pct_fixed"

  d <- inner_join(x, y, by = c("impact","effect")) %>% 
    mutate(nrows = 1:nrow(.), lab_set = nrows%%2)

  p <- ggplot(d, aes(x = pct_polymorphic, y = pct_fixed)) +
    geom_point() +
    geom_abline(color = "red", slope = 1) +
    theme_bw(base_size = 15) +
    ggrepel::geom_label_repel(data = filter(d, lab_set == 0),  aes(label = effect, fill = impact) , 
                              xlim = c(50,NA), direction = "y") +
    ggrepel::geom_label_repel(data = filter(d, lab_set == 1),  aes(label = effect, fill = impact) , 
                              xlim = c(25,NA), direction = "y") +
    ggtitle(pop) +
    theme(plot.title = element_text(hjust = 0.5))
  
  fl_nm <- paste0("scatter_plot_cor_pct_effects_fixed_vs_polym_",pop,".png")
  
  png(here("00_dashboard/figures", fl_nm), res = 300, width = 3500, height = 3500, units = "px")
  print(p)
  dev.off()
  }


# export each plot 
make_export_cor_plot("DUK")
make_export_cor_plot("DUC")
make_export_cor_plot("DU6")
make_export_cor_plot("DU6P")
make_export_cor_plot("DUhLB")
make_export_cor_plot("FZTDU")
```

```{r display_cor_pct_effect_plots_fixed_vs_polym, out.width="50%", fig.align="center"}

include_graphics(
  list.files(here("00_dashboard/figures"), full.names = TRUE) %>% .[grep("scatter_plot_cor_pct_effects_fixed_vs_polym_", .)]
  )

```

### Genes
```{r tabulate_genes_with_polymorphic_SNPs_and_export, eval=FALSE}

alt_af_polym_gn_cts <- lapply(alt_af_polym, function(d){

  gr <- d %>% 
	  mutate(tmp = POS) %>% 
	  dplyr::rename(seqnames = CHROM, start = tmp, end = POS) %>% 
	  dplyr::select(seqnames, start, end) %>% 
	  makeGRangesFromDataFrame()

  subsetByOverlaps(mmu_gene_set_gr, gr, minoverlap = 1) %>% 
    as.data.frame() %>% 
    group_by(mcols.gene_biotype) %>% 
    summarise(n = n()) %>% 
    janitor::adorn_totals()
  }) %>% 
  bind_rows(.id="pop") %>% 
  dplyr::rename(gene_biotype = mcols.gene_biotype) %>% 
  reshape2::dcast(gene_biotype ~ pop, value.var = "n")

write.csv(alt_af_polym_gn_cts, here("00_dashboard/data/alt_af_polym_gn_cts.csv"))

```

```{r display_genes_with_polymorphic_SNPs_counts}
read.csv(here("00_dashboard/data/alt_af_polym_gn_cts.csv")) %>% 
  dplyr::select(-X) %>% 
  datatable(caption = "Genes Types with Polymorphic SNPs", options = list(pageLength = 100))

```

### Pi_distribution
```{r pi_distribution_histogram, eval = FALSE}

p1 <- pi_dat %>% 
  ungroup() %>% 
  dplyr::select(-is_x,-z_pi) %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  as_tibble() %>% 
  ggplot(aes(x = PI)) +
    geom_histogram() +
    theme_bw(base_size = 15) +
    facet_wrap(~pop, scale = "free_y", nrow = 6, strip.position = "right") +
    xlab(expression(pi)) 



p2 <- pi_dat %>% 
  ungroup() %>% 
  dplyr::select(-is_x,-PI) %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  as_tibble() %>% 
  ggplot(aes(x = z_pi)) +
    geom_histogram() +
    theme_bw(base_size = 15) +
    facet_wrap(~pop, scale = "free_y", nrow = 6, strip.position = "right") +
    xlab(bquote(z ~ pi ))


png(here("00_dashboard/figures/histogram_pi_distribution.png"), res = 300, units = "px", height = 2000, width = 2000)
ggpubr::ggarrange(p1,p2,ncol = 2, nrow = 1)
dev.off()

```

```{r zpi_quantiles, include=FALSE}

z_pi_quantiles <- pi_dat %>% 
  summarise(q95 = quantile(z_pi, 0.95),
            q99 = quantile(z_pi, 0.99)) %>% 
  arrange(is_x, pop) %>% 
  mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
  dplyr::select(-is_x) %>% 
  ungroup() %>% 
  reshape2::melt(id.vars = c("pop","chr_class"), variable.name = "quantile", value_name = "quantile_value") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU")))
  
z_pi_quantiles
```

```{r pi_distribution_boxplot_violin, eval=FALSE}

png(here("00_dashboard/figures/violin_boxplot_pi_distribution.png"), res = 300, units = "px",height = 2900, width = 2300)

pi_dat %>% 
  #ungroup() %>% 
  #dplyr::select(-is_x, -PI) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop"), variable.name = "pi_class") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
  ggplot(data = .) + 
    geom_violin(aes(x = z_pi, y = NA), alpha = 0.4, fill = "red", color = "red") +
    geom_boxplot(aes(x = z_pi, y = NA), width = 0.2, color = "blue", outlier.colour = "black") +
    ylab(NULL) +
    xlab(NULL) +
    labs(
      title = "Windowed Pi/z_Pi Distribution"
    ) +
    theme_bw(base_size = 15) +
    theme(
      plot.title = element_text(hjust = 0.5),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      strip.text.y = element_text(size = 10)
    ) +
    facet_grid(pop~chr_class, scale = "free") +
    geom_vline(data = z_pi_quantiles, aes(xintercept = value, color = quantile), linetype = "dashed")
    #facet_wrap(~pop, ncol = 1, strip.position = "left")

dev.off()
```

```{r display_pi_distribution_plots, out.width="50%", fig.align="center"}

include_graphics(c(here("00_dashboard/figures/histogram_pi_distribution.png") , 
                   here("00_dashboard/figures/violin_boxplot_pi_distribution.png")))

```

```{r display_extremely_diverse_windows_duk_duc}

x_chr_zpi_above_20 <- pi_dat_outliers_labeled %>% 
  filter(z_pi > 20) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov <- findOverlaps(query = x_chr_zpi_above_20, subject =  mmu_gene_set_gr, minoverlap = 1)

bind_cols(
  x_chr_zpi_above_20[queryHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(window_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(window_locus, pop, z_pi),
  mmu_gene_set_gr[subjectHits(ov)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol,  mcols.gene_biotype)
  ) %>% 
  kable(caption = "Genes extreme regions X chr - DUK DUC") %>% 
  kable_styling(full_width = F)

```




### Genomewide_Manhattan
```{r make_and_export_zPi_manhattan, eval = FALSE}

make_z_pi_manhattan <- function(mouse_line){

  #mouse_line="FZTDU"
  
  manhattan_pi_dat <- pi_dat %>% 
    filter(pop == mouse_line) %>% 
    mutate(CHROM = factor(CHROM, levels = c(1:19,"X"))) %>% 
    #--mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
    group_by(pop, CHROM) %>% 
    arrange(pop, CHROM, BIN_START) %>% 
    mutate(
      genomic_index = NA, 
      chr_index = 1:n(), 
      center = BIN_START + (BIN_END - BIN_START)/2,
      center_mb = center/1e6,
      chr_color = ifelse(CHROM %in% seq(1,19,2), "black", "grey")
    ) %>% 
    mutate(chr_class = ifelse(is_x, "X", "Autosome")) %>% 
    ungroup() %>% 
    group_by(pop, is_x) %>% 
    mutate(is_top_5_pct = (z_pi > quantile(z_pi, 0.95)),
           is_top_1_pct = (z_pi > quantile(z_pi, 0.99))) 
  
  manhattan_pi_dat$genomic_index <- 1:nrow(manhattan_pi_dat)
  
  
  chr_offset_un <- 1000
  len_out <- length(unique(manhattan_pi_dat$CHROM))-1
  chr_offset <- c(0, seq(chr_offset_un, chr_offset_un*len_out, length.out = len_out))
  names(chr_offset) <- unique(manhattan_pi_dat$CHROM)
  
  
  manhattan_pi_dat <- lapply(unique(manhattan_pi_dat$CHROM), function(chr){
    #chr_offset[chr] * 100
    manhattan_pi_dat %>% 
      filter(CHROM == chr) %>% 
      mutate(genomic_index_offset = genomic_index + chr_offset[chr]) 
    }) %>% 
    bind_rows()
  
  ticks <- manhattan_pi_dat %>% 
    group_by(pop, CHROM) %>% 
    summarise(chr_min = min(genomic_index_offset), 
              chr_max = max(genomic_index_offset),
              chr_sep = chr_max + chr_offset_un/2,
              chr_center = chr_min + (chr_max-chr_min)/2) 
  
  
  #d <- manhattan_pi_dat %>% .[sample(1:nrow(.), 1000),]
  d <- manhattan_pi_dat
  
  p <- d %>% 
    filter(!is_top_5_pct & !is_top_1_pct) %>% 
    ggplot(data = ., aes(x = genomic_index_offset, y = z_pi)) +
      geom_point(alpha = 0.5, color = "darkgrey", size = 1) +
    
      geom_point(filter(d, is_top_5_pct & !is_top_1_pct),
                 mapping = aes(x = genomic_index_offset, y = z_pi), 
                 alpha = 0.5, color = "blue", size = 1) +
    
      geom_point(filter(d, is_top_1_pct),
                 mapping = aes(x = genomic_index_offset, y = z_pi), 
                 alpha = 0.5, color = "red", size = 1) +
  
      theme_bw(base_size = 15) +
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank()
      ) +
      scale_x_continuous(breaks = ticks$chr_center, labels = ticks$CHROM, expand = c(0.01,0.01)) + 
      scale_y_continuous(expand = c(0.05,0.05), labels = scales::number_format(accuracy = 1)) +
      geom_vline(filter(ticks, CHROM != "X"), mapping = aes(xintercept = chr_sep), color = "black", linetype = "dashed")  +
      xlab(NULL) +
      ylab("Window zPi") +
      facet_wrap(~pop, strip.position = "right")

    return(p)
  
}


png(here("00_dashboard/figures/manhattan_pi/manhattan_genomewide_zPi.png"), units = "px", res = 300, height = 3000, width = 5000)
grid.arrange(make_z_pi_manhattan("DUK"), make_z_pi_manhattan("DUC"),
             make_z_pi_manhattan("DU6"), make_z_pi_manhattan("DU6P"),
             make_z_pi_manhattan("DUhLB"), make_z_pi_manhattan("FZTDU"),
             ncol = 1)
dev.off()


```

```{r display_genomewide_zPi_manhattan, out.width="50%", fig.align="center"}
include_graphics(here("00_dashboard/figures/manhattan_pi/manhattan_genomewide_zPi.png"))
```

### Chr_Manhattan
```{r make_and_export_zPi_chr_manhattan, eval = FALSE} 

make_z_pi_manhattan_chr<- function(chr){ 
  #--mouse_line="FZTDU"
  #--chr=1
  
  manhattan_pi_dat <- pi_dat %>% 
    filter(CHROM == chr) %>% 
    mutate(CHROM = paste0("Chromosome ", CHROM)) %>% 
    mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
    mutate(
      center = BIN_START + (BIN_END - BIN_START)/2,
      center_mb = center/1e6
    ) %>% 
    group_by(is_x) %>% 
    mutate(is_top_5_pct = (z_pi > quantile(z_pi, 0.95)),
           is_top_1_pct = (z_pi > quantile(z_pi, 0.99))) 
  
  #d <- manhattan_pi_dat %>% .[sample(1:nrow(.), 1000),]
  d <- manhattan_pi_dat
  
  p <- d %>% 
    filter(!is_top_5_pct & !is_top_1_pct) %>% 
    ggplot(data = ., aes(x = center_mb, y = z_pi)) +
      geom_point(alpha = 0.5, color = "darkgrey", size = 1) +
    
      geom_point(filter(d, is_top_5_pct & !is_top_1_pct),
                 mapping = aes(x = center_mb, y = z_pi), 
                 alpha = 0.5, color = "blue", size = 1) +
    
      geom_point(filter(d, is_top_1_pct),
                 mapping = aes(x = center_mb, y = z_pi), 
                 alpha = 0.5, color = "red", size = 1) +
  
      theme_bw(base_size = 12) +
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank()
      ) +
      scale_y_continuous(expand = c(0.05,0.05), labels = scales::number_format(accuracy = 1)) +
      xlab("Mb") +
      ylab(bquote(z ~ pi )) +
      facet_grid(pop~CHROM)

    return(p)
}


lapply(c(1:19,"X"), function(chr){ 
  #--chr=1
  fl_nm <- paste0("manhattan_genomewide_zPi_chr", chr,".png")
  png(here("00_dashboard/figures/manhattan_pi", fl_nm), units = "px", res = 300, height = 2000, width = 2500)
  make_z_pi_manhattan_chr(chr) %>% print()
  dev.off()
})


#make_z_pi_manhattan_chr("2") 

```

```{r display_genomewide_zPi_chr_manhattan, out.width="50%", fig.align="center"}

fls <- sapply(c(1:19,"X"), function(x){
  fl_nm <- paste0("manhattan_genomewide_zPi_chr", x, ".png")
  fl_nm_full <- here("00_dashboard/figures/manhattan_pi", fl_nm)
  return(fl_nm_full)
})

include_graphics(fls)
```



### Genes_RHD
```{r find_genes_in_diversity_outliers_top_1_pct, include = FALSE} 

outliers_gr_top_1_pct <- pi_dat_outliers_labeled %>% 
  ungroup() %>% 
  dplyr::select(CHROM, BIN_START, BIN_END,is_top_1_pct, pop) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop")) %>% 
  filter(is_top_1_pct) %>% 
  dplyr::select(-is_top_1_pct) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov_top_1_pct <- findOverlaps(query = outliers_gr_top_1_pct, subject = mmu_gene_set_gr, minoverlap = 1)


outliers_and_genes_1_pct <- bind_cols(
  outliers_gr_top_1_pct[queryHits(ov_top_1_pct)] %>% 
    as.data.frame() %>% 
    dplyr::select(seqnames, start, end, pop),
  
  mmu_gene_set_gr[subjectHits(ov_top_1_pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol, mcols.gene_biotype) %>% 
    dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, biotype = mcols.gene_biotype) 
  
) 
```

```{r display_gene_counts_in_highest_diverse_regions_top_1_pct}
outliers_and_genes_1_pct %>% 
  ungroup() %>% 
  #filter(outlier_class == "is_top_1_pct") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  group_by(pop, biotype) %>% 
  summarise(n = n()) %>% 
  reshape2::dcast(biotype ~ pop, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  kable(caption = "Gene Types in top 1% most diverse regions") %>% 
  kable_styling(full_width = F)
  #datatable(caption = "Gene Types in top 1% most diverse regions", options = list(pageLength = 100))
```

```{r find_genes_in_diversity_outliers_top_5_pct, include = FALSE}

outliers_gr_top_5_pct <- pi_dat_outliers_labeled %>% 
  ungroup() %>% 
  dplyr::select(CHROM, BIN_START, BIN_END,is_top_5_pct, pop) %>% 
  #reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END","pop")) %>% 
  filter(is_top_5_pct) %>% 
  dplyr::select(-is_top_5_pct) %>% 
  dplyr::rename(seqnames = CHROM, start = BIN_START, end = BIN_END) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

ov_top_5_pct <- findOverlaps(query = outliers_gr_top_5_pct, subject = mmu_gene_set_gr, minoverlap = 1)


outliers_and_genes_5_pct <- bind_cols(
  outliers_gr_top_5_pct[queryHits(ov_top_5_pct)] %>% 
    as.data.frame() %>% 
    dplyr::select(seqnames, start, end, pop),
  
  mmu_gene_set_gr[subjectHits(ov_top_5_pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames, ":", start, "-", end)) %>% 
    dplyr::select(gene_locus, mcols.gene_id, mcols.gene_symbol, mcols.gene_biotype) %>% 
    dplyr::rename(gene_id = mcols.gene_id, gene_symbol = mcols.gene_symbol, biotype = mcols.gene_biotype) 
  
) 
```

```{r display_gene_counts_in_highest_diverse_regions_top_5_pct}
outliers_and_genes_5_pct %>% 
  ungroup() %>% 
  #filter(outlier_class == "is_top_5_pct") %>% 
  mutate(pop = factor(pop, levels = c("DUK","DUC","DU6","DU6P","DUhLB","FZTDU"))) %>% 
  group_by(pop, biotype) %>% 
  summarise(n = n()) %>% 
  reshape2::dcast(biotype ~ pop, value.var = "n") %>% 
  janitor::adorn_totals() %>% 
  kable(caption = "Gene Types in top 5% most diverse regions") %>% 
  kable_styling(full_width = F)
  #datatable(caption = "Gene Types in top 1% most diverse regions", options = list(pageLength = 100))
```


### RHD_ORA_top1pct 
```{r RHD_GOBP_top1pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP") 
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
)  #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/GOBP")
)
```

```{r RHD_KEGG_top1pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top1pct/KEGG")
)
```

```{r RHD_GOBP_import_results_top1pct, include=FALSE}

RHD_GOBP_res_top1pct <- here("00_dashboard/data/WebgestaltR_RHD_top1pct/GOBP") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) 


RHD_GOBP_res_top1pct[[4]] # replace this empty element for row binding
RHD_GOBP_res_top1pct[[4]] <- RHD_GOBP_res_top1pct[[1]]
RHD_GOBP_res_top1pct[[4]] <- lapply(RHD_GOBP_res_top1pct[[4]], function(x){
  x <- NA
  }) %>% 
  as.data.frame() %>% 
  mutate(pop = "DUhLB")

RHD_GOBP_res_top1pct <- RHD_GOBP_res_top1pct %>% bind_rows() 


#RHD_GOBP_res %>% datatable(caption = "Enriched GOBP terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))
```

```{r RHD_KEGG_import_results_top1pct, include=FALSE}

RHD_KEGG_res_top1pct <- here("00_dashboard/data/WebgestaltR_RHD_top1pct/KEGG") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) %>% 
  bind_rows()

#RHD_KEGG_res %>% bind_rows() %>% datatable(caption = "Enriched KEGG terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))

```

```{r make_and_export_RHD_ORA_sig_heatmaps_top1pct, eval = FALSE} 
RHD_GOBP_res_sig_top1pct <- RHD_GOBP_res_top1pct %>% 
  mutate(is_sig = (FDR <= 0.1 ) %>% as.integer()) %>% 
  filter(!is.na(description)) %>% 
  reshape2::dcast(description ~ pop, value.var = "is_sig") %>% 
  column_to_rownames(., var = "description")

RHD_GOBP_res_sig_top1pct[is.na(RHD_GOBP_res_sig_top1pct)] <- 0

png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct.png"), res = 300, height = 3000, width = 3000, units = "px")
pheatmap(RHD_GOBP_res_sig_top1pct, main = "RHD_GOBP_res_sig_top1pct")
dev.off()

RHD_KEGG_res_sig_top1pct <- RHD_KEGG_res_top1pct %>% 
  mutate(is_sig = (FDR <= 0.1 ) %>% as.integer()) %>% 
  filter(!is.na(description)) %>% 
  reshape2::dcast(description ~ pop, value.var = "is_sig") %>% 
  column_to_rownames(., var = "description") 

RHD_KEGG_res_sig_top1pct[is.na(RHD_KEGG_res_sig_top1pct)] <- 0

png(here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct.png"), res = 300, height = 3000, width = 3000, units = "px")
pheatmap(RHD_KEGG_res_sig_top1pct, main = "RHD_KEGG_res_sig_top1pct")
dev.off()
```

```{r RHD_GOBP_FDR1, eval=FALSE}

RHD_GOBP_res_sig_top1pct_terms <- RHD_GOBP_res_sig_top1pct %>% rownames()

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")
)

setdiff(
  RHD_GOBP_res_sig_top1pct_terms,
  here(
    "00_dashboard/data/WebGestaltR_RHD_lenient/GOBP/Project_WebGestaltR_FZTDU/geneontology_Biological_Process_sig_results.csv"
    ) %>% 
    vroom() %>%
    .$description
  )



```

```{r make_heatmap_enrichmentRatio_GOBP, eval = FALSE}

dir_nm <- here("00_dashboard/data/WebGestaltR_RHD_lenient/GOBP")


prep_ora_dat <- function(pop){
  #pop="DUK"

  nm <- paste0("Project_WebGestaltR_",pop)

  d <- file.path(dir_nm, nm, "geneontology_Biological_Process_sig_results.csv") %>% 
    vroom() %>% 
    dplyr::select(description, enrichmentRatio) %>% 
    filter(description %in% RHD_GOBP_res_sig_top1pct_terms)
  
  names(d)[names(d) == "enrichmentRatio"] <- pop
  
  d
  
}

png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct_enrichmentRatio.png"), 
    res = 300, height = 3000, width = 3000, units = "px")


full_join(prep_ora_dat("DUK"),prep_ora_dat("DUC"),by = "description") %>% 
  full_join(prep_ora_dat("DU6"), by = "description") %>% 
  full_join(prep_ora_dat("DU6P"), by = "description") %>% 
  full_join(prep_ora_dat("DUhLB"), by = "description") %>% 
  full_join(prep_ora_dat("FZTDU"), by = "description") %>% 
  reshape2::melt(id.vars = "description") %>% 
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  reshape2::dcast(description ~ variable, value.var = "value") %>% 
  column_to_rownames(var = "description") %>% 
  #as.matrix() %>% 
  pheatmap(., main = "enrichmentRatio - GOBP")
  
dev.off()



```

```{r RHD_KEGG_FDR1, eval=FALSE}

RHD_KEGG_res_sig_top1pct_terms <- RHD_KEGG_res_sig_top1pct %>% rownames()

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
) 

run_WebGestaltR2(
  fdrThr = 1,
  interestGene=outliers_and_genes_1_pct$gene_id[outliers_and_genes_1_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")
)

```

```{r make_heatmap_enrichmentRatio_KEGG, eval = FALSE}

dir_nm <- here("00_dashboard/data/WebGestaltR_RHD_lenient/KEGG")


prep_ora_dat <- function(pop){
  #pop="DUK"

  nm <- paste0("Project_WebGestaltR_",pop)

  d <- file.path(dir_nm, nm, "pathway_KEGG_sig_results.csv") %>% 
    vroom() %>% 
    dplyr::select(description, enrichmentRatio) %>% 
    filter(description %in% RHD_KEGG_res_sig_top1pct_terms)
  
  names(d)[names(d) == "enrichmentRatio"] <- pop
  
  d
  
}

png(here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct_enrichmentRatio.png"), 
    res = 300, height = 3000, width = 3000, units = "px")


full_join(prep_ora_dat("DUK"),prep_ora_dat("DUC"),by = "description") %>% 
  full_join(prep_ora_dat("DU6"), by = "description") %>% 
  full_join(prep_ora_dat("DU6P"), by = "description") %>% 
  full_join(prep_ora_dat("DUhLB"), by = "description") %>% 
  full_join(prep_ora_dat("FZTDU"), by = "description") %>% 
  reshape2::melt(id.vars = "description") %>% 
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  reshape2::dcast(description ~ variable, value.var = "value") %>% 
  column_to_rownames(var = "description") %>% 
  #as.matrix() %>% 
  pheatmap(., main = "enrichmentRatio - KEGG")
  
dev.off()



```

```{r display_RHD_ORA_sig_heatmaps_top1pct, out.width="50%", fig.align="center"}

include_graphics(
  c(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct.png") , 
    here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct.png"),
    here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top1pct_enrichmentRatio.png"),
    here("00_dashboard/figures/diversity/RHD_KEGG_res_sig_top1pct_enrichmentRatio.png"))
  )

```

#### **Description**

Overrepresentation analysis of RDD genes. 

First two figures represent GOBP and KEGG results. In there, terms are either 1 (signficant) or 0 (not-significant).

Figures 3 and 4 show enrichment scores accross all significant terms shown in figures 1 and 2. Enrichment scores were collected by re-running ORA setting FDR=1 and then subsetting by the score already detected as sifnificant (FDR <= 0.1).



### RHD_ORA_top5pct
```{r RHD_GOBP_top5pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
) 

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/GOBP")
)
```

```{r RHD_KEGG_top5pct, eval = FALSE}

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUK"],
  projectName="WebGestaltR_DUK",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUC"],
  projectName="WebGestaltR_DUC",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
) #No significant gene set is identified based on FDR 0.1!

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6"],
  projectName="WebGestaltR_DU6",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DU6P"],
  projectName="WebGestaltR_DU6P",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "DUhLB"],
  projectName="WebGestaltR_DUhLB",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

run_WebGestaltR(
  interestGene=outliers_and_genes_5_pct$gene_id[outliers_and_genes_5_pct$pop == "FZTDU"],
  projectName="WebGestaltR_FZTDU",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_dashboard/data/WebGestaltR_RHD_top5pct/KEGG")
)

```

```{r RHD_GOBP_import_results_top5pct, include=FALSE}

RHD_GOBP_res_top5pct <- here("00_dashboard/data/WebgestaltR_RHD_top5pct/GOBP") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  }) %>% 
  bind_rows() 


#RHD_GOBP_res %>% datatable(caption = "Enriched GOBP terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))
```

```{r RHD_KEGG_import_results_top5pct, include=FALSE}

RHD_KEGG_res_top5pct <- here("00_dashboard/data/WebgestaltR_RHD_top5pct/KEGG") %>% 
  list.files(full.names = TRUE) %>% 
  lapply(function(x){
    
    pop <- basename(x) %>% str_remove("Project_WebGestaltR_")
    
    fl <- x %>% list.files(full.names = TRUE, pattern = "enrichment_results") 
    
    if(length(fl) == 0){
      
      paste0("There are no results for: ", pop)
    }else{
      
      fl %>% 
        vroom() %>% 
        mutate(pop = pop) %>% 
        dplyr::select(pop, everything())
    }
  })

RHD_KEGG_res_top5pct[[3]] # replace this empty element for row binding
RHD_KEGG_res_top5pct[[3]] <- RHD_KEGG_res_top5pct[[1]]
RHD_KEGG_res_top5pct[[3]] <- lapply(RHD_KEGG_res_top5pct[[3]], function(x){
  x <- NA
  }) %>% 
  as.data.frame() %>% 
  mutate(pop = "DUC")

RHD_KEGG_res_top5pct <- RHD_KEGG_res_top5pct %>% bind_rows() 

#RHD_KEGG_res %>% bind_rows() %>% datatable(caption = "Enriched KEGG terms by RHD genes (FDR < 0.1)", options = list(pageLength = 100))

```

```{r find_common_GOBP_terms, include=FALSE}

RHD_GOBP_res_top5pct_common <- RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUK"] %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUC"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DU6"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DU6P"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "DUhLB"]) %>% 
  intersect(RHD_GOBP_res_top5pct$description[RHD_GOBP_res_top5pct$pop == "FZTDU"]) 

```

```{r find_common_KEGG_terms, eval = FALSE}
RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUK"] %>% 
  #intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUC"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DU6"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DU6P"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "DUhLB"]) %>% 
  intersect(RHD_KEGG_res_top5pct$description[RHD_KEGG_res_top5pct$pop == "FZTDU"]) %>% 
  length() # no intersection
```

```{r heatmap_common_GOBP, eval=FALSE}
png(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top5pct_common.png"), res = 300, height = 1500, width = 2000, units = "px")

RHD_GOBP_res_top5pct %>% 
  filter(description %in% RHD_GOBP_res_top5pct_common) %>% 
  filter(FDR < 0.1) %>% 
  reshape2::dcast(description~pop, value.var = "overlap") %>% 
  column_to_rownames(., var = "description") %>% 
  as.matrix() %>% 
  pheatmap(., main = "RHD_GOBP_res_top5pct_common (n genes)", display_numbers = TRUE, number_format = "%.0f")

dev.off()
```

```{r display_heatmap_common_GOBP, out.width="50%", fig.align="center"}

include_graphics(here("00_dashboard/figures/diversity/RHD_GOBP_res_sig_top5pct_common.png"))
```







### GOBP_immune_system_process

```{r prep_go_genes_for_overlaps, include = FALSE}

# data was downloaded from http://www.informatics.jax.org/go/term/GO:0002376
# This correspond of the root in the tree: http://www.informatics.jax.org/vocab/gene_ontology/GO:0008150
# on 14/01/2020


GOBP_immune_system_process <- here("00_dashboard/data/GO0002376_immune_system_process/GO_term_summary_20210114_063757.txt") %>% 
  vroom() %>% 
  # add ensembl gene set information
  inner_join(mmu_gene_set, by = c("Symbol" = "gene_symbol")) %>% 
  # prepare data for granges
  dplyr::select(seqname, start, end, everything()) %>% 
  dplyr::rename(seqnames = seqname) %>% 
  dplyr::select(-Chr) %>% 
  # make granges
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  sort()
  

```

```{r find_overlaps_goGenes_vs_RHD, include = FALSE}

# RHD=regions of high diversity

# top 1 pct most divese
ov1pct <- findOverlaps(query = outliers_gr_top_1_pct, subject = GOBP_immune_system_process, minoverlap = 1)
immune_system_process_genes_in_1pctRHD <- bind_cols(
  
  outliers_gr_top_1_pct[queryHits(ov1pct)] %>% 
    as.data.frame() %>% 
    mutate(RHD_locus = paste0(seqnames,":",start,"-",end)) %>% 
    dplyr::select(pop, RHD_locus),
  
  GOBP_immune_system_process[subjectHits(ov1pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames,":",start,"-",end)) %>% 
    dplyr::select(-seqnames, -start, -end) %>% 
    dplyr::select(gene_locus, gene_id, gene_biotype, everything())
    
) 

# top 5 pct most divese
ov5pct <- findOverlaps(query = outliers_gr_top_5_pct, subject = GOBP_immune_system_process, minoverlap = 1)
immune_system_process_genes_in_5pctRHD <- bind_cols(
  
  outliers_gr_top_5_pct[queryHits(ov5pct)] %>% 
    as.data.frame() %>% 
    mutate(RHD_locus = paste0(seqnames,":",start,"-",end)) %>% 
    dplyr::select(pop, RHD_locus),
  
  GOBP_immune_system_process[subjectHits(ov5pct)] %>% 
    as.data.frame() %>% 
    mutate(gene_locus = paste0(seqnames,":",start,"-",end)) %>% 
    dplyr::select(-seqnames, -start, -end) %>% 
    dplyr::select(gene_locus, gene_id, gene_biotype, everything())
    
)

```

```{r display_nr_genes_and_terms_summary}
immune_system_process_genes_in_1pctRHD %>% 
  group_by(pop) %>% 
  summarise(n_genes = length(unique(gene_id)), n_terms = length(unique(Annotated.Term))) %>% 
  kable(caption = "Nr of Immune Genes and GOBP terms in extreme top 1% most diverse regions") %>% 
  kable_styling(full_width = F)

immune_system_process_genes_in_5pctRHD %>% 
  group_by(pop) %>% 
  summarise(n_genes = length(unique(gene_id)), n_terms = length(unique(Annotated.Term))) %>% 
  kable(caption = "Nr of Immune Genes and GOBP terms in extreme top 5% most diverse regions") %>% 
  kable_styling(full_width = F)
```

```{r display_nr_genes_and_terms_innate_immunity}

immune_system_process_genes_in_1pctRHD %>% 
  filter(str_detect(Annotated.Term, "innate") ) %>%
  group_by(pop) %>% 
  mutate(gene_symbol_name = paste0(Symbol,"(",Name,")")) %>% 
  summarise(n_genes = length(unique(gene_symbol_name)), n_terms = length(unique(Annotated.Term)),
            terms = paste(unique(Annotated.Term), collapse = " / "),
            gene_details = paste(unique(gene_symbol_name), collapse = " / ")) %>% 
  kable(caption = "Nr of (Innate)Immune Genes in extreme top 1% most diverse regions") %>% 
  kable_styling(full_width = F)


immune_system_process_genes_in_5pctRHD %>% 
  filter(str_detect(Annotated.Term, "innate") ) %>%
  group_by(pop) %>% 
  mutate(gene_symbol_name = paste0(Symbol,"(",Name,")")) %>% 
  summarise(n_genes = length(unique(gene_symbol_name)), n_terms = length(unique(Annotated.Term)),
            terms = paste(unique(Annotated.Term), collapse = " / "),
            gene_details = paste(unique(gene_symbol_name), collapse = " / ")) %>% 
  kable(caption = "Nr of (Innate)Immune Genes in extreme top 5% most diverse regions") %>% 
  kable_styling(full_width = F)

```


